// @nuvem/primeng-components/esm5/lib/crud/components/validation-message/validation-message.component.js
import { __decorate } from "tslib";
import { Component, Input } from "@angular/core";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function ValidationMessageComponent_ng_container_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1, " campo obrigat\u00F3rio ");
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ValidationMessageComponent_ng_container_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r958 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" tamanho m\u00EDnimo ", ctx_r958.control.errors.minlength.requiredLength, " ");
} }
function ValidationMessageComponent_ng_container_0_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r959 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" tamanho m\u00E1ximo ", ctx_r959.control.errors.maxlength.requiredLength, " ");
} }
function ValidationMessageComponent_ng_container_0_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r960 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" valor m\u00EDnimo ", ctx_r960.control.errors.min.min, " ");
} }
function ValidationMessageComponent_ng_container_0_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r961 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" valor m\u00E1ximo ", ctx_r961.control.errors.max.max, " ");
} }
function ValidationMessageComponent_ng_container_0_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1, " campo inv\u00E1lido ");
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ValidationMessageComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 1);
    ɵngcc0.ɵɵtemplate(2, ValidationMessageComponent_ng_container_0_ng_container_2_Template, 2, 0, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(3, ValidationMessageComponent_ng_container_0_ng_container_3_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(4, ValidationMessageComponent_ng_container_0_ng_container_4_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(5, ValidationMessageComponent_ng_container_0_ng_container_5_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(6, ValidationMessageComponent_ng_container_0_ng_container_6_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(7, ValidationMessageComponent_ng_container_0_ng_container_7_Template, 2, 0, "ng-container", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r956 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r956.control.errors.required);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r956.control.errors.minlength);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r956.control.errors.maxlength);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r956.control.errors.min);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r956.control.errors.max);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r956.control.errors.email || ctx_r956.control.errors.pattern);
} }
var ValidationMessageComponent = /** @class */ (function () {
    function ValidationMessageComponent() {
    }
    __decorate([
        Input()
    ], ValidationMessageComponent.prototype, "control", void 0);
ValidationMessageComponent.ɵfac = function ValidationMessageComponent_Factory(t) { return new (t || ValidationMessageComponent)(); };
ValidationMessageComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ValidationMessageComponent, selectors: [["nValidationMessage"]], inputs: { control: "control" }, decls: 1, vars: 1, consts: [[4, "ngIf"], [1, "form-notification-msg", "ui-message", "ui-messages-error", "ui-corner-all"]], template: function ValidationMessageComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ValidationMessageComponent_ng_container_0_Template, 8, 6, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.control.invalid && (ctx.control.dirty || ctx.control.touched) && ctx.control.errors);
    } }, directives: [ɵngcc1.NgIf], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ValidationMessageComponent, [{
        type: Component,
        args: [{
                selector: 'nValidationMessage',
                template: "<ng-container *ngIf=\"control.invalid && (control.dirty || control.touched) && control.errors\">\n    <div class=\"form-notification-msg ui-message ui-messages-error ui-corner-all\" >\n        <ng-container *ngIf=\"control.errors.required\">\n            campo obrigat\u00F3rio\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.minlength\">\n            tamanho m\u00EDnimo {{ control.errors.minlength.requiredLength }}\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.maxlength\">\n            tamanho m\u00E1ximo {{ control.errors.maxlength.requiredLength }}\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.min\">\n            valor m\u00EDnimo {{ control.errors.min.min }}\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.max\">\n            valor m\u00E1ximo {{ control.errors.max.max }}\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.email || control.errors.pattern\">\n            campo inv\u00E1lido\n        </ng-container>\n    </div>\n</ng-container>\n"
            }]
    }], function () { return []; }, { control: [{
            type: Input
        }] }); })();
    return ValidationMessageComponent;
}());
export { ValidationMessageComponent };
