// @nuvem/primeng-components/esm2015/lib/page-notification/page-notification.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
/**
 * Class PageNotificationComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'primeng/toast';
let PageNotificationComponent = class PageNotificationComponent {
    constructor() {
        /**
         * duration property
         * @type {number}
         */
        this.duration = 3000;
    }
};
PageNotificationComponent.ɵfac = function PageNotificationComponent_Factory(t) { return new (t || PageNotificationComponent)(); };
PageNotificationComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: PageNotificationComponent, selectors: [["basis-page-notification"]], inputs: { duration: "duration" }, decls: 1, vars: 0, template: function PageNotificationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "p-toast");
    } }, directives: [ɵngcc1.Toast], encapsulation: 2 });
__decorate([
    Input()
], PageNotificationComponent.prototype, "duration", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PageNotificationComponent, [{
        type: Component,
        args: [{
                selector: 'basis-page-notification',
                template: "<p-toast></p-toast>\n"
            }]
    }], function () { return []; }, { duration: [{
            type: Input
        }] }); })();
export { PageNotificationComponent };
