// @nuvem/primeng-components/fesm2015/nuvem-primeng-components.js
import { __decorate, __param } from 'tslib';
import { HttpParams, HttpClient, HttpClientModule, HTTP_INTERCEPTORS, HttpResponse } from '@angular/common/http';
import { ɵɵdefineInjectable, ɵɵinject, Injectable, EventEmitter, Input, Output, Component, ContentChildren, ContentChild, TemplateRef, Directive, ChangeDetectorRef, ViewChild, ViewEncapsulation, NgModule, CUSTOM_ELEMENTS_SCHEMA, InjectionToken, Inject, ElementRef, Injector, forwardRef, Renderer2, IterableDiffers, Pipe } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PrimeTemplate, SplitButtonModule, MultiSelectModule, CheckboxModule, InputTextModule, MessageService, ButtonModule as ButtonModule$1, TooltipModule, BlockUIModule, ProgressSpinnerModule, PaginatorModule, AutoCompleteModule } from 'primeng';
import { Table, TableModule } from 'primeng/table';
import { Authorization, NotificationProvider, BlockUiService, BlockUiInterceptor, FormNotificationComponentSuperclass, FormNotificationBaseModule, ErrorStackService, ClipboardModule, ErrorProvider, ErrorStackProviderService, WrappedFormControlSuperclass, provideValueAccessor, FieldFormNotification, FormNotificationService, AccessbilityService } from '@nuvem/angular-base';
import { ButtonModule } from 'primeng/button';
import { DropdownModule } from 'primeng/dropdown';
import { FormsModule, ReactiveFormsModule, ControlContainer, NG_VALUE_ACCESSOR } from '@angular/forms';
import { Subject, forkJoin } from 'rxjs';
import { debounceTime, tap, filter, distinctUntilChanged } from 'rxjs/operators';
import { MessageService as MessageService$1, ConfirmationService } from 'primeng/api';
import { ToastModule } from 'primeng/toast';
import { ActivatedRoute, Router, RouterModule, NavigationEnd } from '@angular/router';
import { InputTextModule as InputTextModule$1 } from 'primeng/inputtext';
import { CalendarModule } from 'primeng/calendar';
import { InputTextareaModule } from 'primeng/inputtextarea';
import { InputMaskModule } from 'primeng/inputmask';
import { RadioButtonModule } from 'primeng/radiobutton';
import { ConfirmDialogModule } from 'primeng/confirmdialog';
import { AutoComplete } from 'primeng/autocomplete';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { trigger, state, style, transition, animate } from '@angular/animations';

/**
 * Class DatatableService
 * @class
 */
let DatatableService = class DatatableService {
    /**
     * constructor method
     * @constructor
     * @param {HttpClient} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * search method
     * @param {LazyLoadEvent} event
     * @param {String} url
     * @param {number} rows
     * @param {boolean} orderInSort
     * @param {any} query
     * @return Observable
     */
    search(event, url, rows, orderInSort, query) {
        let page = 0;
        if (event !== undefined && event.first > 0) {
            page = Math.floor(event.first / rows);
        }
        let order = event.sortOrder === 1 ? 'asc' : 'desc';
        let params = new HttpParams()
            .set('page', page.toString())
            .set('size', rows.toString());
        if (orderInSort) {
            if (event.sortField !== undefined) {
                params = params.set('sort', event.sortField + ',' + order);
            }
        }
        else {
            if (event.sortField !== undefined) {
                params = params
                    .set('sort', event.sortField)
                    .set('order', order);
            }
        }
        if ('string' === typeof query) {
            params = params.set('query', query);
        }
        if ('object' === typeof query) {
            Object.keys(query).forEach(key => params = params.set(key, query[key]));
        }
        return this.http.get(`${url}?${params.toString()}`, { observe: 'response' });
    }
};
DatatableService.ctorParameters = () => [
    { type: HttpClient }
];
DatatableService.ɵprov = ɵɵdefineInjectable({ factory: function DatatableService_Factory() { return new DatatableService(ɵɵinject(HttpClient)); }, token: DatatableService, providedIn: "root" });
DatatableService = __decorate([
    Injectable({
        providedIn: "root"
    })
], DatatableService);

/**
 * Class ServerDataProvider
 * @class
 */
class ServerDataProvider {
    /**
     * constructor method
     * @constructor
     * @param {DatatableService} datatableService
     * @param {DatatableComponent} component
     */
    constructor(datatableService, component) {
        this.datatableService = datatableService;
        this.component = component;
        this.component.pDatatableComponent.onLazyLoad.subscribe((event) => this.onLazyLoad(event));
        this.component.pDatatableComponent.lazy = true;
    }
    /**
     * onLazyLoad method
     * @param {LazyLoadEvent} onLazyLoad
     * @returns void
     */
    onLazyLoad(event) {
        this.currentLazyLoadEvent = event;
        this.load();
    }
    /**
     * load method
     * @param {any} query
     * @returns void
     */
    load(query) {
        query = this.retrieveAndSaveQuery(query);
        this.component.startedLoading.emit();
        this.datatableService.search(this.currentLazyLoadEvent, this.component.url, this.component.pDatatableComponent.rows, this.component.orderInSort, query)
            .toPromise()
            .then((response) => {
            this.treatLoadResponse(response);
            this.component.finishedLoading.emit();
        }, (error) => {
            this.component.finishedLoading.emit(error);
        });
    }
    /**
     * treatLoadResponse method
     * @param {any} response
     * @return void
     */
    treatLoadResponse(response) {
        const data = response.body;
        if (this.component.paginationParameters) {
            this.treatLoadResponseWithPaginationParameters(data);
        }
        else {
            this.treatLoadResponseWithoutPaginationParameters(response, data);
        }
    }
    /**
     * treatLoadResponseWithPaginationParameters method
     * @param {any} data
     * @returns void
     */
    treatLoadResponseWithPaginationParameters(data) {
        let contentIndex = this.component.paginationParameters.contentIndex;
        let totalIndex = this.component.paginationParameters.totalElementsIndex;
        this.component.value = (data[contentIndex]) ? data[contentIndex] : data;
        this.component.pDatatableComponent.totalRecords = (data[totalIndex]) ? data[totalIndex] : this.component.value.length;
    }
    /**
     * treatLoadResponseWithoutPaginationParameters method
     * @param {any} response
     * @param {any} data
     * @returns void
     */
    treatLoadResponseWithoutPaginationParameters(response, data) {
        if (response.headers.has('x-total-count')) {
            this.component.pDatatableComponent.totalRecords = parseInt(response.headers.get('x-total-count'), 10);
            this.component.value = data;
        }
        else {
            this.component.pDatatableComponent.totalRecords = data.totalElements;
            this.component.value = data.content;
        }
    }
    /**
     * retrieveAndSaveQuery method
     * @param {any} query
     * @returns any
     */
    retrieveAndSaveQuery(query) {
        const oldQuery = this._query;
        const currentQuery = this.doRetrieveAndSaveQuery(query);
        this.resetToFirstPageIfQueryChanged(oldQuery, currentQuery);
        return currentQuery;
    }
    /**
     * doRetrieveAndSaveQuery method
     * @param {any} query
     * @returns any
     */
    doRetrieveAndSaveQuery(query) {
        if ('object' === typeof query && query.value) {
            this._query = query.value;
        }
        else if (query) {
            this._query = query;
        }
        return this._query;
    }
    /**
     * resetToFirstPageIfQueryChanged method
     * @param {any} oldQuery
     * @param {any} currentQuery
     * @returns void
     */
    resetToFirstPageIfQueryChanged(oldQuery, currentQuery) {
        if (oldQuery !== currentQuery) {
            this.currentLazyLoadEvent.first = 0;
            this.component.pDatatableComponent.first = 0;
        }
    }
    /**
     * reset method
     * @returns void
     */
    reset() {
        this._query = undefined;
    }
}

/**
 * Class MemoryDataProvider
 * @class
 */
class MemoryDataProvider {
    /**
     * constructor method
     * @param {DatatableComponent} component
     * @constructor
     */
    constructor(component) {
        this.component = component;
    }
    /**
     * load method
     * @param {any} query
     * @returns void
     */
    load(query) {
        this.component.pDatatableComponent.reset();
        for (let prop in query) {
            let matchMode = 'contains';
            if (query[prop] instanceof Array) {
                matchMode = 'in';
            }
            this.component.pDatatableComponent.filter(query[prop], prop, matchMode);
        }
        this.component.pDatatableComponent.globalFilterFields = query;
        this.component.pDatatableComponent._filter();
    }
    /**
     * reset method
     * @returns void
     */
    reset() {
        this.component.pDatatableComponent.globalFilterFields = undefined;
    }
}

/**
 * Enum DataProviderType
 * @enum
 */
var DataProviderType;
(function (DataProviderType) {
    DataProviderType["Memory"] = "memory";
    DataProviderType["Server"] = "server";
})(DataProviderType || (DataProviderType = {}));
/**
 * Class DataProviderFactory
 * @class
 */
let DataProviderFactory = class DataProviderFactory {
    /**
     * constructor method
     * @constructor
     * @param {DatatableService} datatableService
     */
    constructor(datatableService) {
        this.datatableService = datatableService;
    }
    /**
     * create method
     * @param {DataProviderType} type
     * @param {DatatableComponent} component
     * @public
     * @returns DataProvider
     */
    create(type, component) {
        if (DataProviderType.Server === type) {
            return new ServerDataProvider(this.datatableService, component);
        }
        if (DataProviderType.Memory === type) {
            return new MemoryDataProvider(component);
        }
    }
};
DataProviderFactory.ctorParameters = () => [
    { type: DatatableService }
];
DataProviderFactory.ɵprov = ɵɵdefineInjectable({ factory: function DataProviderFactory_Factory() { return new DataProviderFactory(ɵɵinject(DatatableService)); }, token: DataProviderFactory, providedIn: "root" });
DataProviderFactory = __decorate([
    Injectable({
        providedIn: "root"
    })
], DataProviderFactory);

/**
 * Class DatatableButtonComponent
 * @class
 */
let DatatableButtonComponent = class DatatableButtonComponent {
    constructor() {
        /**
         * allowMultipleSelection property
         * @type {boolean}
         */
        this.allowMultipleSelection = true;
        /**
         * disabled property
         * @type boolean
         */
        this.disabled = false;
        /**
         * click property
         * @type {EventEmitter<any>}
         */
        this.click = new EventEmitter();
    }
    /**
     * onClick method
     * @param {any} event
     * @return void
     */
    onClick(event) {
        this.click.emit(event);
        event.stopPropagation();
    }
};
__decorate([
    Input()
], DatatableButtonComponent.prototype, "icon", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "class", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "bTooltip", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "name", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "allowMultipleSelection", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "disabled", void 0);
__decorate([
    Output()
], DatatableButtonComponent.prototype, "click", void 0);
DatatableButtonComponent = __decorate([
    Component({
        selector: 'basis-datatable-button',
        template: "<p pTooltip=\"{{ bTooltip }}\">\n    <button\n        type=\"button\" tooltipPosition=\"bottom\" class=\"{{ class }}\"\n        icon=\"ui-icon-{{ icon }}\" (click)=\"onClick($event)\" [disabled]=\"disabled\" pButton>\n    </button>\n</p>\n"
    })
], DatatableButtonComponent);

/**
 * Class DatatableClickEvent
 * @class
 */
class DatatableClickEvent {
    /**
     * constructor method
     * @constructor
     * @param {string} button
     * @param {any} selection
     */
    constructor(button, selection) {
        this.button = button;
        this.selection = selection;
    }
}

let Column = class Column {
    constructor() {
        this.filterType = 'text';
        this.exportable = true;
        this.resizable = true;
        this.sortFunction = new EventEmitter();
    }
    ngAfterContentInit() {
        this.templates.forEach((item) => {
            switch (item.getType()) {
                case 'header':
                    this.headerTemplate = item.template;
                    break;
                case 'body':
                    this.bodyTemplate = item.template;
                    break;
                case 'footer':
                    this.footerTemplate = item.template;
                    break;
                case 'filter':
                    this.filterTemplate = item.template;
                    break;
                case 'editor':
                    this.editorTemplate = item.template;
                    break;
                case 'input':
                    this.inputTemplate = item.template;
                    break;
                case 'output':
                    this.outputTemplate = item.template;
                    break;
                default:
                    this.bodyTemplate = item.template;
                    break;
            }
        });
    }
};
__decorate([
    Input()
], Column.prototype, "field", void 0);
__decorate([
    Input()
], Column.prototype, "colId", void 0);
__decorate([
    Input()
], Column.prototype, "sortField", void 0);
__decorate([
    Input()
], Column.prototype, "filterField", void 0);
__decorate([
    Input()
], Column.prototype, "header", void 0);
__decorate([
    Input()
], Column.prototype, "footer", void 0);
__decorate([
    Input()
], Column.prototype, "sortable", void 0);
__decorate([
    Input()
], Column.prototype, "editable", void 0);
__decorate([
    Input()
], Column.prototype, "filter", void 0);
__decorate([
    Input()
], Column.prototype, "filterMatchMode", void 0);
__decorate([
    Input()
], Column.prototype, "filterType", void 0);
__decorate([
    Input()
], Column.prototype, "excludeGlobalFilter", void 0);
__decorate([
    Input()
], Column.prototype, "rowspan", void 0);
__decorate([
    Input()
], Column.prototype, "colspan", void 0);
__decorate([
    Input()
], Column.prototype, "scope", void 0);
__decorate([
    Input()
], Column.prototype, "style", void 0);
__decorate([
    Input()
], Column.prototype, "styleClass", void 0);
__decorate([
    Input()
], Column.prototype, "exportable", void 0);
__decorate([
    Input()
], Column.prototype, "headerStyle", void 0);
__decorate([
    Input()
], Column.prototype, "headerStyleClass", void 0);
__decorate([
    Input()
], Column.prototype, "bodyStyle", void 0);
__decorate([
    Input()
], Column.prototype, "bodyStyleClass", void 0);
__decorate([
    Input()
], Column.prototype, "footerStyle", void 0);
__decorate([
    Input()
], Column.prototype, "footerStyleClass", void 0);
__decorate([
    Input()
], Column.prototype, "hidden", void 0);
__decorate([
    Input()
], Column.prototype, "expander", void 0);
__decorate([
    Input()
], Column.prototype, "selectionMode", void 0);
__decorate([
    Input()
], Column.prototype, "filterPlaceholder", void 0);
__decorate([
    Input()
], Column.prototype, "filterMaxlength", void 0);
__decorate([
    Input()
], Column.prototype, "frozen", void 0);
__decorate([
    Input()
], Column.prototype, "resizable", void 0);
__decorate([
    Output()
], Column.prototype, "sortFunction", void 0);
__decorate([
    ContentChildren(PrimeTemplate)
], Column.prototype, "templates", void 0);
__decorate([
    ContentChild(TemplateRef)
], Column.prototype, "template", void 0);
Column = __decorate([
    Directive({
        selector: 'p-column'
    })
], Column);

/**
 * Class DatatableComponent
 * @class
 */
let DatatableComponent = class DatatableComponent {
    /**
     * constructor method
     * @constructor
     * @param {ChangeDetectorRef} changeDetectorRef
     * @param {DataProviderFactory} dataProviderFactory
     * @param {Authorization} authorizationService
     */
    constructor(changeDetectorRef, dataProviderFactory, authorizationService) {
        this.changeDetectorRef = changeDetectorRef;
        this.dataProviderFactory = dataProviderFactory;
        this.authorizationService = authorizationService;
        /**
         * filterParams property
         * @type {any}
         */
        this.filterParams = {};
        /**
         * footerMsg property
         * @type {string}
         */
        this.footerMsg = '';
        /**
         * orderInSort property
         * @type {boolean}
         */
        this.orderInSort = false;
        /**
         * value property
         * @type {any}
         */
        this.value = [];
        /**
         * type property
         * @type DataProviderType
         */
        this.type = DataProviderType.Server;
        /**
         * rows property
         * @type {number}
         */
        this.rows = 5;
        /**
         * showPaginationFooter property
         * @type {boolean}
         */
        this.showPaginationFooter = false;
        /**
         * disableEdit property
         * @type {boolean}
         */
        this.disableEdit = false;
        /**
         * disableView property
         * @type {boolean}
         */
        this.disableView = false;
        /**
         * disableDelete property
         * @type {boolean}
         */
        this.disableDelete = false;
        /**
         * selectionMode property
         * @type {string}
         */
        this.selectionMode = 'single';
        /**
         * enableButtonsScroll property
         * @type {boolean}
         */
        this.enableButtonsScroll = false;
        /**
         * emptyMessage property
         * @type {string}
         */
        this.emptyMessage = 'Nenhum registro encontrado.';
        /**
         * verticalButtons property
         * @type {boolean}
         */
        this.verticalButtons = true;
        /**
         * enableScroll property
         * @type {boolean}
         */
        this.enableScroll = false;
        /**
         * disableLoadingBlockUI property
         * @type {boolean}
         */
        this.disableLoadingBlockUI = false;
        /**
         * filterOnColumn property
         */
        this.filterOnColumn = false;
        /**
         * filterSelectOptions property
         */
        this.filterSelectOptions = [];
        /**
         * updateFiltroHistorico property
         */
        this.updateFilterColumn = false;
        /**
         * customFilterOptions property
         */
        this.customFilterOptions = {};
        /**
         * buttonClick property
         * @type {EventEmitter<any>}
         */
        this.buttonClick = new EventEmitter();
        /**
         * startedLoading property
         * @type {EventEmitter<void>}
         */
        this.startedLoading = new EventEmitter();
        /**
         * finishedLoading property
         * @type {EventEmitter<HttpErrorResponse>}
         */
        this.finishedLoading = new EventEmitter();
        /**
         * onRowSelected property
         * @type {EventEmitter<any>}
         */
        this.onRowSelected = new EventEmitter();
        /**
         * showHorizontalButtonsHeader property
         * @type {boolean}
         */
        this.showHorizontalButtonsHeader = false;
        /**
         * blockContent
         * @type {boolean}
         */
        this.blockContent = false;
        /**
         * visibleColumns property
         * @type {Object}
         */
        this.visibleColumns = {};
        /**
         * showVisibleColumnsControl property
         * @type {boolean}
         */
        this.showVisibleColumnsControl = false;
        /**
         * dataKey property
         * Identificador único para cada registro da tabela
         * @type {string}
         */
        this.dataKey = 'id';
        /**
         * onRowEditEvent property
         * Evento disparado quando o usuário realiza ações de edição em uma linha da tabela
         * @type {EventEmitter<void>}
         */
        this.onRowEditEvent = new EventEmitter();
        this.clonedRowData = {};
        /**
         * errorMessage property
         * @type {string}
         */
        this.errorMessage = 'Ocorreu um erro ao carregar os dados da tabela. Tente novamente.';
        this.editingRow = false;
    }
    /**
     * ngOnInit method
     * @returns void
     */
    ngOnInit() {
        this.dataProvider = this.dataProviderFactory.create(this.type, this);
        this.pDatatableComponent.rows = this.rows;
        this.subscribeToLoadingEvents();
        this.initRowsPerPageOptionsIfPresent();
        this.defineIfShouldShowHorizontalButtonsHeader();
        this.updateOnValueChange();
        this.finishedLoading.subscribe(() => {
            this.updateOnValueChange();
        });
    }
    /**
     * subscribeFromValueChange method
     * @returns void
     */
    updateOnValueChange() {
        this.filterSelectOptions = [];
        if (this.value) {
            this.value.map((item) => {
                this.filterSelectOptions.push(item);
            });
        }
    }
    /**
     * subscribeToLoadingEvents method
     * @returns void
     */
    subscribeToLoadingEvents() {
        if (!this.disableLoadingBlockUI) {
            this.originalEmptyMessage = this.emptyMessage;
            this.startedLoading.subscribe(() => this.onStartedLoading());
            this.finishedLoading.subscribe(error => this.onFinishedLoading(error));
        }
    }
    /**
     * onStartedLoading method
     * @returns void
     */
    onStartedLoading() {
        this.blockContent = true;
        this.emptyMessage = this.originalEmptyMessage;
    }
    /**
     * onFinishedLoading method
     * @param {HttpErrorResponse} error
     * @returns void
     */
    onFinishedLoading(error) {
        this.blockContent = false;
        if (error) {
            this.emptyMessage = this.errorMessage;
        }
    }
    /**
     * initRowsPerPageOptionsIfPresent method
     * @returns void
     */
    initRowsPerPageOptionsIfPresent() {
        if (this.rowsPerPageOptions) {
            this.pDatatableComponent.rowsPerPageOptions = this.rowsPerPageOptions;
        }
    }
    /**
     * defineIfShouldShowHorizontalButtonsHeader method
     * @returns void
     */
    defineIfShouldShowHorizontalButtonsHeader() {
        if (this.horizontalButtonsHeader) {
            this.showHorizontalButtonsHeader = true;
        }
    }
    /**
     * ngAfterViewInit method
     * @returns void
     */
    ngAfterViewInit() {
        this.setExtraParams();
        this.pDatatableComponent.columns = this.columns;
        this.pDatatableComponent.ngAfterContentInit();
        this.setAllColumnsVisible();
        this.changeDetectorRef.detectChanges();
    }
    /**
     * setExtraParams method
     * @returns void
     */
    setExtraParams() {
        for (let i in this.extraParams) {
            this.pDatatableComponent[i] = this.extraParams[i];
        }
    }
    /**
     * ngAfterContentInit method
     * @returns void
     */
    ngAfterContentInit() {
        this.extraButtons.forEach((button) => {
            button.click.subscribe(event => {
                this.onClick(button.name, event);
            });
        });
    }
    /**
     * onClick method
     * @param {string} button
     * @param {any} event
     * @return void
     */
    onClick(button, event) {
        this.buttonClick.emit(new DatatableClickEvent(button, this.selectedRow));
        event.stopPropagation();
        this.resetSelectedRow(button);
    }
    /**
     * onSelectChange method
     * @param {any} event
     * @returns void
     */
    onSelectChange(event) {
        this.onRowSelected.emit(this.selectedRow);
        if (Array.isArray(this.selectedRow) && this.selectedRow.length > 1) {
            this.updateMultipleButtonsStatus(true);
        }
        else {
            this.updateMultipleButtonsStatus(false);
        }
    }
    /**
     * updateMultipleButtonsStatus method
     * @param {boolean} hasMultipleSelection
     * @return void
     */
    updateMultipleButtonsStatus(hasMultipleSelection) {
        this.extraButtons.forEach((button) => {
            if (hasMultipleSelection && !button.allowMultipleSelection) {
                button.disabled = true;
            }
            else {
                button.disabled = false;
            }
        });
    }
    /**
     * refresh method
     * @param query
     * @returns void
     */
    refresh(query) {
        this.dataProvider.load(query);
    }
    /**
     * resetSelectedRow method
     * @param button
     * @returns void
     */
    resetSelectedRow(button) {
        if (button === 'delete') {
            this.selectedRow = null;
        }
    }
    /**
     * reset method
     * @returns void
     */
    reset() {
        this.pDatatableComponent.first = 0;
        this.dataProvider.reset();
        this.refresh();
    }
    /**
     * paginationFooterText method
     * @param {string} msg
     * @returns string
     */
    paginationFooterText(msg) {
        const registros = this.totalRecords() === 1 ? 'registro' : 'registros';
        if (msg === '') {
            return `Exibindo: ${this.currentFirstIndex()} a ${this.currentLastIndex()} de ${this.totalRecords()} ${registros}`;
        }
        msg = msg.replace('{totalRecords}', this.totalRecords().toString());
        msg = msg.replace('{currentFirstIndex}', this.currentFirstIndex().toString());
        msg = msg.replace('{currentLastIndex}', this.currentLastIndex().toString());
        return msg;
    }
    /**
     * currentFirstIndex method
     * @returns number
     */
    currentFirstIndex() {
        return this.pDatatableComponent.first + 1;
    }
    /**
     * currentLastIndex method
     * @returns number
     */
    currentLastIndex() {
        const currentLastIndex = this.pDatatableComponent.first + this.pDatatableComponent.rows;
        if (currentLastIndex > this.totalRecords()) {
            return this.totalRecords();
        }
        return currentLastIndex;
    }
    /**
     * totalRecords method
     * @returns number
     */
    totalRecords() {
        return this.pDatatableComponent.totalRecords;
    }
    /**
     * checkRolesToButton method
     * @param {string} button
     * @return boolean
     */
    checkRolesToButton(button) {
        if (this[button]) {
            if (button === 'rulesToHide' || button === 'rulesToShowEdit' || button === 'rulesToShowDelete') {
                this.checkUseButtons(button);
                return this.authorizationService.hasRole(this[button]);
            }
            return !this.authorizationService.hasRole(this[button]);
        }
        return true;
    }
    /**
     * checkUseButtons method
     * @param {string} button
     * @returns void
     */
    checkUseButtons(button) {
        if (this[button] && this[button.replace('Show', 'Hide')]) {
            throw new Error('Não é possível definir dois valores de roles para os botões padrões.');
        }
    }
    /**
     * filter method
     * @returns void
     */
    filter() {
        for (let prop in this.filterParams) {
            if (this.filterParams[prop] === '' || (this.filterParams[prop].length === 0)) {
                delete this.filterParams[prop];
            }
        }
        if (Object.getOwnPropertyNames(this.filterParams).length === 0) {
            this.reset();
        }
        else {
            this.refresh(this.filterParams);
        }
    }
    /**
     * Método para filtrar por colunas
     * @param column chave da coluna na tabela
     * @param filters valor do filtro
     */
    filterColumn(column, filters) {
        const values = [];
        if (filters instanceof Array) {
            values[column] = [];
            for (let item in filters) {
                let val = (filters[item] instanceof Object) ? filters[item].value : filters[item];
                values[column].push(val);
            }
        }
        else {
            values[column] = filters;
        }
        this.filterParams = Object.assign(Object.assign({}, this.filterParams), values);
        return this.filter();
    }
    /**
     * Inicia todas as colunas da tabela como visíveis
     */
    setAllColumnsVisible() {
        this.columns.forEach((col) => {
            this.visibleColumns[col.field] = 'table-cell';
        });
    }
    /**
     * Verifica se uma coluna deve estar visível
     * @param column string
     * @param visibleColumns Array
     */
    visibleColumnCheck(column, visibleColumns) {
        return visibleColumns.some((item) => {
            return (item.field) ? item.field == column : true;
        });
    }
    /**
     * Atualzia os valores da variável 'visibleColuns' para mostrar ou não as colunas
     * @param columns array
     */
    updateVisibleColumns(columns) {
        for (let col in this.visibleColumns) {
            if (this.visibleColumnCheck(col, columns)) {
                this.visibleColumns[col] = 'table-cell';
            }
            else {
                this.visibleColumns[col] = 'none';
            }
        }
    }
    onRowEditInitEvent(event, rowData) {
        const tr = event.srcElement.closest('tr');
        setTimeout(() => {
            this.editingRow = true;
            const top = tr.offsetTop;
            const height = tr.offsetHeight;
            const editButtons = tr.querySelector('.editar-campos');
            editButtons.style.top = (top + height - 75) + 'px';
            editButtons.style.height = '75px';
            editButtons.style.backgroundColor = '#d0d0d0';
            editButtons.style.opacity = 1;
            this.clonedRowData[rowData.id] = Object.assign({}, rowData);
            this.onRowEditEvent.emit({ event: "init", data: rowData });
        }, null, tr);
    }
    onRowEditSaveEvent(event, rowData) {
        this.onRowEditEvent.emit({ event: "save", data: rowData });
        this.resetEditingMode(event);
    }
    onRowEditCancelEvent(event, rowData, index) {
        this.value[index] = this.clonedRowData[rowData.id];
        delete this.clonedRowData[rowData[this.dataKey]];
        this.onRowEditEvent.emit({ event: "cancel", data: rowData });
        this.resetEditingMode(event);
    }
    resetEditingMode(event) {
        const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
        editButtons.style.opacity = 0;
        editButtons.style.backgroundColor = '#0000006b';
        editButtons.style.display = 'none';
        this.editingRow = false;
    }
    editarLinhaOver(event) {
        if (!this.editingRow && this.editMode == 'row') {
            const top = event.srcElement.closest('tr').offsetTop;
            const height = event.srcElement.closest('tr').offsetHeight;
            const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
            editButtons.style.display = 'flex';
            editButtons.style.top = top + 'px';
            editButtons.style.height = height + 'px';
            editButtons.style.backgroundColor = '#0000006b';
            editButtons.style.opacity = 1;
        }
    }
    editarLinhaOut(event) {
        if (!this.editingRow) {
            const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
            editButtons.style.display = 'none';
            editButtons.style.opacity = 0;
        }
    }
};
DatatableComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: DataProviderFactory },
    { type: Authorization }
];
__decorate([
    Input()
], DatatableComponent.prototype, "splitButton", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "value", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "type", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rows", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rowsPerPageOptions", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "url", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "extraParams", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "paginationParameters", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "showPaginationFooter", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "disableEdit", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "disableView", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "disableDelete", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "selectionMode", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "enableButtonsScroll", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "emptyMessage", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "verticalButtons", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "enableScroll", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "scrollHeight", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "scrollWidth", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToHide", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToShowEdit", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToShowDelete", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToHideView", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToHideEdit", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToHideDelete", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "disableLoadingBlockUI", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rowStyleClass", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "filterOnColumn", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "customFilterOptions", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "buttonClick", void 0);
__decorate([
    ViewChild(Table, { static: true })
], DatatableComponent.prototype, "pDatatableComponent", void 0);
__decorate([
    ContentChildren(DatatableButtonComponent)
], DatatableComponent.prototype, "extraButtons", void 0);
__decorate([
    ContentChildren(Column)
], DatatableComponent.prototype, "columns", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "startedLoading", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "finishedLoading", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "onRowSelected", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "showVisibleColumnsControl", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "editMode", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "dataKey", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "onRowEditEvent", void 0);
DatatableComponent = __decorate([
    Component({
        selector: 'basis-datatable',
        template: "<p-blockUI [blocked]=\"blockContent\">\n    <div class=\"center-spinner\">\n        <p-progressSpinner></p-progressSpinner>\n    </div>\n</p-blockUI>\n<div [ngClass]=\"{'ui-g-2':verticalButtons, 'ui-g-5':!verticalButtons}\">&nbsp;</div>\n<div [ngClass]=\"verticalButtons ? 'ui-g-11' : 'ui-g-12'\">\n    <basis-show-column *ngIf=\"showVisibleColumnsControl\" [columns]=\"columns\" (updateVisibleColumns)=\"updateVisibleColumns($event)\"></basis-show-column>\n</div>\n<div [ngClass]=\"{\n        'ui-g-11' : verticalButtons,\n        'ui-g-12' : !verticalButtons,\n        'edicao-tabela' : editMode == 'row',\n        'editing-row': editingRow\n    }\">\n    <p-table\n        [value]=\"value\"\n        [rows]=\"rows\"\n        [paginator]=\"true\"\n        [selectionMode]=\"selectionMode\"\n        [responsive]=\"true\"\n        [scrollable]=\"enableScroll\"\n        [scrollHeight]=\"scrollHeight\"\n        [(selection)]=\"selectedRow\"\n        [editMode]=\"editMode\"\n        (onRowSelect)=\"onSelectChange($event)\"\n        (onRowUnselect)=\"onSelectChange($event)\"\n        #datatable\n        [dataKey]=\"dataKey\">\n        <ng-template pTemplate=\"header\" let-columns>\n            <div class=\"ui-g\" *ngIf=\"!verticalButtons\">\n                <div *ngIf=\"showHorizontalButtonsHeader\" class=\"ui-g-4 ui-sm-12\">\n                    <span class=\"horizontal-buttons-header\">{{ horizontalButtonsHeader }}</span>\n                </div>\n                <div [ngClass]=\"{\n                    'horizontal-buttons': true,\n                    'ui-sm-12': true,\n                    'ui-g-8': showHorizontalButtonsHeader,\n                    'ui-g-12': !showHorizontalButtonsHeader}\">\n                    <ng-container *ngTemplateOutlet=\"defaultButtons\"></ng-container>\n                    <ng-content select=\"horizontalButtons\"></ng-content>\n                </div>\n            </div>\n            <tr>\n                <th *ngFor=\"let col of columns\"\n                    [pSortableColumn]=\"col.sortable ? col.field : null\"\n                    [ngStyle]=\"{'display': visibleColumns[col.field]}\">\n                    {{col.header}}\n                    <p-sortIcon *ngIf=\"col.sortable\" [field]=\"col.field\"></p-sortIcon>\n                </th>\n            </tr>\n            <tr *ngIf=\"filterOnColumn\">\n                <th *ngFor=\"let col of columns\" [ngStyle]=\"{'display': visibleColumns[col.field]}\">\n                    <basis-filter-column\n                        [listaOriginal]=\"filterSelectOptions\"\n                        [lista]=\"filterSelectOptions\"\n                        [field]=\"col.field\"\n                        [updateFiltro]=\"updateFilterColumn\"\n                        (outputEvent)=\"filterColumn(col.field, $event)\"\n                        [filterType]=\"col.filterType\"\n                        [tableType]=\"type\"\n                        [customFilterOptions]=\"customFilterOptions\">\n                    </basis-filter-column>\n                </th>\n            </tr>\n        </ng-template>\n        <ng-template pTemplate=\"body\" let-rowData let-columns=\"columns\" let-editing=\"editing\" let-ri=\"rowIndex\">\n            <tr [pSelectableRow]=\"rowData\"\n                [pContextMenuRow]=\"rowData\"\n                [pEditableRow]=\"rowData\"\n                [ngClass]=\"{'editing': editing}\"\n                (mouseover)=\"editarLinhaOver($event)\"\n                (mouseleave)=\"editarLinhaOut($event)\">\n                <td *ngFor=\"let col of columns; last as isLast\" [ngStyle]=\"{'display': visibleColumns[col.field]}\" >\n                    <ng-container *ngIf=\"!editMode\">\n                        <span *ngIf=\"!col.bodyTemplate\" [innerHTML]=\"rowData[col.field]\"></span>\n                        <ng-container\n                            *ngIf=\"col.bodyTemplate\"\n                            [ngTemplateOutlet]=\"col.bodyTemplate\"\n                            [ngTemplateOutletContext]=\"{ rowData: rowData, col: col }\">\n                        </ng-container>\n                    </ng-container>\n\n                    <ng-container *ngIf=\"editMode\">\n                        <p-cellEditor>\n                            <ng-template pTemplate=\"input\">\n                                <span *ngIf=\"!col.outputTemplate && col.editable == false\" [innerHTML]=\"rowData[col.field]\"></span>\n                                <div class=\"ui-md-12\" *ngIf=\"!col.inputTemplate && col.editable != false\">\n                                    <div class=\"ui-g form-group\">\n                                        <div class=\"ui-md-12\">\n                                            <span class=\"md-inputfield\">\n                                                <input pInputText type=\"text\" [(ngModel)]=\"rowData[col.field]\" [style]=\"{'width':'100%'}\">\n                                                <label>{{ col.header }}</label>\n                                            </span>\n                                        </div>\n                                    </div>\n                                </div>\n                                <ng-container\n                                    *ngIf=\"col.inputTemplate\"\n                                    [ngTemplateOutlet]=\"col.inputTemplate\"\n                                    [ngTemplateOutletContext]=\"{ 'rowData': rowData, 'col': col }\">\n                                </ng-container>\n                            </ng-template>\n                            <ng-template pTemplate=\"output\">\n                                <span *ngIf=\"!col.outputTemplate\" [innerHTML]=\"rowData[col.field]\"></span>\n                                <ng-container\n                                    *ngIf=\"col.outputTemplate\"\n                                    [ngTemplateOutlet]=\"col.outputTemplate\"\n                                    [ngTemplateOutletContext]=\"{ rowData: rowData, col: col }\">\n                                </ng-container>\n                            </ng-template>\n                        </p-cellEditor>\n                    </ng-container>\n                    <div *ngIf=\"isLast\" class=\"editar-campos\">\n                        <button *ngIf=\"!editing\" pButton type=\"button\" pInitEditableRow (click)=\"onRowEditInitEvent($event, rowData)\">Editar</button>\n                        <button *ngIf=\"editing\" pButton type=\"button\" pSaveEditableRow class=\"ui-button-success\" style=\"margin-right: .5em\" (click)=\"onRowEditSaveEvent($event, rowData)\">Salvar</button>\n                        <button *ngIf=\"editing\" pButton type=\"button\" pCancelEditableRow class=\"ui-button-danger\" (click)=\"onRowEditCancelEvent($event, rowData, ri)\">Cancelar</button>\n                    </div>\n                </td>\n            </tr>\n        </ng-template>\n        <ng-template *ngIf=\"emptyMessage\">\n            {{ emptyMessage }}\n        </ng-template>\n        <p-footer *ngIf=\"showPaginationFooter && totalRecords() > 0\">\n            {{ paginationFooterText(footerMsg) }}\n        </p-footer>\n\n    </p-table>\n</div>\n<div *ngIf=\"verticalButtons\"\n    class=\"ui-g-1 ui-g-nopad\"\n    [style.position]=\"enableButtonsScroll ? 'sticky' : ''\"\n    [style.top]=\"enableButtonsScroll ? '15%' : ''\">\n        <ng-container *ngTemplateOutlet=\"defaultButtons\"></ng-container>\n        <ng-content></ng-content>\n</div>\n\n<ng-template #defaultButtons>\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Editar' : 'Escolha um registro para editar.' \"\n        [class.horizontal-button]=\"!verticalButtons\" class=\"ui-button-success\" icon=\"edit\" (click)=\"onClick('edit', $event)\"\n        *ngIf=\"!disableEdit && checkRolesToButton('rulesToShowView') && checkRolesToButton('rulesToHideView')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Visualizar' : 'Escolha um registro para visualizar.'\"\n        [class.horizontal-button]=\"!verticalButtons\" icon=\"remove-red-eye\" (click)=\"onClick('view', $event)\"\n        *ngIf=\"!disableView && checkRolesToButton('rulesToShowEdit') && checkRolesToButton('rulesToHideEdit')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Excluir' : 'Escolha um registro para excluir.'\"\n        [class.horizontal-button]=\"!verticalButtons\" class=\"ui-button-danger\" icon=\"delete\" (click)=\"onClick('delete', $event)\"\n       *ngIf=\"!disableDelete && checkRolesToButton('rulesToShowDelete') && checkRolesToButton('rulesToHideDelete')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <datatable-splitbutton\n        [icon]=\"splitButton.icon\"\n        [class]=\"splitButton.class\"\n        [items]=\"splitButton.itens\"\n        [label]=\"splitButton.label\"\n        [method]=\"splitButton.method\"\n        *ngIf=\"splitButton\">\n    </datatable-splitbutton>\n</ng-template>\n",
        encapsulation: ViewEncapsulation.None,
        styles: ["body .ui-datatable .ui-datatable-data tr td{word-wrap:break-word}.horizontal-buttons{display:flex;flex-flow:row wrap;justify-content:flex-end}.horizontal-button{padding:0 3px}.horizontal-buttons-header{font-size:1.3em;word-wrap:break-word}.center-spinner{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}.text-right{text-align:right}basis-show-column .ui-multiselect-panel.ui-widget{width:100%;max-width:100%}.edicao-tabela,.edita-linha{position:relative}tr.editing td{padding-bottom:75px!important}.editar-campos{opacity:0;position:absolute;width:100%;background-color:#0000006b;left:0;margin-top:0;display:flex;align-items:center;justify-content:center;min-height:35px;transition:.3s ease-in-out;z-index:1}.editar-campos button{color:#fff;font-size:14px;padding:6px 30px 10px;border-radius:5px;height:28px;width:120px!important}.bt-cancelar,.bt-salvar{margin:0 10px}"]
    })
], DatatableComponent);

/**
 * Class DatatableSplitbuttonComponent
 * @class
 */
let DatatableSplitbuttonComponent = class DatatableSplitbuttonComponent {
    /**
     * ngOnInit method
     * @returns void
     */
    ngOnInit() {
    }
};
__decorate([
    Input()
], DatatableSplitbuttonComponent.prototype, "items", void 0);
__decorate([
    Input()
], DatatableSplitbuttonComponent.prototype, "label", void 0);
__decorate([
    Input()
], DatatableSplitbuttonComponent.prototype, "class", void 0);
__decorate([
    Input()
], DatatableSplitbuttonComponent.prototype, "icon", void 0);
__decorate([
    Input()
], DatatableSplitbuttonComponent.prototype, "method", void 0);
DatatableSplitbuttonComponent = __decorate([
    Component({
        selector: 'datatable-splitbutton',
        template: "<p-splitButton label=\"{{label}}\" icon=\"{{icon}}\" (onClick)=\"method\" [model]=\"items\" styleClass=\"{{class}}\">\n</p-splitButton>"
    })
], DatatableSplitbuttonComponent);

/**
 * Class DatatableSplitbuttonModule
 * @class
 */
let DatatableSplitbuttonModule = class DatatableSplitbuttonModule {
};
DatatableSplitbuttonModule = __decorate([
    NgModule({
        imports: [
            SplitButtonModule
        ],
        declarations: [
            DatatableSplitbuttonComponent
        ],
        providers: [],
        exports: [
            DatatableSplitbuttonComponent,
            SplitButtonModule
        ],
        schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
], DatatableSplitbuttonModule);

let FilterColumn = class FilterColumn {
    constructor() {
        this.listaOriginal = [];
        this.lista = [];
        this.labelDefaultImplementation = false;
        this.updateFiltro = false;
        this.customFilterOptions = {};
        this.template = 'filter_text';
        this.outputEvent = new EventEmitter();
        this.debouncer = new Subject();
        this.resultado = [];
        this.listaFields = [];
    }
    ngOnInit() {
        this.debouncer
            .pipe(debounceTime((this.tableType === 'memory') ? 300 : 1000))
            .subscribe((value) => this.outputEvent.emit(value));
        this.template = `filter_${this.filterType}`;
    }
    ngOnChanges(changes) {
        if (this.checkChanges(changes, 'listaOriginal')) {
            this.preencherFields();
        }
    }
    checkChanges(changes, property) {
        return changes[property] && changes[property].previousValue != changes[property].currentValue;
    }
    /**
     * Método para propagar o evento de filtragem
     * @param event filtro
     */
    filtrar(event) {
        this.debouncer.next(this.filtro);
    }
    /**
     * Preenche os itens que serão exibidos no multiselect, removendo duplicidades
     */
    preencherFields() {
        this.listaFields = [];
        this.listaOriginal.filter(pi => pi[this.field] !== null).forEach(row => {
            let item = { label: row[this.field], value: row[this.field] };
            if (!this.verificarSeExisteField(item)) {
                this.listaFields.push(item);
            }
        });
    }
    /**
     *
     * @param field Verifica duplicidade de fields nas opções de seleção
     */
    verificarSeExisteField(field) {
        return this.listaFields.some(item => {
            return item.value == field.value;
        });
    }
    /**
     * Caso o usuaŕio não forneça uma lista para o options, será criada uma lista com
     * os valores presentes na tabela
     */
    getCustomFilterOptions() {
        if (this.customFilterOptions.hasOwnProperty(this.field)) {
            return this.customFilterOptions[this.field];
        }
        return this.listaFields;
    }
};
__decorate([
    Input()
], FilterColumn.prototype, "listaOriginal", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "lista", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "field", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "header", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "labelDefaultImplementation", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "labelDefaultName", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "updateFiltro", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "filterType", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "tableType", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "customFilterOptions", void 0);
__decorate([
    Input()
], FilterColumn.prototype, "template", void 0);
__decorate([
    Output()
], FilterColumn.prototype, "outputEvent", void 0);
FilterColumn = __decorate([
    Component({
        selector: 'basis-filter-column',
        template: "<div [ngSwitch]=\"filterType\">\n    <div *ngSwitchCase=\"'text'\">\n        <div class=\"ui-g form-group\">\n            <div class=\"ui-md-12\">\n                <span class=\"md-inputfield\">\n                    <input type=\"text\" pInputText (input)=\"filtrar($event)\" [(ngModel)]=\"filtro\">\n                    <label>{{ (header) ? header : 'Filtro' }}</label>\n                </span>\n            </div>\n        </div>\n    </div>\n    <div *ngSwitchCase=\"'select'\">\n        <p-multiSelect\n            [options]=\"getCustomFilterOptions()\"\n            [defaultLabel]=\"(header)? header: '&nbsp;'\"\n            [optionLabel]=\"'label'\"\n            appendTo=\"body\"\n            [ngModelOptions]=\"{standalone: true}\"\n            (onChange)=\"filtrar($event)\"\n            [(ngModel)]=\"filtro\"\n            [style]=\"{'width':'100%'}\">\n        </p-multiSelect>\n    </div>\n</div>\n"
    })
], FilterColumn);

let FilterColumnModule = class FilterColumnModule {
};
FilterColumnModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            MultiSelectModule,
            ButtonModule,
            CheckboxModule,
            DropdownModule,
            FormsModule,
            ReactiveFormsModule,
            InputTextModule
        ],
        declarations: [
            FilterColumn
        ],
        exports: [
            FilterColumn
        ],
        schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
], FilterColumnModule);

/**
 * Class PageNotificationService
 * @class
 */
let PageNotificationService = class PageNotificationService extends NotificationProvider {
    /**
     * constructor method
     * @constructor
     * @param {MessageService} messageService
     */
    constructor(messageService) {
        super();
        this.messageService = messageService;
        /**
         * successSeverity property
         * @type {string}
         */
        this.successSeverity = 'success';
        /**
         * infoSeverity property
         * @type {string}
         */
        this.infoSeverity = 'info';
        /**
         * warnSeverity property
         * @type {string}
         */
        this.warnSeverity = 'warn';
        /**
         * errorSeverity property
         * @type {string}
         */
        this.errorSeverity = 'error';
        /**
         * createMsg property
         * @type {string}
         */
        this.createMsg = 'Registro incluído com sucesso!';
        /**
         * updateMsg property
         * @type {string}
         */
        this.updateMsg = 'Registro alterado com sucesso!';
        /**
         * deleteMsg property
         * @type {string}
         */
        this.deleteMsg = 'Registro excluído com sucesso!';
    }
    /**
     * addCreateMsg method
     * @param {string} title
     */
    addCreateMsg(title) {
        this.addSuccessMessage(this.createMsg, title);
    }
    /**
     * addSuccessMessage method
     * @param {string} message
     * @param {string} title
     */
    addSuccessMessage(message, title) {
        this.addMsg(this.successSeverity, message, title);
    }
    /**
     * addMsg method
     * @param {string} severity
     * @param {string} msg
     * @param {string} title
     */
    addMsg(severity, msg, title) {
        this.messageService.add({
            severity: severity,
            summary: title,
            detail: msg
        });
    }
    /**
     * addUpdateMsg method
     * @param {string} title
     */
    addUpdateMsg(title) {
        this.addSuccessMessage(this.updateMsg, title);
    }
    /**
     * addDeleteMsg method
     * @param {string} title
     */
    addDeleteMsg(title) {
        this.addSuccessMessage(this.deleteMsg, title);
    }
    /**
     * addInfoMessage method
     * @param {string} message
     * @param {string} title
     */
    addInfoMessage(message, title) {
        this.addMsg(this.infoSeverity, message, title);
    }
    /**
     * addWarnMessage method
     * @param {string} message
     * @param {string} title
     */
    addWarnMessage(message, title) {
        this.addMsg(this.warnSeverity, message, title);
    }
    /**
     * addErrorMessage method
     * @param {string} message
     * @param {string} title
     */
    addErrorMessage(message, title) {
        this.addMsg(this.errorSeverity, message, title);
    }
    /**
     * addErrorProblem method
     * @param {ApplicationProblemType} problem
     */
    addErrorProblem(problem) {
        this.addMsg(this.errorSeverity, problem.detail, problem.title);
    }
};
PageNotificationService.ctorParameters = () => [
    { type: MessageService }
];
PageNotificationService.ɵprov = ɵɵdefineInjectable({ factory: function PageNotificationService_Factory() { return new PageNotificationService(ɵɵinject(MessageService$1)); }, token: PageNotificationService, providedIn: "root" });
PageNotificationService = __decorate([
    Injectable({
        providedIn: 'root'
    })
], PageNotificationService);

let ShowColumn = class ShowColumn {
    constructor(pageNotificationService) {
        this.pageNotificationService = pageNotificationService;
        /**
         * lista de colunas disponívels na tabela
         * @type {Array}
         */
        this.options = [];
        /**
         * Valores do select de colunas
         * @type {Array}
         */
        this.values = [];
        /**
         * Último valor válido para o select.
         * @type {Array}
         */
        this.lastValue = [];
        /**
         * Evento para atualizar a visibilidade das colunas da tabela
         * @type {EventEmitter}
         */
        this.updateVisibleColumns = new EventEmitter();
    }
    ngOnInit() {
        this.columns.forEach((column) => {
            const col = { field: column.field, filterType: column.filterType, header: column.header };
            this.options.push(col);
            this.values.push(col);
            this.lastValue.push(col);
        });
    }
    mostrarColunas(event) {
        if (this.values.length) {
            this.lastValue = event.value;
            this.updateVisibleColumns.emit(this.values);
        }
        else {
            this.lastValue.map((item) => this.values.push(item));
            this.pageNotificationService.addErrorMessage('Não é possível exibir menos de uma coluna');
        }
    }
};
ShowColumn.ctorParameters = () => [
    { type: PageNotificationService }
];
__decorate([
    Input()
], ShowColumn.prototype, "columns", void 0);
__decorate([
    Output()
], ShowColumn.prototype, "updateVisibleColumns", void 0);
ShowColumn = __decorate([
    Component({
        selector: 'basis-show-column',
        template: "<div class=\"ui-g-12 ui-g-nopad\">\n    <div class=\"ui-g-3 ui-md-offset-6\">\n        <div class=\"text-right\">Visibilidade das Colunas:</div>\n    </div>\n    <div class=\"ui-g-3\">\n        <p-multiSelect\n            class=\"column\"\n            defaultLabel=\"Selecione as colunas a serem exibidas\"\n            [optionLabel]=\"'header'\"\n            [displaySelectedLabel]=\"false\"\n            [ngModelOptions]=\"{ standalone: true }\"\n            (onChange)=\"mostrarColunas($event)\"\n            [options]=\"options\"\n            [(ngModel)]=\"values\"\n            [style]=\"{ width: '100%' }\"\n        >\n        </p-multiSelect>\n    </div>\n</div>\n"
    })
], ShowColumn);

let ShowColumnModule = class ShowColumnModule {
};
ShowColumnModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            MultiSelectModule,
            ButtonModule,
            CheckboxModule,
            DropdownModule,
            FormsModule,
            ReactiveFormsModule,
            InputTextModule
        ],
        declarations: [
            ShowColumn
        ],
        exports: [
            ShowColumn
        ],
        schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
], ShowColumnModule);

/**
 * Class DatatableModule
 * @class
 */
let DatatableModule = class DatatableModule {
};
DatatableModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            ButtonModule$1,
            SplitButtonModule,
            TooltipModule,
            TableModule,
            BlockUIModule,
            ProgressSpinnerModule,
            DatatableSplitbuttonModule,
            HttpClientModule,
            PaginatorModule,
            FilterColumnModule,
            ShowColumnModule,
            InputTextModule
        ],
        declarations: [
            DatatableComponent,
            DatatableButtonComponent,
            Column
        ],
        providers: [
            { provide: DatatableService, useClass: DatatableService, deps: [HttpClient] },
            DataProviderFactory
        ],
        exports: [
            DatatableComponent,
            DatatableButtonComponent,
            FilterColumnModule,
            ShowColumnModule,
            Column
        ]
    })
], DatatableModule);

/**
 * Class BlockUiComponent
 * @class
 */
let BlockUiComponent = class BlockUiComponent {
    /**
     * constructor method
     * @constructor
     * @param {BlockUiService} service
     */
    constructor(service) {
        this.service = service;
        /**
         * status property
         * @type {boolean}
         * @public
         */
        this.status = false;
    }
    /**
     * ngOnInit method
     * @returns void
     */
    ngOnInit() {
        this.subscription = this.service.loaderStatus.subscribe((blockui) => {
            this.status = blockui.status;
        });
    }
    /**
     * ngOnDestroy methods
     * @returns void
     */
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    /**
     * isBlocked method
     * @returns boolean
     */
    isBlocked() {
        return this.status;
    }
    /**
     * show method
     * @returns false
     */
    show() {
        this.status = true;
    }
    /**
     * hide methods
     * @returns void
     */
    hide() {
        this.status = false;
    }
};
BlockUiComponent.ctorParameters = () => [
    { type: BlockUiService }
];
BlockUiComponent = __decorate([
    Component({
        selector: 'basis-block-ui',
        template: "<p-blockUI [blocked]=\"isBlocked()\">\n    <div class=\"center-screen\">\n        <p-progressSpinner></p-progressSpinner>\n    </div>\n</p-blockUI>\n",
        styles: [".center-screen{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}"]
    })
], BlockUiComponent);

/**
 * Class BlockUiModule
 * @class
 */
let BlockUiModule = class BlockUiModule {
};
BlockUiModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            BlockUIModule,
            ProgressSpinnerModule,
        ],
        declarations: [
            BlockUiComponent
        ],
        providers: [
            {
                provide: HTTP_INTERCEPTORS,
                useClass: BlockUiInterceptor,
                multi: true
            },
            BlockUiService
        ],
        exports: [
            BlockUIModule,
            BlockUiComponent,
            ProgressSpinnerModule
        ]
    })
], BlockUiModule);

/**
 * Class PageNotificationComponent
 * @class
 */
let PageNotificationComponent = class PageNotificationComponent {
    constructor() {
        /**
         * duration property
         * @type {number}
         */
        this.duration = 3000;
    }
};
__decorate([
    Input()
], PageNotificationComponent.prototype, "duration", void 0);
PageNotificationComponent = __decorate([
    Component({
        selector: 'basis-page-notification',
        template: "<p-toast></p-toast>\n"
    })
], PageNotificationComponent);

/**
 * Class PageNotificationModule
 * @class
 */
let PageNotificationModule = class PageNotificationModule {
};
PageNotificationModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            ToastModule
        ],
        declarations: [
            PageNotificationComponent
        ],
        exports: [
            PageNotificationComponent
        ],
        providers: [
            MessageService$1,
            PageNotificationService,
            {
                provide: NotificationProvider,
                useExisting: PageNotificationService
            }
        ]
    })
], PageNotificationModule);

let FormNotificationComponent = class FormNotificationComponent extends FormNotificationComponentSuperclass {
};
FormNotificationComponent = __decorate([
    Component({
        selector: '[form-notification]',
        template: "<ng-content></ng-content>\n<div class=\"form-notification-msg ui-message ui-messages-error ui-corner-all\" >{{message}}</div>\n"
    })
], FormNotificationComponent);

/**
 * Módulo para adicionar funcionalidade de notificação de erros de formulários.
 * @module
 */
let FormNotificationModule = class FormNotificationModule {
};
FormNotificationModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            FormNotificationBaseModule
        ],
        declarations: [
            FormNotificationComponent
        ],
        exports: [
            FormNotificationComponent
        ]
    })
], FormNotificationModule);

/**
 * Class ErrorStackComponent
 * @class
 */
let ErrorStackComponent = class ErrorStackComponent {
    /**
     * constructor method
     * @constructor
     * @param {ErrorStackService} errorStack
     */
    constructor(errorStack) {
        this.errorStack = errorStack;
        /**
         * title property
         * @type {string}
         */
        this.title = 'Error Stack';
        /**
         * errors property
         * @type {NamedErrorType}
         */
        this.errors = [];
    }
    /**
     * ngOnInit method
     * @returns void
     */
    ngOnInit() {
        this.errorSubscription = this.errorStack.errors.subscribe((error) => {
            this.errors = [error, ...this.errors];
        });
    }
    /**
     * ngOnDestroy method
     * @returns void
     */
    ngOnDestroy() {
        this.errorSubscription.unsubscribe();
    }
};
ErrorStackComponent.ctorParameters = () => [
    { type: ErrorStackService }
];
__decorate([
    Input()
], ErrorStackComponent.prototype, "title", void 0);
__decorate([
    Input()
], ErrorStackComponent.prototype, "template", void 0);
ErrorStackComponent = __decorate([
    Component({
        selector: 'basis-error-stack',
        template: "<ng-template #defaultTemplate>\n    <div class=\"ui-g\">\n        <div class=\"ui-g-12\">\n            <div class=\"card no-margin\">\n                <h1>{{ title }}</h1>\n                <p-table [value]=\"errors\" [paginator]=\"false\" [rows]=\"5\" first=\"0\">\n                    <ng-template pTemplate=\"header\">\n                        <tr>\n                            <th style=\"width: 20%; text-align: center\">Data</th>\n                            <th style=\"width: 20%; text-align: center\">C\u00F3digo de identifica\u00E7\u00E3o</th>\n                            <th>Mensagem</th>\n                            <th style=\"width: 10%; text-align: center\">A\u00E7\u00E3o</th>\n                        </tr>\n                    </ng-template>\n                    <ng-template pTemplate=\"body\" let-error>\n                        <tr>\n                            <td style=\"width: 20%; text-align: center\">{{error.createdAt|date: 'dd/MM/y H:mm:ss'}}</td>\n                            <td style=\"width: 20%; text-align: center\">{{error.correlationId || 'Indispon\u00EDvel'}}</td>\n                            <td>{{error.message}}</td>\n                            <td style=\"width: 10%; text-align: center\">\n                                <button pButton type=\"button\" pTooltip=\"Copiar erro para \u00E1rea de transfer\u00EAncia\" tooltipPosition=\"left\" class=\"ui-button-primary\" icon=\"ui-icon-content-copy\" [clipboard]=\"error.fullMessage\"></button>\n                            </td>\n                        </tr>\n                    </ng-template>\n                    <ng-template pTemplate=\"emptymessage\" let-columns>\n                        <tr>\n                            <td colspan=\"4\">\n                                Nenhum erro foi registrado at\u00E9 o momento.\n                            </td>\n                        </tr>\n                    </ng-template>\n                </p-table>\n            </div>\n        </div>\n    </div>\n</ng-template>\n<ng-container *ngTemplateOutlet=\"template? template: defaultTemplate\"></ng-container>\n"
    })
], ErrorStackComponent);

/**
 * Class ErrorStackModule
 * @class
 */
let ErrorStackModule = class ErrorStackModule {
};
ErrorStackModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            TableModule,
            ButtonModule$1,
            TooltipModule,
            ClipboardModule
        ],
        declarations: [
            ErrorStackComponent
        ],
        providers: [
            ErrorStackService,
            {
                provide: ErrorProvider,
                useClass: ErrorStackProviderService,
                multi: true
            }
        ],
        exports: [
            ErrorStackComponent
        ]
    })
], ErrorStackModule);

let CrudFormComponent = class CrudFormComponent {
    constructor(route) {
        this.route = route;
        this.disableSave = false;
        this.listRoute = "../../list";
    }
    ngOnInit() {
        this.route.paramMap.subscribe((params) => {
            this.action = params.get('action');
            if (params.get('id')) {
                this.route.data.subscribe((data) => {
                    if (data.value) {
                        this.form.patchValue(data.value);
                    }
                });
            }
            else {
                this.listRoute = "../list";
            }
            if (this.action === "view") {
                this.form.disable();
            }
            else if (this.form.disabled) {
                this.form.enable();
            }
        });
    }
};
CrudFormComponent.ctorParameters = () => [
    { type: ActivatedRoute }
];
__decorate([
    Input()
], CrudFormComponent.prototype, "form", void 0);
__decorate([
    Input()
], CrudFormComponent.prototype, "service", void 0);
__decorate([
    Input()
], CrudFormComponent.prototype, "disableSave", void 0);
CrudFormComponent = __decorate([
    Component({
        selector: 'nCrudForm',
        template: "<nForm [form]=\"form\" [action]=\"action\" [route]=\"[listRoute]\" [disableSave]=\"disableSave\">\n    <ng-content></ng-content>\n    <div nButton class=\"ui-g-12 ui-lg-2 ui-md-4 ui-sm-12\">\n        <button pButton type=\"button\" [routerLink]=\"[listRoute]\" label=\"Voltar\"></button>\n    </div>\n</nForm>\n"
    })
], CrudFormComponent);

const CRUD_SERVICE = new InjectionToken("CRUD_SERVICE");

let CrudListComponent = class CrudListComponent {
    constructor(crudService, confirmationService, route, router) {
        this.crudService = crudService;
        this.confirmationService = confirmationService;
        this.route = route;
        this.router = router;
        this.action = "findAll";
        this.findLabel = "Pesquisar";
        this.disableSave = false;
        this.disableEdit = false;
        this.disableView = false;
        this.disableDelete = false;
        this.find = new EventEmitter();
    }
    ngOnInit() {
        if (this.service) {
            this.crudService = this.service;
        }
        if (!this.url) {
            this.url = this.crudService.uri;
        }
    }
    datatableClick(event) {
        if (event.selection) {
            switch (event.button) {
                case "edit":
                case "view":
                    this.router.navigate([`../${event.button}`, event.selection.id], { relativeTo: this.route });
                    break;
                case "delete":
                    this.confirmationService.confirm({
                        message: 'Você tem certeza que deseja excluir o registro?',
                        accept: () => this.crudService.delete(event.selection.id)
                            .subscribe(() => this.datatable.filter())
                    });
                    break;
            }
        }
    }
    findAll(event) {
        if (this.find.observers.length) {
            this.find.emit(this.form.value);
        }
        else {
            this.datatable.filterParams = {};
            Object.keys(this.form.value).forEach(element => {
                let value = this.form.value[element];
                if (value) {
                    this.datatable.filterParams[element] =
                        value instanceof Date ? JSON.stringify(value).replace(/"/g, "") : value;
                }
            });
            this.datatable.filter();
        }
    }
    ngAfterContentInit() {
        const datatableNgAfterContentInit = this.datatable.pDatatableComponent.ngAfterContentInit;
        this.datatable.pDatatableComponent.ngAfterContentInit = () => {
            this.datatable.pDatatableComponent.columns = this.columns.toArray();
            datatableNgAfterContentInit.call(this.datatable.pDatatableComponent);
        };
    }
    ngAfterViewInit() {
        for (let i in this.datatableExtraParams) {
            this.datatable[i] = this.datatableExtraParams[i];
        }
    }
};
CrudListComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CRUD_SERVICE,] }] },
    { type: ConfirmationService },
    { type: ActivatedRoute },
    { type: Router }
];
__decorate([
    Input()
], CrudListComponent.prototype, "form", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "action", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "findLabel", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "service", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "url", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "disableSave", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "disableEdit", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "disableView", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "disableDelete", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "datatableExtraParams", void 0);
__decorate([
    Output()
], CrudListComponent.prototype, "find", void 0);
__decorate([
    ViewChild('datatable', { static: true })
], CrudListComponent.prototype, "datatable", void 0);
__decorate([
    ContentChildren(Column)
], CrudListComponent.prototype, "columns", void 0);
CrudListComponent = __decorate([
    Component({
        selector: 'nCrudList',
        template: "<nForm [form]=\"form\" (save)=\"findAll($event)\" [saveLabel]=\"findLabel\">\n    <ng-content select=\"[nForm]\"></ng-content>\n    <div nButton class=\"ui-g-12 ui-lg-2 ui-md-4 ui-sm-12\">\n        <button pButton type=\"button\" [routerLink]=\"['../save']\" label=\"Novo\" *ngIf=\"!disableSave\"></button>\n    </div>\n</nForm>\n\n<basis-datatable #datatable [value]=\"result\" (buttonClick)=\"datatableClick($event)\" [url]=\"url\"\n    [disableEdit]=\"disableEdit\" [disableView]=\"disableView\" [disableDelete]=\"disableDelete\">\n    <ng-content select=\"p-column\"></ng-content>\n</basis-datatable>\n\n<p-confirmDialog header=\"Confirmar Exclus\u00E3o\" acceptLabel=\"Sim\" rejectLabel=\"N\u00E3o\" icon=\"pi pi-exclamation-triangle\"></p-confirmDialog>\n"
    }),
    __param(0, Inject(CRUD_SERVICE))
], CrudListComponent);

let CrudResolveGuard = class CrudResolveGuard {
    constructor(crudService) {
        this.crudService = crudService;
    }
    resolve(route) {
        return this.crudService.find(route.paramMap.get('id'));
    }
};
CrudResolveGuard.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CRUD_SERVICE,] }] }
];
CrudResolveGuard = __decorate([
    Injectable(),
    __param(0, Inject(CRUD_SERVICE))
], CrudResolveGuard);

function crudRouteBuilder(route, parent, list, form) {
    return {
        path: route,
        component: parent,
        data: { breadcrumb: route },
        children: [
            {
                path: 'list',
                component: list,
                data: { breadcrumb: 'list' },
            },
            {
                path: ':action',
                component: form,
                data: { breadcrumb: ':action' },
            },
            {
                path: ':action/:id',
                component: form,
                resolve: {
                    value: CrudResolveGuard
                },
                data: { breadcrumb: ':action' },
            }
        ]
    };
}

class CrudServiceNuvem {
    constructor(uri, http) {
        this.uri = uri;
        this.http = http;
    }
    save(entity) {
        return this.http.post(this.uri, entity);
    }
    edit(entity) {
        return this.http.put(this.uri, entity);
    }
    find(id) {
        return this.http.get(`${this.uri}/${id}`);
    }
    delete(id) {
        return this.http.delete(`${this.uri}/${id}`);
    }
    findAll(entity) {
        let params = new HttpParams();
        for (let field in entity) {
            params = params.append(field, entity[field]);
        }
        return this.http.get(`${this.uri}`, { params: params });
    }
}

var STATUS;
(function (STATUS) {
    STATUS[STATUS["OK"] = 200] = "OK";
    STATUS[STATUS["CREATED"] = 201] = "CREATED";
})(STATUS || (STATUS = {}));
let SuccessMessageHandlerInterceptor = class SuccessMessageHandlerInterceptor {
    constructor(notification) {
        this.notification = notification;
    }
    intercept(req, next) {
        return next.handle(req).pipe(tap((event) => {
            if (event instanceof HttpResponse) {
                if (event.headers.has("X-nuvem-alert")) {
                    this.notification.addSuccessMessage(event.headers.get("X-nuvem-alert"));
                }
                else if ((event.status === STATUS.OK || event.status === STATUS.CREATED) && req.method != "GET") {
                    this.notification.addSuccessMessage("Operação realizada com sucesso!");
                }
            }
        }));
    }
};
SuccessMessageHandlerInterceptor.ctorParameters = () => [
    { type: PageNotificationService }
];
SuccessMessageHandlerInterceptor.ɵprov = ɵɵdefineInjectable({ factory: function SuccessMessageHandlerInterceptor_Factory() { return new SuccessMessageHandlerInterceptor(ɵɵinject(PageNotificationService)); }, token: SuccessMessageHandlerInterceptor, providedIn: "root" });
SuccessMessageHandlerInterceptor = __decorate([
    Injectable({ providedIn: 'root' })
], SuccessMessageHandlerInterceptor);
const httpInterceptorProviders = [
    { provide: HTTP_INTERCEPTORS, useClass: SuccessMessageHandlerInterceptor, multi: true },
];

let FieldComponent = class FieldComponent extends WrappedFormControlSuperclass {
    constructor(controlContainer, injector) {
        super(injector);
        this.controlContainer = controlContainer;
        this.injector = injector;
        this.onBlur = new EventEmitter();
        this.onSelect = new EventEmitter();
        this.onFocus = new EventEmitter();
    }
    get name() {
        return this._name ? this._name : this.formControlName;
    }
    set name(value) {
        this._name = value;
    }
    get label() {
        return this.required ? `${this._label}*` : this._label;
    }
    set label(value) {
        this._label = value;
    }
    ngOnInit() {
        this.setValidators();
        this.setRequired();
    }
    setRequired() {
        if (this.required) {
            if (this.input instanceof ElementRef) {
                this.input.nativeElement.required = true;
            }
            else {
                this.input.required = true;
            }
        }
    }
    setElementRefAttr(attr, value) {
        if (value && this.input instanceof ElementRef) {
            this.input.nativeElement[attr] = value;
        }
    }
    setFormControlValidators() {
        if (this.controlContainer && this.formControlName) {
            let validator = this.controlContainer.control.get(this.formControlName).validator;
            if (validator) {
                this.formControl.setValidators(validator);
            }
        }
    }
    setValidators() {
        if (this.controlContainer && this.formControlName) {
            let validator = this.controlContainer.control.get(this.formControlName).validator;
            if (validator) {
                this.formControl.setValidators(validator);
            }
        }
    }
    ngAfterViewInit() {
        for (let i in this.extraParams) {
            this.input[i] = this.extraParams[i];
        }
    }
    onBlurEvent(event) {
        this.onBlur.emit(event);
    }
    onSelectEvent(event) {
        this.onSelect.emit(event);
    }
    onFocusEvent(event) {
        this.onFocus.emit(event);
    }
};
FieldComponent.ctorParameters = () => [
    { type: ControlContainer },
    { type: Injector }
];
__decorate([
    Input()
], FieldComponent.prototype, "form", void 0);
__decorate([
    Input()
], FieldComponent.prototype, "formControlName", void 0);
__decorate([
    Input()
], FieldComponent.prototype, "extraParams", void 0);
__decorate([
    Input()
], FieldComponent.prototype, "required", void 0);
__decorate([
    Output()
], FieldComponent.prototype, "onBlur", void 0);
__decorate([
    Output()
], FieldComponent.prototype, "onSelect", void 0);
__decorate([
    Output()
], FieldComponent.prototype, "onFocus", void 0);
__decorate([
    ViewChild("input", { static: false })
], FieldComponent.prototype, "input", void 0);
__decorate([
    Input()
], FieldComponent.prototype, "name", null);
__decorate([
    Input()
], FieldComponent.prototype, "label", null);
FieldComponent = __decorate([
    Directive({
        selector: "[nField]"
    })
], FieldComponent);

var InputTextComponent_1;
let InputTextComponent = InputTextComponent_1 = class InputTextComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.type = "text";
    }
    ngOnInit() {
        super.ngOnInit();
        this.setElementRefAttr("min", this.min);
        this.setElementRefAttr("max", this.max);
        this.setElementRefAttr("minLength", this.minLength);
        this.setElementRefAttr("maxLength", this.maxLength);
        this.setElementRefAttr("pattern", this.pattern);
    }
};
__decorate([
    Input()
], InputTextComponent.prototype, "minLength", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "maxLength", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "min", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "max", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "pattern", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "type", void 0);
InputTextComponent = InputTextComponent_1 = __decorate([
    Component({
        selector: "[nInputText]",
        template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <input #input\n        pInputText\n        [ngClass]=\"{ 'ng-invalid ng-dirty': notification.message }\"\n        [type]=\"type\"\n        [name]=\"name\"\n        (blur)=\"onBlurEvent($event)\"\n        (focus)=\"onFocusEvent($event)\"\n        [formControl]=\"formControl\" />\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
        providers: [provideValueAccessor(InputTextComponent_1)]
    })
], InputTextComponent);

const CALENDAR_LOCALE = {
    firstDayOfWeek: 0,
    dayNames: ["Domingo", "Segunda", "Terça", "Quarta", "Quinta", "Sexta", "Sábado"],
    dayNamesShort: ["Dom", "Seg", "Ter", "Qua", "Qui", "Sex", "Sab"],
    dayNamesMin: ["D", "S", "T", "Q", "Q", "S", "S"],
    monthNames: ["Janeiro", "Fevereiro", "Março", "Abril", "Maio", "Junho", "Julho", "Agosto", "Setembro", "Outubro", "Novembro", "Dezembro"],
    monthNamesShort: ["Jan", "Fev", "Mar", "Abr", "Mai", "Jun", "Jul", "Ago", "Set", "Out", "Nov", "Dez"],
    today: 'Hoje',
    clear: 'Limpar',
    dateFormat: 'dd/mm/yy',
    weekHeader: 'Semana'
};

var CalendarComponent_1;
let CalendarComponent = CalendarComponent_1 = class CalendarComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.maxDate = null;
        this.yearRange = null;
        this.monthNavigator = false;
        this.yearNavigator = false;
        this.locale = CALENDAR_LOCALE;
    }
    outerToInner(value) {
        return value ? new Date(value) : null;
    }
};
__decorate([
    Input()
], CalendarComponent.prototype, "maxDate", void 0);
__decorate([
    Input()
], CalendarComponent.prototype, "yearRange", void 0);
__decorate([
    Input()
], CalendarComponent.prototype, "monthNavigator", void 0);
__decorate([
    Input()
], CalendarComponent.prototype, "yearNavigator", void 0);
__decorate([
    Input()
], CalendarComponent.prototype, "locale", void 0);
CalendarComponent = CalendarComponent_1 = __decorate([
    Component({
        selector: "[nCalendar]",
        template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-calendar #input\n        [ngClass]=\"{ 'ng-invalid ng-dirty': notification.message }\"\n        [name]=\"name\"\n        [formControl]=\"formControl\"\n        [locale]=\"locale\"\n        [dateFormat]=\"locale.dateFormat\"\n        [yearRange]=\"yearRange\"\n        [maxDate]=\"maxDate\"\n        [monthNavigator]=\"monthNavigator\"\n        [yearNavigator]=\"yearNavigator\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        (onSelect)=\"onSelectEvent($event)\">\n    </p-calendar>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
        providers: [provideValueAccessor(CalendarComponent_1)]
    })
], CalendarComponent);

var TextAreaComponent_1;
let TextAreaComponent = TextAreaComponent_1 = class TextAreaComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.autoResize = false;
    }
};
__decorate([
    Input()
], TextAreaComponent.prototype, "autoResize", void 0);
__decorate([
    Input()
], TextAreaComponent.prototype, "rows", void 0);
__decorate([
    Input()
], TextAreaComponent.prototype, "cols", void 0);
__decorate([
    Input()
], TextAreaComponent.prototype, "minLength", void 0);
__decorate([
    Input()
], TextAreaComponent.prototype, "maxLength", void 0);
TextAreaComponent = TextAreaComponent_1 = __decorate([
    Component({
        selector: "[nTextArea]",
        template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <textarea #input\n        pInputTextarea\n        [ngClass]=\"{ 'ng-invalid ng-dirty': notification.message }\"\n        type=\"text\"\n        [name]=\"name\"\n        [autoResize]=\"autoResize\"\n        [rows]=\"rows\"\n        [cols]=\"cols\"\n        [minLength]=\"minLength\"\n        [maxLength]=\"maxLength\"\n        [formControl]=\"formControl\"\n        (blur)=\"onBlurEvent($event)\"\n        (focus)=\"onFocusEvent($event)\">\n    </textarea>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
        providers: [provideValueAccessor(TextAreaComponent_1)]
    })
], TextAreaComponent);

var DropdownComponent_1;
let DropdownComponent = DropdownComponent_1 = class DropdownComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.virtualScroll = false;
        this.filter = false;
        this.itemSize = 30;
        this.options = [];
        this.onChange = new EventEmitter();
    }
    ngOnInit() {
        super.ngOnInit();
        if (this.service) {
            this.service.findAll(null).subscribe(data => {
                this.options = data;
            });
        }
    }
    onChangeEvent(event) {
        this.onChange.emit(event);
    }
    outerToInner(value) {
        if (value && value.id) {
            this.service.find(value.id).subscribe((data) => {
                this.formControl.setValue(data);
            });
        }
        return value;
    }
};
__decorate([
    Input()
], DropdownComponent.prototype, "optionLabel", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "virtualScroll", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "filter", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "itemSize", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "service", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "options", void 0);
__decorate([
    Output()
], DropdownComponent.prototype, "onChange", void 0);
DropdownComponent = DropdownComponent_1 = __decorate([
    Component({
        selector: "[nDropdown]",
        template: "<span #notification form-notification class=\"md-inputfield custom-select\" [name]=\"name\" [form]=\"form\">\n    <p-dropdown #input\n        [formControl]=\"formControl\"\n        [optionLabel]=\"optionLabel\"\n        [options]=\"options\"\n        [filter]=\"filter\"\n        [virtualScroll]=\"virtualScroll\"\n        [itemSize]=\"itemSize\"\n        (onChange)=\"onChangeEvent($event)\">\n    </p-dropdown>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
        providers: [provideValueAccessor(DropdownComponent_1)]
    })
], DropdownComponent);

var InputMaskComponent_1;
let InputMaskComponent = InputMaskComponent_1 = class InputMaskComponent extends FieldComponent {
};
__decorate([
    Input()
], InputMaskComponent.prototype, "mask", void 0);
__decorate([
    Input()
], InputMaskComponent.prototype, "placeholder", void 0);
InputMaskComponent = InputMaskComponent_1 = __decorate([
    Component({
        selector: "[nInputMask]",
        template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-inputMask #input\n        pInputText\n        [ngClass]=\"{ 'ng-invalid ng-dirty': notification.message }\"\n        [name]=\"name\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        [formControl]=\"formControl\"\n        [mask]=\"mask\"\n        [placeholder]=\"placeholder\"></p-inputMask>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
        providers: [provideValueAccessor(InputMaskComponent_1)]
    })
], InputMaskComponent);

var CheckComponent_1;
let CheckComponent = CheckComponent_1 = class CheckComponent extends FieldComponent {
    /**
     * Método construtor responsável por carregar o serviço de validação
     * @param formNotificationService
     */
    constructor(controlContainer, injector, formNotificationService) {
        super(controlContainer, injector);
        this.formNotificationService = formNotificationService;
    }
    /**
     * Método Inicial para subscrição no endpoint responsável por validar o formulário
     */
    ngOnInit() {
        super.ngOnInit();
        this.subscription = this.formNotificationService
            .notifications(this.form, this.name)
            .subscribe((notification) => {
            if (notification instanceof FieldFormNotification) {
                this.message = notification.message;
            }
            else {
                this.message = null;
            }
        });
    }
    /**
     * Método responsável por realizar a saida das subscrições do endpont de validação
     */
    ngOnDestroy() {
        super.ngOnDestroy();
        this.subscription.unsubscribe();
    }
};
CheckComponent.ctorParameters = () => [
    { type: ControlContainer },
    { type: Injector },
    { type: FormNotificationService }
];
CheckComponent = CheckComponent_1 = __decorate([
    Component({
        selector: "[nCheck]",
        template: "<div class=\"ui-g-12 semborda\">\n    <div class=\"ui-g-4\" [innerHTML]=\"label\"></div>\n    <div class=\"ui-g-4\">\n        <p-radioButton\n            [ngClass]=\"{ 'ng-invalid ng-dirty': message }\"\n            id=\"sim\"\n            type=\"radio\"\n            [value]=\"true\"\n            [name]=\"name\"\n            [formControl]=\"formControl\"\n            pRadioButton></p-radioButton>\n        <label>Sim</label>\n    </div>\n    <div class=\"ui-g-4\">\n        <p-radioButton\n            [ngClass]=\"{ 'ng-invalid ng-dirty': message }\"\n            id=\"n\u00E3o\"\n            type=\"radio\"\n            [value]=\"false\"\n            name=\"correspondencia\"\n            [formControl]=\"formControl\"\n            pRadioButton></p-radioButton>\n        <label>N\u00E3o</label>\n    </div>\n    <ng-content></ng-content>\n</div>\n",
        providers: [provideValueAccessor(CheckComponent_1)]
    })
], CheckComponent);

let FormComponent = class FormComponent {
    constructor(crudService, activatedRoute, router) {
        this.crudService = crudService;
        this.activatedRoute = activatedRoute;
        this.router = router;
        this.action = "save";
        this.saveLabel = "Salvar";
        this.disableSave = false;
        this.save = new EventEmitter();
        this.reset = new EventEmitter();
    }
    ngOnInit() {
        if (this.service) {
            this.crudService = this.service;
        }
    }
    submit() {
        if (this.save.observers.length) {
            this.save.emit(this.form.value);
        }
        else {
            this.crudService[this.action](this.form.value)
                .subscribe(() => {
                if (this.route) {
                    this.router.navigate(this.route, { relativeTo: this.activatedRoute });
                }
                else if (this.reset.observers.length) {
                    this.reset.emit(this.form);
                }
                else {
                    this.form.reset();
                }
            });
        }
    }
};
FormComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CRUD_SERVICE,] }] },
    { type: ActivatedRoute },
    { type: Router }
];
__decorate([
    Input()
], FormComponent.prototype, "form", void 0);
__decorate([
    Input()
], FormComponent.prototype, "route", void 0);
__decorate([
    Input()
], FormComponent.prototype, "action", void 0);
__decorate([
    Input()
], FormComponent.prototype, "saveLabel", void 0);
__decorate([
    Input()
], FormComponent.prototype, "service", void 0);
__decorate([
    Input()
], FormComponent.prototype, "disableSave", void 0);
__decorate([
    Output()
], FormComponent.prototype, "save", void 0);
__decorate([
    Output()
], FormComponent.prototype, "reset", void 0);
FormComponent = __decorate([
    Component({
        selector: 'nForm',
        template: "<div>\n    <form (submit)=\"submit()\" class=\"ui-g form-group\">\n        <ng-content></ng-content>\n        <div class=\"ui-g-12 ui-lg-2 ui-md-4 ui-sm-12\" *ngIf=\"!disableSave\">\n            <button pButton type=\"submit\" [label]=\"saveLabel\" [disabled]=\"form.disabled || form.invalid\"></button>\n        </div>\n        <ng-content select=\"[nButton]\"></ng-content>\n    </form>\n</div>"
    }),
    __param(0, Inject(CRUD_SERVICE))
], FormComponent);

let ValidationMessageComponent = class ValidationMessageComponent {
};
__decorate([
    Input()
], ValidationMessageComponent.prototype, "control", void 0);
ValidationMessageComponent = __decorate([
    Component({
        selector: 'nValidationMessage',
        template: "<ng-container *ngIf=\"control.invalid && (control.dirty || control.touched) && control.errors\">\n    <div class=\"form-notification-msg ui-message ui-messages-error ui-corner-all\" >\n        <ng-container *ngIf=\"control.errors.required\">\n            campo obrigat\u00F3rio\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.minlength\">\n            tamanho m\u00EDnimo {{ control.errors.minlength.requiredLength }}\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.maxlength\">\n            tamanho m\u00E1ximo {{ control.errors.maxlength.requiredLength }}\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.min\">\n            valor m\u00EDnimo {{ control.errors.min.min }}\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.max\">\n            valor m\u00E1ximo {{ control.errors.max.max }}\n        </ng-container>\n        <ng-container *ngIf=\"control.errors.email || control.errors.pattern\">\n            campo inv\u00E1lido\n        </ng-container>\n    </div>\n</ng-container>\n"
    })
], ValidationMessageComponent);

var AutoCompleteComponent_1;
let AutoCompleteComponent = AutoCompleteComponent_1 = class AutoCompleteComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.virtualScroll = false;
        this.itemSize = 30;
    }
    selectItem(option) {
        this.formControl.setValue(option);
    }
    findAll(event) {
        let query = {};
        query[this.field] = event.query;
        this.service.findAll(query).subscribe(data => {
            this.results = data;
        });
    }
    outerToInner(value) {
        if (value && value.id && !value[this.field]) {
            this.service.find(value.id).subscribe((data) => {
                this.formControl.setValue(data);
            });
        }
        return value;
    }
};
__decorate([
    Input()
], AutoCompleteComponent.prototype, "field", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "virtualScroll", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "itemSize", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "service", void 0);
AutoCompleteComponent = AutoCompleteComponent_1 = __decorate([
    Component({
        selector: "[nAutoComplete]",
        template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-autoCompleteCustom #input\n        [formControl]=\"formControl\"\n        [suggestions]=\"results\"\n        [virtualScroll]=\"virtualScroll\"\n        [itemSize]=\"itemSize\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onSelect)=\"onSelectEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        (completeMethod)=\"findAll($event)\"\n        [field]=\"field\">\n    </p-autoCompleteCustom>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
        providers: [provideValueAccessor(AutoCompleteComponent_1)]
    })
], AutoCompleteComponent);

const AUTOCOMPLETE_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => AutoCompleteCustomComponent),
    multi: true
};
let AutoCompleteCustomComponent = class AutoCompleteCustomComponent extends AutoComplete {
    constructor(el, renderer, cd, differs) {
        super(el, renderer, cd, differs);
        this.el = el;
        this.renderer = renderer;
        this.cd = cd;
        this.differs = differs;
    }
};
AutoCompleteCustomComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: ChangeDetectorRef },
    { type: IterableDiffers }
];
__decorate([
    Input()
], AutoCompleteCustomComponent.prototype, "virtualScroll", void 0);
__decorate([
    Input()
], AutoCompleteCustomComponent.prototype, "itemSize", void 0);
AutoCompleteCustomComponent = __decorate([
    Component({
        selector: 'p-autoCompleteCustom',
        template: "<span [ngClass]=\"{'ui-autocomplete ui-widget':true,'ui-autocomplete-dd':dropdown,'ui-autocomplete-multiple':multiple}\" [ngStyle]=\"style\" [class]=\"styleClass\">\n    <input *ngIf=\"!multiple\" #in [attr.type]=\"type\" [attr.id]=\"inputId\" [ngStyle]=\"inputStyle\" [class]=\"inputStyleClass\" [autocomplete]=\"autocomplete\" [attr.required]=\"required\" [attr.name]=\"name\"\n        [ngClass]=\"'ui-inputtext ui-widget ui-state-default ui-corner-all ui-autocomplete-input'\" [value]=\"inputFieldValue\" aria-autocomplete=\"list\" role=\"combobox\" [attr.aria-expanded]=\"overlayVisible\" aria-haspopup=\"true\" [attr.aria-activedescendant]=\"'p-highlighted-option'\"\n        (click)=\"onInputClick($event)\" (input)=\"onInput($event)\" (keydown)=\"onKeydown($event)\" (keyup)=\"onKeyup($event)\" [attr.autofocus]=\"autofocus\" (focus)=\"onInputFocus($event)\" (blur)=\"onInputBlur($event)\" (change)=\"onInputChange($event)\" (paste)=\"onInputPaste($event)\"\n        [attr.placeholder]=\"placeholder\" [attr.size]=\"size\" [attr.maxlength]=\"maxlength\" [attr.tabindex]=\"tabindex\" [readonly]=\"readonly\" [disabled]=\"disabled\" [attr.aria-label]=\"ariaLabel\" [attr.aria-labelledby]=\"ariaLabelledBy\" [attr.aria-required]=\"required\">\n\n    <ul *ngIf=\"multiple\" #multiContainer class=\"ui-autocomplete-multiple-container ui-widget ui-inputtext ui-state-default ui-corner-all\" [ngClass]=\"{'ui-state-disabled':disabled,'ui-state-focus':focus}\" (click)=\"multiIn.focus()\">\n        <li #token *ngFor=\"let val of value\" class=\"ui-autocomplete-token ui-state-highlight ui-corner-all\">\n            <span class=\"ui-autocomplete-token-icon pi pi-fw pi-times\" (click)=\"removeItem(token)\" *ngIf=\"!disabled\"></span>\n            <span *ngIf=\"!selectedItemTemplate\" class=\"ui-autocomplete-token-label\">{{resolveFieldData(val)}}</span>\n            <ng-container *ngTemplateOutlet=\"selectedItemTemplate; context: {$implicit: val}\"></ng-container>\n        </li>\n        <li class=\"ui-autocomplete-input-token\">\n            <input #multiIn [attr.type]=\"type\" [attr.id]=\"inputId\" [disabled]=\"disabled\" [attr.placeholder]=\"(value&&value.length ? null : placeholder)\" [attr.tabindex]=\"tabindex\" (input)=\"onInput($event)\"  (click)=\"onInputClick($event)\"\n                (keydown)=\"onKeydown($event)\" [readonly]=\"readonly\" (keyup)=\"onKeyup($event)\" [attr.autofocus]=\"autofocus\" (focus)=\"onInputFocus($event)\" (blur)=\"onInputBlur($event)\" (change)=\"onInputChange($event)\" (paste)=\"onInputPaste($event)\" [autocomplete]=\"autocomplete\"\n                [ngStyle]=\"inputStyle\" [class]=\"inputStyleClass\" [attr.aria-label]=\"ariaLabel\" [attr.aria-labelledby]=\"ariaLabelledBy\" [attr.aria-required]=\"required\"\n                aria-autocomplete=\"list\" role=\"combobox\" [attr.aria-expanded]=\"overlayVisible\" aria-haspopup=\"true\" [attr.aria-activedescendant]=\"'p-highlighted-option'\">\n        </li>\n    </ul>\n\n    <i *ngIf=\"loading\" class=\"ui-autocomplete-loader pi pi-spinner pi-spin\"></i>\n    <button #ddBtn type=\"button\" pButton [icon]=\"dropdownIcon\" class=\"ui-autocomplete-dropdown\" [disabled]=\"disabled\"\n        (click)=\"handleDropdownClick($event)\" *ngIf=\"dropdown\" [attr.tabindex]=\"tabindex\">\n    </button>\n\n    <div #panel *ngIf=\"overlayVisible\" [ngClass]=\"['ui-autocomplete-panel ui-widget ui-widget-content ui-corner-all ui-shadow']\"\n        [style.max-height]=\"virtualScroll ? 'auto' : (scrollHeight||'auto')\" [ngStyle]=\"panelStyle\" [class]=\"panelStyleClass\"\n        [@overlayAnimation]=\"{value: 'visible', params: {showTransitionParams: showTransitionOptions, hideTransitionParams: hideTransitionOptions}}\" (@overlayAnimation.start)=\"onOverlayAnimationStart($event)\" (@overlayAnimation.done)=\"onOverlayAnimationDone($event)\">\n        <ul role=\"listbox\" class=\"ui-autocomplete-items ui-autocomplete-list ui-widget-content ui-widget ui-corner-all ui-helper-reset\">\n            <ng-container *ngIf=\"!virtualScroll; else virtualScrollList\">\n                <p-autoCompleteItem *ngFor=\"let option of suggestions; let idx = index\"\n                    [option]=\"option\" [(highlightOption)]=\"highlightOption\" [index]=\"idx\"\n                    (onClick)=\"selectItem($event.option)\" [itemSize]=\"itemSize\"\n                    [template]=\"itemTemplate\" [resolvedFieldData]=\"resolveFieldData(option)\">\n                </p-autoCompleteItem>\n            </ng-container>\n            <ng-template #virtualScrollList>\n                <cdk-virtual-scroll-viewport *ngIf=\"virtualScroll && suggestions && suggestions.length\"\n                    [ngStyle]=\"{height: scrollHeight}\" [itemSize]=\"itemSize\">\n                    <ng-container *cdkVirtualFor=\"let option of suggestions; let idx = index\">\n                        <p-autoCompleteItem\n                            [option]=\"option\" [(highlightOption)]=\"highlightOption\" [index]=\"idx\"\n                            (onClick)=\"selectItem($event.option)\" [itemSize]=\"itemSize\"\n                            [template]=\"itemTemplate\" [resolvedFieldData]=\"resolveFieldData(option)\"\n                        ></p-autoCompleteItem>\n                    </ng-container>\n                </cdk-virtual-scroll-viewport>\n            </ng-template>\n\n            <li *ngIf=\"noResults && emptyMessage\" class=\"ui-autocomplete-emptymessage ui-autocomplete-list-item ui-corner-all\">{{emptyMessage}}</li>\n        </ul>\n    </div>\n</span>\n",
        providers: [AUTOCOMPLETE_VALUE_ACCESSOR]
    })
], AutoCompleteCustomComponent);

let AutoCompleteItem = class AutoCompleteItem {
    constructor() {
        this.highlightOptionChange = new EventEmitter();
        this.onClick = new EventEmitter();
    }
    selectItem(event) {
        this.onClick.emit({
            originalEvent: event,
            option: this.option
        });
    }
};
__decorate([
    Input()
], AutoCompleteItem.prototype, "option", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "resolvedFieldData", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "highlightOption", void 0);
__decorate([
    Output()
], AutoCompleteItem.prototype, "highlightOptionChange", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "index", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "itemSize", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "template", void 0);
__decorate([
    Output()
], AutoCompleteItem.prototype, "onClick", void 0);
AutoCompleteItem = __decorate([
    Component({
        selector: 'p-autoCompleteItem',
        template: `
        <li role="option" [ngClass]="{'ui-autocomplete-list-item ui-corner-all':true,'ui-state-highlight':(highlightOption==option)}"
            (mouseenter)="highlightOptionChange.emit(option)" (mouseleave)="highlightOptionChange.emit(null)"
            [id]="highlightOption == option ? 'p-highlighted-option':''" (click)="selectItem($event)"
            [ngStyle]="{'height': itemSize + 'px', 'display': 'table', 'width': '100%'}">
            <span *ngIf="!template">{{resolvedFieldData}}</span>
            <ng-container *ngTemplateOutlet="template; context: {$implicit: option, index: index}"></ng-container>
        </li>
    `
    })
], AutoCompleteItem);

var AutoCompleteMultipleComponent_1;
let AutoCompleteMultipleComponent = AutoCompleteMultipleComponent_1 = class AutoCompleteMultipleComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.virtualScroll = false;
        this.itemSize = 30;
    }
    findAll(event) {
        let query = {};
        query[this.field] = event.query;
        this.service.findAll(query).subscribe(data => {
            this.results = data;
        });
    }
    outerToInner(value) {
        if (value && value.length > 0 && !value[0][this.field]) {
            let obs = [];
            value.forEach(element => {
                if (element && element.id) {
                    obs.push(this.service.find(element.id));
                }
            });
            forkJoin(obs).subscribe((data) => {
                this.formControl.setValue(data);
            });
        }
        return value;
    }
};
__decorate([
    Input()
], AutoCompleteMultipleComponent.prototype, "field", void 0);
__decorate([
    Input()
], AutoCompleteMultipleComponent.prototype, "service", void 0);
__decorate([
    Input()
], AutoCompleteMultipleComponent.prototype, "virtualScroll", void 0);
__decorate([
    Input()
], AutoCompleteMultipleComponent.prototype, "itemSize", void 0);
AutoCompleteMultipleComponent = AutoCompleteMultipleComponent_1 = __decorate([
    Component({
        selector: "[nAutoCompleteMultiple]",
        template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-autoCompleteCustom #input\n        [formControl]=\"formControl\"\n        [suggestions]=\"results\"\n        (completeMethod)=\"findAll($event)\"\n        [field]=\"field\"\n        [virtualScroll]=\"virtualScroll\"\n        [itemSize]=\"itemSize\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onSelect)=\"onSelectEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        [multiple]=\"true\">\n    </p-autoCompleteCustom>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
        providers: [provideValueAccessor(AutoCompleteMultipleComponent_1)]
    })
], AutoCompleteMultipleComponent);

let CheckPipe = class CheckPipe {
    transform(value, ...args) {
        return value ? "Sim" : "Não";
    }
};
CheckPipe = __decorate([
    Pipe({
        name: "check"
    })
], CheckPipe);

let CrudModule = class CrudModule {
};
CrudModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            FormsModule,
            ReactiveFormsModule,
            FormNotificationModule,
            InputTextModule$1,
            InputTextareaModule,
            CalendarModule,
            AutoCompleteModule,
            ButtonModule,
            DatatableModule,
            HttpClientModule,
            RouterModule,
            ConfirmDialogModule,
            DropdownModule,
            ScrollingModule,
            InputMaskModule,
            RadioButtonModule,
        ],
        declarations: [
            CalendarComponent,
            InputTextComponent,
            TextAreaComponent,
            FormComponent,
            AutoCompleteComponent,
            AutoCompleteMultipleComponent,
            CrudListComponent,
            CrudFormComponent,
            FieldComponent,
            ValidationMessageComponent,
            DropdownComponent,
            AutoCompleteItem,
            AutoCompleteCustomComponent,
            CheckComponent,
            InputMaskComponent,
            CheckPipe,
        ],
        exports: [
            CalendarComponent,
            InputTextComponent,
            TextAreaComponent,
            FormComponent,
            AutoCompleteComponent,
            AutoCompleteMultipleComponent,
            CrudListComponent,
            CrudFormComponent,
            FieldComponent,
            ValidationMessageComponent,
            DropdownComponent,
            CheckComponent,
            InputMaskComponent,
            CheckPipe,
        ],
        providers: [
            httpInterceptorProviders,
            ConfirmationService,
        ]
    })
], CrudModule);

let BreadcrumbService = class BreadcrumbService {
    constructor() {
        this.itemsSource = new Subject();
        this.itemsHandler = this.itemsSource.asObservable();
    }
    setItems(items) {
        this.itemsSource.next(items);
    }
    reset() {
        this.itemsSource.next([]);
    }
};
BreadcrumbService = __decorate([
    Injectable()
], BreadcrumbService);

let BreadcrumbComponent = class BreadcrumbComponent {
    constructor(router, activatedRoute, breadcrumbService, messages, accessibilityService) {
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.breadcrumbService = breadcrumbService;
        this.messages = messages;
        this.accessibilityService = accessibilityService;
        this.highContrastEnabled = false;
        this.subscription = breadcrumbService.itemsHandler.subscribe(response => {
            this.items = response;
        });
    }
    ngOnInit() {
        this.router.events.pipe(filter((event) => event instanceof NavigationEnd), distinctUntilChanged()).subscribe((event) => {
            this.items = this.buildBreadCrumb(this.activatedRoute.root);
        });
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    enableHighContrast() {
        this.highContrastEnabled = true;
        this.accessibilityService.enableHighContrast();
    }
    disableHighContrast() {
        this.highContrastEnabled = false;
        this.accessibilityService.disableHighContrast();
    }
    increaseFontSize() {
        this.accessibilityService.increaseFontSize();
    }
    decreaseFontSize() {
        this.accessibilityService.decreaseFontSize();
    }
    ngAfterViewInit() {
        this.messages.messageObserver.subscribe((msg) => {
            this.accessibilityService.addAccessibilityMessages({ severity: msg.severity });
        });
        this.accessibilityService.addAccessibilityIcons();
    }
    buildBreadCrumb(route, url = '', breadcrumbs = []) {
        let label = route.routeConfig && route.routeConfig.data ? route.routeConfig.data.breadcrumb : '';
        let path = route.routeConfig && route.routeConfig.data ? route.routeConfig.path : '';
        const lastRoutePart = path.split('/').pop();
        const isDynamicRoute = lastRoutePart.startsWith(':');
        if (isDynamicRoute && !!route.snapshot) {
            for (let key in route.snapshot.params) {
                path = path.replace(`:${key}`, route.snapshot.params[key]);
            }
            label = Object.values(route.snapshot.params).join(' ');
        }
        const nextUrl = path ? `${url}/${path}` : url;
        const menuItem = {
            "label": label,
            "routerLink": nextUrl
        };
        const newBreadcrumbs = menuItem.label ? [...breadcrumbs, menuItem] : [...breadcrumbs];
        if (route.firstChild) {
            return this.buildBreadCrumb(route.firstChild, nextUrl, newBreadcrumbs);
        }
        return newBreadcrumbs;
    }
};
BreadcrumbComponent.ctorParameters = () => [
    { type: Router },
    { type: ActivatedRoute },
    { type: BreadcrumbService },
    { type: MessageService },
    { type: AccessbilityService }
];
BreadcrumbComponent = __decorate([
    Component({
        selector: 'breadcrumb',
        template: "<div class=\"layout-breadcrumb\">\n    <ul>\n        <li>\n            <a routerLink=\"\">\n                <i class=\"material-icons\">home</i>\n            </a>\n        </li>\n        <li>/</li>\n        <ng-template ngFor let-item let-last=\"last\" [ngForOf]=\"items\">\n            <li>\n                <a [routerLink]=\"item.routerLink\" *ngIf=\"item.routerLink\">{{item.label}}</a>\n\n                <ng-container *ngIf=\"!item.routerLink\">{{item.label}}</ng-container>\n            </li>\n            <li *ngIf=\"!last\">/</li>\n        </ng-template>\n    </ul>\n    <div class=\"layout-breadcrumb-options\">\n        <a (click)=\"highContrastEnabled ? disableHighContrast() : enableHighContrast()\" [ngStyle]=\"{'cursor': 'pointer'}\"\n            title=\"{{ highContrastEnabled ? 'Desabilitar alto contraste' : 'Habilitar alto contraste' }}\">\n            <i class=\"material-icons\">{{ highContrastEnabled ? 'visibility_off' : 'visibility' }}</i>\n        </a>\n        <a (click)=\"increaseFontSize()\" title=\"Aumentar fonte\" [ngStyle]=\"{'cursor': 'pointer'}\">\n            <i class=\"material-icons\">exposure_plus_1</i>\n        </a>\n        <a (click)=\"decreaseFontSize()\" title=\"Diminuir fonte\" [ngStyle]=\"{'cursor': 'pointer'}\">\n            <i class=\"material-icons\">exposure_neg_1</i>\n        </a>\n    </div>\n    <ng-content></ng-content>\n</div>\n",
        styles: [""]
    })
], BreadcrumbComponent);

let BreadcrumbModule = class BreadcrumbModule {
};
BreadcrumbModule = __decorate([
    NgModule({
        declarations: [BreadcrumbComponent],
        imports: [
            CommonModule,
            RouterModule
        ],
        exports: [
            BreadcrumbComponent
        ],
        providers: [
            BreadcrumbService
        ]
    })
], BreadcrumbModule);

var MenuOrientation;
(function (MenuOrientation) {
    MenuOrientation[MenuOrientation["STATIC"] = 0] = "STATIC";
    MenuOrientation[MenuOrientation["OVERLAY"] = 1] = "OVERLAY";
    MenuOrientation[MenuOrientation["SLIM"] = 2] = "SLIM";
    MenuOrientation[MenuOrientation["HORIZONTAL"] = 3] = "HORIZONTAL";
})(MenuOrientation || (MenuOrientation = {}));
/**
 * Class MenuService
 * @class
 */
let MenusService = class MenusService {
    constructor() {
        this.layoutMode = MenuOrientation.STATIC;
        this.itens = [];
    }
    isOverlay() {
        return this.layoutMode === MenuOrientation.OVERLAY;
    }
    isHorizontal() {
        return this.layoutMode === MenuOrientation.HORIZONTAL;
    }
    isSlim() {
        return this.layoutMode === MenuOrientation.SLIM;
    }
    changeToStaticMenu() {
        this.layoutMode = MenuOrientation.STATIC;
    }
    changeToOverlayMenu() {
        this.layoutMode = MenuOrientation.OVERLAY;
    }
    changeToHorizontalMenu() {
        this.layoutMode = MenuOrientation.HORIZONTAL;
    }
    changeToSlimMenu() {
        this.layoutMode = MenuOrientation.SLIM;
    }
};
MenusService.ɵprov = ɵɵdefineInjectable({ factory: function MenusService_Factory() { return new MenusService(); }, token: MenusService, providedIn: "root" });
MenusService = __decorate([
    Injectable({
        providedIn: 'root'
    })
], MenusService);

let MenuComponent = class MenuComponent {
    constructor(menuService) {
        this.menuService = menuService;
    }
};
MenuComponent.ctorParameters = () => [
    { type: MenusService }
];
MenuComponent = __decorate([
    Component({
        selector: 'app-menu',
        template: "<ul submenu\n    [item]=\"menuService.itens\"\n    root=\"true\"\n    class=\"ultima-menu ultima-main-menu clearfix\"\n    [reset]=\"menuService.resetMenu\"\n    visible=\"true\"\n    parentActive=\"true\">\n</ul>\n"
    })
], MenuComponent);

let SubMenuComponent = class SubMenuComponent {
    constructor(menuService) {
        this.menuService = menuService;
    }
    /**
     * Alterar o menu root entre aberto e fechado
     */
    rootMenuToggle() {
        if (this.root) {
            this.menuService.menuHoverActive = !this.menuService.menuHoverActive;
        }
    }
    /**
     * Activa o menu por index e desativa o item ativo
     * @param index
     */
    activeMenuByIndex(index) {
        this.activeIndex = (this.activeIndex === index) ? null : index;
        return this.activeIndex;
    }
    /**
     * Verifica se foi informado um comando para o item de menu e executa caso necessário
     * @param item menu
     */
    execCommand(event, item) {
        if (item.command) {
            item.command({ 'originalEvent': event, 'item': item });
            return true;
        }
        return false;
    }
    /**
     * Esconte o menu caso este não possua itens
     * @param item menu
     */
    hideMenu(item) {
        if (!item.items) {
            if (this.menuService.isHorizontal() || this.menuService.isSlim()) {
                this.menuService.resetMenu = true;
            }
            else {
                this.menuService.resetMenu = false;
            }
            this.menuService.overlayMenuActive = false;
            this.menuService.staticMenuMobileActive = false;
            this.menuService.menuHoverActive = !this.menuService.menuHoverActive;
            return true;
        }
        return false;
    }
    itemClick(event, item, index) {
        this.rootMenuToggle();
        if (item.disabled) {
            event.preventDefault();
            return false;
        }
        this.activeMenuByIndex(index);
        this.execCommand(event, item);
        if (item.items || (!item.url && !item.routerLink)) {
            event.preventDefault();
        }
        this.hideMenu(item);
        return true;
    }
    onMouseEnter(index) {
        if (this.root && this.menuService.menuHoverActive && (this.menuService.isHorizontal() || this.menuService.isSlim())) {
            this.activeIndex = index;
        }
    }
    isActive(index) {
        return this.activeIndex === index;
    }
    get reset() {
        return this._reset;
    }
    set reset(val) {
        this._reset = val;
        if (this._reset && (this.menuService.isHorizontal() || this.menuService.isSlim())) {
            this.activeIndex = null;
        }
    }
    get parentActive() {
        return this._parentActive;
    }
    set parentActive(val) {
        this._parentActive = val;
        if (!this._parentActive) {
            this.activeIndex = null;
        }
    }
};
SubMenuComponent.ctorParameters = () => [
    { type: MenusService }
];
__decorate([
    Input()
], SubMenuComponent.prototype, "item", void 0);
__decorate([
    Input()
], SubMenuComponent.prototype, "root", void 0);
__decorate([
    Input()
], SubMenuComponent.prototype, "visible", void 0);
__decorate([
    Input()
], SubMenuComponent.prototype, "reset", null);
__decorate([
    Input()
], SubMenuComponent.prototype, "parentActive", null);
SubMenuComponent = __decorate([
    Component({
        selector: '[submenu]',
        template: "<ng-template ngFor let-child let-i=\"index\" [ngForOf]=\"(root ? item : item.items)\">\n    <li [ngClass]=\"{'active-menuitem': isActive(i)}\"\n        [class]=\"child.badgeStyleClass\"\n        *ngIf=\"child.visible === false ? false : true\">\n        <a [href]=\"child.url||'#'\" (click)=\"itemClick($event,child,i)\" (mouseenter)=\"onMouseEnter(i)\"\n           class=\"ripplelink\" *ngIf=\"!child.routerLink\"\n            [attr.tabindex]=\"!visible ? '-1' : null\" [attr.target]=\"child.target\">\n            <i *ngIf=\"child.icon\" class=\"material-icons\">{{child.icon}}</i>\n            <span>{{child.label}}</span>\n            <span class=\"menuitem-badge\" *ngIf=\"child.badge\">{{child.badge}}</span>\n            <i class=\"material-icons submenu-icon\" *ngIf=\"child.items\">keyboard_arrow_down</i>\n        </a>\n\n        <a (click)=\"itemClick($event,child,i)\"\n            (mouseenter)=\"onMouseEnter(i)\"\n            class=\"ripplelink\"\n            *ngIf=\"child.routerLink\"\n            [routerLink]=\"child.routerLink\"\n            routerLinkActive=\"active-menuitem-routerlink\"\n            [routerLinkActiveOptions]=\"{exact: true}\" [attr.tabindex]=\"!visible ? '-1' : null\" [attr.target]=\"child.target\">\n            <i *ngIf=\"child.icon\" class=\"material-icons\">{{child.icon}}</i>\n            <span>{{child.label}}</span>\n            <span class=\"menuitem-badge\" *ngIf=\"child.badge\">{{child.badge}}</span>\n            <i class=\"material-icons submenu-icon\" *ngIf=\"child.items\">keyboard_arrow_down</i>\n        </a>\n        <div class=\"layout-menu-tooltip\">\n            <div class=\"layout-menu-tooltip-arrow\"></div>\n            <div class=\"layout-menu-tooltip-text\">{{child.label}}</div>\n        </div>\n        <ul submenu\n            [item]=\"child\"\n            *ngIf=\"child.items\"\n            [visible]=\"isActive(i)\"\n            [reset]=\"reset\"\n            [parentActive]=\"isActive(i)\"\n            [@children]=\"(menuService.isSlim() || menuService.isHorizontal()) && root ? (isActive(i) ?\n            'visible' : 'hidden') : (isActive(i) ? 'visibleAnimated' : 'hiddenAnimated')\">\n        </ul>\n    </li>\n</ng-template>\n",
        animations: [
            trigger('children', [
                state('hiddenAnimated', style({
                    height: '0px'
                })),
                state('visibleAnimated', style({
                    height: '*'
                })),
                state('visible', style({
                    height: '*',
                    'z-index': 100
                })),
                state('hidden', style({
                    height: '0px',
                    'z-index': '*'
                })),
                transition('visibleAnimated => hiddenAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)')),
                transition('hiddenAnimated => visibleAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)'))
            ])
        ]
    })
], SubMenuComponent);

let MenuModule = class MenuModule {
};
MenuModule = __decorate([
    NgModule({
        declarations: [
            MenuComponent,
            SubMenuComponent
        ],
        imports: [
            CommonModule,
            RouterModule
        ],
        exports: [
            MenuComponent,
            SubMenuComponent
        ],
        providers: [
            MenusService
        ]
    })
], MenuModule);

/*
 * Public API Surface of primeng-components
 */

/**
 * Generated bundle index. Do not edit.
 */

export { AutoCompleteComponent, AutoCompleteMultipleComponent, BlockUiComponent, BlockUiModule, BreadcrumbComponent, BreadcrumbModule, BreadcrumbService, CALENDAR_LOCALE, CRUD_SERVICE, CalendarComponent, CheckComponent, CheckPipe, Column, CrudFormComponent, CrudListComponent, CrudModule, CrudResolveGuard, CrudServiceNuvem, DataProviderFactory, DataProviderType, DatatableButtonComponent, DatatableClickEvent, DatatableComponent, DatatableModule, DatatableService, DatatableSplitbuttonComponent, DatatableSplitbuttonModule, DropdownComponent, ErrorStackComponent, ErrorStackModule, FieldComponent, FilterColumn, FilterColumnModule, FormComponent, FormNotificationComponent, FormNotificationModule, InputMaskComponent, InputTextComponent, MemoryDataProvider, MenuComponent, MenuModule, MenuOrientation, MenusService, PageNotificationComponent, PageNotificationModule, PageNotificationService, ServerDataProvider, ShowColumn, ShowColumnModule, SubMenuComponent, SuccessMessageHandlerInterceptor, TextAreaComponent, ValidationMessageComponent, crudRouteBuilder, httpInterceptorProviders, AutoCompleteItem as ɵa, AUTOCOMPLETE_VALUE_ACCESSOR as ɵb, AutoCompleteCustomComponent as ɵc };
