// @nuvem/primeng-components/esm5/lib/breadcrumb/breadcrumb.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
var BreadcrumbService = /** @class */ (function () {
    function BreadcrumbService() {
        this.itemsSource = new Subject();
        this.itemsHandler = this.itemsSource.asObservable();
    }
    BreadcrumbService.prototype.setItems = function (items) {
        this.itemsSource.next(items);
    };
    BreadcrumbService.prototype.reset = function () {
        this.itemsSource.next([]);
    };
BreadcrumbService.ɵfac = function BreadcrumbService_Factory(t) { return new (t || BreadcrumbService)(); };
BreadcrumbService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: BreadcrumbService, factory: function (t) { return BreadcrumbService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BreadcrumbService, [{
        type: Injectable
    }], function () { return []; }, null); })();
    return BreadcrumbService;
}());
export { BreadcrumbService };
