// @nuvem/primeng-components/esm5/lib/breadcrumb/breadcrumb.component.js
import { __decorate, __read, __spread } from "tslib";
import { Component } from '@angular/core';
import { MenuItem, MessageService, Message } from 'primeng';
import { BreadcrumbService } from './breadcrumb.service';
import { AccessbilityService } from '@nuvem/angular-base';
import { ActivatedRoute, Router, NavigationEnd } from '@angular/router';
import { filter, distinctUntilChanged } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from 'primeng';
import * as ɵngcc3 from '@nuvem/angular-base';
import * as ɵngcc4 from '@angular/common';

function BreadcrumbComponent_ng_template_8_a_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "a", 10);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r1032 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("routerLink", item_r1032.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(item_r1032.label);
} }
function BreadcrumbComponent_ng_template_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r1032 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(item_r1032.label);
} }
function BreadcrumbComponent_ng_template_8_li_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵtext(1, "/");
    ɵngcc0.ɵɵelementEnd();
} }
function BreadcrumbComponent_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵtemplate(1, BreadcrumbComponent_ng_template_8_a_1_Template, 2, 2, "a", 8);
    ɵngcc0.ɵɵtemplate(2, BreadcrumbComponent_ng_template_8_ng_container_2_Template, 2, 1, "ng-container", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, BreadcrumbComponent_ng_template_8_li_3_Template, 2, 0, "li", 9);
} if (rf & 2) {
    var item_r1032 = ctx.$implicit;
    var last_r1033 = ctx.last;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r1032.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !item_r1032.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !last_r1033);
} }
var _c0 = function () { return { "cursor": "pointer" }; };
var _c1 = ["*"];
var BreadcrumbComponent = /** @class */ (function () {
    function BreadcrumbComponent(router, activatedRoute, breadcrumbService, messages, accessibilityService) {
        var _this = this;
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.breadcrumbService = breadcrumbService;
        this.messages = messages;
        this.accessibilityService = accessibilityService;
        this.highContrastEnabled = false;
        this.subscription = breadcrumbService.itemsHandler.subscribe(function (response) {
            _this.items = response;
        });
    }
    BreadcrumbComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.router.events.pipe(filter(function (event) { return event instanceof NavigationEnd; }), distinctUntilChanged()).subscribe(function (event) {
            _this.items = _this.buildBreadCrumb(_this.activatedRoute.root);
        });
    };
    BreadcrumbComponent.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    BreadcrumbComponent.prototype.enableHighContrast = function () {
        this.highContrastEnabled = true;
        this.accessibilityService.enableHighContrast();
    };
    BreadcrumbComponent.prototype.disableHighContrast = function () {
        this.highContrastEnabled = false;
        this.accessibilityService.disableHighContrast();
    };
    BreadcrumbComponent.prototype.increaseFontSize = function () {
        this.accessibilityService.increaseFontSize();
    };
    BreadcrumbComponent.prototype.decreaseFontSize = function () {
        this.accessibilityService.decreaseFontSize();
    };
    BreadcrumbComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.messages.messageObserver.subscribe(function (msg) {
            _this.accessibilityService.addAccessibilityMessages({ severity: msg.severity });
        });
        this.accessibilityService.addAccessibilityIcons();
    };
    BreadcrumbComponent.prototype.buildBreadCrumb = function (route, url, breadcrumbs) {
        if (url === void 0) { url = ''; }
        if (breadcrumbs === void 0) { breadcrumbs = []; }
        var label = route.routeConfig && route.routeConfig.data ? route.routeConfig.data.breadcrumb : '';
        var path = route.routeConfig && route.routeConfig.data ? route.routeConfig.path : '';
        var lastRoutePart = path.split('/').pop();
        var isDynamicRoute = lastRoutePart.startsWith(':');
        if (isDynamicRoute && !!route.snapshot) {
            for (var key in route.snapshot.params) {
                path = path.replace(":" + key, route.snapshot.params[key]);
            }
            label = Object.values(route.snapshot.params).join(' ');
        }
        var nextUrl = path ? url + "/" + path : url;
        var menuItem = {
            "label": label,
            "routerLink": nextUrl
        };
        var newBreadcrumbs = menuItem.label ? __spread(breadcrumbs, [menuItem]) : __spread(breadcrumbs);
        if (route.firstChild) {
            return this.buildBreadCrumb(route.firstChild, nextUrl, newBreadcrumbs);
        }
        return newBreadcrumbs;
    };
    BreadcrumbComponent.ctorParameters = function () { return [
        { type: Router },
        { type: ActivatedRoute },
        { type: BreadcrumbService },
        { type: MessageService },
        { type: AccessbilityService }
    ]; };
BreadcrumbComponent.ɵfac = function BreadcrumbComponent_Factory(t) { return new (t || BreadcrumbComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(BreadcrumbService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.MessageService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AccessbilityService)); };
BreadcrumbComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: BreadcrumbComponent, selectors: [["breadcrumb"]], ngContentSelectors: _c1, decls: 20, vars: 9, consts: [[1, "layout-breadcrumb"], ["routerLink", ""], [1, "material-icons"], ["ngFor", "", 3, "ngForOf"], [1, "layout-breadcrumb-options"], [3, "ngStyle", "title", "click"], ["title", "Aumentar fonte", 3, "ngStyle", "click"], ["title", "Diminuir fonte", 3, "ngStyle", "click"], [3, "routerLink", 4, "ngIf"], [4, "ngIf"], [3, "routerLink"]], template: function BreadcrumbComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "ul");
        ɵngcc0.ɵɵelementStart(2, "li");
        ɵngcc0.ɵɵelementStart(3, "a", 1);
        ɵngcc0.ɵɵelementStart(4, "i", 2);
        ɵngcc0.ɵɵtext(5, "home");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "li");
        ɵngcc0.ɵɵtext(7, "/");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, BreadcrumbComponent_ng_template_8_Template, 4, 3, "ng-template", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 4);
        ɵngcc0.ɵɵelementStart(10, "a", 5);
        ɵngcc0.ɵɵlistener("click", function BreadcrumbComponent_Template_a_click_10_listener() { return ctx.highContrastEnabled ? ctx.disableHighContrast() : ctx.enableHighContrast(); });
        ɵngcc0.ɵɵelementStart(11, "i", 2);
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "a", 6);
        ɵngcc0.ɵɵlistener("click", function BreadcrumbComponent_Template_a_click_13_listener() { return ctx.increaseFontSize(); });
        ɵngcc0.ɵɵelementStart(14, "i", 2);
        ɵngcc0.ɵɵtext(15, "exposure_plus_1");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "a", 7);
        ɵngcc0.ɵɵlistener("click", function BreadcrumbComponent_Template_a_click_16_listener() { return ctx.decreaseFontSize(); });
        ɵngcc0.ɵɵelementStart(17, "i", 2);
        ɵngcc0.ɵɵtext(18, "exposure_neg_1");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(19);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.items);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ctx.highContrastEnabled ? "Desabilitar alto contraste" : "Habilitar alto contraste");
        ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction0(6, _c0));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.highContrastEnabled ? "visibility_off" : "visibility");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction0(7, _c0));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction0(8, _c0));
    } }, directives: [ɵngcc1.RouterLinkWithHref, ɵngcc4.NgForOf, ɵngcc4.NgStyle, ɵngcc4.NgIf], styles: [""] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BreadcrumbComponent, [{
        type: Component,
        args: [{
                selector: 'breadcrumb',
                template: "<div class=\"layout-breadcrumb\">\n    <ul>\n        <li>\n            <a routerLink=\"\">\n                <i class=\"material-icons\">home</i>\n            </a>\n        </li>\n        <li>/</li>\n        <ng-template ngFor let-item let-last=\"last\" [ngForOf]=\"items\">\n            <li>\n                <a [routerLink]=\"item.routerLink\" *ngIf=\"item.routerLink\">{{item.label}}</a>\n\n                <ng-container *ngIf=\"!item.routerLink\">{{item.label}}</ng-container>\n            </li>\n            <li *ngIf=\"!last\">/</li>\n        </ng-template>\n    </ul>\n    <div class=\"layout-breadcrumb-options\">\n        <a (click)=\"highContrastEnabled ? disableHighContrast() : enableHighContrast()\" [ngStyle]=\"{'cursor': 'pointer'}\"\n            title=\"{{ highContrastEnabled ? 'Desabilitar alto contraste' : 'Habilitar alto contraste' }}\">\n            <i class=\"material-icons\">{{ highContrastEnabled ? 'visibility_off' : 'visibility' }}</i>\n        </a>\n        <a (click)=\"increaseFontSize()\" title=\"Aumentar fonte\" [ngStyle]=\"{'cursor': 'pointer'}\">\n            <i class=\"material-icons\">exposure_plus_1</i>\n        </a>\n        <a (click)=\"decreaseFontSize()\" title=\"Diminuir fonte\" [ngStyle]=\"{'cursor': 'pointer'}\">\n            <i class=\"material-icons\">exposure_neg_1</i>\n        </a>\n    </div>\n    <ng-content></ng-content>\n</div>\n",
                styles: [""]
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc1.ActivatedRoute }, { type: BreadcrumbService }, { type: ɵngcc2.MessageService }, { type: ɵngcc3.AccessbilityService }]; }, null); })();
    return BreadcrumbComponent;
}());
export { BreadcrumbComponent };
