// @nuvem/angular-base/esm5/lib/error/error.module.js
import { __decorate } from "tslib";
import { ErrorHandler, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ErrorService } from './error.service';
import { ErrorProvider } from './providers/error.provider';
import { HttpApplicationProblemErrorService } from './services/http-application-problem-error.service';
import { HttpGenericErrorService } from './services/http-generic-error.service';
/**
 * O modulo ErrorModule contem funcionalidades para a captura e apresentação dos erros que possam ocorrer na aplicação.
 * @module
 */
import * as ɵngcc0 from '@angular/core';
var ErrorModule = /** @class */ (function () {
    function ErrorModule() {
    }
ErrorModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ErrorModule });
ErrorModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ErrorModule_Factory(t) { return new (t || ErrorModule)(); }, providers: [
        {
            provide: ErrorProvider,
            useClass: HttpGenericErrorService,
            multi: true
        },
        {
            provide: ErrorProvider,
            useClass: HttpApplicationProblemErrorService,
            multi: true
        },
        {
            provide: ErrorHandler,
            useClass: ErrorService
        },
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ErrorModule, { imports: function () { return [CommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ErrorModule, [{
        type: NgModule,
        args: [{
                declarations: [],
                imports: [
                    CommonModule
                ],
                providers: [
                    {
                        provide: ErrorProvider,
                        useClass: HttpGenericErrorService,
                        multi: true
                    },
                    {
                        provide: ErrorProvider,
                        useClass: HttpApplicationProblemErrorService,
                        multi: true
                    },
                    {
                        provide: ErrorHandler,
                        useClass: ErrorService
                    },
                ]
            }]
    }], function () { return []; }, null); })();
    return ErrorModule;
}());
export { ErrorModule };
