// @nuvem/angular-base/esm5/lib/error/error.service.js
import { __decorate, __extends, __param } from "tslib";
import { ErrorHandler, Inject, Injectable } from '@angular/core';
import { ErrorProvider } from './providers/error.provider';
import * as i0 from "@angular/core";
import * as i1 from "./providers/error.provider";
/**
 * Class ErrorService
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var ErrorService = /** @class */ (function (_super) {
    __extends(ErrorService, _super);
    /**
     * Metodo construtor responável por injetar o serviço ErrorProvider
     * @param {ErrorProvider} providers
     */
    function ErrorService(providers) {
        var _this = _super.call(this) || this;
        _this.providers = providers;
        return _this;
    }
    /**
     * Metodo responsável por idetificar e obter os erros lançados de requisições http/https
     * @param {Error} error
     * @returns void
     */
    ErrorService.prototype.handleError = function (error) {
        try {
            this.providers.forEach(function (p) {
                if (p.shouldHandle(error)) {
                    p.handle(error);
                }
            });
        }
        catch (e) {
        }
        _super.prototype.handleError.call(this, error);
    };
    ErrorService.ctorParameters = function () { return [
        { type: Array, decorators: [{ type: Inject, args: [ErrorProvider,] }] }
    ]; };
    ErrorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ErrorService_Factory() { return new ErrorService(i0.ɵɵinject(i1.ErrorProvider)); }, token: ErrorService, providedIn: "root" });
    ErrorService = __decorate([ __param(0, Inject(ErrorProvider))
    ], ErrorService);
ErrorService.ɵfac = function ErrorService_Factory(t) { return new (t || ErrorService)(ɵngcc0.ɵɵinject(ErrorProvider)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ErrorService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: Array, decorators: [{
                type: Inject,
                args: [ErrorProvider]
            }] }]; }, null); })();
    return ErrorService;
}(ErrorHandler));
export { ErrorService };
