// @nuvem/primeng-components/esm5/lib/datatable/show-column/show-column.module.js
import { __decorate } from "tslib";
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule } from 'primeng/button';
import { DropdownModule } from 'primeng/dropdown';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CheckboxModule, MultiSelectModule, InputTextModule, } from 'primeng';
import { ShowColumn } from './show-column.component';
import * as ɵngcc0 from '@angular/core';
var ShowColumnModule = /** @class */ (function () {
    function ShowColumnModule() {
    }
ShowColumnModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ShowColumnModule });
ShowColumnModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ShowColumnModule_Factory(t) { return new (t || ShowColumnModule)(); }, imports: [[
            CommonModule,
            MultiSelectModule,
            ButtonModule,
            CheckboxModule,
            DropdownModule,
            FormsModule,
            ReactiveFormsModule,
            InputTextModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ShowColumnModule, { declarations: function () { return [ShowColumn]; }, imports: function () { return [CommonModule,
        MultiSelectModule,
        ButtonModule,
        CheckboxModule,
        DropdownModule,
        FormsModule,
        ReactiveFormsModule,
        InputTextModule]; }, exports: function () { return [ShowColumn]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ShowColumnModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MultiSelectModule,
                    ButtonModule,
                    CheckboxModule,
                    DropdownModule,
                    FormsModule,
                    ReactiveFormsModule,
                    InputTextModule
                ],
                declarations: [
                    ShowColumn
                ],
                exports: [
                    ShowColumn
                ],
                schemas: [CUSTOM_ELEMENTS_SCHEMA]
            }]
    }], function () { return []; }, null); })();
    return ShowColumnModule;
}());
export { ShowColumnModule };
