// @nuvem/primeng-components/esm5/lib/datatable/show-column/show-column.component.js
import { __decorate } from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { PageNotificationService } from './../../page-notification/page-notification.service';
var ShowColumn = /** @class */ (function () {
    function ShowColumn(pageNotificationService) {
        this.pageNotificationService = pageNotificationService;
        /**
         * lista de colunas disponívels na tabela
         * @type {Array}
         */
        this.options = [];
        /**
         * Valores do select de colunas
         * @type {Array}
         */
        this.values = [];
        /**
         * Último valor válido para o select.
         * @type {Array}
         */
        this.lastValue = [];
        /**
         * Evento para atualizar a visibilidade das colunas da tabela
         * @type {EventEmitter}
         */
        this.updateVisibleColumns = new EventEmitter();
    }
    ShowColumn.prototype.ngOnInit = function () {
        var _this = this;
        this.columns.forEach(function (column) {
            var col = { field: column.field, filterType: column.filterType, header: column.header };
            _this.options.push(col);
            _this.values.push(col);
            _this.lastValue.push(col);
        });
    };
    ShowColumn.prototype.mostrarColunas = function (event) {
        var _this = this;
        if (this.values.length) {
            this.lastValue = event.value;
            this.updateVisibleColumns.emit(this.values);
        }
        else {
            this.lastValue.map(function (item) { return _this.values.push(item); });
            this.pageNotificationService.addErrorMessage('Não é possível exibir menos de uma coluna');
        }
    };
    ShowColumn.ctorParameters = function () { return [
        { type: PageNotificationService }
    ]; };
    __decorate([
        Input()
    ], ShowColumn.prototype, "columns", void 0);
    __decorate([
        Output()
    ], ShowColumn.prototype, "updateVisibleColumns", void 0);
    ShowColumn = __decorate([
        Component({
            selector: 'basis-show-column',
            template: "<div class=\"ui-g-12 ui-g-nopad\">\n    <div class=\"ui-g-3 ui-md-offset-6\">\n        <div class=\"text-right\">Visibilidade das Colunas:</div>\n    </div>\n    <div class=\"ui-g-3\">\n        <p-multiSelect\n            class=\"column\"\n            defaultLabel=\"Selecione as colunas a serem exibidas\"\n            [optionLabel]=\"'header'\"\n            [displaySelectedLabel]=\"false\"\n            [ngModelOptions]=\"{ standalone: true }\"\n            (onChange)=\"mostrarColunas($event)\"\n            [options]=\"options\"\n            [(ngModel)]=\"values\"\n            [style]=\"{ width: '100%' }\"\n        >\n        </p-multiSelect>\n    </div>\n</div>\n"
        })
    ], ShowColumn);
    return ShowColumn;
}());
export { ShowColumn };
