// ng-block-ui/fesm2015/ng-block-ui-router.js
import { Injectable, NgModule } from '@angular/core';
import { BlockUIService, BLOCKUI_DEFAULT, BlockUIModule } from 'ng-block-ui';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ng-block-ui';
class BlockUIPreventNavigation {
    /**
     * @param {?} blockUIService
     */
    constructor(blockUIService) {
        this.blockUIService = blockUIService;
    }
    /**
     * @return {?}
     */
    canActivate() {
        return !this.blockUIService.isActive(BLOCKUI_DEFAULT);
    }
    /**
     * @return {?}
     */
    canActivateChild() {
        return !this.blockUIService.isActive(BLOCKUI_DEFAULT);
    }
}
BlockUIPreventNavigation.ɵfac = function BlockUIPreventNavigation_Factory(t) { return new (t || BlockUIPreventNavigation)(ɵngcc0.ɵɵinject(ɵngcc1.BlockUIService)); };
BlockUIPreventNavigation.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: BlockUIPreventNavigation, factory: BlockUIPreventNavigation.ɵfac });
/** @nocollapse */
BlockUIPreventNavigation.ctorParameters = () => [
    { type: BlockUIService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUIPreventNavigation, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.BlockUIService }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class BlockUIRouterModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: BlockUIRouterModule,
            providers: [
                BlockUIPreventNavigation
            ]
        };
    }
}
BlockUIRouterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BlockUIRouterModule });
BlockUIRouterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function BlockUIRouterModule_Factory(t) { return new (t || BlockUIRouterModule)(); }, imports: [[
            BlockUIModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BlockUIRouterModule, { imports: function () { return [BlockUIModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUIRouterModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BlockUIModule
                ]
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { BlockUIRouterModule, BlockUIPreventNavigation };
