// @nuvem/primeng-components/esm5/lib/crud/components/auto-complete-multiple/auto-complete-multiple.component.js
import { __decorate, __extends } from "tslib";
import { Component, Input } from "@angular/core";
import { provideValueAccessor } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
import { forkJoin } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../form-notification/form-notification.component';
import * as ɵngcc2 from '../auto-complete/auto-complete-custom.component';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../validation-message/validation-message.component';

var _c0 = ["nAutoCompleteMultiple", ""];
function AutoCompleteMultipleComponent_nValidationMessage_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "nValidationMessage", 6);
} if (rf & 2) {
    var ctx_r1030 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("control", ctx_r1030.formControl);
} }
var _c1 = ["*"];
var AutoCompleteMultipleComponent = /** @class */ (function (_super) {
    __extends(AutoCompleteMultipleComponent, _super);
    function AutoCompleteMultipleComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.virtualScroll = false;
        _this.itemSize = 30;
        return _this;
    }
    AutoCompleteMultipleComponent_1 = AutoCompleteMultipleComponent;
    AutoCompleteMultipleComponent.prototype.findAll = function (event) {
        var _this = this;
        var query = {};
        query[this.field] = event.query;
        this.service.findAll(query).subscribe(function (data) {
            _this.results = data;
        });
    };
    AutoCompleteMultipleComponent.prototype.outerToInner = function (value) {
        var _this = this;
        if (value && value.length > 0 && !value[0][this.field]) {
            var obs_1 = [];
            value.forEach(function (element) {
                if (element && element.id) {
                    obs_1.push(_this.service.find(element.id));
                }
            });
            forkJoin(obs_1).subscribe(function (data) {
                _this.formControl.setValue(data);
            });
        }
        return value;
    };
    var AutoCompleteMultipleComponent_1;
    __decorate([
        Input()
    ], AutoCompleteMultipleComponent.prototype, "field", void 0);
    __decorate([
        Input()
    ], AutoCompleteMultipleComponent.prototype, "service", void 0);
    __decorate([
        Input()
    ], AutoCompleteMultipleComponent.prototype, "virtualScroll", void 0);
    __decorate([
        Input()
    ], AutoCompleteMultipleComponent.prototype, "itemSize", void 0);
AutoCompleteMultipleComponent.ɵfac = function AutoCompleteMultipleComponent_Factory(t) { return ɵAutoCompleteMultipleComponent_BaseFactory(t || AutoCompleteMultipleComponent); };
AutoCompleteMultipleComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AutoCompleteMultipleComponent, selectors: [["", "nAutoCompleteMultiple", ""]], inputs: { field: "field", service: "service", virtualScroll: "virtualScroll", itemSize: "itemSize" }, features: [ɵngcc0.ɵɵProvidersFeature([provideValueAccessor(AutoCompleteMultipleComponent_1)]), ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c1, decls: 7, vars: 11, consts: [["form-notification", "", 1, "md-inputfield", 3, "name", "form"], ["notification", ""], [3, "formControl", "suggestions", "field", "virtualScroll", "itemSize", "multiple", "completeMethod", "onBlur", "onSelect", "onFocus"], ["input", ""], [3, "for", "innerHTML"], [3, "control", 4, "ngIf"], [3, "control"]], template: function AutoCompleteMultipleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "span", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "p-autoCompleteCustom", 2, 3);
        ɵngcc0.ɵɵlistener("completeMethod", function AutoCompleteMultipleComponent_Template_p_autoCompleteCustom_completeMethod_2_listener($event) { return ctx.findAll($event); })("onBlur", function AutoCompleteMultipleComponent_Template_p_autoCompleteCustom_onBlur_2_listener($event) { return ctx.onBlurEvent($event); })("onSelect", function AutoCompleteMultipleComponent_Template_p_autoCompleteCustom_onSelect_2_listener($event) { return ctx.onSelectEvent($event); })("onFocus", function AutoCompleteMultipleComponent_Template_p_autoCompleteCustom_onFocus_2_listener($event) { return ctx.onFocusEvent($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(4, "label", 4);
        ɵngcc0.ɵɵtemplate(5, AutoCompleteMultipleComponent_nValidationMessage_5_Template, 1, 1, "nValidationMessage", 5);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        var _r1028 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵproperty("name", ctx.name)("form", ctx.form);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("formControl", ctx.formControl)("suggestions", ctx.results)("field", ctx.field)("virtualScroll", ctx.virtualScroll)("itemSize", ctx.itemSize)("multiple", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("for", ctx.name)("innerHTML", ctx.label, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !_r1028.message);
    } }, directives: [ɵngcc1.FormNotificationComponent, ɵngcc2.AutoCompleteCustomComponent, ɵngcc3.NgControlStatus, ɵngcc3.FormControlDirective, ɵngcc4.NgIf, ɵngcc5.ValidationMessageComponent], encapsulation: 2 });
var ɵAutoCompleteMultipleComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(AutoCompleteMultipleComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AutoCompleteMultipleComponent, [{
        type: Component,
        args: [{
                selector: "[nAutoCompleteMultiple]",
                template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-autoCompleteCustom #input\n        [formControl]=\"formControl\"\n        [suggestions]=\"results\"\n        (completeMethod)=\"findAll($event)\"\n        [field]=\"field\"\n        [virtualScroll]=\"virtualScroll\"\n        [itemSize]=\"itemSize\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onSelect)=\"onSelectEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        [multiple]=\"true\">\n    </p-autoCompleteCustom>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
                providers: [provideValueAccessor(AutoCompleteMultipleComponent_1)]
            }]
    }], null, { field: [{
            type: Input
        }], service: [{
            type: Input
        }], virtualScroll: [{
            type: Input
        }], itemSize: [{
            type: Input
        }] }); })();
    return AutoCompleteMultipleComponent;
}(FieldComponent));
export { AutoCompleteMultipleComponent };
