// @nuvem/primeng-components/esm5/lib/crud/components/auto-complete-multiple/auto-complete-multiple.component.js
import { __decorate, __extends } from "tslib";
import { Component, Input } from "@angular/core";
import { provideValueAccessor } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
import { forkJoin } from 'rxjs';
var AutoCompleteMultipleComponent = /** @class */ (function (_super) {
    __extends(AutoCompleteMultipleComponent, _super);
    function AutoCompleteMultipleComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.virtualScroll = false;
        _this.itemSize = 30;
        return _this;
    }
    AutoCompleteMultipleComponent_1 = AutoCompleteMultipleComponent;
    AutoCompleteMultipleComponent.prototype.findAll = function (event) {
        var _this = this;
        var query = {};
        query[this.field] = event.query;
        this.service.findAll(query).subscribe(function (data) {
            _this.results = data;
        });
    };
    AutoCompleteMultipleComponent.prototype.outerToInner = function (value) {
        var _this = this;
        if (value && value.length > 0 && !value[0][this.field]) {
            var obs_1 = [];
            value.forEach(function (element) {
                if (element && element.id) {
                    obs_1.push(_this.service.find(element.id));
                }
            });
            forkJoin(obs_1).subscribe(function (data) {
                _this.formControl.setValue(data);
            });
        }
        return value;
    };
    var AutoCompleteMultipleComponent_1;
    __decorate([
        Input()
    ], AutoCompleteMultipleComponent.prototype, "field", void 0);
    __decorate([
        Input()
    ], AutoCompleteMultipleComponent.prototype, "service", void 0);
    __decorate([
        Input()
    ], AutoCompleteMultipleComponent.prototype, "virtualScroll", void 0);
    __decorate([
        Input()
    ], AutoCompleteMultipleComponent.prototype, "itemSize", void 0);
    AutoCompleteMultipleComponent = AutoCompleteMultipleComponent_1 = __decorate([
        Component({
            selector: "[nAutoCompleteMultiple]",
            template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-autoCompleteCustom #input\n        [formControl]=\"formControl\"\n        [suggestions]=\"results\"\n        (completeMethod)=\"findAll($event)\"\n        [field]=\"field\"\n        [virtualScroll]=\"virtualScroll\"\n        [itemSize]=\"itemSize\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onSelect)=\"onSelectEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        [multiple]=\"true\">\n    </p-autoCompleteCustom>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
            providers: [provideValueAccessor(AutoCompleteMultipleComponent_1)]
        })
    ], AutoCompleteMultipleComponent);
    return AutoCompleteMultipleComponent;
}(FieldComponent));
export { AutoCompleteMultipleComponent };
