// @nuvem/angular-base/esm5/lib/clipboard/clipboard.directive.js
import { __decorate } from "tslib";
import { Directive, Input, Output, EventEmitter, HostListener } from '@angular/core';
import { ClipboardService } from './clipboard.service';
import * as ɵngcc0 from '@angular/core';
var ClipboardDirective = /** @class */ (function () {
    /**
     * constructor method
     * @param {ClipboardService} service
     * @constructor
     */
    function ClipboardDirective(service) {
        this.service = service;
        /**
         * onClipboard variable
         * @type {EventEmitter<boolean>}
         */
        this.onClipboard = new EventEmitter();
    }
    /**
     * copy method
     * @event click
     * @returns void
     */
    ClipboardDirective.prototype.copy = function () {
        try {
            this.service.copy(this.clipboard);
            this.onClipboard.emit(true);
        }
        catch (e) {
            this.onClipboard.emit(false);
            throw e;
        }
    };
    ClipboardDirective.ctorParameters = function () { return [
        { type: ClipboardService }
    ]; };
    __decorate([
        Input()
    ], ClipboardDirective.prototype, "clipboard", void 0);
    __decorate([
        Output()
    ], ClipboardDirective.prototype, "onClipboard", void 0);
    __decorate([
        HostListener('click')
    ], ClipboardDirective.prototype, "copy", null);
ClipboardDirective.ɵfac = function ClipboardDirective_Factory(t) { return new (t || ClipboardDirective)(ɵngcc0.ɵɵdirectiveInject(ClipboardService)); };
ClipboardDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClipboardDirective, selectors: [["", "clipboard", ""]], hostBindings: function ClipboardDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClipboardDirective_click_HostBindingHandler() { return ctx.copy(); });
    } }, inputs: { clipboard: "clipboard" }, outputs: { onClipboard: "onClipboard" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClipboardDirective, [{
        type: Directive,
        args: [{
                selector: '[clipboard]'
            }]
    }], function () { return [{ type: ClipboardService }]; }, { onClipboard: [{
            type: Output
        }], copy: [{
            type: HostListener,
            args: ['click']
        }], clipboard: [{
            type: Input
        }] }); })();
    return ClipboardDirective;
}());
export { ClipboardDirective };
