// @nuvem/angular-base/esm5/lib/directives/directives.module.js
import { __decorate } from "tslib";
import { UnmaskDirective } from './unmask.directive';
import { OnlyNumbersDirective } from './only-numbers.directive';
import { NgModule } from '@angular/core';
/**
 * DIRECTIVES constant
 * @constant
 */
import * as ɵngcc0 from '@angular/core';
var DIRECTIVES = [
    OnlyNumbersDirective,
    UnmaskDirective
];
/**
 * Expõe directivas para reuso em outras palicações
 * @module
 */
var DirectivesModule = /** @class */ (function () {
    function DirectivesModule() {
    }
DirectivesModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DirectivesModule });
DirectivesModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DirectivesModule_Factory(t) { return new (t || DirectivesModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DirectivesModule, { declarations: [OnlyNumbersDirective,
        UnmaskDirective], exports: [OnlyNumbersDirective,
        UnmaskDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DirectivesModule, [{
        type: NgModule,
        args: [{
                declarations: DIRECTIVES,
                exports: DIRECTIVES
            }]
    }], function () { return []; }, null); })();
    return DirectivesModule;
}());
export { DirectivesModule };
