// @nuvem/angular-base/esm5/lib/accessibility/accessibility.service.js
import { __decorate } from "tslib";
import { Injectable, Renderer2, RendererFactory2 } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * AccessbilityService Serviço de acessibilidade, este serviço prove recursos de contraste
 * de tela e controle do tamanho da fonte apresentadas na tela.
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var AccessbilityService = /** @class */ (function () {
    function AccessbilityService(rendererFactory) {
        /**
         * Propriedade que define tamanho da fonte
         * @type {number}
         */
        this.fontSize = 1.0;
        this.renderer = rendererFactory.createRenderer(null, null);
    }
    /**
     * Medoto para habilitar contraste
     * @public
     * @returns void
     */
    AccessbilityService.prototype.enableHighContrast = function () {
        this.renderer.addClass(document.body, 'contraste');
    };
    /**
     * Metodo para desabilitar contraste
     * @public
     * @returns void
     */
    AccessbilityService.prototype.disableHighContrast = function () {
        this.renderer.removeClass(document.body, 'contraste');
    };
    /**
     * Metodo utilizado para aumentar o tamnho da fonte apresentada no display
     * @public
     * @returns void
     */
    AccessbilityService.prototype.increaseFontSize = function () {
        this.fontSize += 0.1;
        this.renderer.setStyle(document.querySelector('div.layout-wrapper'), 'font-size', this.fontSize + 'em');
    };
    /**
     * Metodo utilizado para diminuir o tamnho da fonte apresentada no display
     * @public
     * @returns void
     */
    AccessbilityService.prototype.decreaseFontSize = function () {
        this.fontSize -= 0.1;
        this.renderer.setStyle(document.querySelector('div.layout-wrapper'), 'font-size', this.fontSize + 'em');
    };
    AccessbilityService.prototype.addAccessibilityIcons = function () {
        var elements = document.querySelector('.ui-button-icon-left, ui-button-icon-righ');
        if (elements) {
            this.renderer.setStyle(elements, 'aria-hidden', true);
        }
    };
    AccessbilityService.prototype.addAccessibilityMessages = function (msg) {
        var type = 'polite';
        if (msg.severity.toLowerCase() !== 'custom' && msg.severity.toLowerCase() !== 'success') {
            type = 'assertive';
        }
        this.renderer.setStyle(document.querySelector('p-toast, p-growl'), 'aria-live', type);
    };
    AccessbilityService.ctorParameters = function () { return [
        { type: RendererFactory2 }
    ]; };
    AccessbilityService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AccessbilityService_Factory() { return new AccessbilityService(i0.ɵɵinject(i0.RendererFactory2)); }, token: AccessbilityService, providedIn: "root" });
AccessbilityService.ɵfac = function AccessbilityService_Factory(t) { return new (t || AccessbilityService)(ɵngcc0.ɵɵinject(ɵngcc0.RendererFactory2)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AccessbilityService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc0.RendererFactory2 }]; }, null); })();
    return AccessbilityService;
}());
export { AccessbilityService };
