// @nuvem/angular-base/esm5/lib/accessibility/accessibility.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AccessbilityService } from './accessibility.service';
/**
 * AccessbilityModule expõe o servico AccessbilityService para o uso de outras aplicações
 * @module
 */
import * as ɵngcc0 from '@angular/core';
var AccessbilityModule = /** @class */ (function () {
    function AccessbilityModule() {
    }
AccessbilityModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AccessbilityModule });
AccessbilityModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AccessbilityModule_Factory(t) { return new (t || AccessbilityModule)(); }, providers: [AccessbilityService], imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AccessbilityModule, { imports: function () { return [CommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AccessbilityModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                providers: [AccessbilityService],
                exports: []
            }]
    }], function () { return []; }, null); })();
    return AccessbilityModule;
}());
export { AccessbilityModule };
