// ng-block-ui/__ivy_ngcc__/fesm5/ng-block-ui-router.js
import { Injectable, NgModule } from '@angular/core';
import { BlockUIService, BLOCKUI_DEFAULT, BlockUIModule } from 'ng-block-ui';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ng-block-ui';
var BlockUIPreventNavigation = /** @class */ (function () {
    function BlockUIPreventNavigation(blockUIService) {
        this.blockUIService = blockUIService;
    }
    /**
     * @return {?}
     */
    BlockUIPreventNavigation.prototype.canActivate = /**
     * @return {?}
     */
    function () {
        return !this.blockUIService.isActive(BLOCKUI_DEFAULT);
    };
    /**
     * @return {?}
     */
    BlockUIPreventNavigation.prototype.canActivateChild = /**
     * @return {?}
     */
    function () {
        return !this.blockUIService.isActive(BLOCKUI_DEFAULT);
    };
    /** @nocollapse */
    BlockUIPreventNavigation.ctorParameters = function () { return [
        { type: BlockUIService }
    ]; };
BlockUIPreventNavigation.ɵfac = function BlockUIPreventNavigation_Factory(t) { return new (t || BlockUIPreventNavigation)(ɵngcc0.ɵɵinject(ɵngcc1.BlockUIService)); };
BlockUIPreventNavigation.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: BlockUIPreventNavigation, factory: function (t) { return BlockUIPreventNavigation.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUIPreventNavigation, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.BlockUIService }]; }, null); })();
    return BlockUIPreventNavigation;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var BlockUIRouterModule = /** @class */ (function () {
    function BlockUIRouterModule() {
    }
    /**
     * @return {?}
     */
    BlockUIRouterModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: BlockUIRouterModule,
            providers: [
                BlockUIPreventNavigation
            ]
        };
    };
BlockUIRouterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BlockUIRouterModule });
BlockUIRouterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function BlockUIRouterModule_Factory(t) { return new (t || BlockUIRouterModule)(); }, imports: [[
            BlockUIModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BlockUIRouterModule, { imports: function () { return [BlockUIModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUIRouterModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BlockUIModule
                ]
            }]
    }], function () { return []; }, null); })();
    return BlockUIRouterModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { BlockUIRouterModule, BlockUIPreventNavigation };
