// @nuvem/angular-base/esm5/lib/pipes/cpf.pipe.js
import { __decorate } from "tslib";
import { Pipe } from '@angular/core';
/**
 *
 * Cria PIPE de CPF para valores.
 *
 * Usage:
 * {{ value | cpf }}
*/
import * as ɵngcc0 from '@angular/core';
var CpfPipe = /** @class */ (function () {
    function CpfPipe() {
    }
    CpfPipe.prototype.transform = function (value) {
        if (value === null || value.length !== 11) {
            return value;
        }
        else {
            value = value.replace(/[^0-9A-Za-z]/g, '');
            return value.substring(0, 3) + '.' +
                value.substring(3, 6) + '.' +
                value.substring(6, 9) + '-' +
                value.substring(9, 11);
        }
    };
CpfPipe.ɵfac = function CpfPipe_Factory(t) { return new (t || CpfPipe)(); };
CpfPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "cpf", type: CpfPipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CpfPipe, [{
        type: Pipe,
        args: [{
                name: 'cpf'
            }]
    }], function () { return []; }, null); })();
    return CpfPipe;
}());
export { CpfPipe };
