// @nuvem/angular-base/esm5/lib/pipes/pipe.limit.js
import { __decorate } from "tslib";
import { Pipe } from '@angular/core';
/**
 *
 * Cria PIPE para valores com restrição de valores a serem apresentados.
 *
 * Usage:
 * {{ value | limit:20 }}
 * {{ value | limit:20:"..." }}
*/
import * as ɵngcc0 from '@angular/core';
var LimitPipe = /** @class */ (function () {
    function LimitPipe() {
    }
    LimitPipe.prototype.transform = function (value, limit, ellipsis) {
        if (limit === void 0) { limit = 25; }
        if (ellipsis === void 0) { ellipsis = '...'; }
        return value.length > limit
            ? "" + value.substr(0, limit) + ellipsis
            : value;
    };
LimitPipe.ɵfac = function LimitPipe_Factory(t) { return new (t || LimitPipe)(); };
LimitPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "limit", type: LimitPipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LimitPipe, [{
        type: Pipe,
        args: [{
                name: 'limit'
            }]
    }], function () { return []; }, null); })();
    return LimitPipe;
}());
export { LimitPipe };
