// @nuvem/primeng-components/esm5/lib/menu/menu.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MenuComponent } from './menu.component';
import { SubMenuComponent } from './submenu.component';
import { RouterModule } from '@angular/router';
import { MenusService } from './menu.service';
import * as ɵngcc0 from '@angular/core';
var MenuModule = /** @class */ (function () {
    function MenuModule() {
    }
MenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MenuModule });
MenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MenuModule_Factory(t) { return new (t || MenuModule)(); }, providers: [
        MenusService
    ], imports: [[
            CommonModule,
            RouterModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MenuModule, { declarations: function () { return [MenuComponent,
        SubMenuComponent]; }, imports: function () { return [CommonModule,
        RouterModule]; }, exports: function () { return [MenuComponent,
        SubMenuComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MenuModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    MenuComponent,
                    SubMenuComponent
                ],
                imports: [
                    CommonModule,
                    RouterModule
                ],
                exports: [
                    MenuComponent,
                    SubMenuComponent
                ],
                providers: [
                    MenusService
                ]
            }]
    }], function () { return []; }, null); })();
    return MenuModule;
}());
export { MenuModule };
