// @nuvem/primeng-components/esm5/lib/menu/submenu.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { MenusService } from './menu.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@angular/router';

var _c0 = ["submenu", ""];
function SubMenuComponent_ng_template_0_li_0_a_1_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "i", 13);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var child_r1040 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(child_r1040.icon);
} }
function SubMenuComponent_ng_template_0_li_0_a_1_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 14);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var child_r1040 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(child_r1040.badge);
} }
function SubMenuComponent_ng_template_0_li_0_a_1_i_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "i", 15);
    ɵngcc0.ɵɵtext(1, "keyboard_arrow_down");
    ɵngcc0.ɵɵelementEnd();
} }
function SubMenuComponent_ng_template_0_li_0_a_1_Template(rf, ctx) { if (rf & 1) {
    var _r1052 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 9);
    ɵngcc0.ɵɵlistener("click", function SubMenuComponent_ng_template_0_li_0_a_1_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r1052); var ctx_r1053 = ɵngcc0.ɵɵnextContext(2); var child_r1040 = ctx_r1053.$implicit; var i_r1041 = ctx_r1053.index; var ctx_r1051 = ɵngcc0.ɵɵnextContext(); return ctx_r1051.itemClick($event, child_r1040, i_r1041); })("mouseenter", function SubMenuComponent_ng_template_0_li_0_a_1_Template_a_mouseenter_0_listener() { ɵngcc0.ɵɵrestoreView(_r1052); var i_r1041 = ɵngcc0.ɵɵnextContext(2).index; var ctx_r1054 = ɵngcc0.ɵɵnextContext(); return ctx_r1054.onMouseEnter(i_r1041); });
    ɵngcc0.ɵɵtemplate(1, SubMenuComponent_ng_template_0_li_0_a_1_i_1_Template, 2, 1, "i", 10);
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, SubMenuComponent_ng_template_0_li_0_a_1_span_4_Template, 2, 1, "span", 11);
    ɵngcc0.ɵɵtemplate(5, SubMenuComponent_ng_template_0_li_0_a_1_i_5_Template, 2, 0, "i", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var child_r1040 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var ctx_r1043 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("href", child_r1040.url || "#", ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵattribute("tabindex", !ctx_r1043.visible ? "-1" : null)("target", child_r1040.target);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(child_r1040.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.badge);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.items);
} }
function SubMenuComponent_ng_template_0_li_0_a_2_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "i", 13);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var child_r1040 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(child_r1040.icon);
} }
function SubMenuComponent_ng_template_0_li_0_a_2_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 14);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var child_r1040 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(child_r1040.badge);
} }
function SubMenuComponent_ng_template_0_li_0_a_2_i_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "i", 15);
    ɵngcc0.ɵɵtext(1, "keyboard_arrow_down");
    ɵngcc0.ɵɵelementEnd();
} }
var _c1 = function () { return { exact: true }; };
function SubMenuComponent_ng_template_0_li_0_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r1063 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 16);
    ɵngcc0.ɵɵlistener("click", function SubMenuComponent_ng_template_0_li_0_a_2_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r1063); var ctx_r1064 = ɵngcc0.ɵɵnextContext(2); var child_r1040 = ctx_r1064.$implicit; var i_r1041 = ctx_r1064.index; var ctx_r1062 = ɵngcc0.ɵɵnextContext(); return ctx_r1062.itemClick($event, child_r1040, i_r1041); })("mouseenter", function SubMenuComponent_ng_template_0_li_0_a_2_Template_a_mouseenter_0_listener() { ɵngcc0.ɵɵrestoreView(_r1063); var i_r1041 = ɵngcc0.ɵɵnextContext(2).index; var ctx_r1065 = ɵngcc0.ɵɵnextContext(); return ctx_r1065.onMouseEnter(i_r1041); });
    ɵngcc0.ɵɵtemplate(1, SubMenuComponent_ng_template_0_li_0_a_2_i_1_Template, 2, 1, "i", 10);
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, SubMenuComponent_ng_template_0_li_0_a_2_span_4_Template, 2, 1, "span", 11);
    ɵngcc0.ɵɵtemplate(5, SubMenuComponent_ng_template_0_li_0_a_2_i_5_Template, 2, 0, "i", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var child_r1040 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var ctx_r1044 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("routerLink", child_r1040.routerLink)("routerLinkActiveOptions", ɵngcc0.ɵɵpureFunction0(8, _c1));
    ɵngcc0.ɵɵattribute("tabindex", !ctx_r1044.visible ? "-1" : null)("target", child_r1040.target);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(child_r1040.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.badge);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.items);
} }
function SubMenuComponent_ng_template_0_li_0_ul_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "ul", 17);
} if (rf & 2) {
    var ctx_r1068 = ɵngcc0.ɵɵnextContext(2);
    var child_r1040 = ctx_r1068.$implicit;
    var i_r1041 = ctx_r1068.index;
    var ctx_r1045 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("item", child_r1040)("visible", ctx_r1045.isActive(i_r1041))("reset", ctx_r1045.reset)("parentActive", ctx_r1045.isActive(i_r1041))("@children", (ctx_r1045.menuService.isSlim() || ctx_r1045.menuService.isHorizontal()) && ctx_r1045.root ? ctx_r1045.isActive(i_r1041) ? "visible" : "hidden" : ctx_r1045.isActive(i_r1041) ? "visibleAnimated" : "hiddenAnimated");
} }
var _c2 = function (a0) { return { "active-menuitem": a0 }; };
function SubMenuComponent_ng_template_0_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 2);
    ɵngcc0.ɵɵtemplate(1, SubMenuComponent_ng_template_0_li_0_a_1_Template, 6, 7, "a", 3);
    ɵngcc0.ɵɵtemplate(2, SubMenuComponent_ng_template_0_li_0_a_2_Template, 6, 9, "a", 4);
    ɵngcc0.ɵɵelementStart(3, "div", 5);
    ɵngcc0.ɵɵelement(4, "div", 6);
    ɵngcc0.ɵɵelementStart(5, "div", 7);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, SubMenuComponent_ng_template_0_li_0_ul_7_Template, 1, 5, "ul", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1069 = ɵngcc0.ɵɵnextContext();
    var child_r1040 = ctx_r1069.$implicit;
    var i_r1041 = ctx_r1069.index;
    var ctx_r1042 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMap(child_r1040.badgeStyleClass);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(7, _c2, ctx_r1042.isActive(i_r1041)));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !child_r1040.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.routerLink);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(child_r1040.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.items);
} }
function SubMenuComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, SubMenuComponent_ng_template_0_li_0_Template, 8, 9, "li", 1);
} if (rf & 2) {
    var child_r1040 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngIf", child_r1040.visible === false ? false : true);
} }
var SubMenuComponent = /** @class */ (function () {
    function SubMenuComponent(menuService) {
        this.menuService = menuService;
    }
    /**
     * Alterar o menu root entre aberto e fechado
     */
    SubMenuComponent.prototype.rootMenuToggle = function () {
        if (this.root) {
            this.menuService.menuHoverActive = !this.menuService.menuHoverActive;
        }
    };
    /**
     * Activa o menu por index e desativa o item ativo
     * @param index
     */
    SubMenuComponent.prototype.activeMenuByIndex = function (index) {
        this.activeIndex = (this.activeIndex === index) ? null : index;
        return this.activeIndex;
    };
    /**
     * Verifica se foi informado um comando para o item de menu e executa caso necessário
     * @param item menu
     */
    SubMenuComponent.prototype.execCommand = function (event, item) {
        if (item.command) {
            item.command({ 'originalEvent': event, 'item': item });
            return true;
        }
        return false;
    };
    /**
     * Esconte o menu caso este não possua itens
     * @param item menu
     */
    SubMenuComponent.prototype.hideMenu = function (item) {
        if (!item.items) {
            if (this.menuService.isHorizontal() || this.menuService.isSlim()) {
                this.menuService.resetMenu = true;
            }
            else {
                this.menuService.resetMenu = false;
            }
            this.menuService.overlayMenuActive = false;
            this.menuService.staticMenuMobileActive = false;
            this.menuService.menuHoverActive = !this.menuService.menuHoverActive;
            return true;
        }
        return false;
    };
    SubMenuComponent.prototype.itemClick = function (event, item, index) {
        this.rootMenuToggle();
        if (item.disabled) {
            event.preventDefault();
            return false;
        }
        this.activeMenuByIndex(index);
        this.execCommand(event, item);
        if (item.items || (!item.url && !item.routerLink)) {
            event.preventDefault();
        }
        this.hideMenu(item);
        return true;
    };
    SubMenuComponent.prototype.onMouseEnter = function (index) {
        if (this.root && this.menuService.menuHoverActive && (this.menuService.isHorizontal() || this.menuService.isSlim())) {
            this.activeIndex = index;
        }
    };
    SubMenuComponent.prototype.isActive = function (index) {
        return this.activeIndex === index;
    };
    Object.defineProperty(SubMenuComponent.prototype, "reset", {
        get: function () {
            return this._reset;
        },
        set: function (val) {
            this._reset = val;
            if (this._reset && (this.menuService.isHorizontal() || this.menuService.isSlim())) {
                this.activeIndex = null;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubMenuComponent.prototype, "parentActive", {
        get: function () {
            return this._parentActive;
        },
        set: function (val) {
            this._parentActive = val;
            if (!this._parentActive) {
                this.activeIndex = null;
            }
        },
        enumerable: true,
        configurable: true
    });
    SubMenuComponent.ctorParameters = function () { return [
        { type: MenusService }
    ]; };
    __decorate([
        Input()
    ], SubMenuComponent.prototype, "item", void 0);
    __decorate([
        Input()
    ], SubMenuComponent.prototype, "root", void 0);
    __decorate([
        Input()
    ], SubMenuComponent.prototype, "visible", void 0);
    __decorate([
        Input()
    ], SubMenuComponent.prototype, "reset", null);
    __decorate([
        Input()
    ], SubMenuComponent.prototype, "parentActive", null);
SubMenuComponent.ɵfac = function SubMenuComponent_Factory(t) { return new (t || SubMenuComponent)(ɵngcc0.ɵɵdirectiveInject(MenusService)); };
SubMenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: SubMenuComponent, selectors: [["", "submenu", ""]], inputs: { reset: "reset", parentActive: "parentActive", item: "item", root: "root", visible: "visible" }, attrs: _c0, decls: 1, vars: 1, consts: [["ngFor", "", 3, "ngForOf"], [3, "ngClass", "class", 4, "ngIf"], [3, "ngClass"], ["class", "ripplelink", 3, "href", "click", "mouseenter", 4, "ngIf"], ["class", "ripplelink", "routerLinkActive", "active-menuitem-routerlink", 3, "routerLink", "routerLinkActiveOptions", "click", "mouseenter", 4, "ngIf"], [1, "layout-menu-tooltip"], [1, "layout-menu-tooltip-arrow"], [1, "layout-menu-tooltip-text"], ["submenu", "", 3, "item", "visible", "reset", "parentActive", 4, "ngIf"], [1, "ripplelink", 3, "href", "click", "mouseenter"], ["class", "material-icons", 4, "ngIf"], ["class", "menuitem-badge", 4, "ngIf"], ["class", "material-icons submenu-icon", 4, "ngIf"], [1, "material-icons"], [1, "menuitem-badge"], [1, "material-icons", "submenu-icon"], ["routerLinkActive", "active-menuitem-routerlink", 1, "ripplelink", 3, "routerLink", "routerLinkActiveOptions", "click", "mouseenter"], ["submenu", "", 3, "item", "visible", "reset", "parentActive"]], template: function SubMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, SubMenuComponent_ng_template_0_Template, 1, 1, "ng-template", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ctx.root ? ctx.item : ctx.item.items);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc2.RouterLinkWithHref, ɵngcc2.RouterLinkActive, SubMenuComponent], encapsulation: 2, data: { animation: [
            trigger('children', [
                state('hiddenAnimated', style({
                    height: '0px'
                })),
                state('visibleAnimated', style({
                    height: '*'
                })),
                state('visible', style({
                    height: '*',
                    'z-index': 100
                })),
                state('hidden', style({
                    height: '0px',
                    'z-index': '*'
                })),
                transition('visibleAnimated => hiddenAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)')),
                transition('hiddenAnimated => visibleAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)'))
            ])
        ] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SubMenuComponent, [{
        type: Component,
        args: [{
                selector: '[submenu]',
                template: "<ng-template ngFor let-child let-i=\"index\" [ngForOf]=\"(root ? item : item.items)\">\n    <li [ngClass]=\"{'active-menuitem': isActive(i)}\"\n        [class]=\"child.badgeStyleClass\"\n        *ngIf=\"child.visible === false ? false : true\">\n        <a [href]=\"child.url||'#'\" (click)=\"itemClick($event,child,i)\" (mouseenter)=\"onMouseEnter(i)\"\n           class=\"ripplelink\" *ngIf=\"!child.routerLink\"\n            [attr.tabindex]=\"!visible ? '-1' : null\" [attr.target]=\"child.target\">\n            <i *ngIf=\"child.icon\" class=\"material-icons\">{{child.icon}}</i>\n            <span>{{child.label}}</span>\n            <span class=\"menuitem-badge\" *ngIf=\"child.badge\">{{child.badge}}</span>\n            <i class=\"material-icons submenu-icon\" *ngIf=\"child.items\">keyboard_arrow_down</i>\n        </a>\n\n        <a (click)=\"itemClick($event,child,i)\"\n            (mouseenter)=\"onMouseEnter(i)\"\n            class=\"ripplelink\"\n            *ngIf=\"child.routerLink\"\n            [routerLink]=\"child.routerLink\"\n            routerLinkActive=\"active-menuitem-routerlink\"\n            [routerLinkActiveOptions]=\"{exact: true}\" [attr.tabindex]=\"!visible ? '-1' : null\" [attr.target]=\"child.target\">\n            <i *ngIf=\"child.icon\" class=\"material-icons\">{{child.icon}}</i>\n            <span>{{child.label}}</span>\n            <span class=\"menuitem-badge\" *ngIf=\"child.badge\">{{child.badge}}</span>\n            <i class=\"material-icons submenu-icon\" *ngIf=\"child.items\">keyboard_arrow_down</i>\n        </a>\n        <div class=\"layout-menu-tooltip\">\n            <div class=\"layout-menu-tooltip-arrow\"></div>\n            <div class=\"layout-menu-tooltip-text\">{{child.label}}</div>\n        </div>\n        <ul submenu\n            [item]=\"child\"\n            *ngIf=\"child.items\"\n            [visible]=\"isActive(i)\"\n            [reset]=\"reset\"\n            [parentActive]=\"isActive(i)\"\n            [@children]=\"(menuService.isSlim() || menuService.isHorizontal()) && root ? (isActive(i) ?\n            'visible' : 'hidden') : (isActive(i) ? 'visibleAnimated' : 'hiddenAnimated')\">\n        </ul>\n    </li>\n</ng-template>\n",
                animations: [
                    trigger('children', [
                        state('hiddenAnimated', style({
                            height: '0px'
                        })),
                        state('visibleAnimated', style({
                            height: '*'
                        })),
                        state('visible', style({
                            height: '*',
                            'z-index': 100
                        })),
                        state('hidden', style({
                            height: '0px',
                            'z-index': '*'
                        })),
                        transition('visibleAnimated => hiddenAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)')),
                        transition('hiddenAnimated => visibleAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)'))
                    ])
                ]
            }]
    }], function () { return [{ type: MenusService }]; }, { reset: [{
            type: Input
        }], parentActive: [{
            type: Input
        }], item: [{
            type: Input
        }], root: [{
            type: Input
        }], visible: [{
            type: Input
        }] }); })();
    return SubMenuComponent;
}());
export { SubMenuComponent };
