// @nuvem/primeng-components/esm2015/lib/menu/menu.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { MenusService } from './menu.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './submenu.component';
let MenuComponent = class MenuComponent {
    constructor(menuService) {
        this.menuService = menuService;
    }
};
MenuComponent.ɵfac = function MenuComponent_Factory(t) { return new (t || MenuComponent)(ɵngcc0.ɵɵdirectiveInject(MenusService)); };
MenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MenuComponent, selectors: [["app-menu"]], decls: 1, vars: 2, consts: [["submenu", "", "root", "true", "visible", "true", "parentActive", "true", 1, "ultima-menu", "ultima-main-menu", "clearfix", 3, "item", "reset"]], template: function MenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "ul", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("item", ctx.menuService.itens)("reset", ctx.menuService.resetMenu);
    } }, directives: [ɵngcc1.SubMenuComponent], encapsulation: 2 });
MenuComponent.ctorParameters = () => [
    { type: MenusService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MenuComponent, [{
        type: Component,
        args: [{
                selector: 'app-menu',
                template: "<ul submenu\n    [item]=\"menuService.itens\"\n    root=\"true\"\n    class=\"ultima-menu ultima-main-menu clearfix\"\n    [reset]=\"menuService.resetMenu\"\n    visible=\"true\"\n    parentActive=\"true\">\n</ul>\n"
            }]
    }], function () { return [{ type: MenusService }]; }, null); })();
export { MenuComponent };
