// @nuvem/angular-base/esm5/lib/error/services/error-stack-provider.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { ErrorStackService } from './error-stack.service';
import * as i0 from "@angular/core";
import * as i1 from "./error-stack.service";
/**
 * Class ErrorStackProviderService
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var ErrorStackProviderService = /** @class */ (function () {
    /**
     * Metodo construtor responsável por injetar o serviso de erros
     * @param {ErrorStackService} errorStackService
     */
    function ErrorStackProviderService(errorStackService) {
        this.errorStackService = errorStackService;
    }
    /**
     * Metodo responsável pela identificação de erros
     * @param {Error} error
     * @returns Boolean
     */
    ErrorStackProviderService.prototype.shouldHandle = function (error) {
        return error instanceof HttpErrorResponse;
    };
    /**
     * Metodo pela adição de erros
     * @param {HttpErrorResponse} error
     * @returns void
     */
    ErrorStackProviderService.prototype.handle = function (error) {
        this.errorStackService.create(error);
    };
    ErrorStackProviderService.ctorParameters = function () { return [
        { type: ErrorStackService }
    ]; };
    ErrorStackProviderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ErrorStackProviderService_Factory() { return new ErrorStackProviderService(i0.ɵɵinject(i1.ErrorStackService)); }, token: ErrorStackProviderService, providedIn: "root" });
ErrorStackProviderService.ɵfac = function ErrorStackProviderService_Factory(t) { return new (t || ErrorStackProviderService)(ɵngcc0.ɵɵinject(ErrorStackService)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ErrorStackProviderService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ErrorStackService }]; }, null); })();
    return ErrorStackProviderService;
}());
export { ErrorStackProviderService };
