// @nuvem/angular-base/esm5/lib/error/services/notification.service.js
import { __decorate, __extends } from "tslib";
import { NotificationProvider } from '../providers/notification.provider';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Class NotificationService
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var NotificationService = /** @class */ (function (_super) {
    __extends(NotificationService, _super);
    function NotificationService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Metodo responsável po adicionar mensagem de erro
     * @param {string} message
     * @param {string} title
     * @returns void
     */
    NotificationService.prototype.addErrorMessage = function (message, title) {
        console.error(title + ' ' + message);
    };
    /**
     * Metodo responsável por adicionar o detalhamento do erro no console
     * @param {ApplicationProblemType} problem
     * @returns void
     */
    NotificationService.prototype.addErrorProblem = function (problem) {
        console.error(problem.title + ' ' + problem.detail);
    };
    NotificationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NotificationService_Factory() { return new NotificationService(); }, token: NotificationService, providedIn: "root" });
NotificationService.ɵfac = function NotificationService_Factory(t) { return ɵNotificationService_BaseFactory(t || NotificationService); };
var ɵNotificationService_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(NotificationService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NotificationService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
    return NotificationService;
}(NotificationProvider));
export { NotificationService };
