// primeng/esm5/tabmenu/tabmenu.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@angular/router';

function TabMenu_li_2_a_1_ng_container_1_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 11);
} if (rf & 2) {
    var item_r682 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngClass", item_r682.icon);
} }
function TabMenu_li_2_a_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, TabMenu_li_2_a_1_ng_container_1_span_1_Template, 1, 1, "span", 9);
    ɵngcc0.ɵɵelementStart(2, "span", 10);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r682 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r682.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r682.label);
} }
function TabMenu_li_2_a_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
var _c0 = function (a0, a1) { return { $implicit: a0, index: a1 }; };
function TabMenu_li_2_a_1_Template(rf, ctx) { if (rf & 1) {
    var _r692 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 6);
    ɵngcc0.ɵɵlistener("click", function TabMenu_li_2_a_1_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r692); var item_r682 = ɵngcc0.ɵɵnextContext().$implicit; var ctx_r691 = ɵngcc0.ɵɵnextContext(); return ctx_r691.itemClick($event, item_r682); });
    ɵngcc0.ɵɵtemplate(1, TabMenu_li_2_a_1_ng_container_1_Template, 4, 2, "ng-container", 7);
    ɵngcc0.ɵɵtemplate(2, TabMenu_li_2_a_1_ng_container_2_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r694 = ɵngcc0.ɵɵnextContext();
    var item_r682 = ctx_r694.$implicit;
    var i_r683 = ctx_r694.index;
    var ctx_r684 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("href", item_r682.url, ɵngcc0.ɵɵsanitizeUrl)("tabindex", item_r682.disabled ? null : "0")("target", item_r682.target)("title", item_r682.title)("id", item_r682.id);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r684.itemTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r684.itemTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(8, _c0, item_r682, i_r683));
} }
function TabMenu_li_2_a_2_ng_container_1_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 11);
} if (rf & 2) {
    var item_r682 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngClass", item_r682.icon);
} }
function TabMenu_li_2_a_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, TabMenu_li_2_a_2_ng_container_1_span_1_Template, 1, 1, "span", 9);
    ɵngcc0.ɵɵelementStart(2, "span", 10);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r682 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r682.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r682.label);
} }
function TabMenu_li_2_a_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function TabMenu_li_2_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r701 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 12);
    ɵngcc0.ɵɵlistener("click", function TabMenu_li_2_a_2_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r701); var item_r682 = ɵngcc0.ɵɵnextContext().$implicit; var ctx_r700 = ɵngcc0.ɵɵnextContext(); return ctx_r700.itemClick($event, item_r682); });
    ɵngcc0.ɵɵtemplate(1, TabMenu_li_2_a_2_ng_container_1_Template, 4, 2, "ng-container", 7);
    ɵngcc0.ɵɵtemplate(2, TabMenu_li_2_a_2_ng_container_2_Template, 1, 0, "ng-container", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r703 = ɵngcc0.ɵɵnextContext();
    var item_r682 = ctx_r703.$implicit;
    var i_r683 = ctx_r703.index;
    var ctx_r685 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("routerLink", item_r682.routerLink)("queryParams", item_r682.queryParams)("fragment", item_r682.fragment)("queryParamsHandling", item_r682.queryParamsHandling)("preserveFragment", item_r682.preserveFragment)("skipLocationChange", item_r682.skipLocationChange)("replaceUrl", item_r682.replaceUrl)("state", item_r682.state);
    ɵngcc0.ɵɵattribute("tabindex", item_r682.disabled ? null : "0")("target", item_r682.target)("title", item_r682.title)("id", item_r682.id);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r685.itemTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r685.itemTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(15, _c0, item_r682, i_r683));
} }
var _c1 = function (a1, a2, a3, a4) { return { "ui-tabmenuitem ui-state-default ui-corner-top": true, "ui-state-disabled": a1, "ui-tabmenuitem-hasicon": a2, "ui-state-active": a3, "ui-helper-hidden": a4 }; };
var _c2 = function () { return { exact: false }; };
function TabMenu_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 3);
    ɵngcc0.ɵɵtemplate(1, TabMenu_li_2_a_1_Template, 3, 11, "a", 4);
    ɵngcc0.ɵɵtemplate(2, TabMenu_li_2_a_2_Template, 3, 18, "a", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r682 = ctx.$implicit;
    var ctx_r681 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction4(7, _c1, item_r682.disabled, item_r682.icon, ctx_r681.activeItem == item_r682, item_r682.visible === false))("routerLinkActive", "ui-state-active")("routerLinkActiveOptions", item_r682.routerLinkActiveOptions || ɵngcc0.ɵɵpureFunction0(12, _c2));
    ɵngcc0.ɵɵattribute("aria-selected", ctx_r681.activeItem == item_r682)("aria-expanded", ctx_r681.activeItem == item_r682);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !item_r682.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r682.routerLink);
} }
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule, Component, Input, ContentChildren, ChangeDetectionStrategy } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PrimeTemplate, SharedModule } from 'primeng/api';
import { RouterModule } from '@angular/router';
var TabMenu = /** @class */ (function () {
    function TabMenu() {
    }
    TabMenu.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.templates.forEach(function (item) {
            switch (item.getType()) {
                case 'item':
                    _this.itemTemplate = item.template;
                    break;
                default:
                    _this.itemTemplate = item.template;
                    break;
            }
        });
    };
    TabMenu.prototype.itemClick = function (event, item) {
        if (item.disabled) {
            event.preventDefault();
            return;
        }
        if (item.command) {
            item.command({
                originalEvent: event,
                item: item
            });
        }
        this.activeItem = item;
    };
    __decorate([
        Input()
    ], TabMenu.prototype, "model", void 0);
    __decorate([
        Input()
    ], TabMenu.prototype, "activeItem", void 0);
    __decorate([
        Input()
    ], TabMenu.prototype, "popup", void 0);
    __decorate([
        Input()
    ], TabMenu.prototype, "style", void 0);
    __decorate([
        Input()
    ], TabMenu.prototype, "styleClass", void 0);
    __decorate([
        ContentChildren(PrimeTemplate)
    ], TabMenu.prototype, "templates", void 0);
TabMenu.ɵfac = function TabMenu_Factory(t) { return new (t || TabMenu)(); };
TabMenu.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: TabMenu, selectors: [["p-tabMenu"]], contentQueries: function TabMenu_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, PrimeTemplate, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templates = _t);
    } }, inputs: { activeItem: "activeItem", model: "model", popup: "popup", style: "style", styleClass: "styleClass" }, decls: 3, vars: 5, consts: [[3, "ngClass", "ngStyle"], ["role", "tablist", 1, "ui-tabmenu-nav", "ui-helper-reset", "ui-helper-clearfix", "ui-widget-header", "ui-corner-all"], ["role", "tab", 3, "ngClass", "routerLinkActive", "routerLinkActiveOptions", 4, "ngFor", "ngForOf"], ["role", "tab", 3, "ngClass", "routerLinkActive", "routerLinkActiveOptions"], ["class", "ui-menuitem-link ui-corner-all", "role", "presentation", 3, "click", 4, "ngIf"], ["role", "presentation", "class", "ui-menuitem-link ui-corner-all", 3, "routerLink", "queryParams", "fragment", "queryParamsHandling", "preserveFragment", "skipLocationChange", "replaceUrl", "state", "click", 4, "ngIf"], ["role", "presentation", 1, "ui-menuitem-link", "ui-corner-all", 3, "click"], [4, "ngIf"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], ["class", "ui-menuitem-icon ", 3, "ngClass", 4, "ngIf"], [1, "ui-menuitem-text"], [1, "ui-menuitem-icon", 3, "ngClass"], ["role", "presentation", 1, "ui-menuitem-link", "ui-corner-all", 3, "routerLink", "queryParams", "fragment", "queryParamsHandling", "preserveFragment", "skipLocationChange", "replaceUrl", "state", "click"]], template: function TabMenu_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "ul", 1);
        ɵngcc0.ɵɵtemplate(2, TabMenu_li_2_Template, 3, 13, "li", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.styleClass);
        ɵngcc0.ɵɵproperty("ngClass", "ui-tabmenu ui-widget ui-widget-content ui-corner-all")("ngStyle", ctx.style);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.model);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgStyle, ɵngcc1.NgForOf, ɵngcc2.RouterLinkActive, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc2.RouterLinkWithHref], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TabMenu, [{
        type: Component,
        args: [{
                selector: 'p-tabMenu',
                template: "\n        <div [ngClass]=\"'ui-tabmenu ui-widget ui-widget-content ui-corner-all'\" [ngStyle]=\"style\" [class]=\"styleClass\">\n            <ul class=\"ui-tabmenu-nav ui-helper-reset ui-helper-clearfix ui-widget-header ui-corner-all\" role=\"tablist\">\n                <li *ngFor=\"let item of model; let i = index\" role=\"tab\" [attr.aria-selected]=\"activeItem==item\" [attr.aria-expanded]=\"activeItem==item\"\n                    [ngClass]=\"{'ui-tabmenuitem ui-state-default ui-corner-top':true,'ui-state-disabled':item.disabled,\n                        'ui-tabmenuitem-hasicon':item.icon,'ui-state-active':activeItem==item,'ui-helper-hidden': item.visible === false}\"\n                        [routerLinkActive]=\"'ui-state-active'\" [routerLinkActiveOptions]=\"item.routerLinkActiveOptions||{exact:false}\">\n                    <a *ngIf=\"!item.routerLink\" [attr.href]=\"item.url\" class=\"ui-menuitem-link ui-corner-all\" role=\"presentation\" (click)=\"itemClick($event,item)\" [attr.tabindex]=\"item.disabled ? null : '0'\"\n                        [attr.target]=\"item.target\" [attr.title]=\"item.title\" [attr.id]=\"item.id\">\n                        <ng-container *ngIf=\"!itemTemplate\">\n                            <span class=\"ui-menuitem-icon \" [ngClass]=\"item.icon\" *ngIf=\"item.icon\"></span>\n                            <span class=\"ui-menuitem-text\">{{item.label}}</span>\n                        </ng-container>\n                        <ng-container *ngTemplateOutlet=\"itemTemplate; context: {$implicit: item, index: i}\"></ng-container>\n                    </a>\n                    <a *ngIf=\"item.routerLink\" [routerLink]=\"item.routerLink\" [queryParams]=\"item.queryParams\" role=\"presentation\" class=\"ui-menuitem-link ui-corner-all\" (click)=\"itemClick($event,item)\" [attr.tabindex]=\"item.disabled ? null : '0'\"\n                        [attr.target]=\"item.target\" [attr.title]=\"item.title\" [attr.id]=\"item.id\"\n                        [fragment]=\"item.fragment\" [queryParamsHandling]=\"item.queryParamsHandling\" [preserveFragment]=\"item.preserveFragment\" [skipLocationChange]=\"item.skipLocationChange\" [replaceUrl]=\"item.replaceUrl\" [state]=\"item.state\">\n                        <ng-container *ngIf=\"!itemTemplate\">\n                            <span class=\"ui-menuitem-icon \" [ngClass]=\"item.icon\" *ngIf=\"item.icon\"></span>\n                            <span class=\"ui-menuitem-text\">{{item.label}}</span>\n                        </ng-container>\n                        <ng-container *ngTemplateOutlet=\"itemTemplate; context: {$implicit: item, index: i}\"></ng-container>\n                    </a>\n                </li>\n            </ul>\n        </div>\n    ",
                changeDetection: ChangeDetectionStrategy.Default
            }]
    }], function () { return []; }, { activeItem: [{
            type: Input
        }], model: [{
            type: Input
        }], popup: [{
            type: Input
        }], style: [{
            type: Input
        }], styleClass: [{
            type: Input
        }], templates: [{
            type: ContentChildren,
            args: [PrimeTemplate]
        }] }); })();
    return TabMenu;
}());
export { TabMenu };
var TabMenuModule = /** @class */ (function () {
    function TabMenuModule() {
    }
TabMenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: TabMenuModule });
TabMenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function TabMenuModule_Factory(t) { return new (t || TabMenuModule)(); }, imports: [[CommonModule, RouterModule, SharedModule],
        RouterModule, SharedModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TabMenuModule, { declarations: function () { return [TabMenu]; }, imports: function () { return [CommonModule, RouterModule, SharedModule]; }, exports: function () { return [TabMenu,
        RouterModule, SharedModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TabMenuModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, RouterModule, SharedModule],
                exports: [TabMenu, RouterModule, SharedModule],
                declarations: [TabMenu]
            }]
    }], function () { return []; }, null); })();
    return TabMenuModule;
}());
export { TabMenuModule };
