// @nuvem/angular-base/esm2015/lib/error/services/notification.service.js
import { __decorate } from "tslib";
import { NotificationProvider } from '../providers/notification.provider';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Class NotificationService
 * @class
 */
import * as ɵngcc0 from '@angular/core';
let NotificationService = class NotificationService extends NotificationProvider {
    /**
     * Metodo responsável po adicionar mensagem de erro
     * @param {string} message
     * @param {string} title
     * @returns void
     */
    addErrorMessage(message, title) {
        console.error(title + ' ' + message);
    }
    /**
     * Metodo responsável por adicionar o detalhamento do erro no console
     * @param {ApplicationProblemType} problem
     * @returns void
     */
    addErrorProblem(problem) {
        console.error(problem.title + ' ' + problem.detail);
    }
};
NotificationService.ɵfac = function NotificationService_Factory(t) { return ɵNotificationService_BaseFactory(t || NotificationService); };
NotificationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NotificationService_Factory() { return new NotificationService(); }, token: NotificationService, providedIn: "root" });
const ɵNotificationService_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(NotificationService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NotificationService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
export { NotificationService };
