// @nuvem/primeng-components/esm5/lib/error-stack/error-stack.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TableModule } from 'primeng/table';
import { ButtonModule, TooltipModule } from 'primeng';
import { ErrorStackComponent } from './error-stack.component';
import { ErrorProvider, ErrorStackService, ErrorStackProviderService, ClipboardModule } from '@nuvem/angular-base';
/**
 * Class ErrorStackModule
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var ErrorStackModule = /** @class */ (function () {
    function ErrorStackModule() {
    }
ErrorStackModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ErrorStackModule });
ErrorStackModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ErrorStackModule_Factory(t) { return new (t || ErrorStackModule)(); }, providers: [
        ErrorStackService,
        {
            provide: ErrorProvider,
            useClass: ErrorStackProviderService,
            multi: true
        }
    ], imports: [[
            CommonModule,
            TableModule,
            ButtonModule,
            TooltipModule,
            ClipboardModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ErrorStackModule, { declarations: function () { return [ErrorStackComponent]; }, imports: function () { return [CommonModule,
        TableModule,
        ButtonModule,
        TooltipModule,
        ClipboardModule]; }, exports: function () { return [ErrorStackComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ErrorStackModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    TableModule,
                    ButtonModule,
                    TooltipModule,
                    ClipboardModule
                ],
                declarations: [
                    ErrorStackComponent
                ],
                providers: [
                    ErrorStackService,
                    {
                        provide: ErrorProvider,
                        useClass: ErrorStackProviderService,
                        multi: true
                    }
                ],
                exports: [
                    ErrorStackComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return ErrorStackModule;
}());
export { ErrorStackModule };
