// @nuvem/primeng-components/esm5/lib/error-stack/error-stack.component.js
import { __decorate, __read, __spread } from "tslib";
import { Component, Input } from '@angular/core';
import { ErrorStackService, NamedErrorType } from '@nuvem/angular-base';
/**
 * Class ErrorStackComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@nuvem/angular-base';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'primeng/table';
import * as ɵngcc4 from 'primeng/api';
import * as ɵngcc5 from 'primeng';

function ErrorStackComponent_ng_template_0_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "tr");
    ɵngcc0.ɵɵelementStart(1, "th", 9);
    ɵngcc0.ɵɵtext(2, "Data");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "th", 9);
    ɵngcc0.ɵɵtext(4, "C\u00F3digo de identifica\u00E7\u00E3o");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "th");
    ɵngcc0.ɵɵtext(6, "Mensagem");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "th", 10);
    ɵngcc0.ɵɵtext(8, "A\u00E7\u00E3o");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function ErrorStackComponent_ng_template_0_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "tr");
    ɵngcc0.ɵɵelementStart(1, "td", 9);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "date");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "td", 9);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "td");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "td", 10);
    ɵngcc0.ɵɵelement(9, "button", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var error_r936 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind2(3, 4, error_r936.createdAt, "dd/MM/y H:mm:ss"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(error_r936.correlationId || "Indispon\u00EDvel");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(error_r936.message);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("clipboard", error_r936.fullMessage);
} }
function ErrorStackComponent_ng_template_0_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "tr");
    ɵngcc0.ɵɵelementStart(1, "td", 12);
    ɵngcc0.ɵɵtext(2, " Nenhum erro foi registrado at\u00E9 o momento. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function ErrorStackComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "div", 3);
    ɵngcc0.ɵɵelementStart(2, "div", 4);
    ɵngcc0.ɵɵelementStart(3, "h1");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p-table", 5);
    ɵngcc0.ɵɵtemplate(6, ErrorStackComponent_ng_template_0_ng_template_6_Template, 9, 0, "ng-template", 6);
    ɵngcc0.ɵɵtemplate(7, ErrorStackComponent_ng_template_0_ng_template_7_Template, 10, 7, "ng-template", 7);
    ɵngcc0.ɵɵtemplate(8, ErrorStackComponent_ng_template_0_ng_template_8_Template, 3, 0, "ng-template", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r931 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ctx_r931.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("value", ctx_r931.errors)("paginator", false)("rows", 5);
} }
function ErrorStackComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
var ErrorStackComponent = /** @class */ (function () {
    /**
     * constructor method
     * @constructor
     * @param {ErrorStackService} errorStack
     */
    function ErrorStackComponent(errorStack) {
        this.errorStack = errorStack;
        /**
         * title property
         * @type {string}
         */
        this.title = 'Error Stack';
        /**
         * errors property
         * @type {NamedErrorType}
         */
        this.errors = [];
    }
    /**
     * ngOnInit method
     * @returns void
     */
    ErrorStackComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.errorSubscription = this.errorStack.errors.subscribe(function (error) {
            _this.errors = __spread([error], _this.errors);
        });
    };
    /**
     * ngOnDestroy method
     * @returns void
     */
    ErrorStackComponent.prototype.ngOnDestroy = function () {
        this.errorSubscription.unsubscribe();
    };
    ErrorStackComponent.ctorParameters = function () { return [
        { type: ErrorStackService }
    ]; };
    __decorate([
        Input()
    ], ErrorStackComponent.prototype, "title", void 0);
    __decorate([
        Input()
    ], ErrorStackComponent.prototype, "template", void 0);
ErrorStackComponent.ɵfac = function ErrorStackComponent_Factory(t) { return new (t || ErrorStackComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ErrorStackService)); };
ErrorStackComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ErrorStackComponent, selectors: [["basis-error-stack"]], inputs: { title: "title", template: "template" }, decls: 3, vars: 1, consts: [["defaultTemplate", ""], [4, "ngTemplateOutlet"], [1, "ui-g"], [1, "ui-g-12"], [1, "card", "no-margin"], ["first", "0", 3, "value", "paginator", "rows"], ["pTemplate", "header"], ["pTemplate", "body"], ["pTemplate", "emptymessage"], [2, "width", "20%", "text-align", "center"], [2, "width", "10%", "text-align", "center"], ["pButton", "", "type", "button", "pTooltip", "Copiar erro para \u00E1rea de transfer\u00EAncia", "tooltipPosition", "left", "icon", "ui-icon-content-copy", 1, "ui-button-primary", 3, "clipboard"], ["colspan", "4"]], template: function ErrorStackComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, ErrorStackComponent_ng_template_0_Template, 9, 4, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(2, ErrorStackComponent_ng_container_2_Template, 1, 0, "ng-container", 1);
    } if (rf & 2) {
        var _r930 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.template ? ctx.template : _r930);
    } }, directives: [ɵngcc2.NgTemplateOutlet, ɵngcc3.Table, ɵngcc4.PrimeTemplate, ɵngcc5.ButtonDirective, ɵngcc5.Tooltip, ɵngcc1.ClipboardDirective], pipes: [ɵngcc2.DatePipe], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ErrorStackComponent, [{
        type: Component,
        args: [{
                selector: 'basis-error-stack',
                template: "<ng-template #defaultTemplate>\n    <div class=\"ui-g\">\n        <div class=\"ui-g-12\">\n            <div class=\"card no-margin\">\n                <h1>{{ title }}</h1>\n                <p-table [value]=\"errors\" [paginator]=\"false\" [rows]=\"5\" first=\"0\">\n                    <ng-template pTemplate=\"header\">\n                        <tr>\n                            <th style=\"width: 20%; text-align: center\">Data</th>\n                            <th style=\"width: 20%; text-align: center\">C\u00F3digo de identifica\u00E7\u00E3o</th>\n                            <th>Mensagem</th>\n                            <th style=\"width: 10%; text-align: center\">A\u00E7\u00E3o</th>\n                        </tr>\n                    </ng-template>\n                    <ng-template pTemplate=\"body\" let-error>\n                        <tr>\n                            <td style=\"width: 20%; text-align: center\">{{error.createdAt|date: 'dd/MM/y H:mm:ss'}}</td>\n                            <td style=\"width: 20%; text-align: center\">{{error.correlationId || 'Indispon\u00EDvel'}}</td>\n                            <td>{{error.message}}</td>\n                            <td style=\"width: 10%; text-align: center\">\n                                <button pButton type=\"button\" pTooltip=\"Copiar erro para \u00E1rea de transfer\u00EAncia\" tooltipPosition=\"left\" class=\"ui-button-primary\" icon=\"ui-icon-content-copy\" [clipboard]=\"error.fullMessage\"></button>\n                            </td>\n                        </tr>\n                    </ng-template>\n                    <ng-template pTemplate=\"emptymessage\" let-columns>\n                        <tr>\n                            <td colspan=\"4\">\n                                Nenhum erro foi registrado at\u00E9 o momento.\n                            </td>\n                        </tr>\n                    </ng-template>\n                </p-table>\n            </div>\n        </div>\n    </div>\n</ng-template>\n<ng-container *ngTemplateOutlet=\"template? template: defaultTemplate\"></ng-container>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ErrorStackService }]; }, { title: [{
            type: Input
        }], template: [{
            type: Input
        }] }); })();
    return ErrorStackComponent;
}());
export { ErrorStackComponent };
