// @nuvem/primeng-components/esm2015/lib/crud/crud-list.component.js
import { __decorate, __param } from "tslib";
import { Component, Inject, Input, Output, EventEmitter, ViewChild, ContentChildren } from "@angular/core";
import { CRUD_SERVICE } from "./crud-service.service";
import { Router, ActivatedRoute } from "@angular/router";
import { Column } from '../datatable/p-column.directive';
import { ConfirmationService } from 'primeng/api';
let CrudListComponent = class CrudListComponent {
    constructor(crudService, confirmationService, route, router) {
        this.crudService = crudService;
        this.confirmationService = confirmationService;
        this.route = route;
        this.router = router;
        this.action = "findAll";
        this.findLabel = "Pesquisar";
        this.disableSave = false;
        this.disableEdit = false;
        this.disableView = false;
        this.disableDelete = false;
        this.find = new EventEmitter();
    }
    ngOnInit() {
        if (this.service) {
            this.crudService = this.service;
        }
        if (!this.url) {
            this.url = this.crudService.uri;
        }
    }
    datatableClick(event) {
        if (event.selection) {
            switch (event.button) {
                case "edit":
                case "view":
                    this.router.navigate([`../${event.button}`, event.selection.id], { relativeTo: this.route });
                    break;
                case "delete":
                    this.confirmationService.confirm({
                        message: 'Você tem certeza que deseja excluir o registro?',
                        accept: () => this.crudService.delete(event.selection.id)
                            .subscribe(() => this.datatable.filter())
                    });
                    break;
            }
        }
    }
    findAll(event) {
        if (this.find.observers.length) {
            this.find.emit(this.form.value);
        }
        else {
            this.datatable.filterParams = {};
            Object.keys(this.form.value).forEach(element => {
                let value = this.form.value[element];
                if (value) {
                    this.datatable.filterParams[element] =
                        value instanceof Date ? JSON.stringify(value).replace(/"/g, "") : value;
                }
            });
            this.datatable.filter();
        }
    }
    ngAfterContentInit() {
        const datatableNgAfterContentInit = this.datatable.pDatatableComponent.ngAfterContentInit;
        this.datatable.pDatatableComponent.ngAfterContentInit = () => {
            this.datatable.pDatatableComponent.columns = this.columns.toArray();
            datatableNgAfterContentInit.call(this.datatable.pDatatableComponent);
        };
    }
    ngAfterViewInit() {
        for (let i in this.datatableExtraParams) {
            this.datatable[i] = this.datatableExtraParams[i];
        }
    }
};
CrudListComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CRUD_SERVICE,] }] },
    { type: ConfirmationService },
    { type: ActivatedRoute },
    { type: Router }
];
__decorate([
    Input()
], CrudListComponent.prototype, "form", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "action", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "findLabel", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "service", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "url", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "disableSave", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "disableEdit", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "disableView", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "disableDelete", void 0);
__decorate([
    Input()
], CrudListComponent.prototype, "datatableExtraParams", void 0);
__decorate([
    Output()
], CrudListComponent.prototype, "find", void 0);
__decorate([
    ViewChild('datatable', { static: true })
], CrudListComponent.prototype, "datatable", void 0);
__decorate([
    ContentChildren(Column)
], CrudListComponent.prototype, "columns", void 0);
CrudListComponent = __decorate([
    Component({
        selector: 'nCrudList',
        template: "<nForm [form]=\"form\" (save)=\"findAll($event)\" [saveLabel]=\"findLabel\">\n    <ng-content select=\"[nForm]\"></ng-content>\n    <div nButton class=\"ui-g-12 ui-lg-2 ui-md-4 ui-sm-12\">\n        <button pButton type=\"button\" [routerLink]=\"['../save']\" label=\"Novo\" *ngIf=\"!disableSave\"></button>\n    </div>\n</nForm>\n\n<basis-datatable #datatable [value]=\"result\" (buttonClick)=\"datatableClick($event)\" [url]=\"url\"\n    [disableEdit]=\"disableEdit\" [disableView]=\"disableView\" [disableDelete]=\"disableDelete\">\n    <ng-content select=\"p-column\"></ng-content>\n</basis-datatable>\n\n<p-confirmDialog header=\"Confirmar Exclus\u00E3o\" acceptLabel=\"Sim\" rejectLabel=\"N\u00E3o\" icon=\"pi pi-exclamation-triangle\"></p-confirmDialog>\n"
    }),
    __param(0, Inject(CRUD_SERVICE))
], CrudListComponent);
export { CrudListComponent };
