// @nuvem/primeng-components/esm5/lib/block-ui/block-ui.module.js
import { __decorate } from "tslib";
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { BlockUiInterceptor, BlockUiService } from '@nuvem/angular-base';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BlockUIModule, ProgressSpinnerModule } from 'primeng';
import { BlockUiComponent } from './block-ui.component';
/**
 * Class BlockUiModule
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var BlockUiModule = /** @class */ (function () {
    function BlockUiModule() {
    }
BlockUiModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BlockUiModule });
BlockUiModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function BlockUiModule_Factory(t) { return new (t || BlockUiModule)(); }, providers: [
        {
            provide: HTTP_INTERCEPTORS,
            useClass: BlockUiInterceptor,
            multi: true
        },
        BlockUiService
    ], imports: [[
            CommonModule,
            BlockUIModule,
            ProgressSpinnerModule,
        ],
        BlockUIModule,
        ProgressSpinnerModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BlockUiModule, { declarations: function () { return [BlockUiComponent]; }, imports: function () { return [CommonModule,
        BlockUIModule,
        ProgressSpinnerModule]; }, exports: function () { return [BlockUIModule,
        BlockUiComponent,
        ProgressSpinnerModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUiModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    BlockUIModule,
                    ProgressSpinnerModule,
                ],
                declarations: [
                    BlockUiComponent
                ],
                providers: [
                    {
                        provide: HTTP_INTERCEPTORS,
                        useClass: BlockUiInterceptor,
                        multi: true
                    },
                    BlockUiService
                ],
                exports: [
                    BlockUIModule,
                    BlockUiComponent,
                    ProgressSpinnerModule
                ]
            }]
    }], function () { return []; }, null); })();
    return BlockUiModule;
}());
export { BlockUiModule };
