// @nuvem/primeng-components/esm5/lib/block-ui/block-ui.component.js
import { __decorate } from "tslib";
import { BlockUiStatus, BlockUiService } from '@nuvem/angular-base';
import { Component } from '@angular/core';
/**
 * Class BlockUiComponent
 * @class
 */
var BlockUiComponent = /** @class */ (function () {
    /**
     * constructor method
     * @constructor
     * @param {BlockUiService} service
     */
    function BlockUiComponent(service) {
        this.service = service;
        /**
         * status property
         * @type {boolean}
         * @public
         */
        this.status = false;
    }
    /**
     * ngOnInit method
     * @returns void
     */
    BlockUiComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription = this.service.loaderStatus.subscribe(function (blockui) {
            _this.status = blockui.status;
        });
    };
    /**
     * ngOnDestroy methods
     * @returns void
     */
    BlockUiComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    /**
     * isBlocked method
     * @returns boolean
     */
    BlockUiComponent.prototype.isBlocked = function () {
        return this.status;
    };
    /**
     * show method
     * @returns false
     */
    BlockUiComponent.prototype.show = function () {
        this.status = true;
    };
    /**
     * hide methods
     * @returns void
     */
    BlockUiComponent.prototype.hide = function () {
        this.status = false;
    };
    BlockUiComponent.ctorParameters = function () { return [
        { type: BlockUiService }
    ]; };
    BlockUiComponent = __decorate([
        Component({
            selector: 'basis-block-ui',
            template: "<p-blockUI [blocked]=\"isBlocked()\">\n    <div class=\"center-screen\">\n        <p-progressSpinner></p-progressSpinner>\n    </div>\n</p-blockUI>\n",
            styles: [".center-screen{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}"]
        })
    ], BlockUiComponent);
    return BlockUiComponent;
}());
export { BlockUiComponent };
