// @nuvem/primeng-components/esm5/lib/block-ui/block-ui.component.js
import { __decorate } from "tslib";
import { BlockUiStatus, BlockUiService } from '@nuvem/angular-base';
import { Component } from '@angular/core';
/**
 * Class BlockUiComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@nuvem/angular-base';
import * as ɵngcc2 from 'primeng';
var BlockUiComponent = /** @class */ (function () {
    /**
     * constructor method
     * @constructor
     * @param {BlockUiService} service
     */
    function BlockUiComponent(service) {
        this.service = service;
        /**
         * status property
         * @type {boolean}
         * @public
         */
        this.status = false;
    }
    /**
     * ngOnInit method
     * @returns void
     */
    BlockUiComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription = this.service.loaderStatus.subscribe(function (blockui) {
            _this.status = blockui.status;
        });
    };
    /**
     * ngOnDestroy methods
     * @returns void
     */
    BlockUiComponent.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    /**
     * isBlocked method
     * @returns boolean
     */
    BlockUiComponent.prototype.isBlocked = function () {
        return this.status;
    };
    /**
     * show method
     * @returns false
     */
    BlockUiComponent.prototype.show = function () {
        this.status = true;
    };
    /**
     * hide methods
     * @returns void
     */
    BlockUiComponent.prototype.hide = function () {
        this.status = false;
    };
    BlockUiComponent.ctorParameters = function () { return [
        { type: BlockUiService }
    ]; };
BlockUiComponent.ɵfac = function BlockUiComponent_Factory(t) { return new (t || BlockUiComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BlockUiService)); };
BlockUiComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: BlockUiComponent, selectors: [["basis-block-ui"]], decls: 3, vars: 1, consts: [[3, "blocked"], [1, "center-screen"]], template: function BlockUiComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p-blockUI", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "p-progressSpinner");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("blocked", ctx.isBlocked());
    } }, directives: [ɵngcc2.BlockUI, ɵngcc2.ProgressSpinner], styles: [".center-screen[_ngcontent-%COMP%]{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUiComponent, [{
        type: Component,
        args: [{
                selector: 'basis-block-ui',
                template: "<p-blockUI [blocked]=\"isBlocked()\">\n    <div class=\"center-screen\">\n        <p-progressSpinner></p-progressSpinner>\n    </div>\n</p-blockUI>\n",
                styles: [".center-screen{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}"]
            }]
    }], function () { return [{ type: ɵngcc1.BlockUiService }]; }, null); })();
    return BlockUiComponent;
}());
export { BlockUiComponent };
