// @nuvem/primeng-components/esm5/lib/form-notification/form-notification.component.js
import { __decorate, __extends } from "tslib";
import { Component } from "@angular/core";
import { FormNotificationComponentSuperclass } from '@nuvem/angular-base';
import * as ɵngcc0 from '@angular/core';

var _c0 = ["form-notification", ""];
var _c1 = ["*"];
var FormNotificationComponent = /** @class */ (function (_super) {
    __extends(FormNotificationComponent, _super);
    function FormNotificationComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
FormNotificationComponent.ɵfac = function FormNotificationComponent_Factory(t) { return ɵFormNotificationComponent_BaseFactory(t || FormNotificationComponent); };
FormNotificationComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FormNotificationComponent, selectors: [["", "form-notification", ""]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c1, decls: 3, vars: 1, consts: [[1, "form-notification-msg", "ui-message", "ui-messages-error", "ui-corner-all"]], template: function FormNotificationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.message);
    } }, encapsulation: 2 });
var ɵFormNotificationComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(FormNotificationComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormNotificationComponent, [{
        type: Component,
        args: [{
                selector: '[form-notification]',
                template: "<ng-content></ng-content>\n<div class=\"form-notification-msg ui-message ui-messages-error ui-corner-all\" >{{message}}</div>\n"
            }]
    }], null, null); })();
    return FormNotificationComponent;
}(FormNotificationComponentSuperclass));
export { FormNotificationComponent };
