// @nuvem/primeng-components/esm2015/lib/datatable/filter-column/filter-column.module.js
import { __decorate } from "tslib";
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule } from 'primeng/button';
import { DropdownModule } from 'primeng/dropdown';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CheckboxModule, MultiSelectModule, InputTextModule, } from 'primeng';
import { FilterColumn } from './filter-column.component';
import * as ɵngcc0 from '@angular/core';
let FilterColumnModule = class FilterColumnModule {
};
FilterColumnModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FilterColumnModule });
FilterColumnModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FilterColumnModule_Factory(t) { return new (t || FilterColumnModule)(); }, imports: [[
            CommonModule,
            MultiSelectModule,
            ButtonModule,
            CheckboxModule,
            DropdownModule,
            FormsModule,
            ReactiveFormsModule,
            InputTextModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FilterColumnModule, { declarations: function () { return [FilterColumn]; }, imports: function () { return [CommonModule,
        MultiSelectModule,
        ButtonModule,
        CheckboxModule,
        DropdownModule,
        FormsModule,
        ReactiveFormsModule,
        InputTextModule]; }, exports: function () { return [FilterColumn]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FilterColumnModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MultiSelectModule,
                    ButtonModule,
                    CheckboxModule,
                    DropdownModule,
                    FormsModule,
                    ReactiveFormsModule,
                    InputTextModule
                ],
                declarations: [
                    FilterColumn
                ],
                exports: [
                    FilterColumn
                ],
                schemas: [CUSTOM_ELEMENTS_SCHEMA]
            }]
    }], null, null); })();
export { FilterColumnModule };
