// @nuvem/primeng-components/esm2015/lib/crud/components/check/check.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from '@nuvem/angular-base';
import * as ɵngcc3 from 'primeng/radiobutton';
import * as ɵngcc4 from '@angular/common';

const _c0 = ["nCheck", ""];
const _c1 = function (a0) { return { "ng-invalid ng-dirty": a0 }; };
const _c2 = ["*"];
var CheckComponent_1;
import { __decorate } from "tslib";
import { Component, Injector } from "@angular/core";
import { provideValueAccessor, FormNotificationService, FormNotification, FieldFormNotification } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
import { ControlContainer } from '@angular/forms';
let CheckComponent = CheckComponent_1 = class CheckComponent extends FieldComponent {
    /**
     * Método construtor responsável por carregar o serviço de validação
     * @param formNotificationService
     */
    constructor(controlContainer, injector, formNotificationService) {
        super(controlContainer, injector);
        this.formNotificationService = formNotificationService;
    }
    /**
     * Método Inicial para subscrição no endpoint responsável por validar o formulário
     */
    ngOnInit() {
        super.ngOnInit();
        this.subscription = this.formNotificationService
            .notifications(this.form, this.name)
            .subscribe((notification) => {
            if (notification instanceof FieldFormNotification) {
                this.message = notification.message;
            }
            else {
                this.message = null;
            }
        });
    }
    /**
     * Método responsável por realizar a saida das subscrições do endpont de validação
     */
    ngOnDestroy() {
        super.ngOnDestroy();
        this.subscription.unsubscribe();
    }
};
CheckComponent.ɵfac = function CheckComponent_Factory(t) { return new (t || CheckComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ControlContainer), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormNotificationService)); };
CheckComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CheckComponent, selectors: [["", "nCheck", ""]], features: [ɵngcc0.ɵɵProvidersFeature([provideValueAccessor(CheckComponent_1)]), ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c2, decls: 11, vars: 12, consts: [[1, "ui-g-12", "semborda"], [1, "ui-g-4", 3, "innerHTML"], [1, "ui-g-4"], ["id", "sim", "type", "radio", "pRadioButton", "", 3, "ngClass", "value", "name", "formControl"], ["id", "n\u00E3o", "type", "radio", "name", "correspondencia", "pRadioButton", "", 3, "ngClass", "value", "formControl"]], template: function CheckComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelement(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelement(3, "p-radioButton", 3);
        ɵngcc0.ɵɵelementStart(4, "label");
        ɵngcc0.ɵɵtext(5, "Sim");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 2);
        ɵngcc0.ɵɵelement(7, "p-radioButton", 4);
        ɵngcc0.ɵɵelementStart(8, "label");
        ɵngcc0.ɵɵtext(9, "N\u00E3o");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(10);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("innerHTML", ctx.label, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c1, ctx.message))("value", true)("name", ctx.name)("formControl", ctx.formControl);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(10, _c1, ctx.message))("value", false)("formControl", ctx.formControl);
    } }, directives: [ɵngcc3.RadioButton, ɵngcc4.NgClass, ɵngcc1.NgControlStatus, ɵngcc1.FormControlDirective], encapsulation: 2 });
CheckComponent.ctorParameters = () => [
    { type: ControlContainer },
    { type: Injector },
    { type: FormNotificationService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CheckComponent, [{
        type: Component,
        args: [{
                selector: "[nCheck]",
                template: "<div class=\"ui-g-12 semborda\">\n    <div class=\"ui-g-4\" [innerHTML]=\"label\"></div>\n    <div class=\"ui-g-4\">\n        <p-radioButton\n            [ngClass]=\"{ 'ng-invalid ng-dirty': message }\"\n            id=\"sim\"\n            type=\"radio\"\n            [value]=\"true\"\n            [name]=\"name\"\n            [formControl]=\"formControl\"\n            pRadioButton></p-radioButton>\n        <label>Sim</label>\n    </div>\n    <div class=\"ui-g-4\">\n        <p-radioButton\n            [ngClass]=\"{ 'ng-invalid ng-dirty': message }\"\n            id=\"n\u00E3o\"\n            type=\"radio\"\n            [value]=\"false\"\n            name=\"correspondencia\"\n            [formControl]=\"formControl\"\n            pRadioButton></p-radioButton>\n        <label>N\u00E3o</label>\n    </div>\n    <ng-content></ng-content>\n</div>\n",
                providers: [provideValueAccessor(CheckComponent_1)]
            }]
    }], function () { return [{ type: ɵngcc1.ControlContainer }, { type: ɵngcc0.Injector }, { type: ɵngcc2.FormNotificationService }]; }, null); })();
export { CheckComponent };
