// primeng/bundles/primeng-selectbutton.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common'), require('primeng/utils'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('primeng/selectbutton', ['@angular/core','@angular/common','exports', '@angular/core', '@angular/common', 'primeng/utils', '@angular/forms'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.common,(global.primeng = global.primeng || {}, global.primeng.selectbutton = {}), global.ng.core, global.ng.common, global.primeng.utils, global.ng.forms));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, common, utils, forms) { 
var _c0 = function () { return ["ui-clickable", "ui-button-icon-left"]; };
function SelectButton_div_1_ng_container_2_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 8);
} if (rf & 2) {
    var option_r557 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵclassMap(option_r557.icon);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction0(3, _c0));
} }
function SelectButton_div_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, SelectButton_div_1_ng_container_2_span_1_Template, 1, 4, "span", 6);
    ɵngcc0.ɵɵelementStart(2, "span", 7);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var option_r557 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", option_r557.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(option_r557.label || "ui-btn");
} }
function SelectButton_div_1_ng_template_3_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
var _c1 = function (a0, a1) { return { $implicit: a0, index: a1 }; };
function SelectButton_div_1_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, SelectButton_div_1_ng_template_3_ng_container_0_Template, 1, 0, "ng-container", 9);
} if (rf & 2) {
    var ctx_r567 = ɵngcc0.ɵɵnextContext();
    var option_r557 = ctx_r567.$implicit;
    var i_r558 = ctx_r567.index;
    var ctx_r562 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r562.itemTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c1, option_r557, i_r558));
} }
var _c2 = function (a0, a1, a2, a3, a4) { return { "ui-state-active": a0, "ui-state-disabled": a1, "ui-state-focus": a2, "ui-button-text-icon-left": a3, "ui-button-icon-only": a4 }; };
function SelectButton_div_1_Template(rf, ctx) { if (rf & 1) {
    var _r569 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 2, 3);
    ɵngcc0.ɵɵlistener("click", function SelectButton_div_1_Template_div_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r569); var option_r557 = ctx.$implicit; var i_r558 = ctx.index; var ctx_r568 = ɵngcc0.ɵɵnextContext(); return ctx_r568.onItemClick($event, option_r557, i_r558); })("keydown.enter", function SelectButton_div_1_Template_div_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r569); var option_r557 = ctx.$implicit; var i_r558 = ctx.index; var ctx_r570 = ɵngcc0.ɵɵnextContext(); return ctx_r570.onItemClick($event, option_r557, i_r558); })("focus", function SelectButton_div_1_Template_div_focus_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r569); var ctx_r571 = ɵngcc0.ɵɵnextContext(); return ctx_r571.onFocus($event); })("blur", function SelectButton_div_1_Template_div_blur_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r569); var ctx_r572 = ɵngcc0.ɵɵnextContext(); return ctx_r572.onBlur($event); });
    ɵngcc0.ɵɵtemplate(2, SelectButton_div_1_ng_container_2_Template, 4, 2, "ng-container", 4);
    ɵngcc0.ɵɵtemplate(3, SelectButton_div_1_ng_template_3_Template, 1, 5, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var option_r557 = ctx.$implicit;
    var _r559 = ɵngcc0.ɵɵreference(1);
    var _r561 = ɵngcc0.ɵɵreference(4);
    var ctx_r556 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMapInterpolate1("ui-button ui-widget ui-state-default ui-button-text-only ", option_r557.styleClass, "");
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction5(11, _c2, ctx_r556.isSelected(option_r557), ctx_r556.disabled || option_r557.disabled, _r559 == ctx_r556.focusedItem, option_r557.icon != null, option_r557.icon && !option_r557.label));
    ɵngcc0.ɵɵattribute("aria-pressed", ctx_r556.isSelected(option_r557))("title", option_r557.title)("aria-label", option_r557.label)("tabindex", ctx_r556.tabindex)("aria-labelledby", ctx_r556.ariaLabelledBy);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r556.itemTemplate)("ngIfElse", _r561);
} }
'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __read = (this && this.__read) || function (o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    };
    var __spread = (this && this.__spread) || function () {
        for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
        return ar;
    };
    var SELECTBUTTON_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: core.forwardRef(function () { return SelectButton; }),
        multi: true
    };
    var SelectButton = /** @class */ (function () {
        function SelectButton(cd) {
            this.cd = cd;
            this.tabindex = 0;
            this.onOptionClick = new core.EventEmitter();
            this.onChange = new core.EventEmitter();
            this.onModelChange = function () { };
            this.onModelTouched = function () { };
        }
        Object.defineProperty(SelectButton.prototype, "options", {
            get: function () {
                return this._options;
            },
            set: function (val) {
                //NoOp
            },
            enumerable: true,
            configurable: true
        });
        SelectButton.prototype.ngOnChanges = function (simpleChange) {
            if (simpleChange.options) {
                this._options = this.optionLabel ? utils.ObjectUtils.generateSelectItems(simpleChange.options.currentValue, this.optionLabel) : simpleChange.options.currentValue;
            }
        };
        SelectButton.prototype.writeValue = function (value) {
            this.value = value;
            this.cd.markForCheck();
        };
        SelectButton.prototype.registerOnChange = function (fn) {
            this.onModelChange = fn;
        };
        SelectButton.prototype.registerOnTouched = function (fn) {
            this.onModelTouched = fn;
        };
        SelectButton.prototype.setDisabledState = function (val) {
            this.disabled = val;
        };
        SelectButton.prototype.onItemClick = function (event, option, index) {
            if (this.disabled || option.disabled) {
                return;
            }
            if (this.multiple) {
                var itemIndex_1 = this.findItemIndex(option);
                if (itemIndex_1 != -1)
                    this.value = this.value.filter(function (val, i) { return i != itemIndex_1; });
                else
                    this.value = __spread(this.value || [], [option.value]);
            }
            else {
                this.value = option.value;
            }
            this.onOptionClick.emit({
                originalEvent: event,
                option: option,
                index: index
            });
            this.onModelChange(this.value);
            this.onChange.emit({
                originalEvent: event,
                value: this.value
            });
        };
        SelectButton.prototype.onFocus = function (event) {
            this.focusedItem = event.target;
        };
        SelectButton.prototype.onBlur = function (event) {
            this.focusedItem = null;
            this.onModelTouched();
        };
        SelectButton.prototype.isSelected = function (option) {
            if (this.multiple)
                return this.findItemIndex(option) != -1;
            else
                return utils.ObjectUtils.equals(option.value, this.value, this.dataKey);
        };
        SelectButton.prototype.findItemIndex = function (option) {
            var index = -1;
            if (this.value) {
                for (var i = 0; i < this.value.length; i++) {
                    if (this.value[i] == option.value) {
                        index = i;
                        break;
                    }
                }
            }
            return index;
        };
        SelectButton.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef }
        ]; };
        __decorate([
            core.Input()
        ], SelectButton.prototype, "tabindex", void 0);
        __decorate([
            core.Input()
        ], SelectButton.prototype, "multiple", void 0);
        __decorate([
            core.Input()
        ], SelectButton.prototype, "style", void 0);
        __decorate([
            core.Input()
        ], SelectButton.prototype, "styleClass", void 0);
        __decorate([
            core.Input()
        ], SelectButton.prototype, "ariaLabelledBy", void 0);
        __decorate([
            core.Input()
        ], SelectButton.prototype, "disabled", void 0);
        __decorate([
            core.Input()
        ], SelectButton.prototype, "dataKey", void 0);
        __decorate([
            core.Input()
        ], SelectButton.prototype, "optionLabel", void 0);
        __decorate([
            core.Output()
        ], SelectButton.prototype, "onOptionClick", void 0);
        __decorate([
            core.Output()
        ], SelectButton.prototype, "onChange", void 0);
        __decorate([
            core.ContentChild(core.TemplateRef)
        ], SelectButton.prototype, "itemTemplate", void 0);
        __decorate([
            core.Input()
        ], SelectButton.prototype, "options", null);
SelectButton.ɵfac = function SelectButton_Factory(t) { return new (t || SelectButton)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
SelectButton.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: SelectButton, selectors: [["p-selectButton"]], contentQueries: function SelectButton_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, core.TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.itemTemplate = _t.first);
    } }, inputs: { tabindex: "tabindex", options: "options", disabled: "disabled", multiple: "multiple", style: "style", styleClass: "styleClass", ariaLabelledBy: "ariaLabelledBy", dataKey: "dataKey", optionLabel: "optionLabel" }, outputs: { onOptionClick: "onOptionClick", onChange: "onChange" }, features: [ɵngcc0.ɵɵProvidersFeature([SELECTBUTTON_VALUE_ACCESSOR]), ɵngcc0.ɵɵNgOnChangesFeature], decls: 2, vars: 5, consts: [["role", "group", 3, "ngClass", "ngStyle"], ["role", "button", 3, "class", "ngClass", "click", "keydown.enter", "focus", "blur", 4, "ngFor", "ngForOf"], ["role", "button", 3, "ngClass", "click", "keydown.enter", "focus", "blur"], ["btn", ""], [4, "ngIf", "ngIfElse"], ["customcontent", ""], [3, "ngClass", "class", 4, "ngIf"], [1, "ui-button-text", "ui-clickable"], [3, "ngClass"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"]], template: function SelectButton_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, SelectButton_div_1_Template, 5, 17, "div", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.styleClass);
        ɵngcc0.ɵɵproperty("ngClass", "ui-selectbutton ui-buttonset ui-widget ui-corner-all ui-buttonset-" + (ctx.options ? ctx.options.length : 0))("ngStyle", ctx.style);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.options);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgStyle, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SelectButton, [{
        type: core.Component,
        args: [{
                selector: 'p-selectButton',
                template: "\n        <div [ngClass]=\"'ui-selectbutton ui-buttonset ui-widget ui-corner-all ui-buttonset-' + (options ? options.length : 0)\" [ngStyle]=\"style\" [class]=\"styleClass\"  role=\"group\">\n            <div *ngFor=\"let option of options; let i = index\" #btn class=\"ui-button ui-widget ui-state-default ui-button-text-only {{option.styleClass}}\"  role=\"button\" [attr.aria-pressed]=\"isSelected(option)\"\n                [ngClass]=\"{'ui-state-active':isSelected(option), 'ui-state-disabled': disabled || option.disabled, 'ui-state-focus': btn == focusedItem, \n                'ui-button-text-icon-left': (option.icon != null), 'ui-button-icon-only': (option.icon && !option.label)}\" (click)=\"onItemClick($event,option,i)\" (keydown.enter)=\"onItemClick($event,option,i)\"\n                [attr.title]=\"option.title\" [attr.aria-label]=\"option.label\" (focus)=\"onFocus($event)\" (blur)=\"onBlur($event)\" [attr.tabindex]=\"tabindex\" [attr.aria-labelledby]=\"ariaLabelledBy\">\n                <ng-container *ngIf=\"!itemTemplate else customcontent\">\n                    <span [ngClass]=\"['ui-clickable', 'ui-button-icon-left']\" [class]=\"option.icon\" *ngIf=\"option.icon\"></span>\n                    <span class=\"ui-button-text ui-clickable\">{{option.label||'ui-btn'}}</span>\n                </ng-container>\n                <ng-template #customcontent>\n                    <ng-container *ngTemplateOutlet=\"itemTemplate; context: {$implicit: option, index: i}\"></ng-container>\n                </ng-template>\n            </div>\n        </div>\n    ",
                providers: [SELECTBUTTON_VALUE_ACCESSOR],
                changeDetection: core.ChangeDetectionStrategy.Default
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { tabindex: [{
            type: core.Input
        }], onOptionClick: [{
            type: core.Output
        }], onChange: [{
            type: core.Output
        }], options: [{
            type: core.Input
        }], disabled: [{
            type: core.Input
        }], multiple: [{
            type: core.Input
        }], style: [{
            type: core.Input
        }], styleClass: [{
            type: core.Input
        }], ariaLabelledBy: [{
            type: core.Input
        }], dataKey: [{
            type: core.Input
        }], optionLabel: [{
            type: core.Input
        }], itemTemplate: [{
            type: core.ContentChild,
            args: [core.TemplateRef]
        }] }); })();
        return SelectButton;
    }());
    var SelectButtonModule = /** @class */ (function () {
        function SelectButtonModule() {
        }
SelectButtonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: SelectButtonModule });
SelectButtonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function SelectButtonModule_Factory(t) { return new (t || SelectButtonModule)(); }, imports: [[common.CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SelectButtonModule, { declarations: [SelectButton], imports: [ɵngcc1.CommonModule], exports: [SelectButton] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SelectButtonModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule],
                exports: [SelectButton],
                declarations: [SelectButton]
            }]
    }], function () { return []; }, null); })();
        return SelectButtonModule;
    }());

    exports.SELECTBUTTON_VALUE_ACCESSOR = SELECTBUTTON_VALUE_ACCESSOR;
    exports.SelectButton = SelectButton;
    exports.SelectButtonModule = SelectButtonModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
