// primeng/bundles/primeng-steps.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('@angular/router'),exports, require('@angular/core'), require('@angular/common'), require('@angular/router')) :
    typeof define === 'function' && define.amd ? define('primeng/steps', ['@angular/core','@angular/common','@angular/router','exports', '@angular/core', '@angular/common', '@angular/router'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.common,global.ng.router,(global.primeng = global.primeng || {}, global.primeng.steps = {}), global.ng.core, global.ng.common, global.ng.router));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, common, router) { 
function Steps_li_2_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r761 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 7);
    ɵngcc0.ɵɵlistener("click", function Steps_li_2_a_2_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r761); var ctx_r762 = ɵngcc0.ɵɵnextContext(); var item_r755 = ctx_r762.$implicit; var i_r756 = ctx_r762.index; var ctx_r760 = ɵngcc0.ɵɵnextContext(); return ctx_r760.itemClick($event, item_r755, i_r756); })("keydown.enter", function Steps_li_2_a_2_Template_a_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r761); var ctx_r764 = ɵngcc0.ɵɵnextContext(); var item_r755 = ctx_r764.$implicit; var i_r756 = ctx_r764.index; var ctx_r763 = ɵngcc0.ɵɵnextContext(); return ctx_r763.itemClick($event, item_r755, i_r756); });
    ɵngcc0.ɵɵelementStart(1, "span", 8);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 9);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r765 = ɵngcc0.ɵɵnextContext();
    var item_r755 = ctx_r765.$implicit;
    var i_r756 = ctx_r765.index;
    var ctx_r758 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("href", item_r755.url, ɵngcc0.ɵɵsanitizeUrl)("target", item_r755.target)("id", item_r755.id)("tabindex", item_r755.disabled || i_r756 !== ctx_r758.activeIndex && ctx_r758.readonly ? null : item_r755.tabindex ? item_r755.tabindex : "0");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(i_r756 + 1);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r755.label);
} }
var _c0 = function () { return { exact: false }; };
function Steps_li_2_a_3_Template(rf, ctx) { if (rf & 1) {
    var _r767 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 10);
    ɵngcc0.ɵɵlistener("click", function Steps_li_2_a_3_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r767); var ctx_r768 = ɵngcc0.ɵɵnextContext(); var item_r755 = ctx_r768.$implicit; var i_r756 = ctx_r768.index; var ctx_r766 = ɵngcc0.ɵɵnextContext(); return ctx_r766.itemClick($event, item_r755, i_r756); })("keydown.enter", function Steps_li_2_a_3_Template_a_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r767); var ctx_r770 = ɵngcc0.ɵɵnextContext(); var item_r755 = ctx_r770.$implicit; var i_r756 = ctx_r770.index; var ctx_r769 = ɵngcc0.ɵɵnextContext(); return ctx_r769.itemClick($event, item_r755, i_r756); });
    ɵngcc0.ɵɵelementStart(1, "span", 8);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 9);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r771 = ɵngcc0.ɵɵnextContext();
    var item_r755 = ctx_r771.$implicit;
    var i_r756 = ctx_r771.index;
    var ctx_r759 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("routerLink", item_r755.routerLink)("queryParams", item_r755.queryParams)("routerLinkActive", "ui-menuitem-link-active")("routerLinkActiveOptions", item_r755.routerLinkActiveOptions || ɵngcc0.ɵɵpureFunction0(15, _c0))("fragment", item_r755.fragment)("queryParamsHandling", item_r755.queryParamsHandling)("preserveFragment", item_r755.preserveFragment)("skipLocationChange", item_r755.skipLocationChange)("replaceUrl", item_r755.replaceUrl)("state", item_r755.state);
    ɵngcc0.ɵɵattribute("target", item_r755.target)("id", item_r755.id)("tabindex", item_r755.disabled || i_r756 !== ctx_r759.activeIndex && ctx_r759.readonly ? null : item_r755.tabindex ? item_r755.tabindex : "0");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(i_r756 + 1);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r755.label);
} }
var _c1 = function (a0, a1, a2, a3) { return { "ui-state-highlight ui-steps-current": a0, "ui-state-default": a1, "ui-state-complete": a2, "ui-state-disabled ui-steps-incomplete": a3 }; };
function Steps_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 3, 4);
    ɵngcc0.ɵɵtemplate(2, Steps_li_2_a_2_Template, 5, 6, "a", 5);
    ɵngcc0.ɵɵtemplate(3, Steps_li_2_a_3_Template, 5, 16, "a", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r755 = ctx.$implicit;
    var i_r756 = ctx.index;
    var ctx_r754 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMap(item_r755.styleClass);
    ɵngcc0.ɵɵproperty("ngStyle", item_r755.style)("ngClass", ɵngcc0.ɵɵpureFunction4(8, _c1, i_r756 === ctx_r754.activeIndex, i_r756 !== ctx_r754.activeIndex, i_r756 < ctx_r754.activeIndex, item_r755.disabled || i_r756 !== ctx_r754.activeIndex && ctx_r754.readonly));
    ɵngcc0.ɵɵattribute("aria-selected", i_r756 === ctx_r754.activeIndex)("aria-expanded", i_r756 === ctx_r754.activeIndex);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !item_r755.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r755.routerLink);
} }
var _c2 = function (a1) { return { "ui-steps ui-widget ui-helper-clearfix": true, "ui-steps-readonly": a1 }; };
'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var Steps = /** @class */ (function () {
        function Steps() {
            this.activeIndex = 0;
            this.readonly = true;
            this.activeIndexChange = new core.EventEmitter();
        }
        Steps.prototype.itemClick = function (event, item, i) {
            if (this.readonly || item.disabled) {
                event.preventDefault();
                return;
            }
            this.activeIndexChange.emit(i);
            if (!item.url) {
                event.preventDefault();
            }
            if (item.command) {
                item.command({
                    originalEvent: event,
                    item: item,
                    index: i
                });
            }
        };
        __decorate([
            core.Input()
        ], Steps.prototype, "activeIndex", void 0);
        __decorate([
            core.Input()
        ], Steps.prototype, "model", void 0);
        __decorate([
            core.Input()
        ], Steps.prototype, "readonly", void 0);
        __decorate([
            core.Input()
        ], Steps.prototype, "style", void 0);
        __decorate([
            core.Input()
        ], Steps.prototype, "styleClass", void 0);
        __decorate([
            core.Output()
        ], Steps.prototype, "activeIndexChange", void 0);
Steps.ɵfac = function Steps_Factory(t) { return new (t || Steps)(); };
Steps.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: Steps, selectors: [["p-steps"]], inputs: { activeIndex: "activeIndex", readonly: "readonly", model: "model", style: "style", styleClass: "styleClass" }, outputs: { activeIndexChange: "activeIndexChange" }, decls: 3, vars: 7, consts: [[3, "ngClass", "ngStyle"], ["role", "tablist"], ["class", "ui-steps-item", "role", "tab", 3, "ngStyle", "class", "ngClass", 4, "ngFor", "ngForOf"], ["role", "tab", 1, "ui-steps-item", 3, "ngStyle", "ngClass"], ["menuitem", ""], ["class", "ui-menuitem-link", "role", "presentation", 3, "click", "keydown.enter", 4, "ngIf"], ["role", "presentation", "class", "ui-menuitem-link", 3, "routerLink", "queryParams", "routerLinkActive", "routerLinkActiveOptions", "fragment", "queryParamsHandling", "preserveFragment", "skipLocationChange", "replaceUrl", "state", "click", "keydown.enter", 4, "ngIf"], ["role", "presentation", 1, "ui-menuitem-link", 3, "click", "keydown.enter"], [1, "ui-steps-number"], [1, "ui-steps-title"], ["role", "presentation", 1, "ui-menuitem-link", 3, "routerLink", "queryParams", "routerLinkActive", "routerLinkActiveOptions", "fragment", "queryParamsHandling", "preserveFragment", "skipLocationChange", "replaceUrl", "state", "click", "keydown.enter"]], template: function Steps_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "ul", 1);
        ɵngcc0.ɵɵtemplate(2, Steps_li_2_Template, 4, 13, "li", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.styleClass);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c2, ctx.readonly))("ngStyle", ctx.style);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.model);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgStyle, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc2.RouterLinkWithHref, ɵngcc2.RouterLinkActive], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Steps, [{
        type: core.Component,
        args: [{
                selector: 'p-steps',
                template: "\n        <div [ngClass]=\"{'ui-steps ui-widget ui-helper-clearfix':true,'ui-steps-readonly':readonly}\" [ngStyle]=\"style\" [class]=\"styleClass\">\n            <ul role=\"tablist\">\n                <li *ngFor=\"let item of model; let i = index\" class=\"ui-steps-item\" #menuitem [ngStyle]=\"item.style\" [class]=\"item.styleClass\" role=\"tab\" [attr.aria-selected]=\"i === activeIndex\" [attr.aria-expanded]=\"i === activeIndex\"\n                    [ngClass]=\"{'ui-state-highlight ui-steps-current':(i === activeIndex),\n                        'ui-state-default':(i !== activeIndex),\n                        'ui-state-complete':(i < activeIndex),\n                        'ui-state-disabled ui-steps-incomplete':item.disabled||(i !== activeIndex && readonly)}\">\n                    <a *ngIf=\"!item.routerLink\" [attr.href]=\"item.url\" class=\"ui-menuitem-link\" role=\"presentation\" (click)=\"itemClick($event, item, i)\" (keydown.enter)=\"itemClick($event, item, i)\" [attr.target]=\"item.target\" [attr.id]=\"item.id\" \n                        [attr.tabindex]=\"item.disabled||(i !== activeIndex && readonly) ? null : (item.tabindex ? item.tabindex : '0')\">\n                        <span class=\"ui-steps-number\">{{i + 1}}</span>\n                        <span class=\"ui-steps-title\">{{item.label}}</span>\n                    </a>\n                    <a *ngIf=\"item.routerLink\" [routerLink]=\"item.routerLink\" [queryParams]=\"item.queryParams\" role=\"presentation\" [routerLinkActive]=\"'ui-menuitem-link-active'\" [routerLinkActiveOptions]=\"item.routerLinkActiveOptions||{exact:false}\" class=\"ui-menuitem-link\" \n                        (click)=\"itemClick($event, item, i)\" (keydown.enter)=\"itemClick($event, item, i)\" [attr.target]=\"item.target\" [attr.id]=\"item.id\" [attr.tabindex]=\"item.disabled||(i !== activeIndex && readonly) ? null : (item.tabindex ? item.tabindex : '0')\"\n                        [fragment]=\"item.fragment\" [queryParamsHandling]=\"item.queryParamsHandling\" [preserveFragment]=\"item.preserveFragment\" [skipLocationChange]=\"item.skipLocationChange\" [replaceUrl]=\"item.replaceUrl\" [state]=\"item.state\">\n                        <span class=\"ui-steps-number\">{{i + 1}}</span>\n                        <span class=\"ui-steps-title\">{{item.label}}</span>\n                    </a>\n                </li>\n            </ul>\n        </div>\n    ",
                changeDetection: core.ChangeDetectionStrategy.Default
            }]
    }], function () { return []; }, { activeIndex: [{
            type: core.Input
        }], readonly: [{
            type: core.Input
        }], activeIndexChange: [{
            type: core.Output
        }], model: [{
            type: core.Input
        }], style: [{
            type: core.Input
        }], styleClass: [{
            type: core.Input
        }] }); })();
        return Steps;
    }());
    var StepsModule = /** @class */ (function () {
        function StepsModule() {
        }
StepsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: StepsModule });
StepsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function StepsModule_Factory(t) { return new (t || StepsModule)(); }, imports: [[common.CommonModule, router.RouterModule], ɵngcc2.RouterModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepsModule, { declarations: [Steps], imports: [ɵngcc1.CommonModule, ɵngcc2.RouterModule], exports: [Steps, ɵngcc2.RouterModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(StepsModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule, router.RouterModule],
                exports: [Steps, router.RouterModule],
                declarations: [Steps]
            }]
    }], function () { return []; }, null); })();
        return StepsModule;
    }());

    exports.Steps = Steps;
    exports.StepsModule = StepsModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
