// primeng/bundles/primeng-rating.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('primeng/rating', ['@angular/core','@angular/common','exports', '@angular/core', '@angular/common', '@angular/forms'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.common,(global.primeng = global.primeng || {}, global.primeng.rating = {}), global.ng.core, global.ng.common, global.ng.forms));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, common, forms) { 
function Rating_a_1_Template(rf, ctx) { if (rf & 1) {
    var _r609 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 3);
    ɵngcc0.ɵɵlistener("click", function Rating_a_1_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r609); var ctx_r608 = ɵngcc0.ɵɵnextContext(); return ctx_r608.clear($event); })("keydown.enter", function Rating_a_1_Template_a_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r609); var ctx_r610 = ɵngcc0.ɵɵnextContext(); return ctx_r610.clear($event); });
    ɵngcc0.ɵɵelement(1, "span", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r606 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("tabindex", ctx_r606.disabled ? null : "0");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r606.iconCancelClass)("ngStyle", ctx_r606.iconCancelStyle);
} }
function Rating_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r614 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 5);
    ɵngcc0.ɵɵlistener("click", function Rating_a_2_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r614); var i_r612 = ctx.index; var ctx_r613 = ɵngcc0.ɵɵnextContext(); return ctx_r613.rate($event, i_r612); })("keydown.enter", function Rating_a_2_Template_a_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r614); var i_r612 = ctx.index; var ctx_r615 = ɵngcc0.ɵɵnextContext(); return ctx_r615.rate($event, i_r612); });
    ɵngcc0.ɵɵelement(1, "span", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var i_r612 = ctx.index;
    var ctx_r607 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("tabindex", ctx_r607.disabled ? null : "0");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", !ctx_r607.value || i_r612 >= ctx_r607.value ? ctx_r607.iconOffClass : ctx_r607.iconOnClass)("ngStyle", !ctx_r607.value || i_r612 >= ctx_r607.value ? ctx_r607.iconOffStyle : ctx_r607.iconOnStyle);
} }
var _c0 = function (a0) { return { "ui-state-disabled": a0 }; };
'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var RATING_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: core.forwardRef(function () { return Rating; }),
        multi: true
    };
    var Rating = /** @class */ (function () {
        function Rating(cd) {
            this.cd = cd;
            this.stars = 5;
            this.cancel = true;
            this.iconOnClass = 'pi pi-star';
            this.iconOffClass = 'pi pi-star-o';
            this.iconCancelClass = 'pi pi-ban';
            this.onRate = new core.EventEmitter();
            this.onCancel = new core.EventEmitter();
            this.onModelChange = function () { };
            this.onModelTouched = function () { };
        }
        Rating.prototype.ngOnInit = function () {
            this.starsArray = [];
            for (var i = 0; i < this.stars; i++) {
                this.starsArray[i] = i;
            }
        };
        Rating.prototype.rate = function (event, i) {
            if (!this.readonly && !this.disabled) {
                this.value = (i + 1);
                this.onModelChange(this.value);
                this.onModelTouched();
                this.onRate.emit({
                    originalEvent: event,
                    value: (i + 1)
                });
            }
            event.preventDefault();
        };
        Rating.prototype.clear = function (event) {
            if (!this.readonly && !this.disabled) {
                this.value = null;
                this.onModelChange(this.value);
                this.onModelTouched();
                this.onCancel.emit(event);
            }
            event.preventDefault();
        };
        Rating.prototype.writeValue = function (value) {
            this.value = value;
            this.cd.detectChanges();
        };
        Rating.prototype.registerOnChange = function (fn) {
            this.onModelChange = fn;
        };
        Rating.prototype.registerOnTouched = function (fn) {
            this.onModelTouched = fn;
        };
        Rating.prototype.setDisabledState = function (val) {
            this.disabled = val;
        };
        Rating.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef }
        ]; };
        __decorate([
            core.Input()
        ], Rating.prototype, "disabled", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "readonly", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "stars", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "cancel", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "iconOnClass", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "iconOnStyle", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "iconOffClass", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "iconOffStyle", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "iconCancelClass", void 0);
        __decorate([
            core.Input()
        ], Rating.prototype, "iconCancelStyle", void 0);
        __decorate([
            core.Output()
        ], Rating.prototype, "onRate", void 0);
        __decorate([
            core.Output()
        ], Rating.prototype, "onCancel", void 0);
Rating.ɵfac = function Rating_Factory(t) { return new (t || Rating)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
Rating.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: Rating, selectors: [["p-rating"]], inputs: { stars: "stars", cancel: "cancel", iconOnClass: "iconOnClass", iconOffClass: "iconOffClass", iconCancelClass: "iconCancelClass", disabled: "disabled", readonly: "readonly", iconOnStyle: "iconOnStyle", iconOffStyle: "iconOffStyle", iconCancelStyle: "iconCancelStyle" }, outputs: { onRate: "onRate", onCancel: "onCancel" }, features: [ɵngcc0.ɵɵProvidersFeature([RATING_VALUE_ACCESSOR])], decls: 3, vars: 5, consts: [[1, "ui-rating", 3, "ngClass"], ["class", "ui-rating-cancel", 3, "click", "keydown.enter", 4, "ngIf"], [3, "click", "keydown.enter", 4, "ngFor", "ngForOf"], [1, "ui-rating-cancel", 3, "click", "keydown.enter"], [1, "ui-rating-icon", 3, "ngClass", "ngStyle"], [3, "click", "keydown.enter"]], template: function Rating_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, Rating_a_1_Template, 2, 3, "a", 1);
        ɵngcc0.ɵɵtemplate(2, Rating_a_2_Template, 2, 3, "a", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c0, ctx.disabled));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.cancel);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.starsArray);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgIf, ɵngcc1.NgForOf, ɵngcc1.NgStyle], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Rating, [{
        type: core.Component,
        args: [{
                selector: 'p-rating',
                template: "\n        <div class=\"ui-rating\" [ngClass]=\"{'ui-state-disabled': disabled}\">\n            <a [attr.tabindex]=\"disabled ? null : '0'\" *ngIf=\"cancel\" (click)=\"clear($event)\" (keydown.enter)=\"clear($event)\"  class=\"ui-rating-cancel\">\n                <span class=\"ui-rating-icon\" [ngClass]=\"iconCancelClass\" [ngStyle]=\"iconCancelStyle\"></span>\n            </a>\n            <a [attr.tabindex]=\"disabled ? null : '0'\" *ngFor=\"let star of starsArray;let i=index\" (click)=\"rate($event,i)\" (keydown.enter)=\"rate($event,i)\">\n                <span class=\"ui-rating-icon\" \n                    [ngClass]=\"(!value || i >= value) ? iconOffClass : iconOnClass\"\n                    [ngStyle]=\"(!value || i >= value) ? iconOffStyle : iconOnStyle\"\n                ></span>\n            </a>\n        </div>\n    ",
                providers: [RATING_VALUE_ACCESSOR],
                changeDetection: core.ChangeDetectionStrategy.Default
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { stars: [{
            type: core.Input
        }], cancel: [{
            type: core.Input
        }], iconOnClass: [{
            type: core.Input
        }], iconOffClass: [{
            type: core.Input
        }], iconCancelClass: [{
            type: core.Input
        }], onRate: [{
            type: core.Output
        }], onCancel: [{
            type: core.Output
        }], disabled: [{
            type: core.Input
        }], readonly: [{
            type: core.Input
        }], iconOnStyle: [{
            type: core.Input
        }], iconOffStyle: [{
            type: core.Input
        }], iconCancelStyle: [{
            type: core.Input
        }] }); })();
        return Rating;
    }());
    var RatingModule = /** @class */ (function () {
        function RatingModule() {
        }
RatingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: RatingModule });
RatingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function RatingModule_Factory(t) { return new (t || RatingModule)(); }, imports: [[common.CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(RatingModule, { declarations: [Rating], imports: [ɵngcc1.CommonModule], exports: [Rating] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RatingModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule],
                exports: [Rating],
                declarations: [Rating]
            }]
    }], function () { return []; }, null); })();
        return RatingModule;
    }());

    exports.RATING_VALUE_ACCESSOR = RATING_VALUE_ACCESSOR;
    exports.Rating = Rating;
    exports.RatingModule = RatingModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
