// primeng/bundles/primeng-messages.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('primeng/api'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common'), require('@angular/animations'), require('primeng/api')) :
    typeof define === 'function' && define.amd ? define('primeng/messages', ['@angular/core','primeng/api','@angular/common','exports', '@angular/core', '@angular/common', '@angular/animations', 'primeng/api'], factory) :
    (global = global || self, factory(global.ng.core,global.primeng.api,global.ng.common,(global.primeng = global.primeng || {}, global.primeng.messages = {}), global.ng.core, global.ng.common, global.ng.animations, global.primeng.api));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, common, animations, api) { 
function Messages_div_0_a_1_Template(rf, ctx) { if (rf & 1) {
    var _r286 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 6);
    ɵngcc0.ɵɵlistener("click", function Messages_div_0_a_1_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r286); var ctx_r285 = ɵngcc0.ɵɵnextContext(2); return ctx_r285.clear($event); })("keydown.enter", function Messages_div_0_a_1_Template_a_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r286); var ctx_r287 = ɵngcc0.ɵɵnextContext(2); return ctx_r287.clear($event); });
    ɵngcc0.ɵɵelement(1, "i", 7);
    ɵngcc0.ɵɵelementEnd();
} }
function Messages_div_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function Messages_div_0_ul_4_li_1_div_1_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 13);
} if (rf & 2) {
    var msg_r289 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", msg_r289.summary, ɵngcc0.ɵɵsanitizeHtml);
} }
function Messages_div_0_ul_4_li_1_div_1_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 14);
} if (rf & 2) {
    var msg_r289 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", msg_r289.detail, ɵngcc0.ɵɵsanitizeHtml);
} }
function Messages_div_0_ul_4_li_1_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵtemplate(1, Messages_div_0_ul_4_li_1_div_1_span_1_Template, 1, 1, "span", 11);
    ɵngcc0.ɵɵtemplate(2, Messages_div_0_ul_4_li_1_div_1_span_2_Template, 1, 1, "span", 12);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var msg_r289 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", msg_r289.summary);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", msg_r289.detail);
} }
function Messages_div_0_ul_4_li_1_ng_template_2_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 17);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var msg_r289 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(msg_r289.summary);
} }
function Messages_div_0_ul_4_li_1_ng_template_2_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 18);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var msg_r289 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(msg_r289.detail);
} }
function Messages_div_0_ul_4_li_1_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, Messages_div_0_ul_4_li_1_ng_template_2_span_0_Template, 2, 1, "span", 15);
    ɵngcc0.ɵɵtemplate(1, Messages_div_0_ul_4_li_1_ng_template_2_span_1_Template, 2, 1, "span", 16);
} if (rf & 2) {
    var msg_r289 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngIf", msg_r289.summary);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", msg_r289.detail);
} }
function Messages_div_0_ul_4_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵtemplate(1, Messages_div_0_ul_4_li_1_div_1_Template, 3, 2, "div", 9);
    ɵngcc0.ɵɵtemplate(2, Messages_div_0_ul_4_li_1_ng_template_2_Template, 2, 2, "ng-template", null, 10, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var _r291 = ɵngcc0.ɵɵreference(3);
    var ctx_r288 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r288.escape)("ngIfElse", _r291);
} }
function Messages_div_0_ul_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul");
    ɵngcc0.ɵɵtemplate(1, Messages_div_0_ul_4_li_1_Template, 4, 2, "li", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r284 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r284.value);
} }
var _c0 = function (a0, a1) { return { showTransitionParams: a0, hideTransitionParams: a1 }; };
var _c1 = function (a1) { return { value: "visible", params: a1 }; };
function Messages_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵtemplate(1, Messages_div_0_a_1_Template, 2, 0, "a", 2);
    ɵngcc0.ɵɵelement(2, "span", 3);
    ɵngcc0.ɵɵtemplate(3, Messages_div_0_ng_container_3_Template, 1, 0, "ng-container", 4);
    ɵngcc0.ɵɵtemplate(4, Messages_div_0_ul_4_Template, 2, 1, "ul", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r281 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMap(ctx_r281.styleClass);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r281.getSeverityClass())("ngStyle", ctx_r281.style)("@messageAnimation", ɵngcc0.ɵɵpureFunction1(12, _c1, ɵngcc0.ɵɵpureFunction2(9, _c0, ctx_r281.showTransitionOptions, ctx_r281.hideTransitionOptions)));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r281.closable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r281.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r281.contentTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r281.value && ctx_r281.value.length);
} }
'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    var __read = (this && this.__read) || function (o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    };
    var __spread = (this && this.__spread) || function () {
        for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
        return ar;
    };
    var Messages = /** @class */ (function () {
        function Messages(messageService, el) {
            this.messageService = messageService;
            this.el = el;
            this.closable = true;
            this.enableService = true;
            this.escape = true;
            this.showTransitionOptions = '300ms ease-out';
            this.hideTransitionOptions = '250ms ease-in';
            this.valueChange = new core.EventEmitter();
        }
        Messages.prototype.ngAfterContentInit = function () {
            var _this = this;
            this.templates.forEach(function (item) {
                switch (item.getType()) {
                    case 'content':
                        _this.contentTemplate = item.template;
                        break;
                    default:
                        _this.contentTemplate = item.template;
                        break;
                }
            });
            if (this.messageService && this.enableService && !this.contentTemplate) {
                this.messageSubscription = this.messageService.messageObserver.subscribe(function (messages) {
                    if (messages) {
                        if (messages instanceof Array) {
                            var filteredMessages = messages.filter(function (m) { return _this.key === m.key; });
                            _this.value = _this.value ? __spread(_this.value, filteredMessages) : __spread(filteredMessages);
                        }
                        else if (_this.key === messages.key) {
                            _this.value = _this.value ? __spread(_this.value, [messages]) : [messages];
                        }
                    }
                });
                this.clearSubscription = this.messageService.clearObserver.subscribe(function (key) {
                    if (key) {
                        if (_this.key === key) {
                            _this.value = null;
                        }
                    }
                    else {
                        _this.value = null;
                    }
                });
            }
        };
        Messages.prototype.hasMessages = function () {
            var parentEl = this.el.nativeElement.parentElement;
            if (parentEl && parentEl.offsetParent) {
                return this.contentTemplate != null || this.value && this.value.length > 0;
            }
            return false;
        };
        Messages.prototype.getSeverityClass = function () {
            if (this.severity) {
                return 'ui-messages-' + this.severity;
            }
            else {
                var msg_1 = this.value[0];
                if (msg_1) {
                    var severities = ['info', 'warn', 'error', 'success'];
                    var severity = severities.find(function (item) { return item === msg_1.severity; });
                    return severity && "ui-messages-" + severity;
                }
            }
            return null;
        };
        Messages.prototype.clear = function (event) {
            this.value = [];
            this.valueChange.emit(this.value);
            event.preventDefault();
        };
        Object.defineProperty(Messages.prototype, "icon", {
            get: function () {
                var severity = this.severity || (this.hasMessages() ? this.value[0].severity : null);
                if (this.hasMessages()) {
                    switch (severity) {
                        case 'success':
                            return 'pi-check';
                            break;
                        case 'info':
                            return 'pi-info-circle';
                            break;
                        case 'error':
                            return 'pi-times';
                            break;
                        case 'warn':
                            return 'pi-exclamation-triangle';
                            break;
                        default:
                            return 'pi-info-circle';
                            break;
                    }
                }
                return null;
            },
            enumerable: true,
            configurable: true
        });
        Messages.prototype.ngOnDestroy = function () {
            if (this.messageSubscription) {
                this.messageSubscription.unsubscribe();
            }
            if (this.clearSubscription) {
                this.clearSubscription.unsubscribe();
            }
        };
        Messages.ctorParameters = function () { return [
            { type: api.MessageService, decorators: [{ type: core.Optional }] },
            { type: core.ElementRef }
        ]; };
        __decorate([
            core.Input()
        ], Messages.prototype, "value", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "closable", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "style", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "styleClass", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "enableService", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "key", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "escape", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "severity", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "showTransitionOptions", void 0);
        __decorate([
            core.Input()
        ], Messages.prototype, "hideTransitionOptions", void 0);
        __decorate([
            core.ContentChildren(api.PrimeTemplate)
        ], Messages.prototype, "templates", void 0);
        __decorate([
            core.Output()
        ], Messages.prototype, "valueChange", void 0);
        Messages = __decorate([ __param(0, core.Optional())
        ], Messages);
Messages.ɵfac = function Messages_Factory(t) { return new (t || Messages)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MessageService, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
Messages.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: Messages, selectors: [["p-messages"]], contentQueries: function Messages_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, api.PrimeTemplate, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templates = _t);
    } }, inputs: { closable: "closable", enableService: "enableService", escape: "escape", showTransitionOptions: "showTransitionOptions", hideTransitionOptions: "hideTransitionOptions", value: "value", style: "style", styleClass: "styleClass", key: "key", severity: "severity" }, outputs: { valueChange: "valueChange" }, decls: 1, vars: 1, consts: [["class", "ui-messages ui-widget ui-corner-all", "role", "alert", 3, "ngClass", "ngStyle", "class", 4, "ngIf"], ["role", "alert", 1, "ui-messages", "ui-widget", "ui-corner-all", 3, "ngClass", "ngStyle"], ["tabindex", "0", "class", "ui-messages-close", 3, "click", "keydown.enter", 4, "ngIf"], [1, "ui-messages-icon", "pi", 3, "ngClass"], [4, "ngTemplateOutlet"], [4, "ngIf"], ["tabindex", "0", 1, "ui-messages-close", 3, "click", "keydown.enter"], [1, "pi", "pi-times"], [4, "ngFor", "ngForOf"], [4, "ngIf", "ngIfElse"], ["escapeOut", ""], ["class", "ui-messages-summary", 3, "innerHTML", 4, "ngIf"], ["class", "ui-messages-detail", 3, "innerHTML", 4, "ngIf"], [1, "ui-messages-summary", 3, "innerHTML"], [1, "ui-messages-detail", 3, "innerHTML"], ["class", "ui-messages-summary", 4, "ngIf"], ["class", "ui-messages-detail", 4, "ngIf"], [1, "ui-messages-summary"], [1, "ui-messages-detail"]], template: function Messages_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, Messages_div_0_Template, 5, 14, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasMessages());
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgClass, ɵngcc2.NgStyle, ɵngcc2.NgTemplateOutlet, ɵngcc2.NgForOf], encapsulation: 2, data: { animation: [
            animations.trigger('messageAnimation', [
                animations.state('visible', animations.style({
                    transform: 'translateY(0)',
                    opacity: 1
                })),
                animations.transition('void => *', [
                    animations.style({ transform: 'translateY(-25%)', opacity: 0 }),
                    animations.animate('{{showTransitionParams}}')
                ]),
                animations.transition('* => void', [
                    animations.animate(('{{hideTransitionParams}}'), animations.style({
                        opacity: 0,
                        transform: 'translateY(-25%)'
                    }))
                ])
            ])
        ] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Messages, [{
        type: core.Component,
        args: [{
                selector: 'p-messages',
                template: "\n        <div *ngIf=\"hasMessages()\" class=\"ui-messages ui-widget ui-corner-all\"\n                    [ngClass]=\"getSeverityClass()\" role=\"alert\" [ngStyle]=\"style\" [class]=\"styleClass\"\n                    [@messageAnimation]=\"{value: 'visible', params: {showTransitionParams: showTransitionOptions, hideTransitionParams: hideTransitionOptions}}\">\n            <a tabindex=\"0\" class=\"ui-messages-close\" (click)=\"clear($event)\" (keydown.enter)=\"clear($event)\" *ngIf=\"closable\">\n                <i class=\"pi pi-times\"></i>\n            </a>\n            <span class=\"ui-messages-icon pi\" [ngClass]=\"icon\"></span>\n            <ng-container *ngTemplateOutlet=\"contentTemplate\"></ng-container>\n            <ul *ngIf=\"value && value.length\">\n                <li *ngFor=\"let msg of value\">\n                    <div *ngIf=\"!escape; else escapeOut\">\n                        <span *ngIf=\"msg.summary\" class=\"ui-messages-summary\" [innerHTML]=\"msg.summary\"></span>\n                        <span *ngIf=\"msg.detail\" class=\"ui-messages-detail\" [innerHTML]=\"msg.detail\"></span>\n                    </div>\n                    <ng-template #escapeOut>\n                        <span *ngIf=\"msg.summary\" class=\"ui-messages-summary\">{{msg.summary}}</span>\n                        <span *ngIf=\"msg.detail\" class=\"ui-messages-detail\">{{msg.detail}}</span>\n                    </ng-template>\n                </li>\n            </ul>\n        </div>\n    ",
                animations: [
                    animations.trigger('messageAnimation', [
                        animations.state('visible', animations.style({
                            transform: 'translateY(0)',
                            opacity: 1
                        })),
                        animations.transition('void => *', [
                            animations.style({ transform: 'translateY(-25%)', opacity: 0 }),
                            animations.animate('{{showTransitionParams}}')
                        ]),
                        animations.transition('* => void', [
                            animations.animate(('{{hideTransitionParams}}'), animations.style({
                                opacity: 0,
                                transform: 'translateY(-25%)'
                            }))
                        ])
                    ])
                ],
                changeDetection: core.ChangeDetectionStrategy.Default
            }]
    }], function () { return [{ type: ɵngcc1.MessageService, decorators: [{
                type: core.Optional
            }] }, { type: ɵngcc0.ElementRef }]; }, { closable: [{
            type: core.Input
        }], enableService: [{
            type: core.Input
        }], escape: [{
            type: core.Input
        }], showTransitionOptions: [{
            type: core.Input
        }], hideTransitionOptions: [{
            type: core.Input
        }], valueChange: [{
            type: core.Output
        }], value: [{
            type: core.Input
        }], style: [{
            type: core.Input
        }], styleClass: [{
            type: core.Input
        }], key: [{
            type: core.Input
        }], severity: [{
            type: core.Input
        }], templates: [{
            type: core.ContentChildren,
            args: [api.PrimeTemplate]
        }] }); })();
        return Messages;
    }());
    var MessagesModule = /** @class */ (function () {
        function MessagesModule() {
        }
MessagesModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MessagesModule });
MessagesModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MessagesModule_Factory(t) { return new (t || MessagesModule)(); }, imports: [[common.CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MessagesModule, { declarations: [Messages], imports: [ɵngcc2.CommonModule], exports: [Messages] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MessagesModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule],
                exports: [Messages],
                declarations: [Messages]
            }]
    }], function () { return []; }, null); })();
        return MessagesModule;
    }());

    exports.Messages = Messages;
    exports.MessagesModule = MessagesModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
