// @nuvem/primeng-components/esm2015/lib/crud/components/auto-complete/auto-complete.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../form-notification/form-notification.component';
import * as ɵngcc2 from './auto-complete-custom.component';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../validation-message/validation-message.component';

const _c0 = ["nAutoComplete", ""];
function AutoCompleteComponent_nValidationMessage_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "nValidationMessage", 6);
} if (rf & 2) {
    const ctx_r985 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("control", ctx_r985.formControl);
} }
const _c1 = ["*"];
var AutoCompleteComponent_1;
import { __decorate } from "tslib";
import { Component, Input } from "@angular/core";
import { provideValueAccessor } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
let AutoCompleteComponent = AutoCompleteComponent_1 = class AutoCompleteComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.virtualScroll = false;
        this.itemSize = 30;
    }
    selectItem(option) {
        this.formControl.setValue(option);
    }
    findAll(event) {
        let query = {};
        query[this.field] = event.query;
        this.service.findAll(query).subscribe(data => {
            this.results = data;
        });
    }
    outerToInner(value) {
        if (value && value.id && !value[this.field]) {
            this.service.find(value.id).subscribe((data) => {
                this.formControl.setValue(data);
            });
        }
        return value;
    }
};
AutoCompleteComponent.ɵfac = function AutoCompleteComponent_Factory(t) { return ɵAutoCompleteComponent_BaseFactory(t || AutoCompleteComponent); };
AutoCompleteComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AutoCompleteComponent, selectors: [["", "nAutoComplete", ""]], inputs: { virtualScroll: "virtualScroll", itemSize: "itemSize", field: "field", service: "service" }, features: [ɵngcc0.ɵɵProvidersFeature([provideValueAccessor(AutoCompleteComponent_1)]), ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c1, decls: 7, vars: 10, consts: [["form-notification", "", 1, "md-inputfield", 3, "name", "form"], ["notification", ""], [3, "formControl", "suggestions", "virtualScroll", "itemSize", "field", "onBlur", "onSelect", "onFocus", "completeMethod"], ["input", ""], [3, "for", "innerHTML"], [3, "control", 4, "ngIf"], [3, "control"]], template: function AutoCompleteComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "span", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "p-autoCompleteCustom", 2, 3);
        ɵngcc0.ɵɵlistener("onBlur", function AutoCompleteComponent_Template_p_autoCompleteCustom_onBlur_2_listener($event) { return ctx.onBlurEvent($event); })("onSelect", function AutoCompleteComponent_Template_p_autoCompleteCustom_onSelect_2_listener($event) { return ctx.onSelectEvent($event); })("onFocus", function AutoCompleteComponent_Template_p_autoCompleteCustom_onFocus_2_listener($event) { return ctx.onFocusEvent($event); })("completeMethod", function AutoCompleteComponent_Template_p_autoCompleteCustom_completeMethod_2_listener($event) { return ctx.findAll($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(4, "label", 4);
        ɵngcc0.ɵɵtemplate(5, AutoCompleteComponent_nValidationMessage_5_Template, 1, 1, "nValidationMessage", 5);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r983 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵproperty("name", ctx.name)("form", ctx.form);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("formControl", ctx.formControl)("suggestions", ctx.results)("virtualScroll", ctx.virtualScroll)("itemSize", ctx.itemSize)("field", ctx.field);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("for", ctx.name)("innerHTML", ctx.label, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !_r983.message);
    } }, directives: [ɵngcc1.FormNotificationComponent, ɵngcc2.AutoCompleteCustomComponent, ɵngcc3.NgControlStatus, ɵngcc3.FormControlDirective, ɵngcc4.NgIf, ɵngcc5.ValidationMessageComponent], encapsulation: 2 });
__decorate([
    Input()
], AutoCompleteComponent.prototype, "field", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "virtualScroll", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "itemSize", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "service", void 0);
const ɵAutoCompleteComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(AutoCompleteComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AutoCompleteComponent, [{
        type: Component,
        args: [{
                selector: "[nAutoComplete]",
                template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-autoCompleteCustom #input\n        [formControl]=\"formControl\"\n        [suggestions]=\"results\"\n        [virtualScroll]=\"virtualScroll\"\n        [itemSize]=\"itemSize\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onSelect)=\"onSelectEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        (completeMethod)=\"findAll($event)\"\n        [field]=\"field\">\n    </p-autoCompleteCustom>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
                providers: [provideValueAccessor(AutoCompleteComponent_1)]
            }]
    }], null, { virtualScroll: [{
            type: Input
        }], itemSize: [{
            type: Input
        }], field: [{
            type: Input
        }], service: [{
            type: Input
        }] }); })();
export { AutoCompleteComponent };
