// @nuvem/primeng-components/esm2015/lib/crud/components/auto-complete/auto-complete-item.component.js
import { __decorate } from "tslib";
import { Component, Input, Output, EventEmitter } from "@angular/core";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function AutoCompleteItem_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1046 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1046.resolvedFieldData);
} }
function AutoCompleteItem_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c0 = function (a1) { return { "ui-autocomplete-list-item ui-corner-all": true, "ui-state-highlight": a1 }; };
const _c1 = function (a0) { return { "height": a0, "display": "table", "width": "100%" }; };
const _c2 = function (a0, a1) { return { $implicit: a0, index: a1 }; };
let AutoCompleteItem = class AutoCompleteItem {
    constructor() {
        this.highlightOptionChange = new EventEmitter();
        this.onClick = new EventEmitter();
    }
    selectItem(event) {
        this.onClick.emit({
            originalEvent: event,
            option: this.option
        });
    }
};
AutoCompleteItem.ɵfac = function AutoCompleteItem_Factory(t) { return new (t || AutoCompleteItem)(); };
AutoCompleteItem.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AutoCompleteItem, selectors: [["p-autoCompleteItem"]], inputs: { option: "option", resolvedFieldData: "resolvedFieldData", highlightOption: "highlightOption", index: "index", itemSize: "itemSize", template: "template" }, outputs: { highlightOptionChange: "highlightOptionChange", onClick: "onClick" }, decls: 3, vars: 13, consts: [["role", "option", 3, "ngClass", "id", "ngStyle", "mouseenter", "mouseleave", "click"], [4, "ngIf"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"]], template: function AutoCompleteItem_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "li", 0);
        ɵngcc0.ɵɵlistener("mouseenter", function AutoCompleteItem_Template_li_mouseenter_0_listener() { return ctx.highlightOptionChange.emit(ctx.option); })("mouseleave", function AutoCompleteItem_Template_li_mouseleave_0_listener() { return ctx.highlightOptionChange.emit(null); })("click", function AutoCompleteItem_Template_li_click_0_listener($event) { return ctx.selectItem($event); });
        ɵngcc0.ɵɵtemplate(1, AutoCompleteItem_span_1_Template, 2, 1, "span", 1);
        ɵngcc0.ɵɵtemplate(2, AutoCompleteItem_ng_container_2_Template, 1, 0, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(6, _c0, ctx.highlightOption == ctx.option))("id", ctx.highlightOption == ctx.option ? "p-highlighted-option" : "")("ngStyle", ɵngcc0.ɵɵpureFunction1(8, _c1, ctx.itemSize + "px"));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.template);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.template)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(10, _c2, ctx.option, ctx.index));
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgStyle, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
__decorate([
    Input()
], AutoCompleteItem.prototype, "option", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "resolvedFieldData", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "highlightOption", void 0);
__decorate([
    Output()
], AutoCompleteItem.prototype, "highlightOptionChange", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "index", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "itemSize", void 0);
__decorate([
    Input()
], AutoCompleteItem.prototype, "template", void 0);
__decorate([
    Output()
], AutoCompleteItem.prototype, "onClick", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AutoCompleteItem, [{
        type: Component,
        args: [{
                selector: 'p-autoCompleteItem',
                template: `
        <li role="option" [ngClass]="{'ui-autocomplete-list-item ui-corner-all':true,'ui-state-highlight':(highlightOption==option)}"
            (mouseenter)="highlightOptionChange.emit(option)" (mouseleave)="highlightOptionChange.emit(null)"
            [id]="highlightOption == option ? 'p-highlighted-option':''" (click)="selectItem($event)"
            [ngStyle]="{'height': itemSize + 'px', 'display': 'table', 'width': '100%'}">
            <span *ngIf="!template">{{resolvedFieldData}}</span>
            <ng-container *ngTemplateOutlet="template; context: {$implicit: option, index: index}"></ng-container>
        </li>
    `
            }]
    }], function () { return []; }, { highlightOptionChange: [{
            type: Output
        }], onClick: [{
            type: Output
        }], option: [{
            type: Input
        }], resolvedFieldData: [{
            type: Input
        }], highlightOption: [{
            type: Input
        }], index: [{
            type: Input
        }], itemSize: [{
            type: Input
        }], template: [{
            type: Input
        }] }); })();
export { AutoCompleteItem };
