// @nuvem/primeng-components/esm2015/lib/crud/components/auto-complete/auto-complete.component.js
var AutoCompleteComponent_1;
import { __decorate } from "tslib";
import { Component, Input } from "@angular/core";
import { provideValueAccessor } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
let AutoCompleteComponent = AutoCompleteComponent_1 = class AutoCompleteComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.virtualScroll = false;
        this.itemSize = 30;
    }
    selectItem(option) {
        this.formControl.setValue(option);
    }
    findAll(event) {
        let query = {};
        query[this.field] = event.query;
        this.service.findAll(query).subscribe(data => {
            this.results = data;
        });
    }
    outerToInner(value) {
        if (value && value.id && !value[this.field]) {
            this.service.find(value.id).subscribe((data) => {
                this.formControl.setValue(data);
            });
        }
        return value;
    }
};
__decorate([
    Input()
], AutoCompleteComponent.prototype, "field", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "virtualScroll", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "itemSize", void 0);
__decorate([
    Input()
], AutoCompleteComponent.prototype, "service", void 0);
AutoCompleteComponent = AutoCompleteComponent_1 = __decorate([
    Component({
        selector: "[nAutoComplete]",
        template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-autoCompleteCustom #input\n        [formControl]=\"formControl\"\n        [suggestions]=\"results\"\n        [virtualScroll]=\"virtualScroll\"\n        [itemSize]=\"itemSize\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onSelect)=\"onSelectEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        (completeMethod)=\"findAll($event)\"\n        [field]=\"field\">\n    </p-autoCompleteCustom>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
        providers: [provideValueAccessor(AutoCompleteComponent_1)]
    })
], AutoCompleteComponent);
export { AutoCompleteComponent };
