// @nuvem/primeng-components/esm5/lib/extra-button/datatable-splitbutton.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
/**
 * Class DatatableSplitbuttonComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'primeng';
var DatatableSplitbuttonComponent = /** @class */ (function () {
    function DatatableSplitbuttonComponent() {
    }
    /**
     * ngOnInit method
     * @returns void
     */
    DatatableSplitbuttonComponent.prototype.ngOnInit = function () {
    };
    __decorate([
        Input()
    ], DatatableSplitbuttonComponent.prototype, "items", void 0);
    __decorate([
        Input()
    ], DatatableSplitbuttonComponent.prototype, "label", void 0);
    __decorate([
        Input()
    ], DatatableSplitbuttonComponent.prototype, "class", void 0);
    __decorate([
        Input()
    ], DatatableSplitbuttonComponent.prototype, "icon", void 0);
    __decorate([
        Input()
    ], DatatableSplitbuttonComponent.prototype, "method", void 0);
DatatableSplitbuttonComponent.ɵfac = function DatatableSplitbuttonComponent_Factory(t) { return new (t || DatatableSplitbuttonComponent)(); };
DatatableSplitbuttonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DatatableSplitbuttonComponent, selectors: [["datatable-splitbutton"]], inputs: { items: "items", label: "label", class: "class", icon: "icon", method: "method" }, decls: 1, vars: 4, consts: [[3, "label", "icon", "model", "styleClass", "onClick"]], template: function DatatableSplitbuttonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p-splitButton", 0);
        ɵngcc0.ɵɵlistener("onClick", function DatatableSplitbuttonComponent_Template_p_splitButton_onClick_0_listener() { return ctx.method; });
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵpropertyInterpolate("label", ctx.label);
        ɵngcc0.ɵɵpropertyInterpolate("icon", ctx.icon);
        ɵngcc0.ɵɵpropertyInterpolate("styleClass", ctx.class);
        ɵngcc0.ɵɵproperty("model", ctx.items);
    } }, directives: [ɵngcc1.SplitButton], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DatatableSplitbuttonComponent, [{
        type: Component,
        args: [{
                selector: 'datatable-splitbutton',
                template: "<p-splitButton label=\"{{label}}\" icon=\"{{icon}}\" (onClick)=\"method\" [model]=\"items\" styleClass=\"{{class}}\">\n</p-splitButton>"
            }]
    }], function () { return []; }, { items: [{
            type: Input
        }], label: [{
            type: Input
        }], class: [{
            type: Input
        }], icon: [{
            type: Input
        }], method: [{
            type: Input
        }] }); })();
    return DatatableSplitbuttonComponent;
}());
export { DatatableSplitbuttonComponent };
