// @nuvem/angular-base/esm5/lib/security/authentication/authorization/authorization.service.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { Authorization } from './authorization';
import { Authentication } from '../authentication';
import * as i0 from "@angular/core";
import * as i1 from "../authentication";
/**
 * Serviço de autorização
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var AuthorizationService = /** @class */ (function (_super) {
    __extends(AuthorizationService, _super);
    /**
     * Método construtor para a injeção do serviço de configuração AuthConfig
     * @constructor
     * @param {AuthConfig} config
     */
    function AuthorizationService(authenticationService) {
        var _this = _super.call(this) || this;
        _this.authenticationService = authenticationService;
        return _this;
    }
    /**
     * Método responsável por identificar as regras de acesso
     * @public
     * @param {any} role
     * @returns boolean
     */
    AuthorizationService.prototype.hasRole = function (role) {
        if (role) {
            var user = this.authenticationService.getUser();
            if (role instanceof Array) {
                return user.roles.some(function (userRole) { return role.some(function (checkRole) { return userRole === checkRole; }); });
            }
            return user.roles.some(function (userRole) { return userRole === role; });
        }
        return true;
    };
    AuthorizationService.ctorParameters = function () { return [
        { type: Authentication }
    ]; };
    AuthorizationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AuthorizationService_Factory() { return new AuthorizationService(i0.ɵɵinject(i1.Authentication)); }, token: AuthorizationService, providedIn: "root" });
AuthorizationService.ɵfac = function AuthorizationService_Factory(t) { return new (t || AuthorizationService)(ɵngcc0.ɵɵinject(Authentication)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AuthorizationService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: Authentication }]; }, null); })();
    return AuthorizationService;
}(Authorization));
export { AuthorizationService };
