// @nuvem/angular-base/esm5/lib/security/authentication/authorization/has-role.directive.js
import { __decorate } from "tslib";
import { Directive, Input, TemplateRef, ViewContainerRef, OnDestroy } from '@angular/core';
import { Authorization } from './authorization';
import { Authentication } from '../authentication';
/**
 * HasRoleDirective diretiva para checagem da existência de permissões de acesso para o usuário logado.
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var HasRoleDirective = /** @class */ (function () {
    /**
     * Método construtor responsável por carregar o serviço de autorização e serviço de identificação de mudança
     * @param {TemplateRef<any>} templateRef
     * @param {ViewContainerRef} viewContainer
     * @param {Authorization} authorization
     * @param {Authentication<User>} authentication
     */
    function HasRoleDirective(templateRef, viewContainer, authorization, authentication) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.authorization = authorization;
        this.authentication = authentication;
    }
    Object.defineProperty(HasRoleDirective.prototype, "hasRole", {
        /**
         * Propriedade para identificar regras de acesso aos itens
         * @type {string}
         */
        set: function (hasRole) {
            var _this = this;
            this.viewContainer.clear();
            if (hasRole) {
                if (this.loginNotifications)
                    this.loginNotifications.unsubscribe();
                this.loginNotifications = this.authentication.getLoginNotifications().subscribe(function () {
                    if (_this.authorization.hasRole(hasRole)) {
                        _this.viewContainer.clear();
                        _this.viewContainer.createEmbeddedView(_this.templateRef);
                    }
                });
            }
            else {
                this.viewContainer.createEmbeddedView(this.templateRef);
            }
        },
        enumerable: true,
        configurable: true
    });
    HasRoleDirective.prototype.ngOnDestroy = function () {
        this.loginNotifications.unsubscribe();
    };
    HasRoleDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef },
        { type: Authorization },
        { type: Authentication }
    ]; };
    __decorate([
        Input()
    ], HasRoleDirective.prototype, "hasRole", null);
HasRoleDirective.ɵfac = function HasRoleDirective_Factory(t) { return new (t || HasRoleDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(Authorization), ɵngcc0.ɵɵdirectiveInject(Authentication)); };
HasRoleDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: HasRoleDirective, selectors: [["", "hasRole", ""]], inputs: { hasRole: "hasRole" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HasRoleDirective, [{
        type: Directive,
        args: [{
                selector: '[hasRole]'
            }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }, { type: ɵngcc0.ViewContainerRef }, { type: Authorization }, { type: Authentication }]; }, { hasRole: [{
            type: Input
        }] }); })();
    return HasRoleDirective;
}());
export { HasRoleDirective };
