// @nuvem/angular-base/esm5/lib/security/security.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LogoutDirective } from './authentication/logout.directive';
import { UserDirective } from './authentication/user.directive';
import { LoginSuccessComponent } from './authentication/login-success.component';
import { HasRoleDirective } from './authentication/authorization/has-role.directive';
import { RouterModule } from '@angular/router';
import { HttpClientModule } from '@angular/common/http';
import { HideWhileLoginComponent } from './authentication/hide-while-login.component';
import { AUTH_CONFIG } from './config/auth-config';
import { AuthorizationService } from './authentication/authorization/authorization.service';
import { AuthenticationService } from './authentication/authentication.service';
import { Authorization } from './authentication/authorization/authorization';
import { Authentication } from './authentication/authentication';
import { NotAuthenticatedErrorProvider } from './authentication/not-authenticated-error-provider.service';
import { ErrorProvider } from '../error/providers/error.provider';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { JWTAuthInterceptor } from './token/jwt-auth.interceptor';
import { Token } from './token/token';
import { JWTTokenService } from './token/jwt-token.service';
/**
 * O modulo de segurança contem funcionalidades para autenticação, autorização, redirecionamento
 * para login, gerenciamento de tokens JWT e obtenção e armazenamento de dados de usuário.
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var SecurityModule = /** @class */ (function () {
    function SecurityModule() {
    }
    SecurityModule_1 = SecurityModule;
    /**
     * forRoot method
     * @public
     * @static
     * @returns ModuleWithProviders
     */
    SecurityModule.forRoot = function (config) {
        return {
            ngModule: SecurityModule_1,
            providers: [
                { provide: AUTH_CONFIG, useValue: config },
                config.tokenStorageIndex ?
                    [
                        { provide: Token, useClass: JWTTokenService },
                        { provide: HTTP_INTERCEPTORS, useClass: JWTAuthInterceptor, multi: true }
                    ] : []
            ]
        };
    };
    var SecurityModule_1;
SecurityModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: SecurityModule });
SecurityModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function SecurityModule_Factory(t) { return new (t || SecurityModule)(); }, providers: [
        AuthorizationService,
        AuthenticationService,
        { provide: ErrorProvider, useClass: NotAuthenticatedErrorProvider, multi: true },
        { provide: Authorization, useClass: AuthorizationService },
        { provide: Authentication, useClass: AuthenticationService }
    ], imports: [[
            CommonModule,
            RouterModule,
            HttpClientModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SecurityModule, { declarations: function () { return [LogoutDirective,
        UserDirective,
        LoginSuccessComponent,
        HasRoleDirective,
        HideWhileLoginComponent]; }, imports: function () { return [CommonModule,
        RouterModule,
        HttpClientModule]; }, exports: function () { return [LogoutDirective,
        UserDirective,
        LoginSuccessComponent,
        HasRoleDirective,
        HideWhileLoginComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SecurityModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    RouterModule,
                    HttpClientModule
                ],
                declarations: [
                    LogoutDirective,
                    UserDirective,
                    LoginSuccessComponent,
                    HasRoleDirective,
                    HideWhileLoginComponent
                ],
                providers: [
                    AuthorizationService,
                    AuthenticationService,
                    { provide: ErrorProvider, useClass: NotAuthenticatedErrorProvider, multi: true },
                    { provide: Authorization, useClass: AuthorizationService },
                    { provide: Authentication, useClass: AuthenticationService }
                ],
                exports: [
                    LogoutDirective,
                    UserDirective,
                    LoginSuccessComponent,
                    HasRoleDirective,
                    HideWhileLoginComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return SecurityModule;
}());
export { SecurityModule };
