// @nuvem/angular-base/esm2015/lib/error/error.service.js
import { __decorate, __param } from "tslib";
import { ErrorHandler, Inject, Injectable } from '@angular/core';
import { ErrorProvider } from './providers/error.provider';
import * as i0 from "@angular/core";
import * as i1 from "./providers/error.provider";
/**
 * Class ErrorService
 * @class
 */
import * as ɵngcc0 from '@angular/core';
let ErrorService = class ErrorService extends ErrorHandler {
    /**
     * Metodo construtor responável por injetar o serviço ErrorProvider
     * @param {ErrorProvider} providers
     */
    constructor(providers) {
        super();
        this.providers = providers;
    }
    /**
     * Metodo responsável por idetificar e obter os erros lançados de requisições http/https
     * @param {Error} error
     * @returns void
     */
    handleError(error) {
        try {
            this.providers.forEach(p => {
                if (p.shouldHandle(error)) {
                    p.handle(error);
                }
            });
        }
        catch (e) {
        }
        super.handleError(error);
    }
};
ErrorService.ɵfac = function ErrorService_Factory(t) { return new (t || ErrorService)(ɵngcc0.ɵɵinject(ErrorProvider)); };
ErrorService.ctorParameters = () => [
    { type: Array, decorators: [{ type: Inject, args: [ErrorProvider,] }] }
];
ErrorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ErrorService_Factory() { return new ErrorService(i0.ɵɵinject(i1.ErrorProvider)); }, token: ErrorService, providedIn: "root" });
ErrorService = __decorate([ __param(0, Inject(ErrorProvider))
], ErrorService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ErrorService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: Array, decorators: [{
                type: Inject,
                args: [ErrorProvider]
            }] }]; }, null); })();
export { ErrorService };
