// @nuvem/primeng-components/esm2015/lib/error-stack/error-stack.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TableModule } from 'primeng/table';
import { ButtonModule, TooltipModule } from 'primeng';
import { ErrorStackComponent } from './error-stack.component';
import { ErrorProvider, ErrorStackService, ErrorStackProviderService, ClipboardModule } from '@nuvem/angular-base';
/**
 * Class ErrorStackModule
 * @class
 */
let ErrorStackModule = class ErrorStackModule {
};
ErrorStackModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            TableModule,
            ButtonModule,
            TooltipModule,
            ClipboardModule
        ],
        declarations: [
            ErrorStackComponent
        ],
        providers: [
            ErrorStackService,
            {
                provide: ErrorProvider,
                useClass: ErrorStackProviderService,
                multi: true
            }
        ],
        exports: [
            ErrorStackComponent
        ]
    })
], ErrorStackModule);
export { ErrorStackModule };
