// @nuvem/primeng-components/esm2015/lib/error-stack/error-stack.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
import { ErrorStackService, NamedErrorType } from '@nuvem/angular-base';
/**
 * Class ErrorStackComponent
 * @class
 */
let ErrorStackComponent = class ErrorStackComponent {
    /**
     * constructor method
     * @constructor
     * @param {ErrorStackService} errorStack
     */
    constructor(errorStack) {
        this.errorStack = errorStack;
        /**
         * title property
         * @type {string}
         */
        this.title = 'Error Stack';
        /**
         * errors property
         * @type {NamedErrorType}
         */
        this.errors = [];
    }
    /**
     * ngOnInit method
     * @returns void
     */
    ngOnInit() {
        this.errorSubscription = this.errorStack.errors.subscribe((error) => {
            this.errors = [error, ...this.errors];
        });
    }
    /**
     * ngOnDestroy method
     * @returns void
     */
    ngOnDestroy() {
        this.errorSubscription.unsubscribe();
    }
};
ErrorStackComponent.ctorParameters = () => [
    { type: ErrorStackService }
];
__decorate([
    Input()
], ErrorStackComponent.prototype, "title", void 0);
__decorate([
    Input()
], ErrorStackComponent.prototype, "template", void 0);
ErrorStackComponent = __decorate([
    Component({
        selector: 'basis-error-stack',
        template: "<ng-template #defaultTemplate>\n    <div class=\"ui-g\">\n        <div class=\"ui-g-12\">\n            <div class=\"card no-margin\">\n                <h1>{{ title }}</h1>\n                <p-table [value]=\"errors\" [paginator]=\"false\" [rows]=\"5\" first=\"0\">\n                    <ng-template pTemplate=\"header\">\n                        <tr>\n                            <th style=\"width: 20%; text-align: center\">Data</th>\n                            <th style=\"width: 20%; text-align: center\">C\u00F3digo de identifica\u00E7\u00E3o</th>\n                            <th>Mensagem</th>\n                            <th style=\"width: 10%; text-align: center\">A\u00E7\u00E3o</th>\n                        </tr>\n                    </ng-template>\n                    <ng-template pTemplate=\"body\" let-error>\n                        <tr>\n                            <td style=\"width: 20%; text-align: center\">{{error.createdAt|date: 'dd/MM/y H:mm:ss'}}</td>\n                            <td style=\"width: 20%; text-align: center\">{{error.correlationId || 'Indispon\u00EDvel'}}</td>\n                            <td>{{error.message}}</td>\n                            <td style=\"width: 10%; text-align: center\">\n                                <button pButton type=\"button\" pTooltip=\"Copiar erro para \u00E1rea de transfer\u00EAncia\" tooltipPosition=\"left\" class=\"ui-button-primary\" icon=\"ui-icon-content-copy\" [clipboard]=\"error.fullMessage\"></button>\n                            </td>\n                        </tr>\n                    </ng-template>\n                    <ng-template pTemplate=\"emptymessage\" let-columns>\n                        <tr>\n                            <td colspan=\"4\">\n                                Nenhum erro foi registrado at\u00E9 o momento.\n                            </td>\n                        </tr>\n                    </ng-template>\n                </p-table>\n            </div>\n        </div>\n    </div>\n</ng-template>\n<ng-container *ngTemplateOutlet=\"template? template: defaultTemplate\"></ng-container>\n"
    })
], ErrorStackComponent);
export { ErrorStackComponent };
