// @nuvem/angular-base/esm5/lib/form-utils/directive-superclass.js
import { __extends } from "tslib";
import { ChangeDetectorRef, } from "@angular/core";
import { Subject } from "rxjs";
import { filter, map, startWith } from "rxjs/operators";
import { InjectableSuperclass } from "./injectable-superclass";
/**
 * Extend this when creating a directive (including a component, which is a kind of directive) to gain access to the helpers demonstrated below. **Warning:** You _must_ include a constructor in your subclass.
 *
 * ```ts
 * @Component({
 *   selector: "s-color-text",
 *   template: `
 *     <span [style.background]="color">{{ color }}</span>
 *   `,
 *   // note that `bindToInstance()` works even with OnPush change detection
 *   changeDetection: ChangeDetectionStrategy.OnPush,
 * })
 *  class ColorTextComponent extends DirectiveSuperclass {
 *   @Input() prefix?: string;
 *   @Input() prefix2?: string;
 *   color!: string;
 *
 *   // Even if you don't need extra arguments injector, you must still include a constructor. It is required for angular to provide `Injector`.
 *   constructor(
 *     @Inject("color$") color$: Observable<string>,
 *     injector: Injector,
 *   ) {
 *     super(injector);
 *
 *     // combine everything to calculate `color` and keep it up to date
 *     this.bindToInstance(
 *       "color",
 *       combineLatest(
 *         this.getInput$("prefix"),
 *         this.getInput$("prefix2"),
 *         color$,
 *       ).pipe(map((parts) => parts.filter((p) => p).join(""))),
 *     );
 *   }
 * }
 * ```
 */
import * as ɵngcc0 from '@angular/core';
var DirectiveSuperclass = /** @class */ (function (_super) {
    __extends(DirectiveSuperclass, _super);
    function DirectiveSuperclass(injector) {
        var _this = _super.call(this) || this;
        /**
         *  Emits the set of `@Input()` property names that change during each call to `ngOnChanges()`.
         */
        _this.inputChanges$ = new Subject();
        _this.changeDetectorRef = injector.get(ChangeDetectorRef);
        return _this;
    }
    DirectiveSuperclass.prototype.ngOnChanges = function (changes) {
        this.inputChanges$.next(new Set(Object.getOwnPropertyNames(changes)));
    };
    /**
     * @return an observable of the values for one of this directive's `@Input()` properties
     */
    DirectiveSuperclass.prototype.getInput$ = function (key) {
        var _this = this;
        return this.inputChanges$.pipe(filter(function (keys) { return keys.has(key); }), startWith(undefined), map(function () { return _this[key]; }));
    };
    /**
     * Binds an observable to one of this directive's instance variables. When the observable emits the instance variable will be updated, and change detection will be triggered to propagate any changes. Use this an an alternative to repeating `| async` multiple times in your template.
     */
    DirectiveSuperclass.prototype.bindToInstance = function (key, value$) {
        var _this = this;
        this.subscribeTo(value$, function (value) {
            _this[key] = value;
            _this.changeDetectorRef.markForCheck();
        });
    };
DirectiveSuperclass.ɵfac = function DirectiveSuperclass_Factory(t) { ɵngcc0.ɵɵinvalidFactory(); };
DirectiveSuperclass.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: DirectiveSuperclass, features: [ɵngcc0.ɵɵInheritDefinitionFeature, ɵngcc0.ɵɵNgOnChangesFeature] });

    return DirectiveSuperclass;
}(InjectableSuperclass));
export { DirectiveSuperclass };
