// @nuvem/primeng-components/esm5/lib/datatable/datatable-button.component.js
import { __decorate } from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
/**
 * Class DatatableButtonComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'primeng';
var DatatableButtonComponent = /** @class */ (function () {
    function DatatableButtonComponent() {
        /**
         * allowMultipleSelection property
         * @type {boolean}
         */
        this.allowMultipleSelection = true;
        /**
         * disabled property
         * @type boolean
         */
        this.disabled = false;
        /**
         * click property
         * @type {EventEmitter<any>}
         */
        this.click = new EventEmitter();
    }
    /**
     * onClick method
     * @param {any} event
     * @return void
     */
    DatatableButtonComponent.prototype.onClick = function (event) {
        this.click.emit(event);
        event.stopPropagation();
    };
    __decorate([
        Input()
    ], DatatableButtonComponent.prototype, "icon", void 0);
    __decorate([
        Input()
    ], DatatableButtonComponent.prototype, "class", void 0);
    __decorate([
        Input()
    ], DatatableButtonComponent.prototype, "bTooltip", void 0);
    __decorate([
        Input()
    ], DatatableButtonComponent.prototype, "name", void 0);
    __decorate([
        Input()
    ], DatatableButtonComponent.prototype, "allowMultipleSelection", void 0);
    __decorate([
        Input()
    ], DatatableButtonComponent.prototype, "disabled", void 0);
    __decorate([
        Output()
    ], DatatableButtonComponent.prototype, "click", void 0);
DatatableButtonComponent.ɵfac = function DatatableButtonComponent_Factory(t) { return new (t || DatatableButtonComponent)(); };
DatatableButtonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DatatableButtonComponent, selectors: [["basis-datatable-button"]], inputs: { allowMultipleSelection: "allowMultipleSelection", disabled: "disabled", icon: "icon", class: "class", bTooltip: "bTooltip", name: "name" }, outputs: { click: "click" }, decls: 2, vars: 6, consts: [[3, "pTooltip"], ["type", "button", "tooltipPosition", "bottom", "pButton", "", 3, "icon", "disabled", "click"]], template: function DatatableButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵlistener("click", function DatatableButtonComponent_Template_button_click_1_listener($event) { return ctx.onClick($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵpropertyInterpolate("pTooltip", ctx.bTooltip);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMap(ctx.class);
        ɵngcc0.ɵɵpropertyInterpolate1("icon", "ui-icon-", ctx.icon, "");
        ɵngcc0.ɵɵproperty("disabled", ctx.disabled);
    } }, directives: [ɵngcc1.Tooltip, ɵngcc1.ButtonDirective], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DatatableButtonComponent, [{
        type: Component,
        args: [{
                selector: 'basis-datatable-button',
                template: "<p pTooltip=\"{{ bTooltip }}\">\n    <button\n        type=\"button\" tooltipPosition=\"bottom\" class=\"{{ class }}\"\n        icon=\"ui-icon-{{ icon }}\" (click)=\"onClick($event)\" [disabled]=\"disabled\" pButton>\n    </button>\n</p>\n"
            }]
    }], function () { return []; }, { allowMultipleSelection: [{
            type: Input
        }], disabled: [{
            type: Input
        }], click: [{
            type: Output
        }], icon: [{
            type: Input
        }], class: [{
            type: Input
        }], bTooltip: [{
            type: Input
        }], name: [{
            type: Input
        }] }); })();
    return DatatableButtonComponent;
}());
export { DatatableButtonComponent };
