// @nuvem/primeng-components/esm5/lib/datatable/datatable.module.js
import { __decorate } from "tslib";
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule, TooltipModule, BlockUIModule, ProgressSpinnerModule, SplitButtonModule, PaginatorModule, InputTextModule } from 'primeng';
import { TableModule } from 'primeng/table';
import { DataProviderFactory } from './dataprovider-factory';
import { DatatableComponent } from './datatable.component';
import { DatatableButtonComponent } from './datatable-button.component';
import { DatatableService } from './datatable.service';
import { DatatableSplitbuttonModule } from '../extra-button/datatable-splitbutton.module';
import { FilterColumnModule } from './filter-column/filter-column.module';
import { ShowColumnModule } from './show-column/show-column.module';
import { Column } from './p-column.directive';
/**
 * Class DatatableModule
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var DatatableModule = /** @class */ (function () {
    function DatatableModule() {
    }
DatatableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DatatableModule });
DatatableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DatatableModule_Factory(t) { return new (t || DatatableModule)(); }, providers: [
        { provide: DatatableService, useClass: DatatableService, deps: [HttpClient] },
        DataProviderFactory
    ], imports: [[
            CommonModule,
            ButtonModule,
            SplitButtonModule,
            TooltipModule,
            TableModule,
            BlockUIModule,
            ProgressSpinnerModule,
            DatatableSplitbuttonModule,
            HttpClientModule,
            PaginatorModule,
            FilterColumnModule,
            ShowColumnModule,
            InputTextModule
        ],
        FilterColumnModule,
        ShowColumnModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DatatableModule, { declarations: function () { return [DatatableComponent,
        DatatableButtonComponent,
        Column]; }, imports: function () { return [CommonModule,
        ButtonModule,
        SplitButtonModule,
        TooltipModule,
        TableModule,
        BlockUIModule,
        ProgressSpinnerModule,
        DatatableSplitbuttonModule,
        HttpClientModule,
        PaginatorModule,
        FilterColumnModule,
        ShowColumnModule,
        InputTextModule]; }, exports: function () { return [DatatableComponent,
        DatatableButtonComponent,
        FilterColumnModule,
        ShowColumnModule,
        Column]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DatatableModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    ButtonModule,
                    SplitButtonModule,
                    TooltipModule,
                    TableModule,
                    BlockUIModule,
                    ProgressSpinnerModule,
                    DatatableSplitbuttonModule,
                    HttpClientModule,
                    PaginatorModule,
                    FilterColumnModule,
                    ShowColumnModule,
                    InputTextModule
                ],
                declarations: [
                    DatatableComponent,
                    DatatableButtonComponent,
                    Column
                ],
                providers: [
                    { provide: DatatableService, useClass: DatatableService, deps: [HttpClient] },
                    DataProviderFactory
                ],
                exports: [
                    DatatableComponent,
                    DatatableButtonComponent,
                    FilterColumnModule,
                    ShowColumnModule,
                    Column
                ]
            }]
    }], function () { return []; }, null); })();
    return DatatableModule;
}());
export { DatatableModule };
