// @nuvem/primeng-components/esm5/lib/datatable/p-column.directive.js
import { __decorate } from "tslib";
import { Input, EventEmitter, TemplateRef, Output, ContentChildren, ContentChild, Directive } from '@angular/core';
import { PrimeTemplate } from 'primeng';
import * as ɵngcc0 from '@angular/core';
var Column = /** @class */ (function () {
    function Column() {
        this.filterType = 'text';
        this.exportable = true;
        this.resizable = true;
        this.sortFunction = new EventEmitter();
    }
    Column.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.templates.forEach(function (item) {
            switch (item.getType()) {
                case 'header':
                    _this.headerTemplate = item.template;
                    break;
                case 'body':
                    _this.bodyTemplate = item.template;
                    break;
                case 'footer':
                    _this.footerTemplate = item.template;
                    break;
                case 'filter':
                    _this.filterTemplate = item.template;
                    break;
                case 'editor':
                    _this.editorTemplate = item.template;
                    break;
                case 'input':
                    _this.inputTemplate = item.template;
                    break;
                case 'output':
                    _this.outputTemplate = item.template;
                    break;
                default:
                    _this.bodyTemplate = item.template;
                    break;
            }
        });
    };
    __decorate([
        Input()
    ], Column.prototype, "field", void 0);
    __decorate([
        Input()
    ], Column.prototype, "colId", void 0);
    __decorate([
        Input()
    ], Column.prototype, "sortField", void 0);
    __decorate([
        Input()
    ], Column.prototype, "filterField", void 0);
    __decorate([
        Input()
    ], Column.prototype, "header", void 0);
    __decorate([
        Input()
    ], Column.prototype, "footer", void 0);
    __decorate([
        Input()
    ], Column.prototype, "sortable", void 0);
    __decorate([
        Input()
    ], Column.prototype, "editable", void 0);
    __decorate([
        Input()
    ], Column.prototype, "filter", void 0);
    __decorate([
        Input()
    ], Column.prototype, "filterMatchMode", void 0);
    __decorate([
        Input()
    ], Column.prototype, "filterType", void 0);
    __decorate([
        Input()
    ], Column.prototype, "excludeGlobalFilter", void 0);
    __decorate([
        Input()
    ], Column.prototype, "rowspan", void 0);
    __decorate([
        Input()
    ], Column.prototype, "colspan", void 0);
    __decorate([
        Input()
    ], Column.prototype, "scope", void 0);
    __decorate([
        Input()
    ], Column.prototype, "style", void 0);
    __decorate([
        Input()
    ], Column.prototype, "styleClass", void 0);
    __decorate([
        Input()
    ], Column.prototype, "exportable", void 0);
    __decorate([
        Input()
    ], Column.prototype, "headerStyle", void 0);
    __decorate([
        Input()
    ], Column.prototype, "headerStyleClass", void 0);
    __decorate([
        Input()
    ], Column.prototype, "bodyStyle", void 0);
    __decorate([
        Input()
    ], Column.prototype, "bodyStyleClass", void 0);
    __decorate([
        Input()
    ], Column.prototype, "footerStyle", void 0);
    __decorate([
        Input()
    ], Column.prototype, "footerStyleClass", void 0);
    __decorate([
        Input()
    ], Column.prototype, "hidden", void 0);
    __decorate([
        Input()
    ], Column.prototype, "expander", void 0);
    __decorate([
        Input()
    ], Column.prototype, "selectionMode", void 0);
    __decorate([
        Input()
    ], Column.prototype, "filterPlaceholder", void 0);
    __decorate([
        Input()
    ], Column.prototype, "filterMaxlength", void 0);
    __decorate([
        Input()
    ], Column.prototype, "frozen", void 0);
    __decorate([
        Input()
    ], Column.prototype, "resizable", void 0);
    __decorate([
        Output()
    ], Column.prototype, "sortFunction", void 0);
    __decorate([
        ContentChildren(PrimeTemplate)
    ], Column.prototype, "templates", void 0);
    __decorate([
        ContentChild(TemplateRef)
    ], Column.prototype, "template", void 0);
Column.ɵfac = function Column_Factory(t) { return new (t || Column)(); };
Column.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: Column, selectors: [["p-column"]], contentQueries: function Column_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, TemplateRef, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, PrimeTemplate, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templates = _t);
    } }, inputs: { filterType: "filterType", exportable: "exportable", resizable: "resizable", field: "field", colId: "colId", sortField: "sortField", filterField: "filterField", header: "header", footer: "footer", sortable: "sortable", editable: "editable", filter: "filter", filterMatchMode: "filterMatchMode", excludeGlobalFilter: "excludeGlobalFilter", rowspan: "rowspan", colspan: "colspan", scope: "scope", style: "style", styleClass: "styleClass", headerStyle: "headerStyle", headerStyleClass: "headerStyleClass", bodyStyle: "bodyStyle", bodyStyleClass: "bodyStyleClass", footerStyle: "footerStyle", footerStyleClass: "footerStyleClass", hidden: "hidden", expander: "expander", selectionMode: "selectionMode", filterPlaceholder: "filterPlaceholder", filterMaxlength: "filterMaxlength", frozen: "frozen" }, outputs: { sortFunction: "sortFunction" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Column, [{
        type: Directive,
        args: [{
                selector: 'p-column'
            }]
    }], function () { return []; }, { filterType: [{
            type: Input
        }], exportable: [{
            type: Input
        }], resizable: [{
            type: Input
        }], sortFunction: [{
            type: Output
        }], field: [{
            type: Input
        }], colId: [{
            type: Input
        }], sortField: [{
            type: Input
        }], filterField: [{
            type: Input
        }], header: [{
            type: Input
        }], footer: [{
            type: Input
        }], sortable: [{
            type: Input
        }], editable: [{
            type: Input
        }], filter: [{
            type: Input
        }], filterMatchMode: [{
            type: Input
        }], excludeGlobalFilter: [{
            type: Input
        }], rowspan: [{
            type: Input
        }], colspan: [{
            type: Input
        }], scope: [{
            type: Input
        }], style: [{
            type: Input
        }], styleClass: [{
            type: Input
        }], headerStyle: [{
            type: Input
        }], headerStyleClass: [{
            type: Input
        }], bodyStyle: [{
            type: Input
        }], bodyStyleClass: [{
            type: Input
        }], footerStyle: [{
            type: Input
        }], footerStyleClass: [{
            type: Input
        }], hidden: [{
            type: Input
        }], expander: [{
            type: Input
        }], selectionMode: [{
            type: Input
        }], filterPlaceholder: [{
            type: Input
        }], filterMaxlength: [{
            type: Input
        }], frozen: [{
            type: Input
        }], templates: [{
            type: ContentChildren,
            args: [PrimeTemplate]
        }], template: [{
            type: ContentChild,
            args: [TemplateRef]
        }] }); })();
    return Column;
}());
export { Column };
