// @nuvem/angular-base/esm2015/lib/directives/only-numbers.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, HostListener } from '@angular/core';
/**
 * OnlyNumbersDirective diretiva para que sejam permitidos somente números
 * @class
 */
import * as ɵngcc0 from '@angular/core';
let OnlyNumbersDirective = class OnlyNumbersDirective {
    /**
     * @param {ElementRef} el
     * @constructor
     */
    constructor(el) {
        this.el = el;
        /**
         * Propriedade que carrega expressão regular para numeros
         * @type {RegExp}
         */
        this.regex = new RegExp(/[^0-9]$/g);
        /**
         * Propriedade que lista opções aceitas além de núsmeros
         * @type {Array<string>}
         */
        this.specialKeys = [
            'Backspace',
            'Delete',
            'Tab',
            'End',
            'Home',
            'ArrowLeft',
            'ArrowRight'
        ];
    }
    /**
     * Metodo executado para os evento KeyDown do teclado executa verificação do valor digitado
     * @param {KeyboardEvent} event
     * @returns void
     */
    onKeyDown(event) {
        const current = this.el.nativeElement.value;
        const next = current.concat(event.key);
        if (this.specialKeys.indexOf(event.key) !== -1) {
            return;
        }
        if (next && String(next).match(this.regex)) {
            event.preventDefault();
        }
    }
};
OnlyNumbersDirective.ɵfac = function OnlyNumbersDirective_Factory(t) { return new (t || OnlyNumbersDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
OnlyNumbersDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: OnlyNumbersDirective, selectors: [["input", "onlyNumbers", ""]], hostBindings: function OnlyNumbersDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown", function OnlyNumbersDirective_keydown_HostBindingHandler($event) { return ctx.onKeyDown($event); });
    } } });
OnlyNumbersDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    HostListener('keydown', ['$event'])
], OnlyNumbersDirective.prototype, "onKeyDown", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(OnlyNumbersDirective, [{
        type: Directive,
        args: [{
                selector: 'input[onlyNumbers]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { onKeyDown: [{
            type: HostListener,
            args: ['keydown', ['$event']]
        }] }); })();
export { OnlyNumbersDirective };
