// @nuvem/angular-base/esm2015/lib/directives/unmask.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, OnInit, Input, OnDestroy } from '@angular/core';
import { NgControl } from '@angular/forms';
/**
 *
 * UnmaskDirective diretiva para remoção de mascaras de valores numericos
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
let UnmaskDirective = class UnmaskDirective {
    /**
     * constructor method
     * @param {ElementRef} elementRef
     * @param {NgControl} model
     * @constructor
     */
    constructor(elementRef, model) {
        this.elementRef = elementRef;
        this.model = model;
    }
    /**
     * Metodo executado no carregamento da diretiva, executa verificação do valor e remove as mascaras baseado em um aexpressão regular
     * @returns void
     */
    ngOnInit() {
        this.subscriber = this.model.control.valueChanges.subscribe(() => {
            const newValue = this.elementRef.nativeElement.value.replace(new RegExp(this.appUnmask), '');
            this.model.control.setValue(newValue, {
                emitEvent: false,
                emitModelToViewChange: false,
                emitViewToModelChange: false
            });
        });
    }
    /**
     * ngOnDestroy method
     * @returns void
     */
    ngOnDestroy() {
        this.subscriber.unsubscribe();
    }
};
UnmaskDirective.ɵfac = function UnmaskDirective_Factory(t) { return new (t || UnmaskDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgControl)); };
UnmaskDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: UnmaskDirective, selectors: [["", "unmask", ""]], inputs: { appUnmask: "appUnmask" } });
UnmaskDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgControl }
];
__decorate([
    Input()
], UnmaskDirective.prototype, "appUnmask", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(UnmaskDirective, [{
        type: Directive,
        args: [{
                selector: '[unmask]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.NgControl }]; }, { appUnmask: [{
            type: Input
        }] }); })();
export { UnmaskDirective };
