// @nuvem/primeng-components/esm5/lib/page-notification/page-notification.service.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { MessageService } from 'primeng';
import { NotificationProvider } from '@nuvem/angular-base';
import * as i0 from "@angular/core";
import * as i1 from "primeng/api";
/**
 * Class PageNotificationService
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'primeng';
var PageNotificationService = /** @class */ (function (_super) {
    __extends(PageNotificationService, _super);
    /**
     * constructor method
     * @constructor
     * @param {MessageService} messageService
     */
    function PageNotificationService(messageService) {
        var _this = _super.call(this) || this;
        _this.messageService = messageService;
        /**
         * successSeverity property
         * @type {string}
         */
        _this.successSeverity = 'success';
        /**
         * infoSeverity property
         * @type {string}
         */
        _this.infoSeverity = 'info';
        /**
         * warnSeverity property
         * @type {string}
         */
        _this.warnSeverity = 'warn';
        /**
         * errorSeverity property
         * @type {string}
         */
        _this.errorSeverity = 'error';
        /**
         * createMsg property
         * @type {string}
         */
        _this.createMsg = 'Registro incluído com sucesso!';
        /**
         * updateMsg property
         * @type {string}
         */
        _this.updateMsg = 'Registro alterado com sucesso!';
        /**
         * deleteMsg property
         * @type {string}
         */
        _this.deleteMsg = 'Registro excluído com sucesso!';
        return _this;
    }
    /**
     * addCreateMsg method
     * @param {string} title
     */
    PageNotificationService.prototype.addCreateMsg = function (title) {
        this.addSuccessMessage(this.createMsg, title);
    };
    /**
     * addSuccessMessage method
     * @param {string} message
     * @param {string} title
     */
    PageNotificationService.prototype.addSuccessMessage = function (message, title) {
        this.addMsg(this.successSeverity, message, title);
    };
    /**
     * addMsg method
     * @param {string} severity
     * @param {string} msg
     * @param {string} title
     */
    PageNotificationService.prototype.addMsg = function (severity, msg, title) {
        this.messageService.add({
            severity: severity,
            summary: title,
            detail: msg
        });
    };
    /**
     * addUpdateMsg method
     * @param {string} title
     */
    PageNotificationService.prototype.addUpdateMsg = function (title) {
        this.addSuccessMessage(this.updateMsg, title);
    };
    /**
     * addDeleteMsg method
     * @param {string} title
     */
    PageNotificationService.prototype.addDeleteMsg = function (title) {
        this.addSuccessMessage(this.deleteMsg, title);
    };
    /**
     * addInfoMessage method
     * @param {string} message
     * @param {string} title
     */
    PageNotificationService.prototype.addInfoMessage = function (message, title) {
        this.addMsg(this.infoSeverity, message, title);
    };
    /**
     * addWarnMessage method
     * @param {string} message
     * @param {string} title
     */
    PageNotificationService.prototype.addWarnMessage = function (message, title) {
        this.addMsg(this.warnSeverity, message, title);
    };
    /**
     * addErrorMessage method
     * @param {string} message
     * @param {string} title
     */
    PageNotificationService.prototype.addErrorMessage = function (message, title) {
        this.addMsg(this.errorSeverity, message, title);
    };
    /**
     * addErrorProblem method
     * @param {ApplicationProblemType} problem
     */
    PageNotificationService.prototype.addErrorProblem = function (problem) {
        this.addMsg(this.errorSeverity, problem.detail, problem.title);
    };
    PageNotificationService.ctorParameters = function () { return [
        { type: MessageService }
    ]; };
    PageNotificationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PageNotificationService_Factory() { return new PageNotificationService(i0.ɵɵinject(i1.MessageService)); }, token: PageNotificationService, providedIn: "root" });
PageNotificationService.ɵfac = function PageNotificationService_Factory(t) { return new (t || PageNotificationService)(ɵngcc0.ɵɵinject(ɵngcc1.MessageService)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PageNotificationService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.MessageService }]; }, null); })();
    return PageNotificationService;
}(NotificationProvider));
export { PageNotificationService };
