// @nuvem/primeng-components/esm5/lib/page-notification/page-notification.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ToastModule } from 'primeng/toast';
import { MessageService } from 'primeng/api';
import { PageNotificationService } from './page-notification.service';
import { PageNotificationComponent } from './page-notification.component';
import { NotificationProvider } from '@nuvem/angular-base';
/**
 * Class PageNotificationModule
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var PageNotificationModule = /** @class */ (function () {
    function PageNotificationModule() {
    }
PageNotificationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: PageNotificationModule });
PageNotificationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function PageNotificationModule_Factory(t) { return new (t || PageNotificationModule)(); }, providers: [
        MessageService,
        PageNotificationService,
        {
            provide: NotificationProvider,
            useExisting: PageNotificationService
        }
    ], imports: [[
            CommonModule,
            ToastModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PageNotificationModule, { declarations: function () { return [PageNotificationComponent]; }, imports: function () { return [CommonModule,
        ToastModule]; }, exports: function () { return [PageNotificationComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PageNotificationModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    ToastModule
                ],
                declarations: [
                    PageNotificationComponent
                ],
                exports: [
                    PageNotificationComponent
                ],
                providers: [
                    MessageService,
                    PageNotificationService,
                    {
                        provide: NotificationProvider,
                        useExisting: PageNotificationService
                    }
                ]
            }]
    }], function () { return []; }, null); })();
    return PageNotificationModule;
}());
export { PageNotificationModule };
