// @nuvem/primeng-components/esm5/lib/crud/components/text-area/text-area.component.js
import { __decorate, __extends } from "tslib";
import { Component, Input } from "@angular/core";
import { provideValueAccessor } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../form-notification/form-notification.component';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from 'primeng/inputtextarea';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../validation-message/validation-message.component';

var _c0 = ["nTextArea", ""];
function TextAreaComponent_nValidationMessage_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "nValidationMessage", 6);
} if (rf & 2) {
    var ctx_r948 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("control", ctx_r948.formControl);
} }
var _c1 = function (a0) { return { "ng-invalid ng-dirty": a0 }; };
var _c2 = ["*"];
var TextAreaComponent = /** @class */ (function (_super) {
    __extends(TextAreaComponent, _super);
    function TextAreaComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.autoResize = false;
        return _this;
    }
    TextAreaComponent_1 = TextAreaComponent;
    var TextAreaComponent_1;
    __decorate([
        Input()
    ], TextAreaComponent.prototype, "autoResize", void 0);
    __decorate([
        Input()
    ], TextAreaComponent.prototype, "rows", void 0);
    __decorate([
        Input()
    ], TextAreaComponent.prototype, "cols", void 0);
    __decorate([
        Input()
    ], TextAreaComponent.prototype, "minLength", void 0);
    __decorate([
        Input()
    ], TextAreaComponent.prototype, "maxLength", void 0);
TextAreaComponent.ɵfac = function TextAreaComponent_Factory(t) { return ɵTextAreaComponent_BaseFactory(t || TextAreaComponent); };
TextAreaComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: TextAreaComponent, selectors: [["", "nTextArea", ""]], inputs: { autoResize: "autoResize", rows: "rows", cols: "cols", minLength: "minLength", maxLength: "maxLength" }, features: [ɵngcc0.ɵɵProvidersFeature([provideValueAccessor(TextAreaComponent_1)]), ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c2, decls: 8, vars: 15, consts: [["form-notification", "", 1, "md-inputfield", 3, "name", "form"], ["notification", ""], ["pInputTextarea", "", "type", "text", 3, "ngClass", "name", "autoResize", "rows", "cols", "minLength", "maxLength", "formControl", "blur", "focus"], ["input", ""], [3, "for", "innerHTML"], [3, "control", 4, "ngIf"], [3, "control"]], template: function TextAreaComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "span", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "textarea", 2, 3);
        ɵngcc0.ɵɵlistener("blur", function TextAreaComponent_Template_textarea_blur_2_listener($event) { return ctx.onBlurEvent($event); })("focus", function TextAreaComponent_Template_textarea_focus_2_listener($event) { return ctx.onFocusEvent($event); });
        ɵngcc0.ɵɵtext(4, "    ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(5, "label", 4);
        ɵngcc0.ɵɵtemplate(6, TextAreaComponent_nValidationMessage_6_Template, 1, 1, "nValidationMessage", 5);
        ɵngcc0.ɵɵprojection(7);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        var _r946 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵproperty("name", ctx.name)("form", ctx.form);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(13, _c1, _r946.message))("name", ctx.name)("autoResize", ctx.autoResize)("rows", ctx.rows)("cols", ctx.cols)("minLength", ctx.minLength)("maxLength", ctx.maxLength)("formControl", ctx.formControl);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("for", ctx.name)("innerHTML", ctx.label, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !_r946.message);
    } }, directives: [ɵngcc1.FormNotificationComponent, ɵngcc2.DefaultValueAccessor, ɵngcc3.InputTextarea, ɵngcc4.NgClass, ɵngcc2.NgControlStatus, ɵngcc2.FormControlDirective, ɵngcc4.NgIf, ɵngcc5.ValidationMessageComponent], encapsulation: 2 });
var ɵTextAreaComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(TextAreaComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TextAreaComponent, [{
        type: Component,
        args: [{
                selector: "[nTextArea]",
                template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <textarea #input\n        pInputTextarea\n        [ngClass]=\"{ 'ng-invalid ng-dirty': notification.message }\"\n        type=\"text\"\n        [name]=\"name\"\n        [autoResize]=\"autoResize\"\n        [rows]=\"rows\"\n        [cols]=\"cols\"\n        [minLength]=\"minLength\"\n        [maxLength]=\"maxLength\"\n        [formControl]=\"formControl\"\n        (blur)=\"onBlurEvent($event)\"\n        (focus)=\"onFocusEvent($event)\">\n    </textarea>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
                providers: [provideValueAccessor(TextAreaComponent_1)]
            }]
    }], null, { autoResize: [{
            type: Input
        }], rows: [{
            type: Input
        }], cols: [{
            type: Input
        }], minLength: [{
            type: Input
        }], maxLength: [{
            type: Input
        }] }); })();
    return TextAreaComponent;
}(FieldComponent));
export { TextAreaComponent };
