// @nuvem/angular-base/esm5/lib/version-tag/version-tag.component.js
import { __decorate } from "tslib";
import { Component, Input } from '@angular/core';
/**
 * Componente responsável por expor o seletor <app-version-tag></app-version-tag> para reuso
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var VersionTagComponent = /** @class */ (function () {
    function VersionTagComponent() {
    }
    /**
     * Metodo executado ao carregar o componente responsável por carregar o arquivo 'package.json'
     * @return void
     */
    VersionTagComponent.prototype.ngOnInit = function () {
    };
    __decorate([
        Input()
    ], VersionTagComponent.prototype, "cssClass", void 0);
VersionTagComponent.ɵfac = function VersionTagComponent_Factory(t) { return new (t || VersionTagComponent)(); };
VersionTagComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: VersionTagComponent, selectors: [["app-version-tag"]], inputs: { cssClass: "cssClass" }, decls: 2, vars: 1, consts: [[2, "margin-left", "5px"]], template: function VersionTagComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "span", 0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" Vers\u00E3o: ", ctx.version, "\n");
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VersionTagComponent, [{
        type: Component,
        args: [{
                selector: 'app-version-tag',
                template: "<span style=\"margin-left: 5px;\">\n    Vers\u00E3o: {{version}}\n</span>"
            }]
    }], function () { return []; }, { cssClass: [{
            type: Input
        }] }); })();
    return VersionTagComponent;
}());
export { VersionTagComponent };
