// @nuvem/angular-base/esm5/lib/version-tag/version-tag.module.js
import { __decorate } from "tslib";
import { VersionTagComponent } from './version-tag.component';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
/**
 * O componente VersionTag exibe o atributo `version` do arquivo `package.json` do sistema,
 * com o intuito de informar o usuário a versão atual da aplicação. Inicialmente foi feito para
 * ser utilizado no rodapé das aplicações conforme o exemplo a seguir.
 * @module
 */
import * as ɵngcc0 from '@angular/core';
var VersionTagModule = /** @class */ (function () {
    function VersionTagModule() {
    }
VersionTagModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: VersionTagModule });
VersionTagModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function VersionTagModule_Factory(t) { return new (t || VersionTagModule)(); }, imports: [[
            CommonModule,
            RouterModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(VersionTagModule, { declarations: function () { return [VersionTagComponent]; }, imports: function () { return [CommonModule,
        RouterModule]; }, exports: function () { return [VersionTagComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VersionTagModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    RouterModule
                ],
                declarations: [
                    VersionTagComponent
                ],
                exports: [
                    VersionTagComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return VersionTagModule;
}());
export { VersionTagModule };
