// @nuvem/angular-base/esm5/lib/security/authentication/authentication.service.js
import { __decorate, __extends, __param } from "tslib";
import { Injectable, Inject } from '@angular/core';
import { AUTH_CONFIG } from '../config/auth-config';
import { BehaviorSubject } from 'rxjs';
import { Authentication } from './authentication';
import { HttpClient } from '@angular/common/http';
import { filter } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "../config/auth-config";
import * as i2 from "@angular/common/http";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common/http';
export function isAuthenticated(config) {
    return null !== config.storage.getItem(config.userStorageIndex);
}
;
export function redirect(config) {
    window.location.href = config.loginUrl;
    ;
}
;
/**
 * Serviço de autenticação
 * @class
 */
var AuthenticationService = /** @class */ (function (_super) {
    __extends(AuthenticationService, _super);
    function AuthenticationService(config, http) {
        var _this = _super.call(this) || this;
        _this.config = config;
        _this.http = http;
        _this.loginNotifications = new BehaviorSubject(_this.getUser());
        return _this;
    }
    /**
     * Método para verificar se o usuário esta autenticado na aplicação
     * @returns boolean
     * @public
     */
    AuthenticationService.prototype.isAuthenticated = function () {
        return isAuthenticated(this.config);
    };
    /**
     * Método responsável por realizar redirecionamentos
     * @returns void
     * @public
     */
    AuthenticationService.prototype.redirect = function () {
        redirect(this.config);
    };
    /**
     * Método responsável pelo login da aplicação
     * @public
     * @param {User} user
     * @returns Observable<any>
     */
    AuthenticationService.prototype.login = function () {
        var _this = this;
        this.getUserDetails()
            .subscribe(function (user) {
            _this.setUser(user);
            _this.loginNotifications.next(user);
        });
    };
    /**
     * Método responsável pelo logou da aplicação
     * @public
     * @returns void
     */
    AuthenticationService.prototype.logout = function () {
        this.loginNotifications.next(null);
        this.config.storage.removeItem(this.config.userStorageIndex);
        window.location.href = this.config.logoutUrl;
    };
    /**
     * Método responsável por obter os dados do usuário logado
     * @public
     * @returns Observable<any>
     */
    AuthenticationService.prototype.getUserDetails = function () {
        return this.http.get(this.config.baseUrl + this.config.detailsUrl);
    };
    AuthenticationService.prototype.setUser = function (user) {
        this.config.storage.setItem(this.config.userStorageIndex, JSON.stringify(user));
    };
    AuthenticationService.prototype.getUser = function () {
        return JSON.parse(this.config.storage.getItem(this.config.userStorageIndex));
    };
    AuthenticationService.prototype.getLoginNotifications = function () {
        return this.loginNotifications.pipe(filter(function (value) { return value != null; }));
    };
    AuthenticationService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [AUTH_CONFIG,] }] },
        { type: HttpClient }
    ]; };
    AuthenticationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AuthenticationService_Factory() { return new AuthenticationService(i0.ɵɵinject(i1.AUTH_CONFIG), i0.ɵɵinject(i2.HttpClient)); }, token: AuthenticationService, providedIn: "root" });
    AuthenticationService = __decorate([ __param(0, Inject(AUTH_CONFIG))
    ], AuthenticationService);
AuthenticationService.ɵfac = function AuthenticationService_Factory(t) { return new (t || AuthenticationService)(ɵngcc0.ɵɵinject(AUTH_CONFIG), ɵngcc0.ɵɵinject(ɵngcc1.HttpClient)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AuthenticationService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [AUTH_CONFIG]
            }] }, { type: ɵngcc1.HttpClient }]; }, null); })();
    return AuthenticationService;
}(Authentication));
export { AuthenticationService };
