// @nuvem/angular-base/esm5/lib/security/authentication/hide-while-login.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, OnDestroy, OnInit, Input, AfterViewInit, AfterContentInit } from '@angular/core';
import { Authentication } from './authentication';
import { BlockUiService } from '../../block-ui/block-ui.service';
/**
 *
 * Componente HideWhileLoginComponent, utilizado para não apresentar a tela da aplicação enquanto a fase de login não terminou.
 *
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

var _c0 = ["hide-while-login", ""];
function HideWhileLoginComponent_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r938 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r938.loadingContent);
} }
var _c1 = ["*"];
var HideWhileLoginComponent = /** @class */ (function () {
    /**
     * constructor method
     * @param {ElementRef} hostComponent
     * @param {AbstractAuthentication} auth
     * @param {BlockUiService} blockUiService
     * @constructor
     */
    function HideWhileLoginComponent(hostComponent, auth, blockUiService) {
        this.hostComponent = hostComponent;
        this.auth = auth;
        this.blockUiService = blockUiService;
        /**
         * Carrega conteúdo
         * @type {loadingString}
         */
        this.loadingContent = "";
        this.waitingLogin = false;
    }
    HideWhileLoginComponent.prototype.ngOnInit = function () {
        this.waitingLogin = !this.auth.isAuthenticated();
    };
    HideWhileLoginComponent.prototype.ngAfterContentInit = function () {
        if (!this.auth.isAuthenticated()) {
            this.blockUiService.show();
        }
    };
    HideWhileLoginComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (!this.auth.isAuthenticated()) {
            for (var i = 1; i < this.hostComponent.nativeElement.children.length; i++) {
                this.hostComponent.nativeElement.children[i].style.display = 'none';
            }
        }
        this.loginNotifications = this.auth.getLoginNotifications().subscribe(function () {
            for (var i = 1; i < _this.hostComponent.nativeElement.children.length; i++) {
                _this.hostComponent.nativeElement.children[i].style.display = 'inherit';
            }
            _this.waitingLogin = false;
            _this.blockUiService.hide();
        });
    };
    /**
     * ngOnDestroy method
     * @returns void
     */
    HideWhileLoginComponent.prototype.ngOnDestroy = function () {
        this.loginNotifications.unsubscribe();
    };
    HideWhileLoginComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Authentication },
        { type: BlockUiService }
    ]; };
    __decorate([
        Input()
    ], HideWhileLoginComponent.prototype, "loadingContent", void 0);
HideWhileLoginComponent.ɵfac = function HideWhileLoginComponent_Factory(t) { return new (t || HideWhileLoginComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(Authentication), ɵngcc0.ɵɵdirectiveInject(BlockUiService)); };
HideWhileLoginComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: HideWhileLoginComponent, selectors: [["", "hide-while-login", ""]], inputs: { loadingContent: "loadingContent" }, attrs: _c0, ngContentSelectors: _c1, decls: 2, vars: 1, consts: [[4, "ngIf"]], template: function HideWhileLoginComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, HideWhileLoginComponent_span_0_Template, 2, 1, "span", 0);
        ɵngcc0.ɵɵprojection(1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.waitingLogin);
    } }, directives: [ɵngcc1.NgIf], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HideWhileLoginComponent, [{
        type: Component,
        args: [{
                selector: '[hide-while-login]',
                template: "<span *ngIf=\"waitingLogin\">{{ loadingContent }}</span>\n<ng-content></ng-content>"
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: Authentication }, { type: BlockUiService }]; }, { loadingContent: [{
            type: Input
        }] }); })();
    return HideWhileLoginComponent;
}());
export { HideWhileLoginComponent };
