// @nuvem/angular-base/esm5/lib/security/authentication/login-success.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import { Authentication } from './authentication';
/**
 * Componente de login que será chamado quando a aplicação for logada com sucesso.
 * Deverá ser criada uma rota para esse componente.
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var LoginSuccessComponent = /** @class */ (function () {
    function LoginSuccessComponent(authenticationService) {
        this.authenticationService = authenticationService;
    }
    LoginSuccessComponent.prototype.ngOnInit = function () {
        this.authenticationService.login();
    };
    LoginSuccessComponent.ctorParameters = function () { return [
        { type: Authentication }
    ]; };
LoginSuccessComponent.ɵfac = function LoginSuccessComponent_Factory(t) { return new (t || LoginSuccessComponent)(ɵngcc0.ɵɵdirectiveInject(Authentication)); };
LoginSuccessComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: LoginSuccessComponent, selectors: [["app-login-success"]], decls: 0, vars: 0, template: function LoginSuccessComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LoginSuccessComponent, [{
        type: Component,
        args: [{
                selector: 'app-login-success',
                template: ''
            }]
    }], function () { return [{ type: Authentication }]; }, null); })();
    return LoginSuccessComponent;
}());
export { LoginSuccessComponent };
