// @nuvem/angular-base/esm5/lib/security/authentication/logout.directive.js
import { __decorate } from "tslib";
import { Directive, HostListener } from "@angular/core";
import { Authentication } from "./authentication";
import * as ɵngcc0 from '@angular/core';
var LogoutDirective = /** @class */ (function () {
    function LogoutDirective(authenticationService) {
        this.authenticationService = authenticationService;
    }
    LogoutDirective.prototype.click = function () {
        this.authenticationService.logout();
    };
    LogoutDirective.ctorParameters = function () { return [
        { type: Authentication }
    ]; };
    __decorate([
        HostListener('click')
    ], LogoutDirective.prototype, "click", null);
LogoutDirective.ɵfac = function LogoutDirective_Factory(t) { return new (t || LogoutDirective)(ɵngcc0.ɵɵdirectiveInject(Authentication)); };
LogoutDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LogoutDirective, selectors: [["", "logout", ""]], hostBindings: function LogoutDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function LogoutDirective_click_HostBindingHandler() { return ctx.click(); });
    } } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LogoutDirective, [{
        type: Directive,
        args: [{ selector: '[logout]' }]
    }], function () { return [{ type: Authentication }]; }, { click: [{
            type: HostListener,
            args: ['click']
        }] }); })();
    return LogoutDirective;
}());
export { LogoutDirective };
