// @nuvem/angular-base/esm5/lib/security/authentication/not-authenticated-error-provider.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { Authentication } from './authentication';
import * as i0 from "@angular/core";
import * as i1 from "./authentication";
/**
 * Provedor de error para criação de erros de notificação.
 * @class
 */
var NotAuthenticatedErrorProvider = /** @class */ (function () {
    function NotAuthenticatedErrorProvider(authenticationService) {
        this.authenticationService = authenticationService;
    }
    /**
     * Método para identificar se o usuário está logado
     * @public
     * @param {HttpErrorResponse} error
     * @returns void
     */
    NotAuthenticatedErrorProvider.prototype.shouldHandle = function (error) {
        return error instanceof HttpErrorResponse && error.status == 401;
    };
    /**
     * Método para redirecionar caso o usuário não esteja
     * @public
     * @param {HttpErrorResponse} error
     * @returns void
     */
    NotAuthenticatedErrorProvider.prototype.handle = function (error) {
        this.authenticationService.redirect();
    };
    NotAuthenticatedErrorProvider.ctorParameters = function () { return [
        { type: Authentication }
    ]; };
    NotAuthenticatedErrorProvider.ɵprov = i0.ɵɵdefineInjectable({ factory: function NotAuthenticatedErrorProvider_Factory() { return new NotAuthenticatedErrorProvider(i0.ɵɵinject(i1.Authentication)); }, token: NotAuthenticatedErrorProvider, providedIn: "root" });
    NotAuthenticatedErrorProvider = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], NotAuthenticatedErrorProvider);
    return NotAuthenticatedErrorProvider;
}());
export { NotAuthenticatedErrorProvider };
