// @nuvem/angular-base/esm5/lib/security/authentication/user.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Input, OnInit } from "@angular/core";
import { Authentication } from "./authentication";
import * as ɵngcc0 from '@angular/core';
var UserDirective = /** @class */ (function () {
    function UserDirective(el, authenticationService) {
        this.el = el;
        this.authenticationService = authenticationService;
    }
    UserDirective.prototype.ngOnInit = function () {
        if (this.authenticationService.isAuthenticated()) {
            var user = this.authenticationService.getUser();
            this.el.nativeElement.innerHTML += this.user ? user[this.user] : user.name;
        }
    };
    UserDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Authentication }
    ]; };
    __decorate([
        Input()
    ], UserDirective.prototype, "user", void 0);
UserDirective.ɵfac = function UserDirective_Factory(t) { return new (t || UserDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(Authentication)); };
UserDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: UserDirective, selectors: [["", "user", ""]], inputs: { user: "user" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(UserDirective, [{
        type: Directive,
        args: [{ selector: '[user]' }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: Authentication }]; }, { user: [{
            type: Input
        }] }); })();
    return UserDirective;
}());
export { UserDirective };
