// @nuvem/primeng-components/esm5/lib/crud/crud-form.component.js
import { __decorate } from "tslib";
import { Component, Input } from "@angular/core";
import { ActivatedRoute } from "@angular/router";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from './form.component';
import * as ɵngcc3 from 'primeng';

var _c0 = function (a0) { return [a0]; };
var _c1 = ["*"];
var CrudFormComponent = /** @class */ (function () {
    function CrudFormComponent(route) {
        this.route = route;
        this.disableSave = false;
        this.listRoute = "../../list";
    }
    CrudFormComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.route.paramMap.subscribe(function (params) {
            _this.action = params.get('action');
            if (params.get('id')) {
                _this.route.data.subscribe(function (data) {
                    if (data.value) {
                        _this.form.patchValue(data.value);
                    }
                });
            }
            else {
                _this.listRoute = "../list";
            }
            if (_this.action === "view") {
                _this.form.disable();
            }
            else if (_this.form.disabled) {
                _this.form.enable();
            }
        });
    };
    CrudFormComponent.ctorParameters = function () { return [
        { type: ActivatedRoute }
    ]; };
    __decorate([
        Input()
    ], CrudFormComponent.prototype, "form", void 0);
    __decorate([
        Input()
    ], CrudFormComponent.prototype, "service", void 0);
    __decorate([
        Input()
    ], CrudFormComponent.prototype, "disableSave", void 0);
CrudFormComponent.ɵfac = function CrudFormComponent_Factory(t) { return new (t || CrudFormComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute)); };
CrudFormComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CrudFormComponent, selectors: [["nCrudForm"]], inputs: { disableSave: "disableSave", form: "form", service: "service" }, ngContentSelectors: _c1, decls: 4, vars: 9, consts: [[3, "form", "action", "route", "disableSave"], ["nButton", "", 1, "ui-g-12", "ui-lg-2", "ui-md-4", "ui-sm-12"], ["pButton", "", "type", "button", "label", "Voltar", 3, "routerLink"]], template: function CrudFormComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "nForm", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementStart(2, "div", 1);
        ɵngcc0.ɵɵelement(3, "button", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("form", ctx.form)("action", ctx.action)("route", ɵngcc0.ɵɵpureFunction1(5, _c0, ctx.listRoute))("disableSave", ctx.disableSave);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("routerLink", ɵngcc0.ɵɵpureFunction1(7, _c0, ctx.listRoute));
    } }, directives: [ɵngcc2.FormComponent, ɵngcc3.ButtonDirective, ɵngcc1.RouterLink], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CrudFormComponent, [{
        type: Component,
        args: [{
                selector: 'nCrudForm',
                template: "<nForm [form]=\"form\" [action]=\"action\" [route]=\"[listRoute]\" [disableSave]=\"disableSave\">\n    <ng-content></ng-content>\n    <div nButton class=\"ui-g-12 ui-lg-2 ui-md-4 ui-sm-12\">\n        <button pButton type=\"button\" [routerLink]=\"[listRoute]\" label=\"Voltar\"></button>\n    </div>\n</nForm>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }]; }, { disableSave: [{
            type: Input
        }], form: [{
            type: Input
        }], service: [{
            type: Input
        }] }); })();
    return CrudFormComponent;
}());
export { CrudFormComponent };
