// @nuvem/primeng-components/esm5/lib/crud/crud-resolve.guard.js
import { __decorate, __param } from "tslib";
import { Injectable, Inject } from '@angular/core';
import { CRUD_SERVICE } from "./crud-service.service";
import * as ɵngcc0 from '@angular/core';
var CrudResolveGuard = /** @class */ (function () {
    function CrudResolveGuard(crudService) {
        this.crudService = crudService;
    }
    CrudResolveGuard.prototype.resolve = function (route) {
        return this.crudService.find(route.paramMap.get('id'));
    };
    CrudResolveGuard.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [CRUD_SERVICE,] }] }
    ]; };
    CrudResolveGuard = __decorate([ __param(0, Inject(CRUD_SERVICE))
    ], CrudResolveGuard);
CrudResolveGuard.ɵfac = function CrudResolveGuard_Factory(t) { return new (t || CrudResolveGuard)(ɵngcc0.ɵɵinject(CRUD_SERVICE)); };
CrudResolveGuard.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: CrudResolveGuard, factory: function (t) { return CrudResolveGuard.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CrudResolveGuard, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [CRUD_SERVICE]
            }] }]; }, null); })();
    return CrudResolveGuard;
}());
export { CrudResolveGuard };
