// @nuvem/primeng-components/esm5/lib/crud/form.component.js
import { __decorate, __param } from "tslib";
import { Component, Inject, Input, Output, EventEmitter } from "@angular/core";
import { CRUD_SERVICE } from "./crud-service.service";
import { Router, ActivatedRoute } from "@angular/router";
var FormComponent = /** @class */ (function () {
    function FormComponent(crudService, activatedRoute, router) {
        this.crudService = crudService;
        this.activatedRoute = activatedRoute;
        this.router = router;
        this.action = "save";
        this.saveLabel = "Salvar";
        this.disableSave = false;
        this.save = new EventEmitter();
        this.reset = new EventEmitter();
    }
    FormComponent.prototype.ngOnInit = function () {
        if (this.service) {
            this.crudService = this.service;
        }
    };
    FormComponent.prototype.submit = function () {
        var _this = this;
        if (this.save.observers.length) {
            this.save.emit(this.form.value);
        }
        else {
            this.crudService[this.action](this.form.value)
                .subscribe(function () {
                if (_this.route) {
                    _this.router.navigate(_this.route, { relativeTo: _this.activatedRoute });
                }
                else if (_this.reset.observers.length) {
                    _this.reset.emit(_this.form);
                }
                else {
                    _this.form.reset();
                }
            });
        }
    };
    FormComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [CRUD_SERVICE,] }] },
        { type: ActivatedRoute },
        { type: Router }
    ]; };
    __decorate([
        Input()
    ], FormComponent.prototype, "form", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "route", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "action", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "saveLabel", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "service", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "disableSave", void 0);
    __decorate([
        Output()
    ], FormComponent.prototype, "save", void 0);
    __decorate([
        Output()
    ], FormComponent.prototype, "reset", void 0);
    FormComponent = __decorate([
        Component({
            selector: 'nForm',
            template: "<div>\n    <form (submit)=\"submit()\" class=\"ui-g form-group\">\n        <ng-content></ng-content>\n        <div class=\"ui-g-12 ui-lg-2 ui-md-4 ui-sm-12\" *ngIf=\"!disableSave\">\n            <button pButton type=\"submit\" [label]=\"saveLabel\" [disabled]=\"form.disabled || form.invalid\"></button>\n        </div>\n        <ng-content select=\"[nButton]\"></ng-content>\n    </form>\n</div>"
        }),
        __param(0, Inject(CRUD_SERVICE))
    ], FormComponent);
    return FormComponent;
}());
export { FormComponent };
