// @nuvem/primeng-components/esm5/lib/crud/form.component.js
import { __decorate, __param } from "tslib";
import { Component, Inject, Input, Output, EventEmitter } from "@angular/core";
import { CRUD_SERVICE } from "./crud-service.service";
import { Router, ActivatedRoute } from "@angular/router";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'primeng';

function FormComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelement(1, "button", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r955 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("label", ctx_r955.saveLabel)("disabled", ctx_r955.form.disabled || ctx_r955.form.invalid);
} }
var _c0 = ["*", [["", "nButton", ""]]];
var _c1 = ["*", "[nButton]"];
var FormComponent = /** @class */ (function () {
    function FormComponent(crudService, activatedRoute, router) {
        this.crudService = crudService;
        this.activatedRoute = activatedRoute;
        this.router = router;
        this.action = "save";
        this.saveLabel = "Salvar";
        this.disableSave = false;
        this.save = new EventEmitter();
        this.reset = new EventEmitter();
    }
    FormComponent.prototype.ngOnInit = function () {
        if (this.service) {
            this.crudService = this.service;
        }
    };
    FormComponent.prototype.submit = function () {
        var _this = this;
        if (this.save.observers.length) {
            this.save.emit(this.form.value);
        }
        else {
            this.crudService[this.action](this.form.value)
                .subscribe(function () {
                if (_this.route) {
                    _this.router.navigate(_this.route, { relativeTo: _this.activatedRoute });
                }
                else if (_this.reset.observers.length) {
                    _this.reset.emit(_this.form);
                }
                else {
                    _this.form.reset();
                }
            });
        }
    };
    FormComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [CRUD_SERVICE,] }] },
        { type: ActivatedRoute },
        { type: Router }
    ]; };
    __decorate([
        Input()
    ], FormComponent.prototype, "form", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "route", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "action", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "saveLabel", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "service", void 0);
    __decorate([
        Input()
    ], FormComponent.prototype, "disableSave", void 0);
    __decorate([
        Output()
    ], FormComponent.prototype, "save", void 0);
    __decorate([
        Output()
    ], FormComponent.prototype, "reset", void 0);
    FormComponent = __decorate([ __param(0, Inject(CRUD_SERVICE))
    ], FormComponent);
FormComponent.ɵfac = function FormComponent_Factory(t) { return new (t || FormComponent)(ɵngcc0.ɵɵdirectiveInject(CRUD_SERVICE), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router)); };
FormComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FormComponent, selectors: [["nForm"]], inputs: { action: "action", saveLabel: "saveLabel", disableSave: "disableSave", form: "form", route: "route", service: "service" }, outputs: { save: "save", reset: "reset" }, ngContentSelectors: _c1, decls: 5, vars: 1, consts: [[1, "ui-g", "form-group", 3, "submit"], ["class", "ui-g-12 ui-lg-2 ui-md-4 ui-sm-12", 4, "ngIf"], [1, "ui-g-12", "ui-lg-2", "ui-md-4", "ui-sm-12"], ["pButton", "", "type", "submit", 3, "label", "disabled"]], template: function FormComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵelementStart(1, "form", 0);
        ɵngcc0.ɵɵlistener("submit", function FormComponent_Template_form_submit_1_listener() { return ctx.submit(); });
        ɵngcc0.ɵɵprojection(2);
        ɵngcc0.ɵɵtemplate(3, FormComponent_div_3_Template, 2, 2, "div", 1);
        ɵngcc0.ɵɵprojection(4, 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.disableSave);
    } }, directives: [ɵngcc2.ɵangular_packages_forms_forms_y, ɵngcc2.NgControlStatusGroup, ɵngcc2.NgForm, ɵngcc3.NgIf, ɵngcc4.ButtonDirective], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormComponent, [{
        type: Component,
        args: [{
                selector: 'nForm',
                template: "<div>\n    <form (submit)=\"submit()\" class=\"ui-g form-group\">\n        <ng-content></ng-content>\n        <div class=\"ui-g-12 ui-lg-2 ui-md-4 ui-sm-12\" *ngIf=\"!disableSave\">\n            <button pButton type=\"submit\" [label]=\"saveLabel\" [disabled]=\"form.disabled || form.invalid\"></button>\n        </div>\n        <ng-content select=\"[nButton]\"></ng-content>\n    </form>\n</div>"
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [CRUD_SERVICE]
            }] }, { type: ɵngcc1.ActivatedRoute }, { type: ɵngcc1.Router }]; }, { action: [{
            type: Input
        }], saveLabel: [{
            type: Input
        }], disableSave: [{
            type: Input
        }], save: [{
            type: Output
        }], reset: [{
            type: Output
        }], form: [{
            type: Input
        }], route: [{
            type: Input
        }], service: [{
            type: Input
        }] }); })();
    return FormComponent;
}());
export { FormComponent };
