// @nuvem/primeng-components/esm5/lib/crud/success-message-handler.interceptor.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { HTTP_INTERCEPTORS, HttpResponse } from '@angular/common/http';
import { tap } from 'rxjs/operators';
import { PageNotificationService } from '../page-notification/page-notification.service';
import * as i0 from "@angular/core";
import * as i1 from "../page-notification/page-notification.service";
import * as ɵngcc0 from '@angular/core';
var STATUS;
(function (STATUS) {
    STATUS[STATUS["OK"] = 200] = "OK";
    STATUS[STATUS["CREATED"] = 201] = "CREATED";
})(STATUS || (STATUS = {}));
var SuccessMessageHandlerInterceptor = /** @class */ (function () {
    function SuccessMessageHandlerInterceptor(notification) {
        this.notification = notification;
    }
    SuccessMessageHandlerInterceptor.prototype.intercept = function (req, next) {
        var _this = this;
        return next.handle(req).pipe(tap(function (event) {
            if (event instanceof HttpResponse) {
                if (event.headers.has("X-nuvem-alert")) {
                    _this.notification.addSuccessMessage(event.headers.get("X-nuvem-alert"));
                }
                else if ((event.status === STATUS.OK || event.status === STATUS.CREATED) && req.method != "GET") {
                    _this.notification.addSuccessMessage("Operação realizada com sucesso!");
                }
            }
        }));
    };
    SuccessMessageHandlerInterceptor.ctorParameters = function () { return [
        { type: PageNotificationService }
    ]; };
    SuccessMessageHandlerInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function SuccessMessageHandlerInterceptor_Factory() { return new SuccessMessageHandlerInterceptor(i0.ɵɵinject(i1.PageNotificationService)); }, token: SuccessMessageHandlerInterceptor, providedIn: "root" });
SuccessMessageHandlerInterceptor.ɵfac = function SuccessMessageHandlerInterceptor_Factory(t) { return new (t || SuccessMessageHandlerInterceptor)(ɵngcc0.ɵɵinject(PageNotificationService)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SuccessMessageHandlerInterceptor, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return [{ type: PageNotificationService }]; }, null); })();
    return SuccessMessageHandlerInterceptor;
}());
export { SuccessMessageHandlerInterceptor };
export var httpInterceptorProviders = [
    { provide: HTTP_INTERCEPTORS, useClass: SuccessMessageHandlerInterceptor, multi: true },
];
