// @nuvem/angular-base/esm5/lib/block-ui/block-ui.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * Class BlockUiService
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var BlockUiService = /** @class */ (function () {
    /**
     * constructor method
     * @constructor
     */
    function BlockUiService() {
        /**
         * subject property
         * @private
         * @type {Subject<BlockUiStatus>}
         */
        this.subject = new Subject();
        /**
         * loaderStatus property
         * @public
         * @type {Observable}
         */
        this.loaderStatus = this.subject.asObservable();
    }
    /**
     * show methods
     * @returns void
     */
    BlockUiService.prototype.show = function () {
        this.subject.next({ status: true });
    };
    /**
     * hide methods
     * @returns void
     */
    BlockUiService.prototype.hide = function () {
        this.subject.next({ status: false });
    };
    BlockUiService.ɵprov = i0.ɵɵdefineInjectable({ factory: function BlockUiService_Factory() { return new BlockUiService(); }, token: BlockUiService, providedIn: "root" });
BlockUiService.ɵfac = function BlockUiService_Factory(t) { return new (t || BlockUiService)(); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUiService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
    return BlockUiService;
}());
export { BlockUiService };
