// @nuvem/angular-base/esm5/lib/block-ui/block-ui.interceptor.js
import { __decorate } from "tslib";
import { HttpResponse } from '@angular/common/http';
import { Injectable } from "@angular/core";
import { tap } from 'rxjs/operators';
import { BlockUiService } from './block-ui.service';
import * as i0 from "@angular/core";
import * as i1 from "./block-ui.service";
/**
 * Class BlockUiInterceptor
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var BlockUiInterceptor = /** @class */ (function () {
    /**
     * constructor method
     * @constructor
     * @param {BlockUiService} service
     */
    function BlockUiInterceptor(service) {
        this.service = service;
    }
    /**
     * intercept method
     * @public
     * @param {HttpRequest<any>} req
     * @param {HttpHandler} next
     * @return Observable<HttpEvent<any>>
     */
    BlockUiInterceptor.prototype.intercept = function (req, next) {
        var _this = this;
        if (this.isMutation(req.method)) {
            this.service.show();
        }
        return next.handle(req).pipe(tap(function (event) {
            if (event instanceof HttpResponse) {
                _this.service.hide();
            }
        }, function (err) {
            _this.service.hide();
        }));
    };
    /**
     * isMutation method
     * @private
     * @param {string} method
     * @return boolean
     */
    BlockUiInterceptor.prototype.isMutation = function (method) {
        var mutationsMethods = [
            'DELETE',
            'PATCH',
            'POST',
            'PUT'
        ];
        return mutationsMethods.indexOf(method) !== -1;
    };
    BlockUiInterceptor.ctorParameters = function () { return [
        { type: BlockUiService }
    ]; };
    BlockUiInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function BlockUiInterceptor_Factory() { return new BlockUiInterceptor(i0.ɵɵinject(i1.BlockUiService)); }, token: BlockUiInterceptor, providedIn: "root" });
BlockUiInterceptor.ɵfac = function BlockUiInterceptor_Factory(t) { return new (t || BlockUiInterceptor)(ɵngcc0.ɵɵinject(BlockUiService)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUiInterceptor, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: BlockUiService }]; }, null); })();
    return BlockUiInterceptor;
}());
export { BlockUiInterceptor };
