// @nuvem/angular-base/esm5/lib/common-mask/common-mask.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 *
 * Class CommonMaskService serviço para inserção de mascara de data em campos input
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var CommonMaskService = /** @class */ (function () {
    function CommonMaskService() {
        /**
         * DATE_SEPARATOR property
         * @type {string}
         */
        this.DATE_SEPARATOR = '/';
    }
    /**
     * Metodo para formatar valor digitado no padrão dd/mm/yyyy
     * @param $event
     * @returns void
     */
    CommonMaskService.prototype.dateMask = function ($event) {
        if ($event.target.value) {
            var date = $event.target.value.replace(/[^0-9]/g, '').substr(0, 8);
            if (date.length >= 3) {
                date = date.replace(/^(\d{2})(\d{1,2})/, '$1' + this.DATE_SEPARATOR + '$2');
            }
            if (date.length >= 6) {
                date = date.replace(/^(\S{5})(\d{1,4})/, '$1' + this.DATE_SEPARATOR + '$2');
            }
            $event.target.value = date;
        }
    };
    CommonMaskService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CommonMaskService_Factory() { return new CommonMaskService(); }, token: CommonMaskService, providedIn: "root" });
CommonMaskService.ɵfac = function CommonMaskService_Factory(t) { return new (t || CommonMaskService)(); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CommonMaskService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
    return CommonMaskService;
}());
export { CommonMaskService };
