// @nuvem/primeng-components/esm5/lib/crud/components/calendar/calendar.component.js
import { __decorate, __extends } from "tslib";
import { Component, Input } from "@angular/core";
import { provideValueAccessor } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
import { CALENDAR_LOCALE } from './calendar-locale';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../form-notification/form-notification.component';
import * as ɵngcc2 from 'primeng/calendar';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from '../validation-message/validation-message.component';

var _c0 = ["nCalendar", ""];
function CalendarComponent_nValidationMessage_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "nValidationMessage", 6);
} if (rf & 2) {
    var ctx_r945 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("control", ctx_r945.formControl);
} }
var _c1 = function (a0) { return { "ng-invalid ng-dirty": a0 }; };
var _c2 = ["*"];
var CalendarComponent = /** @class */ (function (_super) {
    __extends(CalendarComponent, _super);
    function CalendarComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.maxDate = null;
        _this.yearRange = null;
        _this.monthNavigator = false;
        _this.yearNavigator = false;
        _this.locale = CALENDAR_LOCALE;
        return _this;
    }
    CalendarComponent_1 = CalendarComponent;
    CalendarComponent.prototype.outerToInner = function (value) {
        return value ? new Date(value) : null;
    };
    var CalendarComponent_1;
    __decorate([
        Input()
    ], CalendarComponent.prototype, "maxDate", void 0);
    __decorate([
        Input()
    ], CalendarComponent.prototype, "yearRange", void 0);
    __decorate([
        Input()
    ], CalendarComponent.prototype, "monthNavigator", void 0);
    __decorate([
        Input()
    ], CalendarComponent.prototype, "yearNavigator", void 0);
    __decorate([
        Input()
    ], CalendarComponent.prototype, "locale", void 0);
CalendarComponent.ɵfac = function CalendarComponent_Factory(t) { return ɵCalendarComponent_BaseFactory(t || CalendarComponent); };
CalendarComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CalendarComponent, selectors: [["", "nCalendar", ""]], inputs: { maxDate: "maxDate", yearRange: "yearRange", monthNavigator: "monthNavigator", yearNavigator: "yearNavigator", locale: "locale" }, features: [ɵngcc0.ɵɵProvidersFeature([provideValueAccessor(CalendarComponent_1)]), ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c2, decls: 7, vars: 16, consts: [["form-notification", "", 1, "md-inputfield", 3, "name", "form"], ["notification", ""], [3, "ngClass", "name", "formControl", "locale", "dateFormat", "yearRange", "maxDate", "monthNavigator", "yearNavigator", "onBlur", "onFocus", "onSelect"], ["input", ""], [3, "for", "innerHTML"], [3, "control", 4, "ngIf"], [3, "control"]], template: function CalendarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "span", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "p-calendar", 2, 3);
        ɵngcc0.ɵɵlistener("onBlur", function CalendarComponent_Template_p_calendar_onBlur_2_listener($event) { return ctx.onBlurEvent($event); })("onFocus", function CalendarComponent_Template_p_calendar_onFocus_2_listener($event) { return ctx.onFocusEvent($event); })("onSelect", function CalendarComponent_Template_p_calendar_onSelect_2_listener($event) { return ctx.onSelectEvent($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(4, "label", 4);
        ɵngcc0.ɵɵtemplate(5, CalendarComponent_nValidationMessage_5_Template, 1, 1, "nValidationMessage", 5);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        var _r943 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵproperty("name", ctx.name)("form", ctx.form);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(14, _c1, _r943.message))("name", ctx.name)("formControl", ctx.formControl)("locale", ctx.locale)("dateFormat", ctx.locale.dateFormat)("yearRange", ctx.yearRange)("maxDate", ctx.maxDate)("monthNavigator", ctx.monthNavigator)("yearNavigator", ctx.yearNavigator);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("for", ctx.name)("innerHTML", ctx.label, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !_r943.message);
    } }, directives: [ɵngcc1.FormNotificationComponent, ɵngcc2.Calendar, ɵngcc3.NgClass, ɵngcc4.NgControlStatus, ɵngcc4.FormControlDirective, ɵngcc3.NgIf, ɵngcc5.ValidationMessageComponent], encapsulation: 2 });
var ɵCalendarComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(CalendarComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CalendarComponent, [{
        type: Component,
        args: [{
                selector: "[nCalendar]",
                template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <p-calendar #input\n        [ngClass]=\"{ 'ng-invalid ng-dirty': notification.message }\"\n        [name]=\"name\"\n        [formControl]=\"formControl\"\n        [locale]=\"locale\"\n        [dateFormat]=\"locale.dateFormat\"\n        [yearRange]=\"yearRange\"\n        [maxDate]=\"maxDate\"\n        [monthNavigator]=\"monthNavigator\"\n        [yearNavigator]=\"yearNavigator\"\n        (onBlur)=\"onBlurEvent($event)\"\n        (onFocus)=\"onFocusEvent($event)\"\n        (onSelect)=\"onSelectEvent($event)\">\n    </p-calendar>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
                providers: [provideValueAccessor(CalendarComponent_1)]
            }]
    }], null, { maxDate: [{
            type: Input
        }], yearRange: [{
            type: Input
        }], monthNavigator: [{
            type: Input
        }], yearNavigator: [{
            type: Input
        }], locale: [{
            type: Input
        }] }); })();
    return CalendarComponent;
}(FieldComponent));
export { CalendarComponent };
