// @nuvem/angular-base/esm5/lib/form-notification/components/form-notification-superclass.component.js
import { __decorate } from "tslib";
import { FormNotificationService } from '../services/form-notification.service';
import { Directive, Input } from "@angular/core";
import { FieldFormNotification } from '../types/field-form-notification.type';
import * as ɵngcc0 from '@angular/core';
var FormNotificationComponentSuperclass = /** @class */ (function () {
    /**
     * Metodo construtor responsavel por carregar o serviço de validação
     * @param formNotificationService
     */
    function FormNotificationComponentSuperclass(formNotificationService) {
        this.formNotificationService = formNotificationService;
    }
    /**
     * Metodo Inicial para subscrição no endpoint responsável por validar o formulário
     */
    FormNotificationComponentSuperclass.prototype.ngOnInit = function () {
        var _this = this;
        this.subscription = this.formNotificationService
            .notifications(this.form, this.name)
            .subscribe(function (notification) {
            if (notification instanceof FieldFormNotification) {
                _this.message = notification.message;
            }
            else {
                _this.message = null;
            }
        });
    };
    /**
     * Metodo responsavel por realizar a saida das subscrições do endpont de validação
     */
    FormNotificationComponentSuperclass.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    FormNotificationComponentSuperclass.ctorParameters = function () { return [
        { type: FormNotificationService }
    ]; };
    __decorate([
        Input()
    ], FormNotificationComponentSuperclass.prototype, "name", void 0);
    __decorate([
        Input()
    ], FormNotificationComponentSuperclass.prototype, "form", void 0);
FormNotificationComponentSuperclass.ɵfac = function FormNotificationComponentSuperclass_Factory(t) { return new (t || FormNotificationComponentSuperclass)(ɵngcc0.ɵɵdirectiveInject(FormNotificationService)); };
FormNotificationComponentSuperclass.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: FormNotificationComponentSuperclass, selectors: [["form-notification-component-superclass"]], inputs: { name: "name", form: "form" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormNotificationComponentSuperclass, [{
        type: Directive,
        args: [{
                selector: 'form-notification-component-superclass'
            }]
    }], function () { return [{ type: FormNotificationService }]; }, { name: [{
            type: Input
        }], form: [{
            type: Input
        }] }); })();
    return FormNotificationComponentSuperclass;
}());
export { FormNotificationComponentSuperclass };
