// @nuvem/angular-base/esm2015/lib/security/authentication/logout.directive.js
import { __decorate } from "tslib";
import { Directive, HostListener } from "@angular/core";
import { Authentication } from "./authentication";
import * as ɵngcc0 from '@angular/core';
let LogoutDirective = class LogoutDirective {
    constructor(authenticationService) {
        this.authenticationService = authenticationService;
    }
    click() {
        this.authenticationService.logout();
    }
};
LogoutDirective.ɵfac = function LogoutDirective_Factory(t) { return new (t || LogoutDirective)(ɵngcc0.ɵɵdirectiveInject(Authentication)); };
LogoutDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LogoutDirective, selectors: [["", "logout", ""]], hostBindings: function LogoutDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function LogoutDirective_click_HostBindingHandler() { return ctx.click(); });
    } } });
LogoutDirective.ctorParameters = () => [
    { type: Authentication }
];
__decorate([
    HostListener('click')
], LogoutDirective.prototype, "click", null);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LogoutDirective, [{
        type: Directive,
        args: [{ selector: '[logout]' }]
    }], function () { return [{ type: Authentication }]; }, { click: [{
            type: HostListener,
            args: ['click']
        }] }); })();
export { LogoutDirective };
