// @nuvem/angular-base/esm2015/lib/security/authentication/not-authenticated-error-provider.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { Authentication } from './authentication';
import * as i0 from "@angular/core";
import * as i1 from "./authentication";
/**
 * Provedor de error para criação de erros de notificação.
 * @class
 */
import * as ɵngcc0 from '@angular/core';
let NotAuthenticatedErrorProvider = class NotAuthenticatedErrorProvider {
    constructor(authenticationService) {
        this.authenticationService = authenticationService;
    }
    /**
     * Método para identificar se o usuário está logado
     * @public
     * @param {HttpErrorResponse} error
     * @returns void
     */
    shouldHandle(error) {
        return error instanceof HttpErrorResponse && error.status == 401;
    }
    /**
     * Método para redirecionar caso o usuário não esteja
     * @public
     * @param {HttpErrorResponse} error
     * @returns void
     */
    handle(error) {
        this.authenticationService.redirect();
    }
};
NotAuthenticatedErrorProvider.ɵfac = function NotAuthenticatedErrorProvider_Factory(t) { return new (t || NotAuthenticatedErrorProvider)(ɵngcc0.ɵɵinject(Authentication)); };
NotAuthenticatedErrorProvider.ctorParameters = () => [
    { type: Authentication }
];
NotAuthenticatedErrorProvider.ɵprov = i0.ɵɵdefineInjectable({ factory: function NotAuthenticatedErrorProvider_Factory() { return new NotAuthenticatedErrorProvider(i0.ɵɵinject(i1.Authentication)); }, token: NotAuthenticatedErrorProvider, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NotAuthenticatedErrorProvider, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: Authentication }]; }, null); })();
export { NotAuthenticatedErrorProvider };
