// @nuvem/angular-base/esm2015/lib/security/authentication/user.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Input, OnInit } from "@angular/core";
import { Authentication } from "./authentication";
import * as ɵngcc0 from '@angular/core';
let UserDirective = class UserDirective {
    constructor(el, authenticationService) {
        this.el = el;
        this.authenticationService = authenticationService;
    }
    ngOnInit() {
        if (this.authenticationService.isAuthenticated()) {
            let user = this.authenticationService.getUser();
            this.el.nativeElement.innerHTML += this.user ? user[this.user] : user.name;
        }
    }
};
UserDirective.ɵfac = function UserDirective_Factory(t) { return new (t || UserDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(Authentication)); };
UserDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: UserDirective, selectors: [["", "user", ""]], inputs: { user: "user" } });
UserDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Authentication }
];
__decorate([
    Input()
], UserDirective.prototype, "user", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(UserDirective, [{
        type: Directive,
        args: [{ selector: '[user]' }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: Authentication }]; }, { user: [{
            type: Input
        }] }); })();
export { UserDirective };
