// @nuvem/angular-base/esm5/lib/security/token/jwt-auth.interceptor.js
import { __decorate } from "tslib";
import { HttpResponse } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { tap } from 'rxjs/operators';
import { Token } from "./token";
import * as i0 from "@angular/core";
import * as i1 from "./token";
/**
 * Implementação de classe para interceptar e tratar requisições http/https
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var JWTAuthInterceptor = /** @class */ (function () {
    /**
     * Metodo construtor responsável por injetar serviço de autenticação
     * @constructor
     * @param {AbstractAuthentication} auth
     */
    function JWTAuthInterceptor(token) {
        this.token = token;
    }
    /**
     * Metodo responsável por interceptar requisições HTTP/HTTPS
     * @public
     * @param {HttpRequest<any>} request
     * @param {HttpHandler} next
     * @returns Observable<HttpEvent<any>>
     */
    JWTAuthInterceptor.prototype.intercept = function (request, next) {
        var _this = this;
        if (this.token.hasToken()) {
            request = this.token.setTokenInHeader(request);
        }
        return next.handle(request)
            .pipe(tap(function (event) {
            if (event instanceof HttpResponse) {
                var token = _this.token.extractTokenFromHeaders(event.headers);
                if (null !== token)
                    _this.token.storeToken(token);
            }
            return event;
        }));
    };
    JWTAuthInterceptor.ctorParameters = function () { return [
        { type: Token }
    ]; };
    JWTAuthInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function JWTAuthInterceptor_Factory() { return new JWTAuthInterceptor(i0.ɵɵinject(i1.Token)); }, token: JWTAuthInterceptor, providedIn: "root" });
JWTAuthInterceptor.ɵfac = function JWTAuthInterceptor_Factory(t) { return new (t || JWTAuthInterceptor)(ɵngcc0.ɵɵinject(Token)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(JWTAuthInterceptor, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: Token }]; }, null); })();
    return JWTAuthInterceptor;
}());
export { JWTAuthInterceptor };
