// @nuvem/angular-base/esm5/lib/security/token/jwt-token.service.js
import { __decorate, __extends, __param } from "tslib";
import { Injectable, Inject } from '@angular/core';
import { HttpHeaders } from "@angular/common/http";
import { Token } from './token';
import { AUTH_CONFIG } from '../config/auth-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/auth-config";
/**
 * Inplementação de serviço JWTTokenService
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var JWTTokenService = /** @class */ (function (_super) {
    __extends(JWTTokenService, _super);
    /**
     * Metodo construtor responsável por injetar as configurações de autenticação
     * @constructor
     * @param {AuthConfig} config
     */
    function JWTTokenService(config) {
        var _this = _super.call(this) || this;
        _this.config = config;
        return _this;
    }
    JWTTokenService_1 = JWTTokenService;
    /**
     * Metodo responsável pela extração de tokes do heaedrs das requisições
     * @public
     * @param {HttpHeaders} headers
     * @returns string
     */
    JWTTokenService.prototype.extractTokenFromHeaders = function (headers) {
        return headers.get(JWTTokenService_1.HEADER_STRING);
    };
    /**
     * Metodo responsável pela adição dos tokens no storage
     * @public
     * @param {String} token
     * @returns void
     */
    JWTTokenService.prototype.storeToken = function (token) {
        this.config.storage.setItem(this.config.tokenStorageIndex, token.toString());
    };
    /**
     * Metodo responsável por injetar token no header das requisições
     * @public
     * @param {HttpRequest<any>} request
     * @returns HttpRequest<any>
     */
    JWTTokenService.prototype.setTokenInHeader = function (request) {
        var _a;
        var req = request.clone({
            headers: new HttpHeaders((_a = {},
                _a[JWTTokenService_1.HEADER_STRING] = this.config.storage.getItem(this.config.tokenStorageIndex),
                _a))
        });
        return req;
    };
    /**
     * Metodo responsável por verificar a existencia de um token de autenticação
     * @returns boolean
     * @public
     */
    JWTTokenService.prototype.hasToken = function () {
        return null !== this.config.storage.getItem(this.config.tokenStorageIndex);
    };
    var JWTTokenService_1;
    /**
     * TOKEN_PREFIX  property
     * @type {string}
     */
    JWTTokenService.TOKEN_PREFIX = "Bearer";
    /**
     * HEADER_STRING property
     * @type {string}
     */
    JWTTokenService.HEADER_STRING = "Authorization";
    JWTTokenService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [AUTH_CONFIG,] }] }
    ]; };
    JWTTokenService.ɵprov = i0.ɵɵdefineInjectable({ factory: function JWTTokenService_Factory() { return new JWTTokenService(i0.ɵɵinject(i1.AUTH_CONFIG)); }, token: JWTTokenService, providedIn: "root" });
    JWTTokenService = JWTTokenService_1 = __decorate([ __param(0, Inject(AUTH_CONFIG))
    ], JWTTokenService);
JWTTokenService.ɵfac = function JWTTokenService_Factory(t) { return new (t || JWTTokenService)(ɵngcc0.ɵɵinject(AUTH_CONFIG)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(JWTTokenService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [AUTH_CONFIG]
            }] }]; }, null); })();
    return JWTTokenService;
}(Token));
export { JWTTokenService };
