// @nuvem/primeng-components/esm2015/lib/crud/components/field.component.js
import { __decorate } from "tslib";
import { Injector, Input, OnInit, ViewChild, AfterViewInit, ElementRef, Output, EventEmitter, Directive } from "@angular/core";
import { WrappedFormControlSuperclass } from "@nuvem/angular-base";
import { ControlContainer } from "@angular/forms";
let FieldComponent = class FieldComponent extends WrappedFormControlSuperclass {
    constructor(controlContainer, injector) {
        super(injector);
        this.controlContainer = controlContainer;
        this.injector = injector;
        this.onBlur = new EventEmitter();
        this.onSelect = new EventEmitter();
        this.onFocus = new EventEmitter();
    }
    get name() {
        return this._name ? this._name : this.formControlName;
    }
    set name(value) {
        this._name = value;
    }
    get label() {
        return this.required ? `${this._label}*` : this._label;
    }
    set label(value) {
        this._label = value;
    }
    ngOnInit() {
        this.setValidators();
        this.setRequired();
    }
    setRequired() {
        if (this.required) {
            if (this.input instanceof ElementRef) {
                this.input.nativeElement.required = true;
            }
            else {
                this.input.required = true;
            }
        }
    }
    setElementRefAttr(attr, value) {
        if (value && this.input instanceof ElementRef) {
            this.input.nativeElement[attr] = value;
        }
    }
    setFormControlValidators() {
        if (this.controlContainer && this.formControlName) {
            let validator = this.controlContainer.control.get(this.formControlName).validator;
            if (validator) {
                this.formControl.setValidators(validator);
            }
        }
    }
    setValidators() {
        if (this.controlContainer && this.formControlName) {
            let validator = this.controlContainer.control.get(this.formControlName).validator;
            if (validator) {
                this.formControl.setValidators(validator);
            }
        }
    }
    ngAfterViewInit() {
        for (let i in this.extraParams) {
            this.input[i] = this.extraParams[i];
        }
    }
    onBlurEvent(event) {
        this.onBlur.emit(event);
    }
    onSelectEvent(event) {
        this.onSelect.emit(event);
    }
    onFocusEvent(event) {
        this.onFocus.emit(event);
    }
};
FieldComponent.ctorParameters = () => [
    { type: ControlContainer },
    { type: Injector }
];
__decorate([
    Input()
], FieldComponent.prototype, "form", void 0);
__decorate([
    Input()
], FieldComponent.prototype, "formControlName", void 0);
__decorate([
    Input()
], FieldComponent.prototype, "extraParams", void 0);
__decorate([
    Input()
], FieldComponent.prototype, "required", void 0);
__decorate([
    Output()
], FieldComponent.prototype, "onBlur", void 0);
__decorate([
    Output()
], FieldComponent.prototype, "onSelect", void 0);
__decorate([
    Output()
], FieldComponent.prototype, "onFocus", void 0);
__decorate([
    ViewChild("input", { static: false })
], FieldComponent.prototype, "input", void 0);
__decorate([
    Input()
], FieldComponent.prototype, "name", null);
__decorate([
    Input()
], FieldComponent.prototype, "label", null);
FieldComponent = __decorate([
    Directive({
        selector: "[nField]"
    })
], FieldComponent);
export { FieldComponent };
