// @nuvem/primeng-components/esm2015/lib/crud/components/input-text/input-text.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../form-notification/form-notification.component';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from 'primeng';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../validation-message/validation-message.component';

const _c0 = ["nInputText", ""];
function InputTextComponent_nValidationMessage_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "nValidationMessage", 6);
} if (rf & 2) {
    const ctx_r962 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("control", ctx_r962.formControl);
} }
const _c1 = function (a0) { return { "ng-invalid ng-dirty": a0 }; };
const _c2 = ["*"];
var InputTextComponent_1;
import { __decorate } from "tslib";
import { Component, Input } from "@angular/core";
import { provideValueAccessor } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
let InputTextComponent = InputTextComponent_1 = class InputTextComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.type = "text";
    }
    ngOnInit() {
        super.ngOnInit();
        this.setElementRefAttr("min", this.min);
        this.setElementRefAttr("max", this.max);
        this.setElementRefAttr("minLength", this.minLength);
        this.setElementRefAttr("maxLength", this.maxLength);
        this.setElementRefAttr("pattern", this.pattern);
    }
};
InputTextComponent.ɵfac = function InputTextComponent_Factory(t) { return ɵInputTextComponent_BaseFactory(t || InputTextComponent); };
InputTextComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: InputTextComponent, selectors: [["", "nInputText", ""]], inputs: { type: "type", minLength: "minLength", maxLength: "maxLength", min: "min", max: "max", pattern: "pattern" }, features: [ɵngcc0.ɵɵProvidersFeature([provideValueAccessor(InputTextComponent_1)]), ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c2, decls: 7, vars: 11, consts: [["form-notification", "", 1, "md-inputfield", 3, "name", "form"], ["notification", ""], ["pInputText", "", 3, "ngClass", "type", "name", "formControl", "blur", "focus"], ["input", ""], [3, "for", "innerHTML"], [3, "control", 4, "ngIf"], [3, "control"]], template: function InputTextComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "span", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "input", 2, 3);
        ɵngcc0.ɵɵlistener("blur", function InputTextComponent_Template_input_blur_2_listener($event) { return ctx.onBlurEvent($event); })("focus", function InputTextComponent_Template_input_focus_2_listener($event) { return ctx.onFocusEvent($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(4, "label", 4);
        ɵngcc0.ɵɵtemplate(5, InputTextComponent_nValidationMessage_5_Template, 1, 1, "nValidationMessage", 5);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r960 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵproperty("name", ctx.name)("form", ctx.form);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c1, _r960.message))("type", ctx.type)("name", ctx.name)("formControl", ctx.formControl);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("for", ctx.name)("innerHTML", ctx.label, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !_r960.message);
    } }, directives: [ɵngcc1.FormNotificationComponent, ɵngcc2.DefaultValueAccessor, ɵngcc3.InputText, ɵngcc4.NgClass, ɵngcc2.NgControlStatus, ɵngcc2.FormControlDirective, ɵngcc4.NgIf, ɵngcc5.ValidationMessageComponent], encapsulation: 2 });
__decorate([
    Input()
], InputTextComponent.prototype, "minLength", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "maxLength", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "min", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "max", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "pattern", void 0);
__decorate([
    Input()
], InputTextComponent.prototype, "type", void 0);
const ɵInputTextComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(InputTextComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(InputTextComponent, [{
        type: Component,
        args: [{
                selector: "[nInputText]",
                template: "<span #notification form-notification class=\"md-inputfield\" [name]=\"name\" [form]=\"form\">\n    <input #input\n        pInputText\n        [ngClass]=\"{ 'ng-invalid ng-dirty': notification.message }\"\n        [type]=\"type\"\n        [name]=\"name\"\n        (blur)=\"onBlurEvent($event)\"\n        (focus)=\"onFocusEvent($event)\"\n        [formControl]=\"formControl\" />\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
                providers: [provideValueAccessor(InputTextComponent_1)]
            }]
    }], null, { type: [{
            type: Input
        }], minLength: [{
            type: Input
        }], maxLength: [{
            type: Input
        }], min: [{
            type: Input
        }], max: [{
            type: Input
        }], pattern: [{
            type: Input
        }] }); })();
export { InputTextComponent };
