// @nuvem/primeng-components/esm2015/lib/datatable/datatable.component.js
import { __decorate } from "tslib";
import { Component, ViewChild, Input, Output, ContentChildren, QueryList, ChangeDetectorRef, AfterViewInit, AfterContentInit, OnInit, EventEmitter, ViewEncapsulation, ElementRef, } from '@angular/core';
import { Table as DataTable } from 'primeng/table';
import { DataProviderFactory, DataProviderType } from './dataprovider-factory';
import { DatatableButtonComponent } from './datatable-button.component';
import { DatatableClickEvent } from './datatable-click-event';
import { Authorization, User } from '@nuvem/angular-base';
import { Column } from './p-column.directive';
/**
 * Class DatatableComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@nuvem/angular-base';
import * as ɵngcc2 from 'primeng';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'primeng/table';
import * as ɵngcc5 from 'primeng/api';
import * as ɵngcc6 from './show-column/show-column.component';
import * as ɵngcc7 from './filter-column/filter-column.component';
import * as ɵngcc8 from '@angular/forms';
import * as ɵngcc9 from './datatable-button.component';
import * as ɵngcc10 from '../extra-button/datatable-splitbutton.component';

function DatatableComponent_basis_show_column_6_Template(rf, ctx) { if (rf & 1) {
    const _r859 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-show-column", 11);
    ɵngcc0.ɵɵlistener("updateVisibleColumns", function DatatableComponent_basis_show_column_6_Template_basis_show_column_updateVisibleColumns_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r859); const ctx_r858 = ɵngcc0.ɵɵnextContext(); return ctx_r858.updateVisibleColumns($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r849 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("columns", ctx_r849.columns);
} }
function DatatableComponent_ng_template_10_div_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵelementStart(1, "span", 18);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r864 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r864.horizontalButtonsHeader);
} }
function DatatableComponent_ng_template_10_div_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c0 = function (a2, a3) { return { "horizontal-buttons": true, "ui-sm-12": true, "ui-g-8": a2, "ui-g-12": a3 }; };
function DatatableComponent_ng_template_10_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_10_div_0_div_1_Template, 3, 1, "div", 15);
    ɵngcc0.ɵɵelementStart(2, "div", 2);
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_10_div_0_ng_container_3_Template, 1, 0, "ng-container", 16);
    ɵngcc0.ɵɵprojection(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r861 = ɵngcc0.ɵɵnextContext(2);
    const _r856 = ɵngcc0.ɵɵreference(16);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r861.showHorizontalButtonsHeader);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(3, _c0, ctx_r861.showHorizontalButtonsHeader, !ctx_r861.showHorizontalButtonsHeader));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r856);
} }
function DatatableComponent_ng_template_10_th_2_p_sortIcon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "p-sortIcon", 21);
} if (rf & 2) {
    const col_r866 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("field", col_r866.field);
} }
const _c1 = function (a0) { return { "display": a0 }; };
function DatatableComponent_ng_template_10_th_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "th", 19);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_10_th_2_p_sortIcon_2_Template, 1, 1, "p-sortIcon", 20);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r866 = ctx.$implicit;
    const ctx_r862 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("pSortableColumn", col_r866.sortable ? col_r866.field : null)("ngStyle", ɵngcc0.ɵɵpureFunction1(4, _c1, ctx_r862.visibleColumns[col_r866.field]));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", col_r866.header, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r866.sortable);
} }
function DatatableComponent_ng_template_10_tr_3_th_1_Template(rf, ctx) { if (rf & 1) {
    const _r872 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "th", 23);
    ɵngcc0.ɵɵelementStart(1, "basis-filter-column", 24);
    ɵngcc0.ɵɵlistener("outputEvent", function DatatableComponent_ng_template_10_tr_3_th_1_Template_basis_filter_column_outputEvent_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r872); const col_r870 = ctx.$implicit; const ctx_r871 = ɵngcc0.ɵɵnextContext(3); return ctx_r871.filterColumn(col_r870.field, $event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r870 = ctx.$implicit;
    const ctx_r869 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(8, _c1, ctx_r869.visibleColumns[col_r870.field]));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("listaOriginal", ctx_r869.filterSelectOptions)("lista", ctx_r869.filterSelectOptions)("field", col_r870.field)("updateFiltro", ctx_r869.updateFilterColumn)("filterType", col_r870.filterType)("tableType", ctx_r869.type)("customFilterOptions", ctx_r869.customFilterOptions);
} }
function DatatableComponent_ng_template_10_tr_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "tr");
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_10_tr_3_th_1_Template, 2, 10, "th", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const columns_r860 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", columns_r860);
} }
function DatatableComponent_ng_template_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_ng_template_10_div_0_Template, 5, 6, "div", 12);
    ɵngcc0.ɵɵelementStart(1, "tr");
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_10_th_2_Template, 3, 6, "th", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_10_tr_3_Template, 2, 1, "tr", 8);
} if (rf & 2) {
    const columns_r860 = ctx.$implicit;
    const ctx_r851 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r851.verticalButtons);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", columns_r860);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r851.filterOnColumn);
} }
function DatatableComponent_ng_template_11_td_1_ng_container_1_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 29);
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const rowData_r874 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", rowData_r874[col_r879.field], ɵngcc0.ɵɵsanitizeHtml);
} }
const _c2 = function (a0, a1) { return { rowData: a0, col: a1 }; };
function DatatableComponent_ng_template_11_td_1_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 30);
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const rowData_r874 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", col_r879.bodyTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c2, rowData_r874, col_r879));
} }
function DatatableComponent_ng_template_11_td_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_11_td_1_ng_container_1_span_1_Template, 1, 1, "span", 27);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_11_td_1_ng_container_1_ng_container_2_Template, 1, 5, "ng-container", 28);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !col_r879.bodyTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r879.bodyTemplate);
} }
function DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 29);
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const rowData_r874 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", rowData_r874[col_r879.field], ɵngcc0.ɵɵsanitizeHtml);
} }
const _c3 = function () { return { "width": "100%" }; };
function DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r900 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 34);
    ɵngcc0.ɵɵelementStart(1, "div", 35);
    ɵngcc0.ɵɵelementStart(2, "div", 34);
    ɵngcc0.ɵɵelementStart(3, "span", 36);
    ɵngcc0.ɵɵelementStart(4, "input", 37);
    ɵngcc0.ɵɵlistener("ngModelChange", function DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_div_1_Template_input_ngModelChange_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r900); const col_r879 = ɵngcc0.ɵɵnextContext(3).$implicit; const rowData_r874 = ɵngcc0.ɵɵnextContext().$implicit; return (rowData_r874[col_r879.field] = $event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "label");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const rowData_r874 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵstyleMap(ɵngcc0.ɵɵpureFunction0(4, _c3));
    ɵngcc0.ɵɵproperty("ngModel", rowData_r874[col_r879.field]);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(col_r879.header);
} }
const _c4 = function (a0, a1) { return { "rowData": a0, "col": a1 }; };
function DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 30);
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const rowData_r874 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", col_r879.inputTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c4, rowData_r874, col_r879));
} }
function DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_span_0_Template, 1, 1, "span", 27);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_div_1_Template, 7, 5, "div", 33);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_ng_container_2_Template, 1, 5, "ng-container", 28);
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("ngIf", !col_r879.outputTemplate && col_r879.editable == false);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !col_r879.inputTemplate && col_r879.editable != false);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r879.inputTemplate);
} }
function DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_3_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 29);
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const rowData_r874 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", rowData_r874[col_r879.field], ɵngcc0.ɵɵsanitizeHtml);
} }
function DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 30);
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const rowData_r874 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", col_r879.outputTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c2, rowData_r874, col_r879));
} }
function DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_3_span_0_Template, 1, 1, "span", 27);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_3_ng_container_1_Template, 1, 5, "ng-container", 28);
} if (rf & 2) {
    const col_r879 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("ngIf", !col_r879.outputTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r879.outputTemplate);
} }
function DatatableComponent_ng_template_11_td_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "p-cellEditor");
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_2_Template, 3, 3, "ng-template", 31);
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_11_td_1_ng_container_2_ng_template_3_Template, 2, 2, "ng-template", 32);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
function DatatableComponent_ng_template_11_td_1_div_3_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r918 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 42);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_11_td_1_div_3_button_1_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r918); const rowData_r874 = ɵngcc0.ɵɵnextContext(3).$implicit; const ctx_r917 = ɵngcc0.ɵɵnextContext(); return ctx_r917.onRowEditInitEvent($event, rowData_r874); });
    ɵngcc0.ɵɵtext(1, "Editar");
    ɵngcc0.ɵɵelementEnd();
} }
function DatatableComponent_ng_template_11_td_1_div_3_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r921 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 43);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_11_td_1_div_3_button_2_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r921); const rowData_r874 = ɵngcc0.ɵɵnextContext(3).$implicit; const ctx_r920 = ɵngcc0.ɵɵnextContext(); return ctx_r920.onRowEditSaveEvent($event, rowData_r874); });
    ɵngcc0.ɵɵtext(1, "Salvar");
    ɵngcc0.ɵɵelementEnd();
} }
function DatatableComponent_ng_template_11_td_1_div_3_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r924 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 44);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_11_td_1_div_3_button_3_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r924); const ctx_r925 = ɵngcc0.ɵɵnextContext(3); const rowData_r874 = ctx_r925.$implicit; const ri_r877 = ctx_r925.rowIndex; const ctx_r923 = ɵngcc0.ɵɵnextContext(); return ctx_r923.onRowEditCancelEvent($event, rowData_r874, ri_r877); });
    ɵngcc0.ɵɵtext(1, "Cancelar");
    ɵngcc0.ɵɵelementEnd();
} }
function DatatableComponent_ng_template_11_td_1_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 38);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_11_td_1_div_3_button_1_Template, 2, 0, "button", 39);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_11_td_1_div_3_button_2_Template, 2, 0, "button", 40);
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_11_td_1_div_3_button_3_Template, 2, 0, "button", 41);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const editing_r876 = ɵngcc0.ɵɵnextContext(2).editing;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !editing_r876);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", editing_r876);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", editing_r876);
} }
function DatatableComponent_ng_template_11_td_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 23);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_11_td_1_ng_container_1_Template, 3, 2, "ng-container", 8);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_11_td_1_ng_container_2_Template, 4, 0, "ng-container", 8);
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_11_td_1_div_3_Template, 4, 3, "div", 26);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r879 = ctx.$implicit;
    const isLast_r880 = ctx.last;
    const ctx_r878 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(4, _c1, ctx_r878.visibleColumns[col_r879.field]));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r878.editMode);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r878.editMode);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", isLast_r880);
} }
const _c5 = function (a0) { return { "editing": a0 }; };
function DatatableComponent_ng_template_11_Template(rf, ctx) { if (rf & 1) {
    const _r928 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr", 25);
    ɵngcc0.ɵɵlistener("mouseover", function DatatableComponent_ng_template_11_Template_tr_mouseover_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r928); const ctx_r927 = ɵngcc0.ɵɵnextContext(); return ctx_r927.editarLinhaOver($event); })("mouseleave", function DatatableComponent_ng_template_11_Template_tr_mouseleave_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r928); const ctx_r929 = ɵngcc0.ɵɵnextContext(); return ctx_r929.editarLinhaOut($event); });
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_11_td_1_Template, 4, 6, "td", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const rowData_r874 = ctx.$implicit;
    const columns_r875 = ctx.columns;
    const editing_r876 = ctx.editing;
    ɵngcc0.ɵɵproperty("pSelectableRow", rowData_r874)("pContextMenuRow", rowData_r874)("pEditableRow", rowData_r874)("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c5, editing_r876));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", columns_r875);
} }
function DatatableComponent_12_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    const ctx_r930 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r930.emptyMessage, " ");
} }
function DatatableComponent_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_12_ng_template_0_Template, 1, 1, "ng-template");
} }
function DatatableComponent_p_footer_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p-footer");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r854 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r854.paginationFooterText(ctx_r854.footerMsg), " ");
} }
function DatatableComponent_div_14_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function DatatableComponent_div_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 45);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_div_14_ng_container_1_Template, 1, 0, "ng-container", 16);
    ɵngcc0.ɵɵprojection(2, 1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r855 = ɵngcc0.ɵɵnextContext();
    const _r856 = ɵngcc0.ɵɵreference(16);
    ɵngcc0.ɵɵstyleProp("position", ctx_r855.enableButtonsScroll ? "sticky" : "")("top", ctx_r855.enableButtonsScroll ? "15%" : "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r856);
} }
function DatatableComponent_ng_template_15_basis_datatable_button_0_Template(rf, ctx) { if (rf & 1) {
    const _r937 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-datatable-button", 50);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_15_basis_datatable_button_0_Template_basis_datatable_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r937); const ctx_r936 = ɵngcc0.ɵɵnextContext(2); return ctx_r936.onClick("edit", $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r932 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("horizontal-button", !ctx_r932.verticalButtons);
    ɵngcc0.ɵɵproperty("bTooltip", ctx_r932.selectedRow != null ? "Editar" : "Escolha um registro para editar.")("disabled", ctx_r932.selectedRow == null);
} }
function DatatableComponent_ng_template_15_basis_datatable_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r939 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-datatable-button", 51);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_15_basis_datatable_button_1_Template_basis_datatable_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r939); const ctx_r938 = ɵngcc0.ɵɵnextContext(2); return ctx_r938.onClick("view", $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r933 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("horizontal-button", !ctx_r933.verticalButtons);
    ɵngcc0.ɵɵproperty("bTooltip", ctx_r933.selectedRow != null ? "Visualizar" : "Escolha um registro para visualizar.")("disabled", ctx_r933.selectedRow == null);
} }
function DatatableComponent_ng_template_15_basis_datatable_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r941 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-datatable-button", 52);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_15_basis_datatable_button_2_Template_basis_datatable_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r941); const ctx_r940 = ɵngcc0.ɵɵnextContext(2); return ctx_r940.onClick("delete", $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r934 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("horizontal-button", !ctx_r934.verticalButtons);
    ɵngcc0.ɵɵproperty("bTooltip", ctx_r934.selectedRow != null ? "Excluir" : "Escolha um registro para excluir.")("disabled", ctx_r934.selectedRow == null);
} }
function DatatableComponent_ng_template_15_datatable_splitbutton_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "datatable-splitbutton", 53);
} if (rf & 2) {
    const ctx_r935 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMap(ctx_r935.splitButton.class);
    ɵngcc0.ɵɵproperty("icon", ctx_r935.splitButton.icon)("items", ctx_r935.splitButton.itens)("label", ctx_r935.splitButton.label)("method", ctx_r935.splitButton.method);
} }
function DatatableComponent_ng_template_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_ng_template_15_basis_datatable_button_0_Template, 1, 4, "basis-datatable-button", 46);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_15_basis_datatable_button_1_Template, 1, 4, "basis-datatable-button", 47);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_15_basis_datatable_button_2_Template, 1, 4, "basis-datatable-button", 48);
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_15_datatable_splitbutton_3_Template, 1, 6, "datatable-splitbutton", 49);
} if (rf & 2) {
    const ctx_r857 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r857.disableEdit && ctx_r857.checkRolesToButton("rulesToShowView") && ctx_r857.checkRolesToButton("rulesToHideView"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r857.disableView && ctx_r857.checkRolesToButton("rulesToShowEdit") && ctx_r857.checkRolesToButton("rulesToHideEdit"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r857.disableDelete && ctx_r857.checkRolesToButton("rulesToShowDelete") && ctx_r857.checkRolesToButton("rulesToHideDelete"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r857.splitButton);
} }
const _c6 = [[["horizontalButtons"]], "*"];
const _c7 = function (a0, a1) { return { "ui-g-2": a0, "ui-g-5": a1 }; };
const _c8 = function (a0, a1, a2, a3) { return { "ui-g-11": a0, "ui-g-12": a1, "edicao-tabela": a2, "editing-row": a3 }; };
const _c9 = ["horizontalButtons", "*"];
let DatatableComponent = class DatatableComponent {
    /**
     * constructor method
     * @constructor
     * @param {ChangeDetectorRef} changeDetectorRef
     * @param {DataProviderFactory} dataProviderFactory
     * @param {Authorization} authorizationService
     */
    constructor(changeDetectorRef, dataProviderFactory, authorizationService) {
        this.changeDetectorRef = changeDetectorRef;
        this.dataProviderFactory = dataProviderFactory;
        this.authorizationService = authorizationService;
        /**
         * filterParams property
         * @type {any}
         */
        this.filterParams = {};
        /**
         * footerMsg property
         * @type {string}
         */
        this.footerMsg = '';
        /**
         * orderInSort property
         * @type {boolean}
         */
        this.orderInSort = false;
        /**
         * value property
         * @type {any}
         */
        this.value = [];
        /**
         * type property
         * @type DataProviderType
         */
        this.type = DataProviderType.Server;
        /**
         * rows property
         * @type {number}
         */
        this.rows = 5;
        /**
         * showPaginationFooter property
         * @type {boolean}
         */
        this.showPaginationFooter = false;
        /**
         * disableEdit property
         * @type {boolean}
         */
        this.disableEdit = false;
        /**
         * disableView property
         * @type {boolean}
         */
        this.disableView = false;
        /**
         * disableDelete property
         * @type {boolean}
         */
        this.disableDelete = false;
        /**
         * selectionMode property
         * @type {string}
         */
        this.selectionMode = 'single';
        /**
         * enableButtonsScroll property
         * @type {boolean}
         */
        this.enableButtonsScroll = false;
        /**
         * emptyMessage property
         * @type {string}
         */
        this.emptyMessage = 'Nenhum registro encontrado.';
        /**
         * verticalButtons property
         * @type {boolean}
         */
        this.verticalButtons = true;
        /**
         * enableScroll property
         * @type {boolean}
         */
        this.enableScroll = false;
        /**
         * disableLoadingBlockUI property
         * @type {boolean}
         */
        this.disableLoadingBlockUI = false;
        /**
         * filterOnColumn property
         */
        this.filterOnColumn = false;
        /**
         * filterSelectOptions property
         */
        this.filterSelectOptions = [];
        /**
         * updateFiltroHistorico property
         */
        this.updateFilterColumn = false;
        /**
         * customFilterOptions property
         */
        this.customFilterOptions = {};
        /**
         * buttonClick property
         * @type {EventEmitter<any>}
         */
        this.buttonClick = new EventEmitter();
        /**
         * startedLoading property
         * @type {EventEmitter<void>}
         */
        this.startedLoading = new EventEmitter();
        /**
         * finishedLoading property
         * @type {EventEmitter<HttpErrorResponse>}
         */
        this.finishedLoading = new EventEmitter();
        /**
         * onRowSelected property
         * @type {EventEmitter<any>}
         */
        this.onRowSelected = new EventEmitter();
        /**
         * showHorizontalButtonsHeader property
         * @type {boolean}
         */
        this.showHorizontalButtonsHeader = false;
        /**
         * blockContent
         * @type {boolean}
         */
        this.blockContent = false;
        /**
         * visibleColumns property
         * @type {Object}
         */
        this.visibleColumns = {};
        /**
         * showVisibleColumnsControl property
         * @type {boolean}
         */
        this.showVisibleColumnsControl = false;
        /**
         * dataKey property
         * Identificador único para cada registro da tabela
         * @type {string}
         */
        this.dataKey = 'id';
        /**
         * onRowEditEvent property
         * Evento disparado quando o usuário realiza ações de edição em uma linha da tabela
         * @type {EventEmitter<void>}
         */
        this.onRowEditEvent = new EventEmitter();
        this.clonedRowData = {};
        /**
         * errorMessage property
         * @type {string}
         */
        this.errorMessage = 'Ocorreu um erro ao carregar os dados da tabela. Tente novamente.';
        this.editingRow = false;
    }
    /**
     * ngOnInit method
     * @returns void
     */
    ngOnInit() {
        this.dataProvider = this.dataProviderFactory.create(this.type, this);
        this.pDatatableComponent.rows = this.rows;
        this.subscribeToLoadingEvents();
        this.initRowsPerPageOptionsIfPresent();
        this.defineIfShouldShowHorizontalButtonsHeader();
        this.updateOnValueChange();
        this.finishedLoading.subscribe(() => {
            this.updateOnValueChange();
        });
    }
    /**
     * subscribeFromValueChange method
     * @returns void
     */
    updateOnValueChange() {
        this.filterSelectOptions = [];
        if (this.value) {
            this.value.map((item) => {
                this.filterSelectOptions.push(item);
            });
        }
    }
    /**
     * subscribeToLoadingEvents method
     * @returns void
     */
    subscribeToLoadingEvents() {
        if (!this.disableLoadingBlockUI) {
            this.originalEmptyMessage = this.emptyMessage;
            this.startedLoading.subscribe(() => this.onStartedLoading());
            this.finishedLoading.subscribe(error => this.onFinishedLoading(error));
        }
    }
    /**
     * onStartedLoading method
     * @returns void
     */
    onStartedLoading() {
        this.blockContent = true;
        this.emptyMessage = this.originalEmptyMessage;
    }
    /**
     * onFinishedLoading method
     * @param {HttpErrorResponse} error
     * @returns void
     */
    onFinishedLoading(error) {
        this.blockContent = false;
        if (error) {
            this.emptyMessage = this.errorMessage;
        }
    }
    /**
     * initRowsPerPageOptionsIfPresent method
     * @returns void
     */
    initRowsPerPageOptionsIfPresent() {
        if (this.rowsPerPageOptions) {
            this.pDatatableComponent.rowsPerPageOptions = this.rowsPerPageOptions;
        }
    }
    /**
     * defineIfShouldShowHorizontalButtonsHeader method
     * @returns void
     */
    defineIfShouldShowHorizontalButtonsHeader() {
        if (this.horizontalButtonsHeader) {
            this.showHorizontalButtonsHeader = true;
        }
    }
    /**
     * ngAfterViewInit method
     * @returns void
     */
    ngAfterViewInit() {
        this.setExtraParams();
        this.pDatatableComponent.columns = this.columns;
        this.pDatatableComponent.ngAfterContentInit();
        this.setAllColumnsVisible();
        this.changeDetectorRef.detectChanges();
    }
    /**
     * setExtraParams method
     * @returns void
     */
    setExtraParams() {
        for (let i in this.extraParams) {
            this.pDatatableComponent[i] = this.extraParams[i];
        }
    }
    /**
     * ngAfterContentInit method
     * @returns void
     */
    ngAfterContentInit() {
        this.extraButtons.forEach((button) => {
            button.click.subscribe(event => {
                this.onClick(button.name, event);
            });
        });
    }
    /**
     * onClick method
     * @param {string} button
     * @param {any} event
     * @return void
     */
    onClick(button, event) {
        this.buttonClick.emit(new DatatableClickEvent(button, this.selectedRow));
        event.stopPropagation();
        this.resetSelectedRow(button);
    }
    /**
     * onSelectChange method
     * @param {any} event
     * @returns void
     */
    onSelectChange(event) {
        this.onRowSelected.emit(this.selectedRow);
        if (Array.isArray(this.selectedRow) && this.selectedRow.length > 1) {
            this.updateMultipleButtonsStatus(true);
        }
        else {
            this.updateMultipleButtonsStatus(false);
        }
    }
    /**
     * updateMultipleButtonsStatus method
     * @param {boolean} hasMultipleSelection
     * @return void
     */
    updateMultipleButtonsStatus(hasMultipleSelection) {
        this.extraButtons.forEach((button) => {
            if (hasMultipleSelection && !button.allowMultipleSelection) {
                button.disabled = true;
            }
            else {
                button.disabled = false;
            }
        });
    }
    /**
     * refresh method
     * @param query
     * @returns void
     */
    refresh(query) {
        this.dataProvider.load(query);
    }
    /**
     * resetSelectedRow method
     * @param button
     * @returns void
     */
    resetSelectedRow(button) {
        if (button === 'delete') {
            this.selectedRow = null;
        }
    }
    /**
     * reset method
     * @returns void
     */
    reset() {
        this.pDatatableComponent.first = 0;
        this.dataProvider.reset();
        this.refresh();
    }
    /**
     * paginationFooterText method
     * @param {string} msg
     * @returns string
     */
    paginationFooterText(msg) {
        const registros = this.totalRecords() === 1 ? 'registro' : 'registros';
        if (msg === '') {
            return `Exibindo: ${this.currentFirstIndex()} a ${this.currentLastIndex()} de ${this.totalRecords()} ${registros}`;
        }
        msg = msg.replace('{totalRecords}', this.totalRecords().toString());
        msg = msg.replace('{currentFirstIndex}', this.currentFirstIndex().toString());
        msg = msg.replace('{currentLastIndex}', this.currentLastIndex().toString());
        return msg;
    }
    /**
     * currentFirstIndex method
     * @returns number
     */
    currentFirstIndex() {
        return this.pDatatableComponent.first + 1;
    }
    /**
     * currentLastIndex method
     * @returns number
     */
    currentLastIndex() {
        const currentLastIndex = this.pDatatableComponent.first + this.pDatatableComponent.rows;
        if (currentLastIndex > this.totalRecords()) {
            return this.totalRecords();
        }
        return currentLastIndex;
    }
    /**
     * totalRecords method
     * @returns number
     */
    totalRecords() {
        return this.pDatatableComponent.totalRecords;
    }
    /**
     * checkRolesToButton method
     * @param {string} button
     * @return boolean
     */
    checkRolesToButton(button) {
        if (this[button]) {
            if (button === 'rulesToHide' || button === 'rulesToShowEdit' || button === 'rulesToShowDelete') {
                this.checkUseButtons(button);
                return this.authorizationService.hasRole(this[button]);
            }
            return !this.authorizationService.hasRole(this[button]);
        }
        return true;
    }
    /**
     * checkUseButtons method
     * @param {string} button
     * @returns void
     */
    checkUseButtons(button) {
        if (this[button] && this[button.replace('Show', 'Hide')]) {
            throw new Error('Não é possível definir dois valores de roles para os botões padrões.');
        }
    }
    /**
     * filter method
     * @returns void
     */
    filter() {
        for (let prop in this.filterParams) {
            if (this.filterParams[prop] === '' || (this.filterParams[prop].length === 0)) {
                delete this.filterParams[prop];
            }
        }
        if (Object.getOwnPropertyNames(this.filterParams).length === 0) {
            this.reset();
        }
        else {
            this.refresh(this.filterParams);
        }
    }
    /**
     * Método para filtrar por colunas
     * @param column chave da coluna na tabela
     * @param filters valor do filtro
     */
    filterColumn(column, filters) {
        const values = [];
        if (filters instanceof Array) {
            values[column] = [];
            for (let item in filters) {
                let val = (filters[item] instanceof Object) ? filters[item].value : filters[item];
                values[column].push(val);
            }
        }
        else {
            values[column] = filters;
        }
        this.filterParams = Object.assign(Object.assign({}, this.filterParams), values);
        return this.filter();
    }
    /**
     * Inicia todas as colunas da tabela como visíveis
     */
    setAllColumnsVisible() {
        this.columns.forEach((col) => {
            this.visibleColumns[col.field] = 'table-cell';
        });
    }
    /**
     * Verifica se uma coluna deve estar visível
     * @param column string
     * @param visibleColumns Array
     */
    visibleColumnCheck(column, visibleColumns) {
        return visibleColumns.some((item) => {
            return (item.field) ? item.field == column : true;
        });
    }
    /**
     * Atualzia os valores da variável 'visibleColuns' para mostrar ou não as colunas
     * @param columns array
     */
    updateVisibleColumns(columns) {
        for (let col in this.visibleColumns) {
            if (this.visibleColumnCheck(col, columns)) {
                this.visibleColumns[col] = 'table-cell';
            }
            else {
                this.visibleColumns[col] = 'none';
            }
        }
    }
    onRowEditInitEvent(event, rowData) {
        const tr = event.srcElement.closest('tr');
        setTimeout(() => {
            this.editingRow = true;
            const top = tr.offsetTop;
            const height = tr.offsetHeight;
            const editButtons = tr.querySelector('.editar-campos');
            editButtons.style.top = (top + height - 75) + 'px';
            editButtons.style.height = '75px';
            editButtons.style.backgroundColor = '#d0d0d0';
            editButtons.style.opacity = 1;
            this.clonedRowData[rowData.id] = Object.assign({}, rowData);
            this.onRowEditEvent.emit({ event: "init", data: rowData });
        }, null, tr);
    }
    onRowEditSaveEvent(event, rowData) {
        this.onRowEditEvent.emit({ event: "save", data: rowData });
        this.resetEditingMode(event);
    }
    onRowEditCancelEvent(event, rowData, index) {
        this.value[index] = this.clonedRowData[rowData.id];
        delete this.clonedRowData[rowData[this.dataKey]];
        this.onRowEditEvent.emit({ event: "cancel", data: rowData });
        this.resetEditingMode(event);
    }
    resetEditingMode(event) {
        const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
        editButtons.style.opacity = 0;
        editButtons.style.backgroundColor = '#0000006b';
        editButtons.style.display = 'none';
        this.editingRow = false;
    }
    editarLinhaOver(event) {
        if (!this.editingRow && this.editMode == 'row') {
            const top = event.srcElement.closest('tr').offsetTop;
            const height = event.srcElement.closest('tr').offsetHeight;
            const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
            editButtons.style.display = 'flex';
            editButtons.style.top = top + 'px';
            editButtons.style.height = height + 'px';
            editButtons.style.backgroundColor = '#0000006b';
            editButtons.style.opacity = 1;
        }
    }
    editarLinhaOut(event) {
        if (!this.editingRow) {
            const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
            editButtons.style.display = 'none';
            editButtons.style.opacity = 0;
        }
    }
};
DatatableComponent.ɵfac = function DatatableComponent_Factory(t) { return new (t || DatatableComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(DataProviderFactory), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Authorization)); };
DatatableComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DatatableComponent, selectors: [["basis-datatable"]], contentQueries: function DatatableComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, DatatableButtonComponent, false);
        ɵngcc0.ɵɵcontentQuery(dirIndex, Column, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.extraButtons = _t);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.columns = _t);
    } }, viewQuery: function DatatableComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(DataTable, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pDatatableComponent = _t.first);
    } }, inputs: { value: "value", type: "type", rows: "rows", showPaginationFooter: "showPaginationFooter", disableEdit: "disableEdit", disableView: "disableView", disableDelete: "disableDelete", selectionMode: "selectionMode", enableButtonsScroll: "enableButtonsScroll", emptyMessage: "emptyMessage", verticalButtons: "verticalButtons", enableScroll: "enableScroll", disableLoadingBlockUI: "disableLoadingBlockUI", filterOnColumn: "filterOnColumn", customFilterOptions: "customFilterOptions", showVisibleColumnsControl: "showVisibleColumnsControl", dataKey: "dataKey", splitButton: "splitButton", rowsPerPageOptions: "rowsPerPageOptions", url: "url", extraParams: "extraParams", paginationParameters: "paginationParameters", scrollHeight: "scrollHeight", scrollWidth: "scrollWidth", rulesToHide: "rulesToHide", rulesToShowEdit: "rulesToShowEdit", rulesToShowDelete: "rulesToShowDelete", rulesToHideView: "rulesToHideView", rulesToHideEdit: "rulesToHideEdit", rulesToHideDelete: "rulesToHideDelete", rowStyleClass: "rowStyleClass", editMode: "editMode" }, outputs: { buttonClick: "buttonClick", startedLoading: "startedLoading", finishedLoading: "finishedLoading", onRowSelected: "onRowSelected", onRowEditEvent: "onRowEditEvent" }, ngContentSelectors: _c9, decls: 17, vars: 26, consts: [[3, "blocked"], [1, "center-spinner"], [3, "ngClass"], [3, "columns", "updateVisibleColumns", 4, "ngIf"], [3, "value", "rows", "paginator", "selectionMode", "responsive", "scrollable", "scrollHeight", "selection", "editMode", "dataKey", "selectionChange", "onRowSelect", "onRowUnselect"], ["datatable", ""], ["pTemplate", "header"], ["pTemplate", "body"], [4, "ngIf"], ["class", "ui-g-1 ui-g-nopad", 3, "position", "top", 4, "ngIf"], ["defaultButtons", ""], [3, "columns", "updateVisibleColumns"], ["class", "ui-g", 4, "ngIf"], [3, "pSortableColumn", "ngStyle", 4, "ngFor", "ngForOf"], [1, "ui-g"], ["class", "ui-g-4 ui-sm-12", 4, "ngIf"], [4, "ngTemplateOutlet"], [1, "ui-g-4", "ui-sm-12"], [1, "horizontal-buttons-header"], [3, "pSortableColumn", "ngStyle"], [3, "field", 4, "ngIf"], [3, "field"], [3, "ngStyle", 4, "ngFor", "ngForOf"], [3, "ngStyle"], [3, "listaOriginal", "lista", "field", "updateFiltro", "filterType", "tableType", "customFilterOptions", "outputEvent"], [3, "pSelectableRow", "pContextMenuRow", "pEditableRow", "ngClass", "mouseover", "mouseleave"], ["class", "editar-campos", 4, "ngIf"], [3, "innerHTML", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext", 4, "ngIf"], [3, "innerHTML"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["pTemplate", "input"], ["pTemplate", "output"], ["class", "ui-md-12", 4, "ngIf"], [1, "ui-md-12"], [1, "ui-g", "form-group"], [1, "md-inputfield"], ["pInputText", "", "type", "text", 3, "ngModel", "ngModelChange"], [1, "editar-campos"], ["pButton", "", "type", "button", "pInitEditableRow", "", 3, "click", 4, "ngIf"], ["pButton", "", "type", "button", "pSaveEditableRow", "", "class", "ui-button-success", "style", "margin-right: .5em", 3, "click", 4, "ngIf"], ["pButton", "", "type", "button", "pCancelEditableRow", "", "class", "ui-button-danger", 3, "click", 4, "ngIf"], ["pButton", "", "type", "button", "pInitEditableRow", "", 3, "click"], ["pButton", "", "type", "button", "pSaveEditableRow", "", 1, "ui-button-success", 2, "margin-right", ".5em", 3, "click"], ["pButton", "", "type", "button", "pCancelEditableRow", "", 1, "ui-button-danger", 3, "click"], [1, "ui-g-1", "ui-g-nopad"], ["class", "ui-button-success", "icon", "edit", 3, "bTooltip", "horizontal-button", "disabled", "click", 4, "ngIf"], ["icon", "remove-red-eye", 3, "bTooltip", "horizontal-button", "disabled", "click", 4, "ngIf"], ["class", "ui-button-danger", "icon", "delete", 3, "bTooltip", "horizontal-button", "disabled", "click", 4, "ngIf"], [3, "icon", "class", "items", "label", "method", 4, "ngIf"], ["icon", "edit", 1, "ui-button-success", 3, "bTooltip", "disabled", "click"], ["icon", "remove-red-eye", 3, "bTooltip", "disabled", "click"], ["icon", "delete", 1, "ui-button-danger", 3, "bTooltip", "disabled", "click"], [3, "icon", "items", "label", "method"]], template: function DatatableComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c6);
        ɵngcc0.ɵɵelementStart(0, "p-blockUI", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "p-progressSpinner");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 2);
        ɵngcc0.ɵɵtext(4, "\u00A0");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 2);
        ɵngcc0.ɵɵtemplate(6, DatatableComponent_basis_show_column_6_Template, 1, 1, "basis-show-column", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 2);
        ɵngcc0.ɵɵelementStart(8, "p-table", 4, 5);
        ɵngcc0.ɵɵlistener("selectionChange", function DatatableComponent_Template_p_table_selectionChange_8_listener($event) { return ctx.selectedRow = $event; })("onRowSelect", function DatatableComponent_Template_p_table_onRowSelect_8_listener($event) { return ctx.onSelectChange($event); })("onRowUnselect", function DatatableComponent_Template_p_table_onRowUnselect_8_listener($event) { return ctx.onSelectChange($event); });
        ɵngcc0.ɵɵtemplate(10, DatatableComponent_ng_template_10_Template, 4, 3, "ng-template", 6);
        ɵngcc0.ɵɵtemplate(11, DatatableComponent_ng_template_11_Template, 2, 7, "ng-template", 7);
        ɵngcc0.ɵɵtemplate(12, DatatableComponent_12_Template, 1, 0, undefined, 8);
        ɵngcc0.ɵɵtemplate(13, DatatableComponent_p_footer_13_Template, 2, 1, "p-footer", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(14, DatatableComponent_div_14_Template, 3, 5, "div", 9);
        ɵngcc0.ɵɵtemplate(15, DatatableComponent_ng_template_15_Template, 4, 4, "ng-template", null, 10, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("blocked", ctx.blockContent);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(18, _c7, ctx.verticalButtons, !ctx.verticalButtons));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ctx.verticalButtons ? "ui-g-11" : "ui-g-12");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showVisibleColumnsControl);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction4(21, _c8, ctx.verticalButtons, !ctx.verticalButtons, ctx.editMode == "row", ctx.editingRow));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("value", ctx.value)("rows", ctx.rows)("paginator", true)("selectionMode", ctx.selectionMode)("responsive", true)("scrollable", ctx.enableScroll)("scrollHeight", ctx.scrollHeight)("selection", ctx.selectedRow)("editMode", ctx.editMode)("dataKey", ctx.dataKey);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.emptyMessage);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showPaginationFooter && ctx.totalRecords() > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.verticalButtons);
    } }, directives: [ɵngcc2.BlockUI, ɵngcc2.ProgressSpinner, ɵngcc3.NgClass, ɵngcc3.NgIf, ɵngcc4.Table, ɵngcc5.PrimeTemplate, ɵngcc6.ShowColumn, ɵngcc3.NgForOf, ɵngcc3.NgTemplateOutlet, ɵngcc4.SortableColumn, ɵngcc3.NgStyle, ɵngcc4.SortIcon, ɵngcc7.FilterColumn, ɵngcc4.SelectableRow, ɵngcc4.ContextMenuRow, ɵngcc4.EditableRow, ɵngcc4.CellEditor, ɵngcc8.DefaultValueAccessor, ɵngcc2.InputText, ɵngcc8.NgControlStatus, ɵngcc8.NgModel, ɵngcc2.ButtonDirective, ɵngcc4.InitEditableRow, ɵngcc4.SaveEditableRow, ɵngcc4.CancelEditableRow, ɵngcc5.Footer, ɵngcc9.DatatableButtonComponent, ɵngcc10.DatatableSplitbuttonComponent], styles: ["body .ui-datatable .ui-datatable-data tr td{word-wrap:break-word}.horizontal-buttons{display:flex;flex-flow:row wrap;justify-content:flex-end}.horizontal-button{padding:0 3px}.horizontal-buttons-header{font-size:1.3em;word-wrap:break-word}.center-spinner{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}.text-right{text-align:right}basis-show-column .ui-multiselect-panel.ui-widget{width:100%;max-width:100%}.edicao-tabela,.edita-linha{position:relative}tr.editing td{padding-bottom:75px!important}.editar-campos{opacity:0;position:absolute;width:100%;background-color:#0000006b;left:0;margin-top:0;display:flex;align-items:center;justify-content:center;min-height:35px;transition:.3s ease-in-out;z-index:1}.editar-campos button{color:#fff;font-size:14px;padding:6px 30px 10px;border-radius:5px;height:28px;width:120px!important}.bt-cancelar,.bt-salvar{margin:0 10px}"], encapsulation: 2 });
DatatableComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: DataProviderFactory },
    { type: Authorization }
];
__decorate([
    Input()
], DatatableComponent.prototype, "splitButton", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "value", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "type", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rows", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rowsPerPageOptions", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "url", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "extraParams", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "paginationParameters", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "showPaginationFooter", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "disableEdit", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "disableView", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "disableDelete", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "selectionMode", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "enableButtonsScroll", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "emptyMessage", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "verticalButtons", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "enableScroll", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "scrollHeight", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "scrollWidth", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToHide", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToShowEdit", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToShowDelete", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToHideView", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToHideEdit", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rulesToHideDelete", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "disableLoadingBlockUI", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "rowStyleClass", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "filterOnColumn", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "customFilterOptions", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "buttonClick", void 0);
__decorate([
    ViewChild(DataTable, { static: true })
], DatatableComponent.prototype, "pDatatableComponent", void 0);
__decorate([
    ContentChildren(DatatableButtonComponent)
], DatatableComponent.prototype, "extraButtons", void 0);
__decorate([
    ContentChildren(Column)
], DatatableComponent.prototype, "columns", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "startedLoading", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "finishedLoading", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "onRowSelected", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "showVisibleColumnsControl", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "editMode", void 0);
__decorate([
    Input()
], DatatableComponent.prototype, "dataKey", void 0);
__decorate([
    Output()
], DatatableComponent.prototype, "onRowEditEvent", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DatatableComponent, [{
        type: Component,
        args: [{
                selector: 'basis-datatable',
                template: "<p-blockUI [blocked]=\"blockContent\">\n    <div class=\"center-spinner\">\n        <p-progressSpinner></p-progressSpinner>\n    </div>\n</p-blockUI>\n<div [ngClass]=\"{'ui-g-2':verticalButtons, 'ui-g-5':!verticalButtons}\">&nbsp;</div>\n<div [ngClass]=\"verticalButtons ? 'ui-g-11' : 'ui-g-12'\">\n    <basis-show-column *ngIf=\"showVisibleColumnsControl\" [columns]=\"columns\" (updateVisibleColumns)=\"updateVisibleColumns($event)\"></basis-show-column>\n</div>\n<div [ngClass]=\"{\n        'ui-g-11' : verticalButtons,\n        'ui-g-12' : !verticalButtons,\n        'edicao-tabela' : editMode == 'row',\n        'editing-row': editingRow\n    }\">\n    <p-table\n        [value]=\"value\"\n        [rows]=\"rows\"\n        [paginator]=\"true\"\n        [selectionMode]=\"selectionMode\"\n        [responsive]=\"true\"\n        [scrollable]=\"enableScroll\"\n        [scrollHeight]=\"scrollHeight\"\n        [(selection)]=\"selectedRow\"\n        [editMode]=\"editMode\"\n        (onRowSelect)=\"onSelectChange($event)\"\n        (onRowUnselect)=\"onSelectChange($event)\"\n        #datatable\n        [dataKey]=\"dataKey\">\n        <ng-template pTemplate=\"header\" let-columns>\n            <div class=\"ui-g\" *ngIf=\"!verticalButtons\">\n                <div *ngIf=\"showHorizontalButtonsHeader\" class=\"ui-g-4 ui-sm-12\">\n                    <span class=\"horizontal-buttons-header\">{{ horizontalButtonsHeader }}</span>\n                </div>\n                <div [ngClass]=\"{\n                    'horizontal-buttons': true,\n                    'ui-sm-12': true,\n                    'ui-g-8': showHorizontalButtonsHeader,\n                    'ui-g-12': !showHorizontalButtonsHeader}\">\n                    <ng-container *ngTemplateOutlet=\"defaultButtons\"></ng-container>\n                    <ng-content select=\"horizontalButtons\"></ng-content>\n                </div>\n            </div>\n            <tr>\n                <th *ngFor=\"let col of columns\"\n                    [pSortableColumn]=\"col.sortable ? col.field : null\"\n                    [ngStyle]=\"{'display': visibleColumns[col.field]}\">\n                    {{col.header}}\n                    <p-sortIcon *ngIf=\"col.sortable\" [field]=\"col.field\"></p-sortIcon>\n                </th>\n            </tr>\n            <tr *ngIf=\"filterOnColumn\">\n                <th *ngFor=\"let col of columns\" [ngStyle]=\"{'display': visibleColumns[col.field]}\">\n                    <basis-filter-column\n                        [listaOriginal]=\"filterSelectOptions\"\n                        [lista]=\"filterSelectOptions\"\n                        [field]=\"col.field\"\n                        [updateFiltro]=\"updateFilterColumn\"\n                        (outputEvent)=\"filterColumn(col.field, $event)\"\n                        [filterType]=\"col.filterType\"\n                        [tableType]=\"type\"\n                        [customFilterOptions]=\"customFilterOptions\">\n                    </basis-filter-column>\n                </th>\n            </tr>\n        </ng-template>\n        <ng-template pTemplate=\"body\" let-rowData let-columns=\"columns\" let-editing=\"editing\" let-ri=\"rowIndex\">\n            <tr [pSelectableRow]=\"rowData\"\n                [pContextMenuRow]=\"rowData\"\n                [pEditableRow]=\"rowData\"\n                [ngClass]=\"{'editing': editing}\"\n                (mouseover)=\"editarLinhaOver($event)\"\n                (mouseleave)=\"editarLinhaOut($event)\">\n                <td *ngFor=\"let col of columns; last as isLast\" [ngStyle]=\"{'display': visibleColumns[col.field]}\" >\n                    <ng-container *ngIf=\"!editMode\">\n                        <span *ngIf=\"!col.bodyTemplate\" [innerHTML]=\"rowData[col.field]\"></span>\n                        <ng-container\n                            *ngIf=\"col.bodyTemplate\"\n                            [ngTemplateOutlet]=\"col.bodyTemplate\"\n                            [ngTemplateOutletContext]=\"{ rowData: rowData, col: col }\">\n                        </ng-container>\n                    </ng-container>\n\n                    <ng-container *ngIf=\"editMode\">\n                        <p-cellEditor>\n                            <ng-template pTemplate=\"input\">\n                                <span *ngIf=\"!col.outputTemplate && col.editable == false\" [innerHTML]=\"rowData[col.field]\"></span>\n                                <div class=\"ui-md-12\" *ngIf=\"!col.inputTemplate && col.editable != false\">\n                                    <div class=\"ui-g form-group\">\n                                        <div class=\"ui-md-12\">\n                                            <span class=\"md-inputfield\">\n                                                <input pInputText type=\"text\" [(ngModel)]=\"rowData[col.field]\" [style]=\"{'width':'100%'}\">\n                                                <label>{{ col.header }}</label>\n                                            </span>\n                                        </div>\n                                    </div>\n                                </div>\n                                <ng-container\n                                    *ngIf=\"col.inputTemplate\"\n                                    [ngTemplateOutlet]=\"col.inputTemplate\"\n                                    [ngTemplateOutletContext]=\"{ 'rowData': rowData, 'col': col }\">\n                                </ng-container>\n                            </ng-template>\n                            <ng-template pTemplate=\"output\">\n                                <span *ngIf=\"!col.outputTemplate\" [innerHTML]=\"rowData[col.field]\"></span>\n                                <ng-container\n                                    *ngIf=\"col.outputTemplate\"\n                                    [ngTemplateOutlet]=\"col.outputTemplate\"\n                                    [ngTemplateOutletContext]=\"{ rowData: rowData, col: col }\">\n                                </ng-container>\n                            </ng-template>\n                        </p-cellEditor>\n                    </ng-container>\n                    <div *ngIf=\"isLast\" class=\"editar-campos\">\n                        <button *ngIf=\"!editing\" pButton type=\"button\" pInitEditableRow (click)=\"onRowEditInitEvent($event, rowData)\">Editar</button>\n                        <button *ngIf=\"editing\" pButton type=\"button\" pSaveEditableRow class=\"ui-button-success\" style=\"margin-right: .5em\" (click)=\"onRowEditSaveEvent($event, rowData)\">Salvar</button>\n                        <button *ngIf=\"editing\" pButton type=\"button\" pCancelEditableRow class=\"ui-button-danger\" (click)=\"onRowEditCancelEvent($event, rowData, ri)\">Cancelar</button>\n                    </div>\n                </td>\n            </tr>\n        </ng-template>\n        <ng-template *ngIf=\"emptyMessage\">\n            {{ emptyMessage }}\n        </ng-template>\n        <p-footer *ngIf=\"showPaginationFooter && totalRecords() > 0\">\n            {{ paginationFooterText(footerMsg) }}\n        </p-footer>\n\n    </p-table>\n</div>\n<div *ngIf=\"verticalButtons\"\n    class=\"ui-g-1 ui-g-nopad\"\n    [style.position]=\"enableButtonsScroll ? 'sticky' : ''\"\n    [style.top]=\"enableButtonsScroll ? '15%' : ''\">\n        <ng-container *ngTemplateOutlet=\"defaultButtons\"></ng-container>\n        <ng-content></ng-content>\n</div>\n\n<ng-template #defaultButtons>\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Editar' : 'Escolha um registro para editar.' \"\n        [class.horizontal-button]=\"!verticalButtons\" class=\"ui-button-success\" icon=\"edit\" (click)=\"onClick('edit', $event)\"\n        *ngIf=\"!disableEdit && checkRolesToButton('rulesToShowView') && checkRolesToButton('rulesToHideView')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Visualizar' : 'Escolha um registro para visualizar.'\"\n        [class.horizontal-button]=\"!verticalButtons\" icon=\"remove-red-eye\" (click)=\"onClick('view', $event)\"\n        *ngIf=\"!disableView && checkRolesToButton('rulesToShowEdit') && checkRolesToButton('rulesToHideEdit')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Excluir' : 'Escolha um registro para excluir.'\"\n        [class.horizontal-button]=\"!verticalButtons\" class=\"ui-button-danger\" icon=\"delete\" (click)=\"onClick('delete', $event)\"\n       *ngIf=\"!disableDelete && checkRolesToButton('rulesToShowDelete') && checkRolesToButton('rulesToHideDelete')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <datatable-splitbutton\n        [icon]=\"splitButton.icon\"\n        [class]=\"splitButton.class\"\n        [items]=\"splitButton.itens\"\n        [label]=\"splitButton.label\"\n        [method]=\"splitButton.method\"\n        *ngIf=\"splitButton\">\n    </datatable-splitbutton>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None,
                styles: ["body .ui-datatable .ui-datatable-data tr td{word-wrap:break-word}.horizontal-buttons{display:flex;flex-flow:row wrap;justify-content:flex-end}.horizontal-button{padding:0 3px}.horizontal-buttons-header{font-size:1.3em;word-wrap:break-word}.center-spinner{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}.text-right{text-align:right}basis-show-column .ui-multiselect-panel.ui-widget{width:100%;max-width:100%}.edicao-tabela,.edita-linha{position:relative}tr.editing td{padding-bottom:75px!important}.editar-campos{opacity:0;position:absolute;width:100%;background-color:#0000006b;left:0;margin-top:0;display:flex;align-items:center;justify-content:center;min-height:35px;transition:.3s ease-in-out;z-index:1}.editar-campos button{color:#fff;font-size:14px;padding:6px 30px 10px;border-radius:5px;height:28px;width:120px!important}.bt-cancelar,.bt-salvar{margin:0 10px}"]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: DataProviderFactory }, { type: ɵngcc1.Authorization }]; }, { value: [{
            type: Input
        }], type: [{
            type: Input
        }], rows: [{
            type: Input
        }], showPaginationFooter: [{
            type: Input
        }], disableEdit: [{
            type: Input
        }], disableView: [{
            type: Input
        }], disableDelete: [{
            type: Input
        }], selectionMode: [{
            type: Input
        }], enableButtonsScroll: [{
            type: Input
        }], emptyMessage: [{
            type: Input
        }], verticalButtons: [{
            type: Input
        }], enableScroll: [{
            type: Input
        }], disableLoadingBlockUI: [{
            type: Input
        }], filterOnColumn: [{
            type: Input
        }], customFilterOptions: [{
            type: Input
        }], buttonClick: [{
            type: Output
        }], startedLoading: [{
            type: Output
        }], finishedLoading: [{
            type: Output
        }], onRowSelected: [{
            type: Output
        }], showVisibleColumnsControl: [{
            type: Input
        }], dataKey: [{
            type: Input
        }], onRowEditEvent: [{
            type: Output
        }], splitButton: [{
            type: Input
        }], rowsPerPageOptions: [{
            type: Input
        }], url: [{
            type: Input
        }], extraParams: [{
            type: Input
        }], paginationParameters: [{
            type: Input
        }], scrollHeight: [{
            type: Input
        }], scrollWidth: [{
            type: Input
        }], rulesToHide: [{
            type: Input
        }], rulesToShowEdit: [{
            type: Input
        }], rulesToShowDelete: [{
            type: Input
        }], rulesToHideView: [{
            type: Input
        }], rulesToHideEdit: [{
            type: Input
        }], rulesToHideDelete: [{
            type: Input
        }], rowStyleClass: [{
            type: Input
        }], pDatatableComponent: [{
            type: ViewChild,
            args: [DataTable, { static: true }]
        }], extraButtons: [{
            type: ContentChildren,
            args: [DatatableButtonComponent]
        }], columns: [{
            type: ContentChildren,
            args: [Column]
        }], editMode: [{
            type: Input
        }] }); })();
export { DatatableComponent };
