// @nuvem/primeng-components/esm2015/lib/datatable/datatable-button.component.js
import { __decorate } from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
/**
 * Class DatatableButtonComponent
 * @class
 */
let DatatableButtonComponent = class DatatableButtonComponent {
    constructor() {
        /**
         * allowMultipleSelection property
         * @type {boolean}
         */
        this.allowMultipleSelection = true;
        /**
         * disabled property
         * @type boolean
         */
        this.disabled = false;
        /**
         * click property
         * @type {EventEmitter<any>}
         */
        this.click = new EventEmitter();
    }
    /**
     * onClick method
     * @param {any} event
     * @return void
     */
    onClick(event) {
        this.click.emit(event);
        event.stopPropagation();
    }
};
__decorate([
    Input()
], DatatableButtonComponent.prototype, "icon", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "class", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "bTooltip", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "name", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "allowMultipleSelection", void 0);
__decorate([
    Input()
], DatatableButtonComponent.prototype, "disabled", void 0);
__decorate([
    Output()
], DatatableButtonComponent.prototype, "click", void 0);
DatatableButtonComponent = __decorate([
    Component({
        selector: 'basis-datatable-button',
        template: "<p pTooltip=\"{{ bTooltip }}\">\n    <button\n        type=\"button\" tooltipPosition=\"bottom\" class=\"{{ class }}\"\n        icon=\"ui-icon-{{ icon }}\" (click)=\"onClick($event)\" [disabled]=\"disabled\" pButton>\n    </button>\n</p>\n"
    })
], DatatableButtonComponent);
export { DatatableButtonComponent };
