// @nuvem/primeng-components/esm2015/lib/datatable/dataprovider-factory.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { DatatableService } from './datatable.service';
import { ServerDataProvider } from './server-dataprovider';
import { MemoryDataProvider } from './memory-dataprovider';
import * as i0 from "@angular/core";
import * as i1 from "./datatable.service";
/**
 * Enum DataProviderType
 * @enum
 */
import * as ɵngcc0 from '@angular/core';
export var DataProviderType;
(function (DataProviderType) {
    DataProviderType["Memory"] = "memory";
    DataProviderType["Server"] = "server";
})(DataProviderType || (DataProviderType = {}));
/**
 * Class DataProviderFactory
 * @class
 */
let DataProviderFactory = class DataProviderFactory {
    /**
     * constructor method
     * @constructor
     * @param {DatatableService} datatableService
     */
    constructor(datatableService) {
        this.datatableService = datatableService;
    }
    /**
     * create method
     * @param {DataProviderType} type
     * @param {DatatableComponent} component
     * @public
     * @returns DataProvider
     */
    create(type, component) {
        if (DataProviderType.Server === type) {
            return new ServerDataProvider(this.datatableService, component);
        }
        if (DataProviderType.Memory === type) {
            return new MemoryDataProvider(component);
        }
    }
};
DataProviderFactory.ɵfac = function DataProviderFactory_Factory(t) { return new (t || DataProviderFactory)(ɵngcc0.ɵɵinject(DatatableService)); };
DataProviderFactory.ctorParameters = () => [
    { type: DatatableService }
];
DataProviderFactory.ɵprov = i0.ɵɵdefineInjectable({ factory: function DataProviderFactory_Factory() { return new DataProviderFactory(i0.ɵɵinject(i1.DatatableService)); }, token: DataProviderFactory, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DataProviderFactory, [{
        type: Injectable,
        args: [{
                providedIn: "root"
            }]
    }], function () { return [{ type: DatatableService }]; }, null); })();
export { DataProviderFactory };
