// @nuvem/primeng-components/esm2015/lib/datatable/p-column.directive.js
import { __decorate } from "tslib";
import { Input, EventEmitter, TemplateRef, Output, ContentChildren, ContentChild, Directive } from '@angular/core';
import { PrimeTemplate } from 'primeng';
let Column = class Column {
    constructor() {
        this.filterType = 'text';
        this.exportable = true;
        this.resizable = true;
        this.sortFunction = new EventEmitter();
    }
    ngAfterContentInit() {
        this.templates.forEach((item) => {
            switch (item.getType()) {
                case 'header':
                    this.headerTemplate = item.template;
                    break;
                case 'body':
                    this.bodyTemplate = item.template;
                    break;
                case 'footer':
                    this.footerTemplate = item.template;
                    break;
                case 'filter':
                    this.filterTemplate = item.template;
                    break;
                case 'editor':
                    this.editorTemplate = item.template;
                    break;
                case 'input':
                    this.inputTemplate = item.template;
                    break;
                case 'output':
                    this.outputTemplate = item.template;
                    break;
                default:
                    this.bodyTemplate = item.template;
                    break;
            }
        });
    }
};
__decorate([
    Input()
], Column.prototype, "field", void 0);
__decorate([
    Input()
], Column.prototype, "colId", void 0);
__decorate([
    Input()
], Column.prototype, "sortField", void 0);
__decorate([
    Input()
], Column.prototype, "filterField", void 0);
__decorate([
    Input()
], Column.prototype, "header", void 0);
__decorate([
    Input()
], Column.prototype, "footer", void 0);
__decorate([
    Input()
], Column.prototype, "sortable", void 0);
__decorate([
    Input()
], Column.prototype, "editable", void 0);
__decorate([
    Input()
], Column.prototype, "filter", void 0);
__decorate([
    Input()
], Column.prototype, "filterMatchMode", void 0);
__decorate([
    Input()
], Column.prototype, "filterType", void 0);
__decorate([
    Input()
], Column.prototype, "excludeGlobalFilter", void 0);
__decorate([
    Input()
], Column.prototype, "rowspan", void 0);
__decorate([
    Input()
], Column.prototype, "colspan", void 0);
__decorate([
    Input()
], Column.prototype, "scope", void 0);
__decorate([
    Input()
], Column.prototype, "style", void 0);
__decorate([
    Input()
], Column.prototype, "styleClass", void 0);
__decorate([
    Input()
], Column.prototype, "exportable", void 0);
__decorate([
    Input()
], Column.prototype, "headerStyle", void 0);
__decorate([
    Input()
], Column.prototype, "headerStyleClass", void 0);
__decorate([
    Input()
], Column.prototype, "bodyStyle", void 0);
__decorate([
    Input()
], Column.prototype, "bodyStyleClass", void 0);
__decorate([
    Input()
], Column.prototype, "footerStyle", void 0);
__decorate([
    Input()
], Column.prototype, "footerStyleClass", void 0);
__decorate([
    Input()
], Column.prototype, "hidden", void 0);
__decorate([
    Input()
], Column.prototype, "expander", void 0);
__decorate([
    Input()
], Column.prototype, "selectionMode", void 0);
__decorate([
    Input()
], Column.prototype, "filterPlaceholder", void 0);
__decorate([
    Input()
], Column.prototype, "filterMaxlength", void 0);
__decorate([
    Input()
], Column.prototype, "frozen", void 0);
__decorate([
    Input()
], Column.prototype, "resizable", void 0);
__decorate([
    Output()
], Column.prototype, "sortFunction", void 0);
__decorate([
    ContentChildren(PrimeTemplate)
], Column.prototype, "templates", void 0);
__decorate([
    ContentChild(TemplateRef)
], Column.prototype, "template", void 0);
Column = __decorate([
    Directive({
        selector: 'p-column'
    })
], Column);
export { Column };
