// @nuvem/angular-base/esm2015/lib/form-notification/interceptor/form-notification.interceptor.js
import { __decorate } from "tslib";
import { Injectable } from "@angular/core";
import { FormNotificationService } from "../services/form-notification.service";
import { CleanFormNotification } from "../types/clean-form-notification.type";
import * as i0 from "@angular/core";
import * as i1 from "../services/form-notification.service";
/**
 * Implementação de classe para interceptar e tratar requisições http/https
 * @class
 */
import * as ɵngcc0 from '@angular/core';
let FormNotificationInterceptor = class FormNotificationInterceptor {
    /**
     * Método construtor
     * @constructor
     * @param {AbstractAuthentication} auth
     */
    constructor(formNotificationService) {
        this.formNotificationService = formNotificationService;
    }
    /**
     * Método responsável por interceptar requisições HTTP/HTTPS
     * @public
     * @param {HttpRequest<any>} request
     * @param {HttpHandler} next
     * @returns Observable<HttpEvent<any>>
     */
    intercept(request, next) {
        this.formNotificationService
            .notify(new CleanFormNotification(request.url));
        return next.handle(request);
    }
};
FormNotificationInterceptor.ɵfac = function FormNotificationInterceptor_Factory(t) { return new (t || FormNotificationInterceptor)(ɵngcc0.ɵɵinject(FormNotificationService)); };
FormNotificationInterceptor.ctorParameters = () => [
    { type: FormNotificationService }
];
FormNotificationInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function FormNotificationInterceptor_Factory() { return new FormNotificationInterceptor(i0.ɵɵinject(i1.FormNotificationService)); }, token: FormNotificationInterceptor, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormNotificationInterceptor, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: FormNotificationService }]; }, null); })();
export { FormNotificationInterceptor };
