// @nuvem/primeng-components/esm2015/lib/block-ui/block-ui.module.js
import { __decorate } from "tslib";
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { BlockUiInterceptor, BlockUiService } from '@nuvem/angular-base';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BlockUIModule, ProgressSpinnerModule } from 'primeng';
import { BlockUiComponent } from './block-ui.component';
/**
 * Class BlockUiModule
 * @class
 */
let BlockUiModule = class BlockUiModule {
};
BlockUiModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            BlockUIModule,
            ProgressSpinnerModule,
        ],
        declarations: [
            BlockUiComponent
        ],
        providers: [
            {
                provide: HTTP_INTERCEPTORS,
                useClass: BlockUiInterceptor,
                multi: true
            },
            BlockUiService
        ],
        exports: [
            BlockUIModule,
            BlockUiComponent,
            ProgressSpinnerModule
        ]
    })
], BlockUiModule);
export { BlockUiModule };
