// @nuvem/primeng-components/esm2015/lib/crud/components/dropdown/dropdown.component.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../form-notification/form-notification.component';
import * as ɵngcc2 from 'primeng/dropdown';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../validation-message/validation-message.component';

const _c0 = ["nDropdown", ""];
function DropdownComponent_nValidationMessage_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "nValidationMessage", 6);
} if (rf & 2) {
    const ctx_r971 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("control", ctx_r971.formControl);
} }
const _c1 = ["*"];
var DropdownComponent_1;
import { __decorate } from "tslib";
import { Component, Input, Output, EventEmitter } from "@angular/core";
import { provideValueAccessor } from "@nuvem/angular-base";
import { FieldComponent } from "../field.component";
let DropdownComponent = DropdownComponent_1 = class DropdownComponent extends FieldComponent {
    constructor() {
        super(...arguments);
        this.virtualScroll = false;
        this.filter = false;
        this.itemSize = 30;
        this.options = [];
        this.onChange = new EventEmitter();
    }
    ngOnInit() {
        super.ngOnInit();
        if (this.service) {
            this.service.findAll(null).subscribe(data => {
                this.options = data;
            });
        }
    }
    onChangeEvent(event) {
        this.onChange.emit(event);
    }
    outerToInner(value) {
        if (value && value.id) {
            this.service.find(value.id).subscribe((data) => {
                this.formControl.setValue(data);
            });
        }
        return value;
    }
};
DropdownComponent.ɵfac = function DropdownComponent_Factory(t) { return ɵDropdownComponent_BaseFactory(t || DropdownComponent); };
DropdownComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DropdownComponent, selectors: [["", "nDropdown", ""]], inputs: { virtualScroll: "virtualScroll", filter: "filter", itemSize: "itemSize", options: "options", optionLabel: "optionLabel", service: "service" }, outputs: { onChange: "onChange" }, features: [ɵngcc0.ɵɵProvidersFeature([provideValueAccessor(DropdownComponent_1)]), ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c1, decls: 7, vars: 11, consts: [["form-notification", "", 1, "md-inputfield", "custom-select", 3, "name", "form"], ["notification", ""], [3, "formControl", "optionLabel", "options", "filter", "virtualScroll", "itemSize", "onChange"], ["input", ""], [3, "for", "innerHTML"], [3, "control", 4, "ngIf"], [3, "control"]], template: function DropdownComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "span", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "p-dropdown", 2, 3);
        ɵngcc0.ɵɵlistener("onChange", function DropdownComponent_Template_p_dropdown_onChange_2_listener($event) { return ctx.onChangeEvent($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(4, "label", 4);
        ɵngcc0.ɵɵtemplate(5, DropdownComponent_nValidationMessage_5_Template, 1, 1, "nValidationMessage", 5);
        ɵngcc0.ɵɵprojection(6);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r969 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵproperty("name", ctx.name)("form", ctx.form);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("formControl", ctx.formControl)("optionLabel", ctx.optionLabel)("options", ctx.options)("filter", ctx.filter)("virtualScroll", ctx.virtualScroll)("itemSize", ctx.itemSize);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("for", ctx.name)("innerHTML", ctx.label, ɵngcc0.ɵɵsanitizeHtml);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !_r969.message);
    } }, directives: [ɵngcc1.FormNotificationComponent, ɵngcc2.Dropdown, ɵngcc3.NgControlStatus, ɵngcc3.FormControlDirective, ɵngcc4.NgIf, ɵngcc5.ValidationMessageComponent], encapsulation: 2 });
__decorate([
    Input()
], DropdownComponent.prototype, "optionLabel", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "virtualScroll", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "filter", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "itemSize", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "service", void 0);
__decorate([
    Input()
], DropdownComponent.prototype, "options", void 0);
__decorate([
    Output()
], DropdownComponent.prototype, "onChange", void 0);
const ɵDropdownComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(DropdownComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DropdownComponent, [{
        type: Component,
        args: [{
                selector: "[nDropdown]",
                template: "<span #notification form-notification class=\"md-inputfield custom-select\" [name]=\"name\" [form]=\"form\">\n    <p-dropdown #input\n        [formControl]=\"formControl\"\n        [optionLabel]=\"optionLabel\"\n        [options]=\"options\"\n        [filter]=\"filter\"\n        [virtualScroll]=\"virtualScroll\"\n        [itemSize]=\"itemSize\"\n        (onChange)=\"onChangeEvent($event)\">\n    </p-dropdown>\n    <label [for]=\"name\" [innerHTML]=\"label\"></label>\n    <nValidationMessage [control]=\"formControl\" *ngIf=\"!notification.message\"></nValidationMessage>\n    <ng-content></ng-content>\n</span>\n",
                providers: [provideValueAccessor(DropdownComponent_1)]
            }]
    }], null, { virtualScroll: [{
            type: Input
        }], filter: [{
            type: Input
        }], itemSize: [{
            type: Input
        }], options: [{
            type: Input
        }], onChange: [{
            type: Output
        }], optionLabel: [{
            type: Input
        }], service: [{
            type: Input
        }] }); })();
export { DropdownComponent };
