// @nuvem/angular-base/esm2015/lib/pipes/cnpj.pipe.js
import { __decorate } from "tslib";
import { Pipe } from '@angular/core';
/**
 *
 * Cria PIPE de CNPJ para valores.
 *
 * Usage:
 * {{ value | cnpj }}
*/
import * as ɵngcc0 from '@angular/core';
let CnpjPipe = class CnpjPipe {
    transform(value) {
        if (value === null || value.length !== 14) {
            return value;
        }
        else {
            value = value.replace(/[^0-9A-Za-z]/g, '');
            return value.substring(0, 2) + '.' +
                value.substring(2, 5) + '.' +
                value.substring(5, 8) + '/' +
                value.substring(8, 12) + '-' +
                value.substring(12, 14);
        }
    }
};
CnpjPipe.ɵfac = function CnpjPipe_Factory(t) { return new (t || CnpjPipe)(); };
CnpjPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "cnpj", type: CnpjPipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CnpjPipe, [{
        type: Pipe,
        args: [{
                name: 'cnpj'
            }]
    }], null, null); })();
export { CnpjPipe };
