// @nuvem/angular-base/esm2015/lib/pipes/pipe.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CpfPipe } from './cpf.pipe';
import { CepPipe } from './cep.pipe';
import { CnpjPipe } from './cnpj.pipe';
import { RegexPipe } from './pipe.regex';
import { LimitPipe } from './pipe.limit';
/**
 *
 *  Export do módulo
 */
import * as ɵngcc0 from '@angular/core';
let PipeModule = class PipeModule {
};
PipeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: PipeModule });
PipeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function PipeModule_Factory(t) { return new (t || PipeModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PipeModule, { declarations: function () { return [CepPipe,
        CpfPipe,
        CnpjPipe,
        RegexPipe,
        LimitPipe]; }, exports: function () { return [CepPipe,
        CpfPipe,
        CnpjPipe,
        RegexPipe,
        LimitPipe]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(PipeModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [
                    CepPipe,
                    CpfPipe,
                    CnpjPipe,
                    RegexPipe,
                    LimitPipe,
                ],
                exports: [
                    CepPipe,
                    CpfPipe,
                    CnpjPipe,
                    RegexPipe,
                    LimitPipe,
                ]
            }]
    }], null, null); })();
export { PipeModule };
