// @nuvem/angular-base/esm5/lib/form-notification/services/form-notification-error-provider.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { FormNotificationService } from './form-notification.service';
import { FieldFormNotification } from '../types/field-form-notification.type';
import * as i0 from "@angular/core";
import * as i1 from "./form-notification.service";
/**
 * Provedor de error para criação de erros de notificação.
 * @class
 */
import * as ɵngcc0 from '@angular/core';
var FormNotificationErrorProvider = /** @class */ (function () {
    /**
     * Método construtor
     * @param {FormtNotificationService} formNotificationService
     * @constructor
     */
    function FormNotificationErrorProvider(formNotificationService) {
        this.formNotificationService = formNotificationService;
    }
    /**
     * Método responsável pela identificação de erros
     * @public
     * @param {Error | HttpErrorResponse} error
     * @returns Boolean
     */
    FormNotificationErrorProvider.prototype.shouldHandle = function (error) {
        return error instanceof HttpErrorResponse &&
            error.headers.get('Content-Type') === 'application/problem+json' &&
            error.error && (error.error.violations || error.error.fieldErrors);
    };
    /**
     * Método para emissão de erros
     * @public
     * @param {HttpErrorResponse} error
     * @returns void
     */
    FormNotificationErrorProvider.prototype.handle = function (error) {
        if (error.error.violations) {
            this.violation(error.url, error.error.violations);
        }
        if (error.error.fieldErrors) {
            this.violation(error.url, error.error.fieldErrors);
        }
    };
    /**
     * Método iteração de violações e emissão de erros
     * @param values
     */
    FormNotificationErrorProvider.prototype.violation = function (form, values) {
        var _this = this;
        values.forEach(function (violation) {
            return _this.formNotificationService.notify(new FieldFormNotification(form, violation.field, violation.message));
        });
    };
    FormNotificationErrorProvider.ctorParameters = function () { return [
        { type: FormNotificationService }
    ]; };
    FormNotificationErrorProvider.ɵprov = i0.ɵɵdefineInjectable({ factory: function FormNotificationErrorProvider_Factory() { return new FormNotificationErrorProvider(i0.ɵɵinject(i1.FormNotificationService)); }, token: FormNotificationErrorProvider, providedIn: "root" });
FormNotificationErrorProvider.ɵfac = function FormNotificationErrorProvider_Factory(t) { return new (t || FormNotificationErrorProvider)(ɵngcc0.ɵɵinject(FormNotificationService)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormNotificationErrorProvider, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: FormNotificationService }]; }, null); })();
    return FormNotificationErrorProvider;
}());
export { FormNotificationErrorProvider };
