// @nuvem/angular-base/esm5/lib/form-notification/form-notification.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { FormNotificationErrorProvider } from './services/form-notification-error-provider.service';
import { FormNotificationService } from './services/form-notification.service';
import { ErrorProvider } from '../error/providers/error.provider';
import { CommonModule } from '@angular/common';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { FormNotificationInterceptor } from './interceptor/form-notification.interceptor';
import { FormNotificationComponentSuperclass } from './components/form-notification-superclass.component';
/**
 * Módulo para adicionar funcionalidade de notificação de erros de formulários.
 * @module
 */
import * as ɵngcc0 from '@angular/core';
var FormNotificationBaseModule = /** @class */ (function () {
    function FormNotificationBaseModule() {
    }
FormNotificationBaseModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FormNotificationBaseModule });
FormNotificationBaseModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FormNotificationBaseModule_Factory(t) { return new (t || FormNotificationBaseModule)(); }, providers: [
        FormNotificationService,
        {
            provide: ErrorProvider,
            useClass: FormNotificationErrorProvider,
            multi: true
        },
        {
            provide: HTTP_INTERCEPTORS,
            useClass: FormNotificationInterceptor,
            multi: true
        },
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormNotificationBaseModule, { declarations: function () { return [FormNotificationComponentSuperclass]; }, imports: function () { return [CommonModule]; }, exports: function () { return [FormNotificationComponentSuperclass]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormNotificationBaseModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [FormNotificationComponentSuperclass],
                exports: [FormNotificationComponentSuperclass],
                providers: [
                    FormNotificationService,
                    {
                        provide: ErrorProvider,
                        useClass: FormNotificationErrorProvider,
                        multi: true
                    },
                    {
                        provide: HTTP_INTERCEPTORS,
                        useClass: FormNotificationInterceptor,
                        multi: true
                    },
                ]
            }]
    }], function () { return []; }, null); })();
    return FormNotificationBaseModule;
}());
export { FormNotificationBaseModule };
