// primeng/esm5/steps/steps.js
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@angular/router';

function Steps_li_2_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r614 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 7);
    ɵngcc0.ɵɵlistener("click", function Steps_li_2_a_2_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r614); var ctx_r615 = ɵngcc0.ɵɵnextContext(); var item_r608 = ctx_r615.$implicit; var i_r609 = ctx_r615.index; var ctx_r613 = ɵngcc0.ɵɵnextContext(); return ctx_r613.itemClick($event, item_r608, i_r609); })("keydown.enter", function Steps_li_2_a_2_Template_a_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r614); var ctx_r617 = ɵngcc0.ɵɵnextContext(); var item_r608 = ctx_r617.$implicit; var i_r609 = ctx_r617.index; var ctx_r616 = ɵngcc0.ɵɵnextContext(); return ctx_r616.itemClick($event, item_r608, i_r609); });
    ɵngcc0.ɵɵelementStart(1, "span", 8);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 9);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r618 = ɵngcc0.ɵɵnextContext();
    var item_r608 = ctx_r618.$implicit;
    var i_r609 = ctx_r618.index;
    var ctx_r611 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵattribute("href", item_r608.url, ɵngcc0.ɵɵsanitizeUrl)("target", item_r608.target)("id", item_r608.id)("tabindex", item_r608.disabled || i_r609 !== ctx_r611.activeIndex && ctx_r611.readonly ? null : item_r608.tabindex ? item_r608.tabindex : "0");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(i_r609 + 1);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r608.label);
} }
var _c0 = function () { return { exact: false }; };
function Steps_li_2_a_3_Template(rf, ctx) { if (rf & 1) {
    var _r620 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 10);
    ɵngcc0.ɵɵlistener("click", function Steps_li_2_a_3_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r620); var ctx_r621 = ɵngcc0.ɵɵnextContext(); var item_r608 = ctx_r621.$implicit; var i_r609 = ctx_r621.index; var ctx_r619 = ɵngcc0.ɵɵnextContext(); return ctx_r619.itemClick($event, item_r608, i_r609); })("keydown.enter", function Steps_li_2_a_3_Template_a_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r620); var ctx_r623 = ɵngcc0.ɵɵnextContext(); var item_r608 = ctx_r623.$implicit; var i_r609 = ctx_r623.index; var ctx_r622 = ɵngcc0.ɵɵnextContext(); return ctx_r622.itemClick($event, item_r608, i_r609); });
    ɵngcc0.ɵɵelementStart(1, "span", 8);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 9);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r624 = ɵngcc0.ɵɵnextContext();
    var item_r608 = ctx_r624.$implicit;
    var i_r609 = ctx_r624.index;
    var ctx_r612 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("routerLink", item_r608.routerLink)("queryParams", item_r608.queryParams)("routerLinkActive", "ui-menuitem-link-active")("routerLinkActiveOptions", item_r608.routerLinkActiveOptions || ɵngcc0.ɵɵpureFunction0(15, _c0))("fragment", item_r608.fragment)("queryParamsHandling", item_r608.queryParamsHandling)("preserveFragment", item_r608.preserveFragment)("skipLocationChange", item_r608.skipLocationChange)("replaceUrl", item_r608.replaceUrl)("state", item_r608.state);
    ɵngcc0.ɵɵattribute("target", item_r608.target)("id", item_r608.id)("tabindex", item_r608.disabled || i_r609 !== ctx_r612.activeIndex && ctx_r612.readonly ? null : item_r608.tabindex ? item_r608.tabindex : "0");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(i_r609 + 1);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r608.label);
} }
var _c1 = function (a0, a1, a2, a3) { return { "ui-state-highlight ui-steps-current": a0, "ui-state-default": a1, "ui-state-complete": a2, "ui-state-disabled ui-steps-incomplete": a3 }; };
function Steps_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 3, 4);
    ɵngcc0.ɵɵtemplate(2, Steps_li_2_a_2_Template, 5, 6, "a", 5);
    ɵngcc0.ɵɵtemplate(3, Steps_li_2_a_3_Template, 5, 16, "a", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r608 = ctx.$implicit;
    var i_r609 = ctx.index;
    var ctx_r607 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMap(item_r608.styleClass);
    ɵngcc0.ɵɵproperty("ngStyle", item_r608.style)("ngClass", ɵngcc0.ɵɵpureFunction4(8, _c1, i_r609 === ctx_r607.activeIndex, i_r609 !== ctx_r607.activeIndex, i_r609 < ctx_r607.activeIndex, item_r608.disabled || i_r609 !== ctx_r607.activeIndex && ctx_r607.readonly));
    ɵngcc0.ɵɵattribute("aria-selected", i_r609 === ctx_r607.activeIndex)("aria-expanded", i_r609 === ctx_r607.activeIndex);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !item_r608.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r608.routerLink);
} }
var _c2 = function (a1) { return { "ui-steps ui-widget ui-helper-clearfix": true, "ui-steps-readonly": a1 }; };
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule, Component, Input, Output, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
var Steps = /** @class */ (function () {
    function Steps() {
        this.activeIndex = 0;
        this.readonly = true;
        this.activeIndexChange = new EventEmitter();
    }
    Steps.prototype.itemClick = function (event, item, i) {
        if (this.readonly || item.disabled) {
            event.preventDefault();
            return;
        }
        this.activeIndexChange.emit(i);
        if (!item.url) {
            event.preventDefault();
        }
        if (item.command) {
            item.command({
                originalEvent: event,
                item: item,
                index: i
            });
        }
    };
    __decorate([
        Input()
    ], Steps.prototype, "activeIndex", void 0);
    __decorate([
        Input()
    ], Steps.prototype, "model", void 0);
    __decorate([
        Input()
    ], Steps.prototype, "readonly", void 0);
    __decorate([
        Input()
    ], Steps.prototype, "style", void 0);
    __decorate([
        Input()
    ], Steps.prototype, "styleClass", void 0);
    __decorate([
        Output()
    ], Steps.prototype, "activeIndexChange", void 0);
Steps.ɵfac = function Steps_Factory(t) { return new (t || Steps)(); };
Steps.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: Steps, selectors: [["p-steps"]], inputs: { activeIndex: "activeIndex", readonly: "readonly", model: "model", style: "style", styleClass: "styleClass" }, outputs: { activeIndexChange: "activeIndexChange" }, decls: 3, vars: 7, consts: [[3, "ngClass", "ngStyle"], ["role", "tablist"], ["class", "ui-steps-item", "role", "tab", 3, "ngStyle", "class", "ngClass", 4, "ngFor", "ngForOf"], ["role", "tab", 1, "ui-steps-item", 3, "ngStyle", "ngClass"], ["menuitem", ""], ["class", "ui-menuitem-link", "role", "presentation", 3, "click", "keydown.enter", 4, "ngIf"], ["role", "presentation", "class", "ui-menuitem-link", 3, "routerLink", "queryParams", "routerLinkActive", "routerLinkActiveOptions", "fragment", "queryParamsHandling", "preserveFragment", "skipLocationChange", "replaceUrl", "state", "click", "keydown.enter", 4, "ngIf"], ["role", "presentation", 1, "ui-menuitem-link", 3, "click", "keydown.enter"], [1, "ui-steps-number"], [1, "ui-steps-title"], ["role", "presentation", 1, "ui-menuitem-link", 3, "routerLink", "queryParams", "routerLinkActive", "routerLinkActiveOptions", "fragment", "queryParamsHandling", "preserveFragment", "skipLocationChange", "replaceUrl", "state", "click", "keydown.enter"]], template: function Steps_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "ul", 1);
        ɵngcc0.ɵɵtemplate(2, Steps_li_2_Template, 4, 13, "li", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.styleClass);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c2, ctx.readonly))("ngStyle", ctx.style);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.model);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgStyle, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc2.RouterLinkWithHref, ɵngcc2.RouterLinkActive], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Steps, [{
        type: Component,
        args: [{
                selector: 'p-steps',
                template: "\n        <div [ngClass]=\"{'ui-steps ui-widget ui-helper-clearfix':true,'ui-steps-readonly':readonly}\" [ngStyle]=\"style\" [class]=\"styleClass\">\n            <ul role=\"tablist\">\n                <li *ngFor=\"let item of model; let i = index\" class=\"ui-steps-item\" #menuitem [ngStyle]=\"item.style\" [class]=\"item.styleClass\" role=\"tab\" [attr.aria-selected]=\"i === activeIndex\" [attr.aria-expanded]=\"i === activeIndex\"\n                    [ngClass]=\"{'ui-state-highlight ui-steps-current':(i === activeIndex),\n                        'ui-state-default':(i !== activeIndex),\n                        'ui-state-complete':(i < activeIndex),\n                        'ui-state-disabled ui-steps-incomplete':item.disabled||(i !== activeIndex && readonly)}\">\n                    <a *ngIf=\"!item.routerLink\" [attr.href]=\"item.url\" class=\"ui-menuitem-link\" role=\"presentation\" (click)=\"itemClick($event, item, i)\" (keydown.enter)=\"itemClick($event, item, i)\" [attr.target]=\"item.target\" [attr.id]=\"item.id\" \n                        [attr.tabindex]=\"item.disabled||(i !== activeIndex && readonly) ? null : (item.tabindex ? item.tabindex : '0')\">\n                        <span class=\"ui-steps-number\">{{i + 1}}</span>\n                        <span class=\"ui-steps-title\">{{item.label}}</span>\n                    </a>\n                    <a *ngIf=\"item.routerLink\" [routerLink]=\"item.routerLink\" [queryParams]=\"item.queryParams\" role=\"presentation\" [routerLinkActive]=\"'ui-menuitem-link-active'\" [routerLinkActiveOptions]=\"item.routerLinkActiveOptions||{exact:false}\" class=\"ui-menuitem-link\" \n                        (click)=\"itemClick($event, item, i)\" (keydown.enter)=\"itemClick($event, item, i)\" [attr.target]=\"item.target\" [attr.id]=\"item.id\" [attr.tabindex]=\"item.disabled||(i !== activeIndex && readonly) ? null : (item.tabindex ? item.tabindex : '0')\"\n                        [fragment]=\"item.fragment\" [queryParamsHandling]=\"item.queryParamsHandling\" [preserveFragment]=\"item.preserveFragment\" [skipLocationChange]=\"item.skipLocationChange\" [replaceUrl]=\"item.replaceUrl\" [state]=\"item.state\">\n                        <span class=\"ui-steps-number\">{{i + 1}}</span>\n                        <span class=\"ui-steps-title\">{{item.label}}</span>\n                    </a>\n                </li>\n            </ul>\n        </div>\n    ",
                changeDetection: ChangeDetectionStrategy.Default
            }]
    }], function () { return []; }, { activeIndex: [{
            type: Input
        }], readonly: [{
            type: Input
        }], activeIndexChange: [{
            type: Output
        }], model: [{
            type: Input
        }], style: [{
            type: Input
        }], styleClass: [{
            type: Input
        }] }); })();
    return Steps;
}());
export { Steps };
var StepsModule = /** @class */ (function () {
    function StepsModule() {
    }
StepsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: StepsModule });
StepsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function StepsModule_Factory(t) { return new (t || StepsModule)(); }, imports: [[CommonModule, RouterModule],
        RouterModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepsModule, { declarations: function () { return [Steps]; }, imports: function () { return [CommonModule, RouterModule]; }, exports: function () { return [Steps,
        RouterModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(StepsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, RouterModule],
                exports: [Steps, RouterModule],
                declarations: [Steps]
            }]
    }], function () { return []; }, null); })();
    return StepsModule;
}());
export { StepsModule };
