// @nuvem/angular-base/esm2015/lib/form-notification/services/form-notification-error-provider.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { FormNotificationService } from './form-notification.service';
import { FieldFormNotification } from '../types/field-form-notification.type';
import * as i0 from "@angular/core";
import * as i1 from "./form-notification.service";
/**
 * Provedor de error para criação de erros de notificação.
 * @class
 */
import * as ɵngcc0 from '@angular/core';
let FormNotificationErrorProvider = class FormNotificationErrorProvider {
    /**
     * Método construtor
     * @param {FormtNotificationService} formNotificationService
     * @constructor
     */
    constructor(formNotificationService) {
        this.formNotificationService = formNotificationService;
    }
    /**
     * Método responsável pela identificação de erros
     * @public
     * @param {Error | HttpErrorResponse} error
     * @returns Boolean
     */
    shouldHandle(error) {
        return error instanceof HttpErrorResponse &&
            error.headers.get('Content-Type') === 'application/problem+json' &&
            error.error && (error.error.violations || error.error.fieldErrors);
    }
    /**
     * Método para emissão de erros
     * @public
     * @param {HttpErrorResponse} error
     * @returns void
     */
    handle(error) {
        if (error.error.violations) {
            this.violation(error.url, error.error.violations);
        }
        if (error.error.fieldErrors) {
            this.violation(error.url, error.error.fieldErrors);
        }
    }
    /**
     * Método iteração de violações e emissão de erros
     * @param values
     */
    violation(form, values) {
        values.forEach(violation => this.formNotificationService.notify(new FieldFormNotification(form, violation.field, violation.message)));
    }
};
FormNotificationErrorProvider.ɵfac = function FormNotificationErrorProvider_Factory(t) { return new (t || FormNotificationErrorProvider)(ɵngcc0.ɵɵinject(FormNotificationService)); };
FormNotificationErrorProvider.ctorParameters = () => [
    { type: FormNotificationService }
];
FormNotificationErrorProvider.ɵprov = i0.ɵɵdefineInjectable({ factory: function FormNotificationErrorProvider_Factory() { return new FormNotificationErrorProvider(i0.ɵɵinject(i1.FormNotificationService)); }, token: FormNotificationErrorProvider, providedIn: "root" });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FormNotificationErrorProvider, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: FormNotificationService }]; }, null); })();
export { FormNotificationErrorProvider };
