// ng-block-ui/bundles/ng-block-ui-router.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('ng-block-ui'),exports, require('@angular/core'), require('ng-block-ui')) :
    typeof define === 'function' && define.amd ? define('ng-block-ui/router', ['@angular/core','ng-block-ui','exports', '@angular/core', 'ng-block-ui'], factory) :
    (factory(global.ng.core,global.ngBlockUi,(global['ng-block-ui'] = global['ng-block-ui'] || {}, global['ng-block-ui'].router = {}),global.ng.core,global['ng-block-ui']));
}(this, (function (ɵngcc0,ɵngcc1,exports,core,ngBlockUi) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var BlockUIPreventNavigation = (function () {
        function BlockUIPreventNavigation(blockUIService) {
            this.blockUIService = blockUIService;
        }
        /**
         * @return {?}
         */
        BlockUIPreventNavigation.prototype.canActivate = /**
         * @return {?}
         */
            function () {
                return !this.blockUIService.isActive(ngBlockUi.BLOCKUI_DEFAULT);
            };
        /**
         * @return {?}
         */
        BlockUIPreventNavigation.prototype.canActivateChild = /**
         * @return {?}
         */
            function () {
                return !this.blockUIService.isActive(ngBlockUi.BLOCKUI_DEFAULT);
            };
        /** @nocollapse */
        BlockUIPreventNavigation.ctorParameters = function () {
            return [
                { type: ngBlockUi.BlockUIService }
            ];
        };
BlockUIPreventNavigation.ɵfac = function BlockUIPreventNavigation_Factory(t) { return new (t || BlockUIPreventNavigation)(ɵngcc0.ɵɵinject(ɵngcc1.BlockUIService)); };
BlockUIPreventNavigation.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: BlockUIPreventNavigation, factory: function (t) { return BlockUIPreventNavigation.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUIPreventNavigation, [{
        type: core.Injectable
    }], function () { return [{ type: ɵngcc1.BlockUIService }]; }, null); })();
        return BlockUIPreventNavigation;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var BlockUIRouterModule = (function () {
        function BlockUIRouterModule() {
        }
        /**
         * @return {?}
         */
        BlockUIRouterModule.forRoot = /**
         * @return {?}
         */
            function () {
                return {
                    ngModule: BlockUIRouterModule,
                    providers: [
                        BlockUIPreventNavigation
                    ]
                };
            };
BlockUIRouterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BlockUIRouterModule });
BlockUIRouterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function BlockUIRouterModule_Factory(t) { return new (t || BlockUIRouterModule)(); }, imports: [[
            ngBlockUi.BlockUIModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BlockUIRouterModule, { imports: [ɵngcc1.BlockUIModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BlockUIRouterModule, [{
        type: core.NgModule,
        args: [{
                imports: [
                    ngBlockUi.BlockUIModule
                ]
            }]
    }], function () { return []; }, null); })();
        return BlockUIRouterModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.BlockUIRouterModule = BlockUIRouterModule;
    exports.BlockUIPreventNavigation = BlockUIPreventNavigation;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
